// @dcl/schemas/dist/dapps/preview/spring-bone-params.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SpringBoneParams = void 0;
const validation_1 = require("../../validation");
/** @alpha */
var SpringBoneParams;
(function (SpringBoneParams) {
    SpringBoneParams.schema = {
        type: 'object',
        properties: {
            stiffness: { type: 'number', minimum: 0, maximum: 5 },
            gravityPower: { type: 'number', minimum: 0, maximum: 10 },
            gravityDir: {
                type: 'array',
                items: [{ type: 'number' }, { type: 'number' }, { type: 'number' }],
                minItems: 3,
                maxItems: 3
            },
            dragForce: { type: 'number', minimum: 0, maximum: 1 },
            center: { type: 'integer', minimum: 0, nullable: true }
        },
        required: ['stiffness', 'gravityPower', 'gravityDir', 'dragForce']
    };
    SpringBoneParams.validate = (0, validation_1.generateLazyValidator)(SpringBoneParams.schema);
})(SpringBoneParams = exports.SpringBoneParams || (exports.SpringBoneParams = {}));
