// @dcl/protocol/out-js/decentraland/social_service/errors.gen.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NotFoundError = exports.ForbiddenError = exports.ConflictingError = exports.ProfileNotFound = exports.InvalidRequest = exports.InternalServerError = exports.InvalidFriendshipAction = exports.protobufPackage = void 0;
/* eslint-disable */
const minimal_1 = __importDefault(require("protobufjs/minimal"));
exports.protobufPackage = "decentraland.social_service";
function createBaseInvalidFriendshipAction() {
    return { message: undefined };
}
var InvalidFriendshipAction;
(function (InvalidFriendshipAction) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.message !== undefined) {
            writer.uint32(10).string(message.message);
        }
        return writer;
    }
    InvalidFriendshipAction.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseInvalidFriendshipAction();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.message = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    InvalidFriendshipAction.decode = decode;
    function fromJSON(object) {
        return { message: isSet(object.message) ? String(object.message) : undefined };
    }
    InvalidFriendshipAction.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.message !== undefined && (obj.message = message.message);
        return obj;
    }
    InvalidFriendshipAction.toJSON = toJSON;
    function create(base) {
        return InvalidFriendshipAction.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    InvalidFriendshipAction.create = create;
    function fromPartial(object) {
        var _a;
        const message = createBaseInvalidFriendshipAction();
        message.message = (_a = object.message) !== null && _a !== void 0 ? _a : undefined;
        return message;
    }
    InvalidFriendshipAction.fromPartial = fromPartial;
})(InvalidFriendshipAction || (exports.InvalidFriendshipAction = InvalidFriendshipAction = {}));
function createBaseInternalServerError() {
    return { message: undefined };
}
var InternalServerError;
(function (InternalServerError) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.message !== undefined) {
            writer.uint32(10).string(message.message);
        }
        return writer;
    }
    InternalServerError.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseInternalServerError();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.message = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    InternalServerError.decode = decode;
    function fromJSON(object) {
        return { message: isSet(object.message) ? String(object.message) : undefined };
    }
    InternalServerError.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.message !== undefined && (obj.message = message.message);
        return obj;
    }
    InternalServerError.toJSON = toJSON;
    function create(base) {
        return InternalServerError.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    InternalServerError.create = create;
    function fromPartial(object) {
        var _a;
        const message = createBaseInternalServerError();
        message.message = (_a = object.message) !== null && _a !== void 0 ? _a : undefined;
        return message;
    }
    InternalServerError.fromPartial = fromPartial;
})(InternalServerError || (exports.InternalServerError = InternalServerError = {}));
function createBaseInvalidRequest() {
    return { message: undefined };
}
var InvalidRequest;
(function (InvalidRequest) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.message !== undefined) {
            writer.uint32(10).string(message.message);
        }
        return writer;
    }
    InvalidRequest.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseInvalidRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.message = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    InvalidRequest.decode = decode;
    function fromJSON(object) {
        return { message: isSet(object.message) ? String(object.message) : undefined };
    }
    InvalidRequest.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.message !== undefined && (obj.message = message.message);
        return obj;
    }
    InvalidRequest.toJSON = toJSON;
    function create(base) {
        return InvalidRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    InvalidRequest.create = create;
    function fromPartial(object) {
        var _a;
        const message = createBaseInvalidRequest();
        message.message = (_a = object.message) !== null && _a !== void 0 ? _a : undefined;
        return message;
    }
    InvalidRequest.fromPartial = fromPartial;
})(InvalidRequest || (exports.InvalidRequest = InvalidRequest = {}));
function createBaseProfileNotFound() {
    return { message: undefined };
}
var ProfileNotFound;
(function (ProfileNotFound) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.message !== undefined) {
            writer.uint32(10).string(message.message);
        }
        return writer;
    }
    ProfileNotFound.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseProfileNotFound();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.message = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    ProfileNotFound.decode = decode;
    function fromJSON(object) {
        return { message: isSet(object.message) ? String(object.message) : undefined };
    }
    ProfileNotFound.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.message !== undefined && (obj.message = message.message);
        return obj;
    }
    ProfileNotFound.toJSON = toJSON;
    function create(base) {
        return ProfileNotFound.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    ProfileNotFound.create = create;
    function fromPartial(object) {
        var _a;
        const message = createBaseProfileNotFound();
        message.message = (_a = object.message) !== null && _a !== void 0 ? _a : undefined;
        return message;
    }
    ProfileNotFound.fromPartial = fromPartial;
})(ProfileNotFound || (exports.ProfileNotFound = ProfileNotFound = {}));
function createBaseConflictingError() {
    return { message: undefined };
}
var ConflictingError;
(function (ConflictingError) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.message !== undefined) {
            writer.uint32(10).string(message.message);
        }
        return writer;
    }
    ConflictingError.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseConflictingError();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.message = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    ConflictingError.decode = decode;
    function fromJSON(object) {
        return { message: isSet(object.message) ? String(object.message) : undefined };
    }
    ConflictingError.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.message !== undefined && (obj.message = message.message);
        return obj;
    }
    ConflictingError.toJSON = toJSON;
    function create(base) {
        return ConflictingError.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    ConflictingError.create = create;
    function fromPartial(object) {
        var _a;
        const message = createBaseConflictingError();
        message.message = (_a = object.message) !== null && _a !== void 0 ? _a : undefined;
        return message;
    }
    ConflictingError.fromPartial = fromPartial;
})(ConflictingError || (exports.ConflictingError = ConflictingError = {}));
function createBaseForbiddenError() {
    return { message: undefined };
}
var ForbiddenError;
(function (ForbiddenError) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.message !== undefined) {
            writer.uint32(10).string(message.message);
        }
        return writer;
    }
    ForbiddenError.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseForbiddenError();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.message = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    ForbiddenError.decode = decode;
    function fromJSON(object) {
        return { message: isSet(object.message) ? String(object.message) : undefined };
    }
    ForbiddenError.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.message !== undefined && (obj.message = message.message);
        return obj;
    }
    ForbiddenError.toJSON = toJSON;
    function create(base) {
        return ForbiddenError.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    ForbiddenError.create = create;
    function fromPartial(object) {
        var _a;
        const message = createBaseForbiddenError();
        message.message = (_a = object.message) !== null && _a !== void 0 ? _a : undefined;
        return message;
    }
    ForbiddenError.fromPartial = fromPartial;
})(ForbiddenError || (exports.ForbiddenError = ForbiddenError = {}));
function createBaseNotFoundError() {
    return { message: undefined };
}
var NotFoundError;
(function (NotFoundError) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.message !== undefined) {
            writer.uint32(10).string(message.message);
        }
        return writer;
    }
    NotFoundError.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseNotFoundError();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.message = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    NotFoundError.decode = decode;
    function fromJSON(object) {
        return { message: isSet(object.message) ? String(object.message) : undefined };
    }
    NotFoundError.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.message !== undefined && (obj.message = message.message);
        return obj;
    }
    NotFoundError.toJSON = toJSON;
    function create(base) {
        return NotFoundError.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    NotFoundError.create = create;
    function fromPartial(object) {
        var _a;
        const message = createBaseNotFoundError();
        message.message = (_a = object.message) !== null && _a !== void 0 ? _a : undefined;
        return message;
    }
    NotFoundError.fromPartial = fromPartial;
})(NotFoundError || (exports.NotFoundError = NotFoundError = {}));
function isSet(value) {
    return value !== null && value !== undefined;
}
