// @dcl/asset-packs/dist/versioning/framework.js
import { Schemas } from '@dcl/ecs';
function mergeMapProperties(targetProps, sourceProps) {
    const result = { ...targetProps };
    for (const [key, sourceProp] of Object.entries(sourceProps)) {
        const targetProp = targetProps[key];
        if (targetProp?.serializationType === 'map' &&
            sourceProp?.serializationType === 'map') {
            result[key] = {
                ...sourceProp,
                properties: mergeMapProperties(targetProp.properties || {}, sourceProp.properties || {}),
            };
        }
        else {
            result[key] = sourceProp;
        }
    }
    return result;
}
function deepMergeSchemas(target, source) {
    const result = { ...target };
    for (const [key, sourceValue] of Object.entries(source)) {
        const targetValue = target[key];
        if (!(key in target)) {
            result[key] = sourceValue;
        }
        else if (targetValue?.jsonSchema?.serializationType === 'map' &&
            sourceValue?.jsonSchema?.serializationType === 'map') {
            const mergedProps = mergeMapProperties(targetValue.jsonSchema.properties || {}, sourceValue.jsonSchema.properties || {});
            const spec = {};
            for (const [k, v] of Object.entries(mergedProps)) {
                spec[k] = Schemas.fromJson(v);
            }
            result[key] = Schemas.Map(spec);
        }
        else if (targetValue?.jsonSchema?.serializationType === 'optional' &&
            sourceValue?.jsonSchema?.serializationType === 'optional') {
            const targetInner = targetValue.jsonSchema.optionalJsonSchema;
            const sourceInner = sourceValue.jsonSchema.optionalJsonSchema;
            if (targetInner?.serializationType === 'map' && sourceInner?.serializationType === 'map') {
                const mergedProps = mergeMapProperties(targetInner.properties || {}, sourceInner.properties || {});
                const spec = {};
                for (const [k, v] of Object.entries(mergedProps)) {
                    spec[k] = Schemas.fromJson(v);
                }
                result[key] = Schemas.Optional(Schemas.Map(spec));
            }
            else if (targetInner?.serializationType === 'array' &&
                sourceInner?.serializationType === 'array') {
                const targetItems = targetInner.items;
                const sourceItems = sourceInner.items;
                if (targetItems?.serializationType === 'map' && sourceItems?.serializationType === 'map') {
                    const mergedProps = mergeMapProperties(targetItems.properties || {}, sourceItems.properties || {});
                    const spec = {};
                    for (const [k, v] of Object.entries(mergedProps)) {
                        spec[k] = Schemas.fromJson(v);
                    }
                    result[key] = Schemas.Optional(Schemas.Array(Schemas.Map(spec)));
                }
                else {
                    result[key] = sourceValue;
                }
            }
            else {
                result[key] = sourceValue;
            }
        }
        else {
            result[key] = sourceValue;
        }
    }
    return result;
}
export function createComponentFramework(registry) {
    const VERSIONS_REGISTRY = Object.fromEntries(Object.entries(registry).map(([baseName, diffs]) => {
        const versions = [];
        let merged = {};
        for (let i = 0; i < diffs.length; i++) {
            merged = deepMergeSchemas(merged, diffs[i]);
            versions.push({
                versionName: i === 0 ? baseName : `${baseName}-v${i}`,
                component: { ...merged },
            });
        }
        return [baseName, versions];
    }));
    function getLatestVersionName(baseName) {
        const versions = VERSIONS_REGISTRY[baseName];
        if (!versions || versions.length === 0) {
            throw new Error(`No versions found for component: ${baseName}`);
        }
        return versions[versions.length - 1].versionName;
    }
    function defineAllComponents(engine) {
        const components = {};
        for (const [baseName, versions] of Object.entries(VERSIONS_REGISTRY)) {
            for (const v of versions) {
                engine.defineComponent(v.versionName, v.component);
            }
            const latest = versions[versions.length - 1];
            components[baseName] = engine.getComponent(latest.versionName);
        }
        return components;
    }
    function migrateAll(engine) {
        for (const [baseName, versions] of Object.entries(VERSIONS_REGISTRY)) {
            migrateVersionedComponent(engine, baseName, versions);
        }
    }
    return {
        VERSIONS_REGISTRY,
        getLatestVersionName,
        defineAllComponents,
        migrateAll,
    };
}
export function migrateVersionedComponent(engine, baseName, versionedComponents) {
    const latestComponentVersion = versionedComponents[versionedComponents.length - 1];
    console.log(`[MIGRATION] component: ${baseName}`);
    console.log(`[MIGRATION] target version: ${latestComponentVersion.versionName}`);
    const componentsToMigrate = [...engine.componentsIter()].filter(c => {
        const isComponentVersion = c.componentName === baseName || c.componentName.startsWith(`${baseName}-v`);
        return isComponentVersion && c.componentName !== latestComponentVersion.versionName;
    });
    console.log(`[MIGRATION] Versions to migrate to latest: ${componentsToMigrate.map(c => c.componentName)}`);
    for (const component of componentsToMigrate) {
        const oldComponent = component;
        const entities = [...engine.getEntitiesWith(oldComponent)];
        console.log(`[MIGRATION] Entities with ${component.componentName}: ${entities.length}`);
        if (entities.length === 0)
            continue;
        const targetComponent = engine.getComponent(latestComponentVersion.versionName);
        for (const [entity, value] of entities) {
            console.log(`[MIGRATION] Migrating entity ${entity} from ${component.componentName} to ${latestComponentVersion.versionName}`);
            console.log('[MIGRATION] Old value:', value);
            oldComponent.deleteFrom(entity);
            targetComponent.createOrReplace(entity, { ...value });
        }
    }
    console.log('[MIGRATION] Finished');
}
