// @dcl-sdk/utils/dist/trigger.js
import { engine, MeshRenderer, Transform, Material, EntityState, TriggerArea, triggerAreaEventsSystem } from '@dcl/sdk/ecs';
import { Vector3, Color4, Color3, Quaternion } from '@dcl/sdk/math';
import { getWorldPosition, getWorldRotation } from './math';
import { priority } from './priority';
export const LAYER_1 = 1;
export const LAYER_2 = 2;
export const LAYER_3 = 4;
export const LAYER_4 = 8;
export const LAYER_5 = 16;
export const LAYER_6 = 32;
export const LAYER_7 = 64;
export const LAYER_8 = 128;
export const ALL_LAYERS = 255;
export const NO_LAYERS = 0;
export let PLAYER_LAYER_ID = LAYER_1;
function createTriggers(targetEngine) {
    const stateByRoot = new Map();
    const areaChildrenByRoot = new Map();
    const childToRoot = new Map();
    const overlapCountByPair = new Map();
    let debugDraw = false;
    function utilsMaskToColliderCustom(mask) {
        let res = 0;
        if (mask & LAYER_1)
            res |= 256;
        if (mask & LAYER_2)
            res |= 512;
        if (mask & LAYER_3)
            res |= 1024;
        if (mask & LAYER_4)
            res |= 2048;
        if (mask & LAYER_5)
            res |= 4096;
        if (mask & LAYER_6)
            res |= 8192;
        if (mask & LAYER_7)
            res |= 16384;
        if (mask & LAYER_8)
            res |= 32768;
        return res;
    }
    function utilsTriggeredByToTriggerAreaMask(triggeredByMask) {
        let res = utilsMaskToColliderCustom(triggeredByMask);
        if (triggeredByMask & PLAYER_LAYER_ID) {
            res |= 4;
        }
        return res;
    }
    function buildInternalAreasFromSpec(areas) {
        if (!areas)
            areas = [{ type: 'box' }];
        const result = [];
        for (const area of areas) {
            if (area.type == 'box') {
                result.push({
                    $case: 'box',
                    value: {
                        position: area.position ? area.position : Vector3.Zero(),
                        scale: area.scale ? area.scale : Vector3.One()
                    }
                });
            }
            else {
                result.push({
                    $case: 'sphere',
                    value: {
                        position: area.position ? area.position : Vector3.Zero(),
                        radius: area.radius ? area.radius : 1
                    }
                });
            }
        }
        return result;
    }
    function setChildTransformFromRoot(child, root, area) {
        const rootWorldPos = getWorldPosition(root);
        const rootWorldRot = getWorldRotation(root);
        const offset = Vector3.rotate(area.value.position, rootWorldRot);
        const position = Vector3.add(rootWorldPos, offset);
        let scale;
        let rotation;
        if (area.$case == 'box') {
            scale = area.value.scale;
            rotation = rootWorldRot;
        }
        else {
            scale = { x: area.value.radius, y: area.value.radius, z: area.value.radius };
            rotation = Quaternion.Identity();
        }
        Transform.createOrReplace(child, {
            position: position,
            scale: scale,
            rotation: rotation
        });
    }
    function attachComponentsToAreaChild(child, area, triggeredByMask, layerMask) {
        const triggerMask = utilsTriggeredByToTriggerAreaMask(triggeredByMask);
        if (area.$case == 'box') {
            TriggerArea.setBox(child, triggerMask);
        }
        else {
            TriggerArea.setSphere(child, triggerMask);
        }
    }
    function subscribeAreaEvents(child, root) {
        triggerAreaEventsSystem.onTriggerEnter(child, (result) => {
            const rootState = stateByRoot.get(root);
            if (!rootState || !rootState.active)
                return;
            let other = result.trigger.entity;
            if (childToRoot.has(other))
                other = childToRoot.get(other);
            if (other == root)
                return;
            let map = overlapCountByPair.get(root);
            if (!map) {
                map = new Map();
                overlapCountByPair.set(root, map);
            }
            const prev = map.get(other) || 0;
            const next = prev + 1;
            map.set(other, next);
            if (prev == 0 && rootState.onEnter)
                rootState.onEnter(other);
        });
        triggerAreaEventsSystem.onTriggerExit(child, (result) => {
            const rootState = stateByRoot.get(root);
            if (!rootState)
                return;
            let other = result.trigger.entity;
            if (childToRoot.has(other))
                other = childToRoot.get(other);
            if (other == root)
                return;
            const map = overlapCountByPair.get(root);
            if (!map)
                return;
            const prev = map.get(other) || 0;
            if (prev <= 0)
                return;
            const next = prev - 1;
            if (next == 0) {
                map.delete(other);
                if (rootState.onExit)
                    rootState.onExit(other);
            }
            else {
                map.set(other, next);
            }
        });
    }
    function createAreaChildren(root) {
        const s = stateByRoot.get(root);
        const areas = buildInternalAreasFromSpec(s.areas);
        const children = [];
        for (const area of areas) {
            const child = targetEngine.addEntity();
            childToRoot.set(child, root);
            setChildTransformFromRoot(child, root, area);
            attachComponentsToAreaChild(child, area, s.triggeredByMask, s.layerMask);
            subscribeAreaEvents(child, root);
            children.push(child);
        }
        areaChildrenByRoot.set(root, children);
    }
    function destroyAreaChildren(root) {
        const children = areaChildrenByRoot.get(root);
        if (!children)
            return;
        for (const child of children) {
            if (debugDraw) {
            }
            childToRoot.delete(child);
            targetEngine.removeEntity(child);
        }
        areaChildrenByRoot.delete(root);
    }
    function updateDebugForRoot(root) {
        const s = stateByRoot.get(root);
        if (!s)
            return;
        const children = areaChildrenByRoot.get(root) || [];
        for (const child of children) {
            if (debugDraw) {
            }
        }
    }
    function setDebugMeshForArea(child, area, color, active) {
        if (area.$case == 'box')
            MeshRenderer.setBox(child);
        else
            MeshRenderer.setSphere(child);
        const c = active ? color : Color3.Black();
        Material.setPbrMaterial(child, { albedoColor: Color4.fromInts(255 * c.r, 255 * c.g, 255 * c.b, 75) });
    }
    function refreshDebugMeshes(root) {
        if (!debugDraw)
            return;
        const s = stateByRoot.get(root);
        const areas = buildInternalAreasFromSpec(s.areas);
        const children = areaChildrenByRoot.get(root) || [];
        for (let i = 0; i < children.length; i++)
            setDebugMeshForArea(children[i], areas[i], s.debugColor, s.active);
    }
    function enableDisableRoot(root, enabled) {
        const s = stateByRoot.get(root);
        s.active = enabled;
        const children = areaChildrenByRoot.get(root) || [];
        for (const child of children) {
            if (enabled) {
            }
        }
        if (!enabled) {
            const map = overlapCountByPair.get(root);
            if (map) {
                const onExit = s.onExit;
                for (const other of map.keys())
                    if (onExit)
                        onExit(other);
                map.clear();
            }
        }
        refreshDebugMeshes(root);
    }
    function system(dt) {
        for (const [root] of stateByRoot) {
            if (targetEngine.getEntityState(root) == EntityState.Removed) {
                destroyAreaChildren(root);
                stateByRoot.delete(root);
                overlapCountByPair.delete(root);
            }
        }
        for (const [root, s] of stateByRoot) {
            const areas = buildInternalAreasFromSpec(s.areas);
            const children = areaChildrenByRoot.get(root) || [];
            for (let i = 0; i < children.length; i++)
                setChildTransformFromRoot(children[i], root, areas[i]);
        }
    }
    targetEngine.addSystem(system, priority.TriggerSystemPriority);
    const triggersInterface = {
        addTrigger(entity, layerMask = NO_LAYERS, triggeredByMask = NO_LAYERS, areas, onEnterCallback, onExitCallback, debugColor) {
            if (layerMask < 0 || layerMask > ALL_LAYERS || !Number.isInteger(layerMask))
                throw new Error(`Bad layerMask: ${layerMask}. Expected a non-negative integer no greater than ${ALL_LAYERS}`);
            if (triggeredByMask < 0 || triggeredByMask > ALL_LAYERS || !Number.isInteger(triggeredByMask))
                throw new Error(`Bad triggeredByMask: ${triggeredByMask}. Expected a non-negative integer no greater than ${ALL_LAYERS}`);
            stateByRoot.set(entity, {
                active: true,
                layerMask: layerMask,
                triggeredByMask: triggeredByMask,
                areas: areas ? areas : [{ type: 'box' }],
                onEnter: onEnterCallback,
                onExit: onExitCallback,
                debugColor: debugColor ? debugColor : Color3.Red()
            });
            createAreaChildren(entity);
            refreshDebugMeshes(entity);
        },
        removeTrigger(entity) {
            const map = overlapCountByPair.get(entity);
            const s = stateByRoot.get(entity);
            if (map && s && s.onExit) {
                for (const other of map.keys())
                    s.onExit(other);
            }
            overlapCountByPair.delete(entity);
            destroyAreaChildren(entity);
            stateByRoot.delete(entity);
        },
        oneTimeTrigger(entity, layerMask = NO_LAYERS, triggeredByMask = NO_LAYERS, areas, onEnterCallback, debugColor) {
            this.addTrigger(entity, layerMask, triggeredByMask, areas, function (e) {
                triggers.removeTrigger(entity);
                if (onEnterCallback)
                    onEnterCallback(e);
            }, undefined, debugColor);
        },
        enableTrigger(entity, enabled) {
            if (!stateByRoot.has(entity))
                return;
            enableDisableRoot(entity, enabled);
            const s = stateByRoot.get(entity);
            const areas = buildInternalAreasFromSpec(s.areas);
            const children = areaChildrenByRoot.get(entity) || [];
            for (let i = 0; i < children.length; i++) {
                const child = children[i];
                if (enabled) {
                    attachComponentsToAreaChild(child, areas[i], s.triggeredByMask, s.layerMask);
                }
                else {
                    if (areas[i].$case == 'box') {
                        TriggerArea.setBox(child, 0);
                    }
                    else {
                        TriggerArea.setSphere(child, 0);
                    }
                }
            }
        },
        isTriggerEnabled(entity) {
            const s = stateByRoot.get(entity);
            return !!(s && s.active);
        },
        getLayerMask(entity) {
            return stateByRoot.get(entity)?.layerMask ?? NO_LAYERS;
        },
        setLayerMask(entity, mask) {
            if (mask < 0 || mask > ALL_LAYERS || !Number.isInteger(mask))
                throw new Error(`Bad layerMask: ${mask}. Expected a non-negative integer no greater than ${ALL_LAYERS}`);
            const s = stateByRoot.get(entity);
            if (!s)
                return;
            s.layerMask = mask;
            refreshDebugMeshes(entity);
        },
        getTriggeredByMask(entity) {
            return stateByRoot.get(entity)?.triggeredByMask ?? NO_LAYERS;
        },
        setTriggeredByMask(entity, mask) {
            if (mask < 0 || mask > ALL_LAYERS || !Number.isInteger(mask))
                throw new Error(`Bad layerMask: ${mask}. Expected a non-negative integer no greater than ${ALL_LAYERS}`);
            const s = stateByRoot.get(entity);
            if (!s)
                return;
            s.triggeredByMask = mask;
            const areas = buildInternalAreasFromSpec(s.areas);
            const children = areaChildrenByRoot.get(entity) || [];
            const la = utilsTriggeredByToTriggerAreaMask(mask);
            for (let i = 0; i < children.length; i++) {
                if (areas[i].$case == 'box')
                    TriggerArea.setBox(children[i], la);
                else
                    TriggerArea.setSphere(children[i], la);
            }
        },
        getAreas(entity) {
            return stateByRoot.get(entity)?.areas ?? [];
        },
        setAreas(entity, areas) {
            const s = stateByRoot.get(entity);
            if (!s)
                return;
            s.areas = areas;
            destroyAreaChildren(entity);
            createAreaChildren(entity);
            refreshDebugMeshes(entity);
        },
        setOnEnterCallback(entity, callback) {
            const s = stateByRoot.get(entity);
            if (!s)
                return;
            s.onEnter = callback;
        },
        setOnExitCallback(entity, callback) {
            const s = stateByRoot.get(entity);
            if (!s)
                return;
            s.onExit = callback;
        },
        enableDebugDraw(enabled) {
            debugDraw = enabled;
            for (const root of stateByRoot.keys())
                refreshDebugMeshes(root);
        },
        isDebugDrawEnabled() {
            return debugDraw;
        }
    };
    return triggersInterface;
}
export const triggers = createTriggers(engine);
