// @dcl/sdk-commands/dist/commands/start/server/routes.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.sceneUpdateClients = void 0;
exports.wireRouter = wireRouter;
const qrcode_1 = __importDefault(require("qrcode"));
const http_server_1 = require("@well-known-components/http-server");
const ws_1 = require("@well-known-components/http-server/dist/ws");
const ws_2 = require("../data-layer/ws");
const endpoints_1 = require("./endpoints");
const realm_1 = require("./realm");
const utils_1 = require("../utils");
exports.sceneUpdateClients = new Set();
async function wireRouter(components, workspace, dataLayer) {
    const router = new http_server_1.Router();
    if (dataLayer) {
        router.get('/data-layer', async (ctx, next) => {
            if (ctx.request.headers.get('upgrade') === 'websocket') {
                return (0, ws_1.upgradeWebSocketResponse)((ws) => (0, ws_2.handleDataLayerWs)(components, ws, dataLayer));
            }
            return next();
        });
    }
    router.get('/', async (ctx, next) => {
        if (ctx.request.headers.get('upgrade') === 'websocket') {
            return (0, ws_1.upgradeWebSocketResponse)((ws) => initWsConnection(ws, exports.sceneUpdateClients));
        }
        return next();
    });
    const localSceneParcels = [];
    const baseParcel = workspace.projects[0].scene.scene.base;
    for (const project of workspace.projects) {
        for (const parcel of project.scene.scene.parcels) {
            localSceneParcels.push(parcel);
        }
    }
    // Mobile preview QR code endpoint
    router.get('/mobile-preview', async (ctx) => {
        const lanUrl = (0, utils_1.getLanUrl)(ctx.url.port);
        if (!lanUrl) {
            return {
                status: 404,
                body: { ok: false, error: 'No LAN IP address found' }
            };
        }
        const [x, y] = baseParcel.split(',');
        const deepLink = `decentraland://open?preview=${lanUrl}&position=${x},${y}`;
        const qrDataUrl = await qrcode_1.default.toDataURL(deepLink);
        return {
            headers: { 'content-type': 'application/json' },
            body: {
                ok: true,
                data: {
                    url: deepLink,
                    qr: qrDataUrl
                }
            }
        };
    });
    (0, realm_1.setupRealmAndComms)(components, router, localSceneParcels);
    await (0, endpoints_1.setupEcs6Endpoints)(components, router, workspace);
    components.server.setContext(components);
    components.server.use(router.allowedMethods());
    components.server.use(router.middleware());
}
const initWsConnection = (ws, clients) => {
    if (ws.readyState === ws.OPEN) {
        clients.add(ws);
    }
    else {
        ws.on('open', () => clients.add(ws));
    }
    ws.on('close', () => clients.delete(ws));
};
