// @dcl/asset-packs/dist/actions.js
import { Material, AudioStream, MeshCollider, getComponentEntityTree, Tween, } from '@dcl/ecs';
import { Quaternion, Vector3 } from '@dcl/sdk/math';
import { getEntityParent, getPlayerPosition, getWorldPosition, getWorldRotation } from './helpers';
import { ActionType, ProximityLayer, TriggerType, TweenType, TeleportMode, clone, getComponents, initVideoPlayerComponentMaterial, } from './definitions';
import { getDefaultValue, isValidState } from './states';
import { getActionEvents, getTriggerEvents } from './events';
import { startInterval, startTimeout, stopAllIntervals, stopAllTimeouts, stopInterval, stopTimeout, } from './timer';
import { getPayload } from './action-types';
import { getUIBackground, getUIText, getUITransform, mapAlignToScreenAlign, showCaptchaPrompt, } from './ui';
import { AlignMode } from './enums';
import { getExplorerComponents } from './components';
import { initTriggers, damageTargets, healTargets } from './triggers';
import { followMap } from './transform';
import { getEasingFunctionFromInterpolation } from './tweens';
import { REWARDS_SERVER_URL } from './admin-toolkit-ui/constants';
import { callScriptMethod } from '~sdk/script-utils';
import { movePlayerTo, triggerEmote, triggerSceneEmote, openExternalUrl, teleportTo, changeRealm, } from '~system/RestrictedActions';
import { getRealm } from '~system/Runtime';
import { signedFetch } from '~system/SignedFetch';
const initedEntities = new Set();
const uiStacks = new Map();
const lastUiEntityClicked = new Map();
const textEntities = new Map();
let internalInitActions = null;
export function initActions(entity) {
    if (internalInitActions) {
        return internalInitActions(entity);
    }
    throw new Error('Cannot call initActions while actionsSystem has not been created');
}
export function createActionsSystem(engine, pointerEventsSystem, sdkHelpers, playersHelper) {
    const { Animator, Transform, AudioSource, AvatarAttach, VisibilityComponent, GltfContainer, UiTransform, UiText, UiBackground, UiInput, UiInputResult, Tween: TweenComponent, TweenSequence, VideoPlayer, LightSource, MainCamera, VirtualCamera, TextShape, InputModifier, SkyboxTime, } = getExplorerComponents(engine);
    const { Actions, States, Counter, Triggers, Rewards } = getComponents(engine);
    internalInitActions = initActions;
    return function actionsSystem(_dt) {
        const entitiesWithActions = engine.getEntitiesWith(Actions);
        for (const [entity] of entitiesWithActions) {
            initActions(entity);
        }
    };
    function initActions(entity) {
        if (!Actions.has(entity) || initedEntities.has(entity)) {
            return;
        }
        const actions = Actions.get(entity);
        const types = actions.value.reduce((types, action) => types.add(action.type), new Set());
        for (const type of types) {
            switch (type) {
                case ActionType.PLAY_ANIMATION: {
                    initPlayAnimation(entity);
                    break;
                }
                default:
                    break;
            }
        }
        const actionEvents = getActionEvents(entity);
        for (const action of actions.value) {
            actionEvents.on(action.name, () => {
                switch (action.type) {
                    case ActionType.PLAY_ANIMATION: {
                        handlePlayAnimation(entity, getPayload(action));
                        break;
                    }
                    case ActionType.STOP_ANIMATION: {
                        handleStopAnimation(entity, getPayload(action));
                        break;
                    }
                    case ActionType.SET_STATE: {
                        handleSetState(entity, getPayload(action));
                        break;
                    }
                    case ActionType.START_TWEEN: {
                        handleStartTween(entity, getPayload(action));
                        break;
                    }
                    case ActionType.STOP_TWEEN: {
                        handleStopTween(entity);
                        break;
                    }
                    case ActionType.SLIDE_TEXTURE: {
                        handleSlideTexture(entity, getPayload(action));
                        break;
                    }
                    case ActionType.SET_COUNTER: {
                        handleSetCounter(entity, getPayload(action));
                        break;
                    }
                    case ActionType.INCREMENT_COUNTER: {
                        handleIncrementCounter(entity, getPayload(action));
                        break;
                    }
                    case ActionType.DECREASE_COUNTER: {
                        handleDecreaseCounter(entity, getPayload(action));
                        break;
                    }
                    case ActionType.PLAY_SOUND: {
                        handlePlaySound(entity, getPayload(action));
                        break;
                    }
                    case ActionType.STOP_SOUND: {
                        handleStopSound(entity, getPayload(action));
                        break;
                    }
                    case ActionType.SET_VISIBILITY: {
                        handleSetVisibility(entity, getPayload(action));
                        break;
                    }
                    case ActionType.ATTACH_TO_PLAYER: {
                        handleAttachToPlayer(entity, getPayload(action));
                        break;
                    }
                    case ActionType.DETACH_FROM_PLAYER: {
                        handleDetachFromPlayer(entity, getPayload(action));
                        break;
                    }
                    case ActionType.PLAY_VIDEO_STREAM: {
                        handlePlayVideo(entity, getPayload(action));
                        break;
                    }
                    case ActionType.STOP_VIDEO_STREAM: {
                        handleStopVideo(entity, getPayload(action));
                        break;
                    }
                    case ActionType.PLAY_AUDIO_STREAM: {
                        handlePlayAudioStream(entity, getPayload(action));
                        break;
                    }
                    case ActionType.STOP_AUDIO_STREAM: {
                        handleStopAudioStream(entity, getPayload(action));
                        break;
                    }
                    case ActionType.TELEPORT_PLAYER: {
                        handleTeleportPlayer(entity, getPayload(action));
                        break;
                    }
                    case ActionType.MOVE_PLAYER: {
                        handleMovePlayer(entity, getPayload(action));
                        break;
                    }
                    case ActionType.PLAY_DEFAULT_EMOTE: {
                        handlePlayDefaultEmote(entity, getPayload(action));
                        break;
                    }
                    case ActionType.PLAY_CUSTOM_EMOTE: {
                        handlePlayCustomEmote(entity, getPayload(action));
                        break;
                    }
                    case ActionType.OPEN_LINK: {
                        handleOpenLink(entity, getPayload(action));
                        break;
                    }
                    case ActionType.SHOW_TEXT: {
                        handleShowText(entity, getPayload(action));
                        break;
                    }
                    case ActionType.HIDE_TEXT: {
                        handleHideText(entity, getPayload(action));
                        break;
                    }
                    case ActionType.START_DELAY: {
                        handleStartDelay(entity, getPayload(action));
                        break;
                    }
                    case ActionType.STOP_DELAY: {
                        handleStopDelay(entity, getPayload(action));
                        break;
                    }
                    case ActionType.START_LOOP: {
                        handleStartLoop(entity, getPayload(action));
                        break;
                    }
                    case ActionType.STOP_LOOP: {
                        handleStopLoop(entity, getPayload(action));
                        break;
                    }
                    case ActionType.CLONE_ENTITY: {
                        handleCloneEntity(entity, getPayload(action));
                        break;
                    }
                    case ActionType.REMOVE_ENTITY: {
                        handleRemoveEntity(entity, getPayload(action));
                        break;
                    }
                    case ActionType.SHOW_IMAGE: {
                        handleShowImage(entity, getPayload(action));
                        break;
                    }
                    case ActionType.HIDE_IMAGE: {
                        handleHideImage(entity, getPayload(action));
                        break;
                    }
                    case ActionType.DAMAGE: {
                        handleDamage(entity, getPayload(action));
                        break;
                    }
                    case ActionType.MOVE_PLAYER_HERE: {
                        handleMovePlayerHere(entity, getPayload(action));
                        break;
                    }
                    case ActionType.PLAYER_FACE_ITEM: {
                        handlePlayerFaceItem(entity, getPayload(action));
                        break;
                    }
                    case ActionType.FOLLOW_PLAYER: {
                        handleFollowPlayer(entity, getPayload(action));
                        break;
                    }
                    case ActionType.STOP_FOLLOWING_PLAYER: {
                        handleStopFollowingPlayer(entity, getPayload(action));
                        break;
                    }
                    case ActionType.PLACE_ON_PLAYER: {
                        handlePlaceOnPlayer(entity, getPayload(action));
                        break;
                    }
                    case ActionType.ROTATE_AS_PLAYER: {
                        handleRotateAsPlayer(entity, getPayload(action));
                        break;
                    }
                    case ActionType.PLACE_ON_CAMERA: {
                        handlePlaceOnCamera(entity, getPayload(action));
                        break;
                    }
                    case ActionType.ROTATE_AS_CAMERA: {
                        handleRotateAsCamera(entity, getPayload(action));
                        break;
                    }
                    case ActionType.SET_POSITION: {
                        handleSetPosition(entity, getPayload(action));
                        break;
                    }
                    case ActionType.SET_ROTATION: {
                        handleSetRotation(entity, getPayload(action));
                        break;
                    }
                    case ActionType.SET_SCALE: {
                        handleSetScale(entity, getPayload(action));
                        break;
                    }
                    case ActionType.RANDOM: {
                        handleRandom(entity, getPayload(action));
                        break;
                    }
                    case ActionType.BATCH: {
                        handleBatch(entity, getPayload(action));
                        break;
                    }
                    case ActionType.HEAL_PLAYER: {
                        handleHealPlayer(entity, getPayload(action));
                        break;
                    }
                    case ActionType.CLAIM_AIRDROP: {
                        handleClaimAirdrop(entity, getPayload(action));
                        break;
                    }
                    case ActionType.LIGHTS_ON: {
                        handleLightsOn(entity);
                        break;
                    }
                    case ActionType.LIGHTS_OFF: {
                        handleLightsOff(entity);
                        break;
                    }
                    case ActionType.LIGHTS_MODIFY: {
                        handleLightsModify(entity, getPayload(action));
                        break;
                    }
                    case ActionType.CHANGE_CAMERA: {
                        handleChangeCamera(entity, getPayload(action));
                        break;
                    }
                    case ActionType.CHANGE_TEXT: {
                        handleChangeText(entity, getPayload(action));
                        break;
                    }
                    case ActionType.FREEZE_PLAYER: {
                        handleFreezePlayer();
                        break;
                    }
                    case ActionType.UNFREEZE_PLAYER: {
                        handleUnfreezePlayer();
                        break;
                    }
                    case ActionType.CHANGE_COLLISIONS: {
                        handleChangeCollisions(entity, getPayload(action));
                        break;
                    }
                    case ActionType.CHANGE_SKYBOX: {
                        handleChangeSkybox(getPayload(action));
                        break;
                    }
                    case ActionType.RESET_SKYBOX: {
                        handleResetSkybox();
                        break;
                    }
                    case ActionType.CALL_SCRIPT_METHOD: {
                        handleCallScriptMethod(entity, getPayload(action));
                        break;
                    }
                    case ActionType.LOG_TO_CONSOLE: {
                        handleLogToConsole(getPayload(action));
                        break;
                    }
                    case ActionType.DELETE: {
                        handleDelete(entity);
                        break;
                    }
                    default:
                        break;
                }
            });
        }
        initedEntities.add(entity);
    }
    function handleLightsOn(entity) {
        const light = LightSource.getMutableOrNull(entity);
        if (light) {
            light.active = true;
        }
    }
    function handleLightsOff(entity) {
        const light = LightSource.getMutableOrNull(entity);
        if (light) {
            light.active = false;
        }
    }
    function handleLightsModify(entity, payload) {
        const light = LightSource.getMutableOrNull(entity);
        if (!light)
            return;
        if (payload.active !== undefined)
            light.active = payload.active;
        if (payload.color)
            light.color = payload.color;
        if (payload.intensity !== undefined)
            light.intensity = payload.intensity;
    }
    function handleChangeCamera(entity, payload) {
        const target = payload.virtualCameraEntity;
        let cameraEntity = target;
        if (cameraEntity === undefined) {
            if (VirtualCamera.has(entity)) {
                cameraEntity = entity;
            }
            else {
                if (MainCamera.has(engine.CameraEntity)) {
                    MainCamera.deleteFrom(engine.CameraEntity);
                }
                return;
            }
        }
        if (cameraEntity === engine.RootEntity) {
            if (MainCamera.has(engine.CameraEntity)) {
                MainCamera.deleteFrom(engine.CameraEntity);
            }
            return;
        }
        if (!VirtualCamera.has(cameraEntity)) {
            return;
        }
        MainCamera.createOrReplace(engine.CameraEntity, {
            virtualCameraEntity: cameraEntity,
        });
    }
    function handleChangeText(entity, payload) {
        const text = TextShape.getMutableOrNull(entity);
        if (!text)
            return;
        text.text = payload.text;
        if (payload.fontSize !== undefined)
            text.fontSize = payload.fontSize;
        if (payload.color)
            text.textColor = payload.color;
    }
    function handleFreezePlayer() {
        InputModifier.createOrReplace(engine.PlayerEntity, {
            mode: InputModifier.Mode.Standard({
                disableAll: true,
            }),
        });
    }
    function handleUnfreezePlayer() {
        InputModifier.createOrReplace(engine.PlayerEntity, {
            mode: InputModifier.Mode.Standard({
                disableAll: false,
            }),
        });
    }
    function handleChangeCollisions(entity, payload) {
        const { visibleCollisions, invisibleCollisions } = payload;
        const gltf = GltfContainer.getMutableOrNull(entity);
        const meshCollider = MeshCollider.getMutableOrNull(entity);
        if (gltf) {
            if (visibleCollisions !== undefined) {
                gltf.visibleMeshesCollisionMask = visibleCollisions;
            }
            if (invisibleCollisions !== undefined) {
                gltf.invisibleMeshesCollisionMask = invisibleCollisions;
            }
        }
        if (meshCollider) {
            if (visibleCollisions !== undefined) {
                meshCollider.collisionMask = visibleCollisions;
            }
            else if (invisibleCollisions !== undefined) {
                meshCollider.collisionMask = invisibleCollisions;
            }
        }
    }
    function handleChangeSkybox(payload) {
        const { time, direction } = payload;
        if (typeof time !== 'number' || isNaN(time) || time < 0 || time > 86400) {
            console.error('Invalid skybox time value:', time);
            return;
        }
        SkyboxTime.createOrReplace(engine.RootEntity, {
            fixedTime: time,
            transitionMode: direction,
        });
    }
    function handleResetSkybox() {
        if (SkyboxTime.has(engine.RootEntity)) {
            SkyboxTime.deleteFrom(engine.RootEntity);
        }
    }
    function initPlayAnimation(entity) {
        if (!Animator.has(entity)) {
            Animator.create(entity, {
                states: [],
            });
            Animator.stopAllAnimations(entity);
        }
    }
    function normalizeAnimationWeight(rawWeight) {
        if (rawWeight < 0 || rawWeight > 1)
            return 1;
        return rawWeight;
    }
    function handlePlayAnimation(entity, payload) {
        const { animation, loop, shouldReset } = payload;
        const animator = Animator.getMutable(entity);
        if (!animator.states.some(($) => $.clip === animation)) {
            animator.states = [
                ...animator.states,
                {
                    clip: animation,
                },
            ];
        }
        Animator.stopAllAnimations(entity);
        try {
            const clip = Animator.getClip(entity, animation);
            clip.playing = true;
            clip.loop = loop ?? false;
            clip.weight = normalizeAnimationWeight(clip.weight ?? 1);
            if (shouldReset !== undefined) {
                clip.shouldReset = shouldReset;
            }
        }
        catch (e) {
            console.error('Error playing animation', e);
        }
    }
    function handleStopAnimation(entity, _payload) {
        if (Animator.has(entity)) {
            Animator.stopAllAnimations(entity);
        }
    }
    function handleSetState(entity, payload) {
        const states = States.getMutableOrNull(entity);
        if (states) {
            const defaultValue = getDefaultValue(states);
            let nextState = payload.state;
            nextState = isValidState(states, nextState) ? nextState : defaultValue;
            const previousValue = states.currentValue ?? defaultValue ?? undefined;
            states.previousValue = previousValue;
            states.currentValue = nextState;
            const triggerEvents = getTriggerEvents(entity);
            triggerEvents.emit(TriggerType.ON_STATE_CHANGE);
        }
    }
    function handleStartTween(entity, payload) {
        if (payload) {
            const initialTween = TweenComponent.getMutableOrNull(entity);
            let tween;
            switch (payload.type) {
                case TweenType.MOVE_ITEM: {
                    tween = handleMoveItem(entity, payload);
                    break;
                }
                case TweenType.ROTATE_ITEM: {
                    tween = handleRotateItem(entity, payload);
                    break;
                }
                case TweenType.SCALE_ITEM: {
                    tween = handleScaleItem(entity, payload);
                    break;
                }
                case TweenType.KEEP_ROTATING_ITEM: {
                    tween = handleKeepRotatingItem(entity, payload);
                    break;
                }
                default: {
                    throw new Error(`Unknown tween type: ${payload.type}`);
                }
            }
            revertTween(entity, initialTween, tween);
        }
    }
    function revertTween(entity, initialTween, tween) {
        const tweenSequence = TweenSequence.getMutableOrNull(entity);
        let _revertTween = {
            ...tween,
        };
        if (!initialTween || !tweenSequence || !tweenSequence.loop)
            return;
        switch (initialTween.mode?.$case) {
            case 'move': {
                _revertTween = {
                    ..._revertTween,
                    mode: Tween.Mode.Move({
                        start: tween.mode.move.end,
                        end: initialTween.mode.move.start,
                    }),
                };
                break;
            }
            case 'rotate': {
                _revertTween = {
                    ..._revertTween,
                    mode: Tween.Mode.Rotate({
                        start: tween.mode.rotate.end,
                        end: initialTween.mode.rotate.start,
                    }),
                };
                break;
            }
            case 'scale': {
                _revertTween = {
                    ..._revertTween,
                    mode: Tween.Mode.Scale({
                        start: tween.mode.scale.end,
                        end: initialTween.mode.scale.start,
                    }),
                };
                break;
            }
            default: {
                throw new Error(`Unknown tween mode: ${initialTween.mode}`);
            }
        }
        initialTween.playing = true;
        tweenSequence.sequence = [_revertTween, initialTween];
    }
    function handleMoveItem(entity, tween) {
        const transform = Transform.get(entity);
        const { duration, interpolationType, relative } = tween;
        const end = Vector3.create(tween.end.x, tween.end.y, tween.end.z);
        const endPosition = relative ? Vector3.add(transform.position, end) : end;
        return TweenComponent.createOrReplace(entity, {
            mode: Tween.Mode.Move({
                start: transform.position,
                end: endPosition,
            }),
            duration: duration * 1000,
            easingFunction: getEasingFunctionFromInterpolation(interpolationType),
        });
    }
    function handleRotateItem(entity, tween) {
        const transform = Transform.get(entity);
        const { duration, interpolationType, relative } = tween;
        const end = Quaternion.fromEulerDegrees(tween.end.x, tween.end.y, tween.end.z);
        const endRotation = relative ? Quaternion.multiply(transform.rotation, end) : end;
        return TweenComponent.createOrReplace(entity, {
            mode: Tween.Mode.Rotate({
                start: transform.rotation,
                end: endRotation,
            }),
            duration: duration * 1000,
            easingFunction: getEasingFunctionFromInterpolation(interpolationType),
        });
    }
    function handleScaleItem(entity, tween) {
        const transform = Transform.get(entity);
        const { duration, interpolationType, relative } = tween;
        const end = Vector3.create(tween.end.x, tween.end.y, tween.end.z);
        const endScale = relative ? Vector3.add(transform.scale, end) : end;
        return TweenComponent.createOrReplace(entity, {
            mode: Tween.Mode.Scale({
                start: transform.scale,
                end: endScale,
            }),
            duration: duration * 1000,
            easingFunction: getEasingFunctionFromInterpolation(interpolationType),
        });
    }
    function handleKeepRotatingItem(entity, tween) {
        const { duration, interpolationType } = tween;
        const dir = tween.direction ?? { x: 0, y: 1, z: 0 };
        const speed = tween.speed ?? 1;
        return TweenComponent.createOrReplace(entity, {
            mode: Tween.Mode.RotateContinuous({
                direction: Quaternion.fromEulerDegrees(dir.x, dir.y, dir.z),
                speed,
            }),
            duration: (duration ?? 0) * 1000,
            easingFunction: getEasingFunctionFromInterpolation(interpolationType),
        });
    }
    function handleStopTween(entity) {
        if (TweenComponent.has(entity)) {
            TweenComponent.deleteFrom(entity);
        }
    }
    function handleSlideTexture(entity, payload) {
        const dir = payload.direction;
        const speed = payload.speed;
        const movementType = payload.movementType;
        const duration = payload.duration ?? 0;
        return TweenComponent.createOrReplace(entity, {
            mode: Tween.Mode.TextureMoveContinuous({
                direction: { x: dir.x, y: dir.y },
                speed,
                movementType,
            }),
            duration: duration * 1000,
            easingFunction: 0,
        });
    }
    function handleSetCounter(entity, payload) {
        const counter = Counter.getMutableOrNull(entity);
        if (counter) {
            counter.value = payload.counter;
            const triggerEvents = getTriggerEvents(entity);
            triggerEvents.emit(TriggerType.ON_COUNTER_CHANGE);
        }
    }
    function handleIncrementCounter(entity, payload) {
        const counter = Counter.getMutableOrNull(entity);
        const amount = payload.amount ?? 1;
        if (counter) {
            counter.value += amount;
            const triggerEvents = getTriggerEvents(entity);
            triggerEvents.emit(TriggerType.ON_COUNTER_CHANGE);
        }
    }
    function handleDecreaseCounter(entity, payload) {
        const counter = Counter.getMutableOrNull(entity);
        const amount = payload.amount ?? 1;
        if (counter) {
            counter.value -= amount;
            const triggerEvents = getTriggerEvents(entity);
            triggerEvents.emit(TriggerType.ON_COUNTER_CHANGE);
        }
    }
    function handlePlaySound(entity, payload) {
        const { src, loop, volume, global } = payload;
        if (AudioSource.has(entity)) {
            AudioSource.playSound(entity, src);
            const audioSource = AudioSource.getMutable(entity);
            audioSource.loop = loop;
            audioSource.volume = volume ?? 1;
            audioSource.global = global ?? false;
        }
        else {
            AudioSource.create(entity, {
                audioClipUrl: src,
                loop,
                playing: true,
                volume: volume ?? 1,
                global,
            });
        }
    }
    function handleStopSound(entity, _payload) {
        const audioSource = AudioSource.getMutableOrNull(entity);
        if (audioSource) {
            audioSource.playing = false;
        }
    }
    function handleSetVisibility(entity, payload) {
        const { visible, collider } = payload;
        VisibilityComponent.createOrReplace(entity, { visible });
        const gltf = GltfContainer.getMutableOrNull(entity);
        const meshCollider = MeshCollider.getMutableOrNull(entity);
        if (collider !== undefined) {
            if (gltf) {
                gltf.invisibleMeshesCollisionMask = collider;
                if (collider === 0) {
                    gltf.visibleMeshesCollisionMask = 0;
                }
            }
            else if (meshCollider) {
                meshCollider.collisionMask = collider;
            }
        }
    }
    function handleAttachToPlayer(entity, payload) {
        const { anchorPointId } = payload;
        const user = playersHelper?.getPlayer()?.userId;
        if (user) {
            AvatarAttach.createOrReplace(entity, { anchorPointId: anchorPointId, avatarId: user });
        }
    }
    function handleDetachFromPlayer(entity, _payload) {
        if (AvatarAttach.has(entity)) {
            AvatarAttach.deleteFrom(entity);
        }
    }
    function handleTeleportPlayer(_entity, payload) {
        const { mode, x, y, realm } = payload;
        if (mode === TeleportMode.TO_WORLD) {
            if (realm) {
                void changeRealm({ realm });
            }
        }
        else {
            if (x !== undefined && y !== undefined) {
                void teleportTo({
                    worldCoordinates: { x, y },
                });
            }
        }
    }
    function handleMovePlayer(_entity, payload) {
        const options = {
            newRelativePosition: payload.position,
            cameraTarget: payload.cameraTarget,
        };
        void movePlayerTo(options);
    }
    function handlePlayDefaultEmote(_entity, payload) {
        const { emote } = payload;
        void triggerEmote({ predefinedEmote: emote });
    }
    function handlePlayCustomEmote(_entity, payload) {
        const { src, loop } = payload;
        void triggerSceneEmote({ src, loop });
    }
    function handleOpenLink(_entity, payload) {
        const { url } = payload;
        void openExternalUrl({ url });
    }
    function handlePlayVideo(entity, payload) {
        const videoSource = VideoPlayer.getMutableOrNull(entity);
        if (!videoSource && payload.src) {
            VideoPlayer.createOrReplace(entity, {
                src: payload.src,
                volume: payload.volume ?? 1,
                loop: payload.loop ?? false,
                playing: true,
            });
            initVideoPlayerComponentMaterial(entity, { Material }, Material.getOrNull(entity));
        }
        if (videoSource) {
            if (videoSource?.src !== payload.src) {
                videoSource.src = payload.src ?? '';
            }
            videoSource.volume = payload.volume ?? videoSource.volume;
            videoSource.loop = payload.loop ?? videoSource.loop;
            videoSource.playing = true;
        }
    }
    function handleStopVideo(entity, _payload) {
        const videoSource = VideoPlayer.getMutableOrNull(entity);
        if (videoSource) {
            videoSource.playing = false;
        }
    }
    function handlePlayAudioStream(entity, payload) {
        const { url, volume } = payload;
        AudioStream.createOrReplace(entity, {
            url,
            playing: true,
            volume: volume ?? 1,
        });
    }
    function handleStopAudioStream(entity, _payload) {
        const audioSource = AudioStream.getMutableOrNull(entity);
        if (audioSource) {
            audioSource.playing = false;
        }
    }
    function handleShowText(entity, payload) {
        const { text, hideAfterSeconds, font, fontSize, textAlign } = payload;
        const textEntity = engine.addEntity();
        textEntities.set(entity, textEntity);
        const uiTransformComponent = getUITransform(UiTransform, textEntity);
        if (uiTransformComponent) {
            uiTransformComponent.parent = entity;
            uiTransformComponent.pointerFilter = 0;
            UiText.createOrReplace(textEntity, {
                value: text,
                font: font,
                fontSize,
                textAlign: textAlign,
            });
            startTimeout(entity, ActionType.HIDE_TEXT, hideAfterSeconds, () => handleHideText(entity, {}));
        }
    }
    function handleHideText(entity, _payload) {
        const textEntity = textEntities.get(entity);
        if (textEntity) {
            UiText.deleteFrom(textEntity);
            UiTransform.deleteFrom(textEntity);
            stopTimeout(entity, ActionType.HIDE_TEXT);
            textEntities.delete(entity);
        }
    }
    function findActionByName(entity, name) {
        const actions = Actions.getOrNull(entity);
        return actions?.value.find(($) => $.name === name);
    }
    function handleStartDelay(entity, payload) {
        const { actions, timeout } = payload;
        for (const actionName of actions) {
            const action = findActionByName(entity, actionName);
            if (action) {
                startTimeout(entity, actionName, timeout, () => {
                    const actionEvents = getActionEvents(entity);
                    actionEvents.emit(action.name, getPayload(action));
                });
            }
        }
    }
    function handleStopDelay(entity, payload) {
        const { action } = payload;
        stopTimeout(entity, action);
    }
    function handleStartLoop(entity, payload) {
        const { actions, interval } = payload;
        for (const actionName of actions) {
            const action = findActionByName(entity, actionName);
            if (action) {
                startInterval(entity, actionName, interval, () => {
                    const actionEvents = getActionEvents(entity);
                    actionEvents.emit(action.name, getPayload(action));
                });
            }
        }
    }
    function handleStopLoop(entity, payload) {
        const { action } = payload;
        stopInterval(entity, action);
    }
    function handleCloneEntity(entity, payload) {
        const { position } = payload;
        const { cloned, entities } = clone(entity, engine, Transform, Triggers, sdkHelpers);
        for (const cloned of entities.values()) {
            initActions(cloned);
            initTriggers(cloned);
            const triggerEvents = getTriggerEvents(cloned);
            triggerEvents.emit(TriggerType.ON_CLONE);
        }
        const transform = Transform.getOrCreateMutable(cloned);
        transform.position = position;
    }
    function handleRemoveEntity(entity, _payload) {
        stopAllTimeouts(entity);
        stopAllIntervals(entity);
        const tree = getComponentEntityTree(engine, entity, Transform);
        for (const entityToRemove of tree) {
            engine.removeEntity(entityToRemove);
        }
    }
    function getUiStack(align) {
        const key = `${align.alignItems},${align.justifyContent}`;
        if (!uiStacks.has(key)) {
            uiStacks.set(key, engine.addEntity());
        }
        return uiStacks.get(key);
    }
    function handleShowImage(entity, payload) {
        const { src, text, hideAfterSeconds, fontSize, align, height, width } = payload;
        getUITransform(UiTransform, engine.RootEntity);
        const screenAlign = mapAlignToScreenAlign(align ?? AlignMode.TAM_MIDDLE_CENTER);
        const uiStack = getUiStack(screenAlign);
        const uiStackTransformComponent = getUITransform(UiTransform, uiStack);
        uiStackTransformComponent.alignItems = screenAlign.alignItems;
        uiStackTransformComponent.justifyContent = screenAlign.justifyContent;
        uiStackTransformComponent.positionType = 1;
        const imageEntity = engine.addEntity();
        const imageTransformComponent = getUITransform(UiTransform, imageEntity, width, height, 1);
        imageTransformComponent.parent = uiStack;
        imageTransformComponent.pointerFilter = 1;
        getUIBackground(UiBackground, imageEntity, src);
        if (text) {
            getUIText(UiText, imageEntity, text, fontSize, width);
        }
        pointerEventsSystem.onPointerDown({
            entity: imageEntity,
            opts: {
                button: 0,
                hoverText: 'Click',
            },
        }, () => {
            lastUiEntityClicked.set(entity, imageEntity);
            const triggerEvents = getTriggerEvents(entity);
            triggerEvents.emit(TriggerType.ON_CLICK_IMAGE);
        });
        if (hideAfterSeconds) {
            startTimeout(entity, ActionType.HIDE_IMAGE, hideAfterSeconds, () => handleHideImage(entity, { imageEntity: imageEntity }));
        }
    }
    function handleHideImage(entity, payload) {
        const { imageEntity } = payload;
        if (imageEntity) {
            engine.removeEntity(imageEntity);
        }
        else {
            const clickedImage = lastUiEntityClicked.get(entity);
            if (clickedImage) {
                engine.removeEntity(clickedImage);
                lastUiEntityClicked.delete(entity);
            }
        }
    }
    function handleDamage(entity, payload) {
        const { radius, layer, hits } = payload;
        const entityPosition = AvatarAttach.has(entity)
            ? getPlayerPosition()
            : getWorldPosition(entity);
        const getRoot = (entity) => {
            const parent = getEntityParent(entity);
            return !parent ? entity : getRoot(parent);
        };
        for (const target of damageTargets) {
            const targetPosition = getWorldPosition(target);
            const distance = Vector3.distance(entityPosition, targetPosition);
            const entityTree = Array.from(getComponentEntityTree(engine, entity, Transform));
            const isPartOfEntityTree = entityTree.some(($) => $ === target);
            if (isPartOfEntityTree) {
                continue;
            }
            if (layer) {
                if (layer === ProximityLayer.PLAYER) {
                    const root = getRoot(target);
                    if (root !== engine.PlayerEntity && root !== engine.CameraEntity) {
                        continue;
                    }
                }
                else if (layer === ProximityLayer.NON_PLAYER) {
                    const root = getRoot(target);
                    if (root === engine.PlayerEntity || root === engine.CameraEntity) {
                        continue;
                    }
                }
            }
            if (distance <= radius) {
                const total = hits === undefined ? 1 : Math.max(hits, 1);
                for (let i = 0; i < total; i++) {
                    const triggerEvents = getTriggerEvents(target);
                    triggerEvents.emit(TriggerType.ON_DAMAGE);
                }
            }
        }
    }
    function handleMovePlayerHere(entity, _payload) {
        const here = getWorldPosition(entity);
        const rotation = getWorldRotation(entity);
        const forward = Vector3.Forward();
        const rotatedDirection = Vector3.rotate(forward, rotation);
        const moveDelta = Vector3.scale(rotatedDirection, 1);
        void movePlayerTo({
            newRelativePosition: here,
            avatarTarget: Vector3.add(here, moveDelta),
        });
        const triggerEvents = getTriggerEvents(entity);
        triggerEvents.emit(TriggerType.ON_PLAYER_SPAWN);
    }
    function handlePlayerFaceItem(entity, _payload) {
        const itemPosition = getWorldPosition(entity);
        const currentPlayerPosition = getPlayerPosition();
        void movePlayerTo({
            newRelativePosition: currentPlayerPosition,
            avatarTarget: itemPosition,
        });
    }
    function handlePlaceOnPlayer(entity, _payload) {
        const transform = Transform.getMutableOrNull(entity);
        const player = Transform.getOrNull(engine.PlayerEntity);
        if (transform && player) {
            transform.position = player.position;
        }
    }
    function handleRotateAsPlayer(entity, _payload) {
        const transform = Transform.getMutableOrNull(entity);
        const player = Transform.getOrNull(engine.PlayerEntity);
        if (transform && player) {
            transform.rotation = player.rotation;
        }
    }
    function handlePlaceOnCamera(entity, _payload) {
        const transform = Transform.getMutableOrNull(entity);
        const camera = Transform.getOrNull(engine.CameraEntity);
        if (transform && camera) {
            transform.position = camera.position;
        }
    }
    function handleRotateAsCamera(entity, _payload) {
        const transform = Transform.getMutableOrNull(entity);
        const camera = Transform.getOrNull(engine.CameraEntity);
        if (transform && camera) {
            transform.rotation = camera.rotation;
        }
    }
    function handleSetPosition(entity, payload) {
        const transform = Transform.getMutableOrNull(entity);
        if (transform) {
            if (payload.relative) {
                transform.position = Vector3.add(transform.position, Vector3.create(payload.x, payload.y, payload.z));
            }
            else {
                transform.position = Vector3.create(payload.x, payload.y, payload.z);
            }
        }
    }
    function handleSetRotation(entity, payload) {
        const transform = Transform.getMutableOrNull(entity);
        if (transform) {
            if (payload.relative) {
                transform.rotation = Quaternion.multiply(transform.rotation, Quaternion.fromEulerDegrees(payload.x, payload.y, payload.z));
            }
            else {
                transform.rotation = Quaternion.fromEulerDegrees(payload.x, payload.y, payload.z);
            }
        }
    }
    function handleSetScale(entity, payload) {
        const transform = Transform.getMutableOrNull(entity);
        if (transform) {
            if (payload.relative) {
                transform.scale = Vector3.add(transform.scale, Vector3.create(payload.x, payload.y, payload.z));
            }
            else {
                transform.scale = Vector3.create(payload.x, payload.y, payload.z);
            }
        }
    }
    function handleFollowPlayer(entity, payload) {
        const { speed, x, y, z, minDistance } = payload;
        followMap.set(entity, {
            target: engine.PlayerEntity,
            speed,
            minDistance,
            axes: { x, y, z },
        });
    }
    function handleStopFollowingPlayer(entity, _payload) {
        followMap.delete(entity);
    }
    function handleRandom(entity, payload) {
        const { actions } = payload;
        const actionEvents = getActionEvents(entity);
        const actionName = actions[Math.floor(Math.random() * actions.length)];
        const action = findActionByName(entity, actionName);
        if (action) {
            actionEvents.emit(action.name, getPayload(action));
        }
    }
    function handleBatch(entity, payload) {
        const { actions } = payload;
        const actionEvents = getActionEvents(entity);
        for (const actionName of actions) {
            const action = findActionByName(entity, actionName);
            if (action) {
                actionEvents.emit(action.name, getPayload(action));
            }
        }
    }
    function handleHealPlayer(entity, payload) {
        const { multiplier } = payload;
        const getRoot = (entity) => {
            const parent = getEntityParent(entity);
            return !parent ? entity : getRoot(parent);
        };
        for (const target of healTargets) {
            const root = getRoot(target);
            if (root === engine.PlayerEntity) {
                const triggerEvents = getTriggerEvents(target);
                const total = Math.max(multiplier ?? 1, 1);
                for (let i = 0; i < total; i++) {
                    triggerEvents.emit(TriggerType.ON_HEAL_PLAYER);
                }
            }
        }
    }
    async function request(url, init) {
        try {
            const response = await signedFetch({
                url: url,
                ...(init ? { init } : {}),
            });
            if (!response || !response.body) {
                return null;
            }
            const json = await JSON.parse(response.body);
            if (!json.ok) {
                return null;
            }
            return json.data;
        }
        catch (error) {
            return null;
        }
    }
    async function fetchCampaignsByDispenserKey(dispenserKey) {
        const url = `${REWARDS_SERVER_URL}/api/campaigns/keys?campaign_key=${encodeURIComponent(dispenserKey)}`;
        const response = await request(url);
        return response;
    }
    async function fetchCaptcha() {
        const response = await request(`${REWARDS_SERVER_URL}/api/captcha`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
        });
        return response;
    }
    async function requestToken(dispenserKey, captcha) {
        const url = `${REWARDS_SERVER_URL}/api/rewards`;
        const realm = await getRealm({});
        const player = playersHelper?.getPlayer();
        const response = await request(url, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify({
                campaign_key: dispenserKey,
                beneficiary: !player?.isGuest ? player?.userId : '',
                catalyst: realm.realmInfo ? realm.realmInfo.baseUrl : '',
                ...(captcha ? { captcha_id: captcha.id, captcha_value: captcha.value } : {}),
            }),
        });
        return response;
    }
    function handleClaimAirdrop(entity, _payload) {
        const rewards = Rewards.getOrNull(entity);
        if (!rewards) {
            return;
        }
        const { testMode, campaignId, dispenserKey } = rewards;
        if (testMode) {
            return;
        }
        fetchCampaignsByDispenserKey(dispenserKey)
            .then(campaigns => {
            const campaign = campaigns.find((c) => c.campaign_id === campaignId);
            if (campaign && campaign.enabled) {
                if (campaign.requires_captcha) {
                    fetchCaptcha()
                        .then(captcha => {
                        if (captcha) {
                            _showCaptchaPrompt(entity, {
                                campaignId,
                                dispenserKey,
                                captcha,
                            });
                        }
                    })
                        .catch(_error => {
                    });
                }
                else {
                    requestToken(dispenserKey);
                }
            }
        })
            .catch(_error => {
        });
    }
    function _showCaptchaPrompt(_entity, data) {
        showCaptchaPrompt(engine, UiTransform, UiBackground, UiText, UiInput, UiInputResult, pointerEventsSystem, data, inputText => {
            requestToken(data.dispenserKey, {
                id: data.captcha.id,
                value: inputText,
            }).then((token) => {
                if (token) {
                }
            });
        });
    }
    function handleCallScriptMethod(entity, payload) {
        const { scriptPath, methodName, params } = payload;
        try {
            const paramValues = Object.values(params || {});
            callScriptMethod(entity, scriptPath, methodName, ...paramValues);
        }
        catch (error) {
            console.error(`[Script Action Error] Failed to call ${methodName} on ${scriptPath}:`, error);
        }
    }
    function handleLogToConsole(payload) {
        console.log(payload.message);
    }
    function handleDelete(entity) {
        stopAllTimeouts(entity);
        stopAllIntervals(entity);
        engine.removeEntityWithChildren(entity);
    }
}
