// @dcl/asset-packs/dist/components.js
import { Animator as defineAnimator, Transform as defineTransform, AudioSource as defineAudioSource, AvatarAttach as defineAvatarAttach, VisibilityComponent as defineVisibilityComponent, GltfContainer as defineGltfContainer, UiTransform as defineUiTransform, UiText as defineUiText, UiBackground as defineUiBackground, UiInput as defineUiInput, UiInputResult as defineUiInputResult, UiCanvasInformation as defineUiCanvasInformation, VideoPlayer as defineVideoPlayer, LightSource as defineLightSource, VirtualCamera as defineVirtualCamera, MainCamera as defineMainCamera, TextShape as defineTextShape, Material as defineMaterial, MeshRenderer as defineMeshRenderer, Billboard as defineBillboard, Name as defineName, Tween as defineTween, TweenSequence as defineTweenSequence, PointerEvents as definePointerEvents, NetworkEntity as defineNetworkEntity, SyncComponents as defineSyncComponents, InputModifier as defineInputModifier, SkyboxTime as defineSkyboxTime, } from '@dcl/ecs/dist/components';
export function getExplorerComponents(engine) {
    return {
        Animator: defineAnimator(engine),
        Transform: defineTransform(engine),
        AudioSource: defineAudioSource(engine),
        AvatarAttach: defineAvatarAttach(engine),
        VisibilityComponent: defineVisibilityComponent(engine),
        GltfContainer: defineGltfContainer(engine),
        UiTransform: defineUiTransform(engine),
        UiText: defineUiText(engine),
        UiBackground: defineUiBackground(engine),
        UiInput: defineUiInput(engine),
        UiInputResult: defineUiInputResult(engine),
        UiCanvasInformation: defineUiCanvasInformation(engine),
        VideoPlayer: defineVideoPlayer(engine),
        LightSource: defineLightSource(engine),
        VirtualCamera: defineVirtualCamera(engine),
        MainCamera: defineMainCamera(engine),
        TextShape: defineTextShape(engine),
        Material: defineMaterial(engine),
        MeshRenderer: defineMeshRenderer(engine),
        Billboard: defineBillboard(engine),
        Name: defineName(engine),
        Tween: defineTween(engine),
        TweenSequence: defineTweenSequence(engine),
        PointerEvents: definePointerEvents(engine),
        NetworkEntity: defineNetworkEntity(engine),
        SyncComponents: defineSyncComponents(engine),
        InputModifier: defineInputModifier(engine),
        SkyboxTime: defineSkyboxTime(engine),
    };
}
