// @dcl/asset-packs/dist/definitions.js
import { Schemas } from '@dcl/ecs';
import { addActionType } from './action-types';
import { ComponentName, TweenType, InterpolationType, ActionType, TriggerConditionType, AlignMode, Font, Colliders, ProximityLayer, TextureMovementType, TeleportMode, } from './enums';
import { getExplorerComponents } from './components';
import { defineAllComponents } from './versioning/registry';
import { BaseComponentNames } from './constants';
export const LIVEKIT_STREAM_SRC = 'livekit-video://current-stream';
export const VIDEO_URL_TYPE = 'https://';
export * from './enums';
export * from './action-types';
export * from './events';
export * from './id';
export * from './states';
export * from './clone';
export * from './lww';
export * from './types';
export * from './versioning';
export const ActionSchemas = {
    [ActionType.PLAY_ANIMATION]: Schemas.Map({
        animation: Schemas.String,
        loop: Schemas.Optional(Schemas.Boolean),
        shouldReset: Schemas.Optional(Schemas.Boolean),
    }),
    [ActionType.STOP_ANIMATION]: Schemas.Map({}),
    [ActionType.SET_STATE]: Schemas.Map({ state: Schemas.String }),
    [ActionType.START_TWEEN]: Schemas.Map({
        type: Schemas.EnumString(TweenType, TweenType.MOVE_ITEM),
        end: Schemas.Vector3,
        interpolationType: Schemas.EnumString(InterpolationType, InterpolationType.LINEAR),
        duration: Schemas.Float,
        relative: Schemas.Boolean,
        direction: Schemas.Optional(Schemas.Vector3),
        speed: Schemas.Optional(Schemas.Float),
    }),
    [ActionType.SET_COUNTER]: Schemas.Map({ counter: Schemas.Int }),
    [ActionType.INCREMENT_COUNTER]: Schemas.Map({
        amount: Schemas.Optional(Schemas.Int),
    }),
    [ActionType.DECREASE_COUNTER]: Schemas.Map({
        amount: Schemas.Optional(Schemas.Int),
    }),
    [ActionType.PLAY_SOUND]: Schemas.Map({
        src: Schemas.String,
        loop: Schemas.Optional(Schemas.Boolean),
        volume: Schemas.Optional(Schemas.Float),
        global: Schemas.Optional(Schemas.Boolean),
    }),
    [ActionType.STOP_SOUND]: Schemas.Map({}),
    [ActionType.SET_VISIBILITY]: Schemas.Map({
        visible: Schemas.Boolean,
        physicsCollider: Schemas.Optional(Schemas.Boolean),
        collider: Schemas.Optional(Schemas.EnumNumber(Colliders, Colliders.CL_POINTER)),
    }),
    [ActionType.ATTACH_TO_PLAYER]: Schemas.Map({
        anchorPointId: Schemas.Int,
    }),
    [ActionType.DETACH_FROM_PLAYER]: Schemas.Map({}),
    [ActionType.PLAY_VIDEO_STREAM]: Schemas.Map({
        src: Schemas.Optional(Schemas.String),
        loop: Schemas.Optional(Schemas.Boolean),
        volume: Schemas.Optional(Schemas.Float),
        dclCast: Schemas.Optional(Schemas.Boolean),
    }),
    [ActionType.STOP_VIDEO_STREAM]: Schemas.Map({}),
    [ActionType.PLAY_AUDIO_STREAM]: Schemas.Map({
        url: Schemas.String,
        volume: Schemas.Optional(Schemas.Float),
    }),
    [ActionType.STOP_AUDIO_STREAM]: Schemas.Map({}),
    [ActionType.TELEPORT_PLAYER]: Schemas.Map({
        mode: Schemas.EnumString(TeleportMode, TeleportMode.TO_COORDINATES),
        x: Schemas.Optional(Schemas.Int),
        y: Schemas.Optional(Schemas.Int),
        realm: Schemas.Optional(Schemas.String),
    }),
    [ActionType.MOVE_PLAYER]: Schemas.Map({
        position: Schemas.Vector3,
        cameraTarget: Schemas.Optional(Schemas.Vector3),
    }),
    [ActionType.PLAY_DEFAULT_EMOTE]: Schemas.Map({
        emote: Schemas.String,
    }),
    [ActionType.PLAY_CUSTOM_EMOTE]: Schemas.Map({
        src: Schemas.String,
        loop: Schemas.Optional(Schemas.Boolean),
    }),
    [ActionType.OPEN_LINK]: Schemas.Map({
        url: Schemas.String,
    }),
    [ActionType.SHOW_TEXT]: Schemas.Map({
        text: Schemas.String,
        hideAfterSeconds: Schemas.Float,
        font: Schemas.EnumNumber(Font, Font.F_SANS_SERIF),
        fontSize: Schemas.Optional(Schemas.Float),
        textAlign: Schemas.EnumNumber(AlignMode, AlignMode.TAM_MIDDLE_CENTER),
    }),
    [ActionType.HIDE_TEXT]: Schemas.Map({}),
    [ActionType.START_DELAY]: Schemas.Map({
        actions: Schemas.Array(Schemas.String),
        timeout: Schemas.Float,
    }),
    [ActionType.STOP_DELAY]: Schemas.Map({
        action: Schemas.String,
    }),
    [ActionType.START_LOOP]: Schemas.Map({
        actions: Schemas.Array(Schemas.String),
        interval: Schemas.Float,
    }),
    [ActionType.STOP_LOOP]: Schemas.Map({
        action: Schemas.String,
    }),
    [ActionType.CLONE_ENTITY]: Schemas.Map({
        position: Schemas.Vector3,
    }),
    [ActionType.REMOVE_ENTITY]: Schemas.Map({}),
    [ActionType.SHOW_IMAGE]: Schemas.Map({
        src: Schemas.String,
        align: Schemas.EnumNumber(AlignMode, AlignMode.TAM_MIDDLE_CENTER),
        height: Schemas.Float,
        width: Schemas.Float,
        hideAfterSeconds: Schemas.Optional(Schemas.Float),
        text: Schemas.Optional(Schemas.String),
        fontSize: Schemas.Optional(Schemas.Float),
    }),
    [ActionType.HIDE_IMAGE]: Schemas.Map({
        imageEntity: Schemas.Optional(Schemas.Int),
    }),
    [ActionType.DAMAGE]: Schemas.Map({
        radius: Schemas.Float,
        layer: Schemas.Optional(Schemas.EnumString(ProximityLayer, ProximityLayer.ALL)),
        hits: Schemas.Optional(Schemas.Int),
    }),
    [ActionType.MOVE_PLAYER_HERE]: Schemas.Map({}),
    [ActionType.PLAYER_FACE_ITEM]: Schemas.Map({}),
    [ActionType.PLACE_ON_PLAYER]: Schemas.Map({}),
    [ActionType.ROTATE_AS_PLAYER]: Schemas.Map({}),
    [ActionType.PLACE_ON_CAMERA]: Schemas.Map({}),
    [ActionType.ROTATE_AS_CAMERA]: Schemas.Map({}),
    [ActionType.SET_POSITION]: Schemas.Map({
        x: Schemas.Float,
        y: Schemas.Float,
        z: Schemas.Float,
        relative: Schemas.Optional(Schemas.Boolean),
    }),
    [ActionType.SET_ROTATION]: Schemas.Map({
        x: Schemas.Float,
        y: Schemas.Float,
        z: Schemas.Float,
        relative: Schemas.Optional(Schemas.Boolean),
    }),
    [ActionType.SET_SCALE]: Schemas.Map({
        x: Schemas.Float,
        y: Schemas.Float,
        z: Schemas.Float,
        relative: Schemas.Optional(Schemas.Boolean),
    }),
    [ActionType.FOLLOW_PLAYER]: Schemas.Map({
        speed: Schemas.Float,
        x: Schemas.Boolean,
        y: Schemas.Boolean,
        z: Schemas.Boolean,
        minDistance: Schemas.Float,
    }),
    [ActionType.STOP_FOLLOWING_PLAYER]: Schemas.Map({}),
    [ActionType.RANDOM]: Schemas.Map({
        actions: Schemas.Array(Schemas.String),
    }),
    [ActionType.BATCH]: Schemas.Map({
        actions: Schemas.Array(Schemas.String),
    }),
    [ActionType.HEAL_PLAYER]: Schemas.Map({
        multiplier: Schemas.Int,
    }),
    [ActionType.CLAIM_AIRDROP]: Schemas.Map({}),
    [ActionType.LIGHTS_ON]: Schemas.Map({}),
    [ActionType.LIGHTS_OFF]: Schemas.Map({}),
    [ActionType.LIGHTS_MODIFY]: Schemas.Map({
        active: Schemas.Optional(Schemas.Boolean),
        color: Schemas.Optional(Schemas.Color3),
        intensity: Schemas.Optional(Schemas.Float),
    }),
    [ActionType.CHANGE_CAMERA]: Schemas.Map({
        virtualCameraEntity: Schemas.Optional(Schemas.Entity),
    }),
    [ActionType.CHANGE_TEXT]: Schemas.Map({
        text: Schemas.String,
        fontSize: Schemas.Optional(Schemas.Float),
        color: Schemas.Optional(Schemas.Color4),
    }),
    [ActionType.STOP_TWEEN]: Schemas.Map({}),
    [ActionType.SLIDE_TEXTURE]: Schemas.Map({
        direction: Schemas.Map({ x: Schemas.Float, y: Schemas.Float }),
        speed: Schemas.Float,
        movementType: Schemas.Optional(Schemas.EnumNumber(TextureMovementType, TextureMovementType.TMT_OFFSET)),
        duration: Schemas.Optional(Schemas.Float),
    }),
    [ActionType.FREEZE_PLAYER]: Schemas.Map({}),
    [ActionType.UNFREEZE_PLAYER]: Schemas.Map({}),
    [ActionType.CHANGE_COLLISIONS]: Schemas.Map({
        visibleCollisions: Schemas.Optional(Schemas.Int),
        invisibleCollisions: Schemas.Optional(Schemas.Int),
    }),
    [ActionType.CHANGE_SKYBOX]: Schemas.Map({
        time: Schemas.Int,
        direction: Schemas.Optional(Schemas.Int),
    }),
    [ActionType.RESET_SKYBOX]: Schemas.Map({}),
    [ActionType.CALL_SCRIPT_METHOD]: Schemas.Map({
        scriptPath: Schemas.String,
        methodName: Schemas.String,
        params: Schemas.Optional(Schemas.Map({})),
    }),
    [ActionType.LOG_TO_CONSOLE]: Schemas.Map({
        message: Schemas.String,
    }),
    [ActionType.DELETE]: Schemas.Map({}),
};
export function getComponent(componentName, engine) {
    try {
        return engine.getComponent(componentName);
    }
    catch (error) {
        console.error(`Error using getComponent with componentName="${componentName}"`);
        throw error;
    }
}
export function getComponents(engine) {
    return {
        Actions: getComponent(ComponentName.ACTIONS, engine),
        States: getComponent(ComponentName.STATES, engine),
        Counter: getComponent(ComponentName.COUNTER, engine),
        Triggers: getComponent(ComponentName.TRIGGERS, engine),
        CounterBar: getComponent(ComponentName.COUNTER_BAR, engine),
        AdminTools: getComponent(ComponentName.ADMIN_TOOLS, engine),
        VideoScreen: getComponent(ComponentName.VIDEO_SCREEN, engine),
        Rewards: getComponent(ComponentName.REWARDS, engine),
        TextAnnouncements: getComponent(ComponentName.TEXT_ANNOUNCEMENTS, engine),
        VideoControlState: getComponent(ComponentName.VIDEO_CONTROL_STATE, engine),
        Script: getComponent(ComponentName.SCRIPT, engine),
        Placeholder: getComponent(ComponentName.PLACEHOLDER, engine),
    };
}
export function createComponents(engine) {
    const components = defineAllComponents(engine);
    return {
        ActionTypes: components[BaseComponentNames.ACTION_TYPES],
        Actions: components[BaseComponentNames.ACTIONS],
        Counter: components[BaseComponentNames.COUNTER],
        Triggers: components[BaseComponentNames.TRIGGERS],
        States: components[BaseComponentNames.STATES],
        CounterBar: components[BaseComponentNames.COUNTER_BAR],
        AdminTools: components[BaseComponentNames.ADMIN_TOOLS],
        Rewards: components[BaseComponentNames.REWARDS],
        TextAnnouncements: components[BaseComponentNames.TEXT_ANNOUNCEMENTS],
        VideoControlState: components[BaseComponentNames.VIDEO_CONTROL_STATE],
        VideoScreen: components[BaseComponentNames.VIDEO_SCREEN],
        Script: components[BaseComponentNames.SCRIPT],
        Placeholder: components[BaseComponentNames.PLACEHOLDER],
    };
}
export function initComponents(engine) {
    const actionTypes = Object.values(ActionType);
    for (const type of actionTypes) {
        const actionType = type;
        addActionType(engine, actionType, ActionSchemas[actionType]);
    }
    const Counter = engine.getComponent(ComponentName.COUNTER);
    const counter = Counter.getOrCreateMutable(engine.RootEntity);
    counter.value = counter.value || 0;
    const { VideoPlayer, Material } = getExplorerComponents(engine);
    initVideoPlayerComponents(engine, { VideoPlayer, Material });
}
function getVideoTexture({ material }) {
    if (material?.$case === 'unlit' && material.unlit.texture?.tex?.$case === 'videoTexture') {
        return material.unlit.texture.tex.videoTexture;
    }
    return undefined;
}
export function initVideoPlayerComponentMaterial(entity, { Material }, material) {
    if (!material || !material.material || material.material.$case !== 'unlit') {
        return null;
    }
    Material.setBasicMaterial(entity, {
        ...material.material.unlit,
        texture: Material.Texture.Video({
            videoPlayerEntity: entity,
        }),
    });
}
function initVideoPlayerComponents(engine, components) {
    function replaceVideoTexture() {
        const { Material, VideoPlayer } = components;
        engine.removeSystem(replaceVideoTexture);
        for (const [entity, material] of engine.getEntitiesWith(Material, VideoPlayer)) {
            const videoTexture = getVideoTexture(material);
            if (videoTexture?.videoPlayerEntity === engine.RootEntity) {
                initVideoPlayerComponentMaterial(entity, components, material);
            }
        }
    }
    engine.addSystem(replaceVideoTexture);
}
export function getConditionTypesByComponentName(componentName) {
    switch (componentName) {
        case ComponentName.STATES: {
            return [
                TriggerConditionType.WHEN_STATE_IS,
                TriggerConditionType.WHEN_STATE_IS_NOT,
                TriggerConditionType.WHEN_PREVIOUS_STATE_IS,
                TriggerConditionType.WHEN_PREVIOUS_STATE_IS_NOT,
            ];
        }
        case ComponentName.COUNTER: {
            return [
                TriggerConditionType.WHEN_COUNTER_EQUALS,
                TriggerConditionType.WHEN_COUNTER_IS_GREATER_THAN,
                TriggerConditionType.WHEN_COUNTER_IS_LESS_THAN,
            ];
        }
        case ComponentName.ACTIONS: {
            return [
                TriggerConditionType.WHEN_DISTANCE_TO_PLAYER_LESS_THAN,
                TriggerConditionType.WHEN_DISTANCE_TO_PLAYER_GREATER_THAN,
            ];
        }
        default: {
            return [];
        }
    }
}
