// @dcl/asset-packs/dist/enums.js
import { getLatestVersionName } from './versioning/registry';
import { BaseComponentNames } from './constants';
export { BaseComponentNames } from './constants';
export const ComponentName = {
    ACTION_TYPES: getLatestVersionName(BaseComponentNames.ACTION_TYPES),
    ACTIONS: getLatestVersionName(BaseComponentNames.ACTIONS),
    COUNTER: getLatestVersionName(BaseComponentNames.COUNTER),
    TRIGGERS: getLatestVersionName(BaseComponentNames.TRIGGERS),
    STATES: getLatestVersionName(BaseComponentNames.STATES),
    COUNTER_BAR: getLatestVersionName(BaseComponentNames.COUNTER_BAR),
    ADMIN_TOOLS: getLatestVersionName(BaseComponentNames.ADMIN_TOOLS),
    VIDEO_SCREEN: getLatestVersionName(BaseComponentNames.VIDEO_SCREEN),
    REWARDS: getLatestVersionName(BaseComponentNames.REWARDS),
    TEXT_ANNOUNCEMENTS: getLatestVersionName(BaseComponentNames.TEXT_ANNOUNCEMENTS),
    VIDEO_CONTROL_STATE: getLatestVersionName(BaseComponentNames.VIDEO_CONTROL_STATE),
    SCRIPT: getLatestVersionName(BaseComponentNames.SCRIPT),
    PLACEHOLDER: getLatestVersionName(BaseComponentNames.PLACEHOLDER),
};
export var TweenType;
(function (TweenType) {
    TweenType["MOVE_ITEM"] = "move_item";
    TweenType["ROTATE_ITEM"] = "rotate_item";
    TweenType["SCALE_ITEM"] = "scale_item";
    TweenType["KEEP_ROTATING_ITEM"] = "keep_rotating_item";
})(TweenType || (TweenType = {}));
export var TeleportMode;
(function (TeleportMode) {
    TeleportMode["TO_COORDINATES"] = "to_coordinates";
    TeleportMode["TO_WORLD"] = "to_world";
})(TeleportMode || (TeleportMode = {}));
export var InterpolationType;
(function (InterpolationType) {
    InterpolationType["LINEAR"] = "linear";
    InterpolationType["EASEINQUAD"] = "easeinquad";
    InterpolationType["EASEOUTQUAD"] = "easeoutquad";
    InterpolationType["EASEQUAD"] = "easequad";
    InterpolationType["EASEINSINE"] = "easeinsine";
    InterpolationType["EASEOUTSINE"] = "easeoutsine";
    InterpolationType["EASESINE"] = "easeinoutsine";
    InterpolationType["EASEINEXPO"] = "easeinexpo";
    InterpolationType["EASEOUTEXPO"] = "easeoutexpo";
    InterpolationType["EASEEXPO"] = "easeinoutexpo";
    InterpolationType["EASEINELASTIC"] = "easeinelastic";
    InterpolationType["EASEOUTELASTIC"] = "easeoutelastic";
    InterpolationType["EASEELASTIC"] = "easeinoutelastic";
    InterpolationType["EASEINBOUNCE"] = "easeinbounce";
    InterpolationType["EASEOUTEBOUNCE"] = "easeoutbounce";
    InterpolationType["EASEBOUNCE"] = "easeinoutbounce";
})(InterpolationType || (InterpolationType = {}));
export var ActionType;
(function (ActionType) {
    ActionType["PLAY_ANIMATION"] = "play_animation";
    ActionType["STOP_ANIMATION"] = "stop_animation";
    ActionType["SET_STATE"] = "set_state";
    ActionType["START_TWEEN"] = "start_tween";
    ActionType["SET_COUNTER"] = "set_counter";
    ActionType["INCREMENT_COUNTER"] = "increment_counter";
    ActionType["DECREASE_COUNTER"] = "decrease_counter";
    ActionType["PLAY_SOUND"] = "play_sound";
    ActionType["STOP_SOUND"] = "stop_sound";
    ActionType["SET_VISIBILITY"] = "set_visibility";
    ActionType["ATTACH_TO_PLAYER"] = "attach_to_player";
    ActionType["DETACH_FROM_PLAYER"] = "detach_from_player";
    ActionType["PLAY_VIDEO_STREAM"] = "play_video_stream";
    ActionType["STOP_VIDEO_STREAM"] = "stop_video_stream";
    ActionType["PLAY_AUDIO_STREAM"] = "play_audio_stream";
    ActionType["STOP_AUDIO_STREAM"] = "stop_audio_stream";
    ActionType["TELEPORT_PLAYER"] = "teleport_player";
    ActionType["MOVE_PLAYER"] = "move_player";
    ActionType["PLAY_DEFAULT_EMOTE"] = "play_default_emote";
    ActionType["PLAY_CUSTOM_EMOTE"] = "play_custom_emote";
    ActionType["OPEN_LINK"] = "open_link";
    ActionType["SHOW_TEXT"] = "show_text";
    ActionType["HIDE_TEXT"] = "hide_text";
    ActionType["START_DELAY"] = "start_delay";
    ActionType["STOP_DELAY"] = "stop_delay";
    ActionType["START_LOOP"] = "start_loop";
    ActionType["STOP_LOOP"] = "stop_loop";
    ActionType["CLONE_ENTITY"] = "clone_entity";
    ActionType["REMOVE_ENTITY"] = "remove_entity";
    ActionType["SHOW_IMAGE"] = "show_image";
    ActionType["HIDE_IMAGE"] = "hide_image";
    ActionType["DAMAGE"] = "damage";
    ActionType["MOVE_PLAYER_HERE"] = "move_player_here";
    ActionType["PLAYER_FACE_ITEM"] = "player_face_item";
    ActionType["PLACE_ON_PLAYER"] = "place_on_player";
    ActionType["ROTATE_AS_PLAYER"] = "rotate_as_player";
    ActionType["PLACE_ON_CAMERA"] = "place_on_camera";
    ActionType["ROTATE_AS_CAMERA"] = "rotate_as_camera";
    ActionType["SET_POSITION"] = "set_position";
    ActionType["SET_ROTATION"] = "set_rotation";
    ActionType["SET_SCALE"] = "set_scale";
    ActionType["FOLLOW_PLAYER"] = "follow_player";
    ActionType["STOP_FOLLOWING_PLAYER"] = "stop_following_player";
    ActionType["RANDOM"] = "random";
    ActionType["BATCH"] = "batch";
    ActionType["HEAL_PLAYER"] = "heal_player";
    ActionType["CLAIM_AIRDROP"] = "claim_airdrop";
    ActionType["LIGHTS_ON"] = "lights_on";
    ActionType["LIGHTS_OFF"] = "lights_off";
    ActionType["LIGHTS_MODIFY"] = "lights_modify";
    ActionType["CHANGE_CAMERA"] = "change_camera";
    ActionType["CHANGE_TEXT"] = "change_text";
    ActionType["STOP_TWEEN"] = "stop_tween";
    ActionType["SLIDE_TEXTURE"] = "slide_texture";
    ActionType["FREEZE_PLAYER"] = "freeze_player";
    ActionType["UNFREEZE_PLAYER"] = "unfreeze_player";
    ActionType["CHANGE_COLLISIONS"] = "change_collisions";
    ActionType["CHANGE_SKYBOX"] = "change_skybox";
    ActionType["RESET_SKYBOX"] = "reset_skybox";
    ActionType["CALL_SCRIPT_METHOD"] = "call_script_method";
    ActionType["LOG_TO_CONSOLE"] = "log_to_console";
    ActionType["DELETE"] = "delete";
})(ActionType || (ActionType = {}));
export { TriggerType, TriggerConditionType, TriggerConditionOperation } from './trigger-enums';
export var AlignMode;
(function (AlignMode) {
    AlignMode[AlignMode["TAM_TOP_LEFT"] = 0] = "TAM_TOP_LEFT";
    AlignMode[AlignMode["TAM_TOP_CENTER"] = 1] = "TAM_TOP_CENTER";
    AlignMode[AlignMode["TAM_TOP_RIGHT"] = 2] = "TAM_TOP_RIGHT";
    AlignMode[AlignMode["TAM_MIDDLE_LEFT"] = 3] = "TAM_MIDDLE_LEFT";
    AlignMode[AlignMode["TAM_MIDDLE_CENTER"] = 4] = "TAM_MIDDLE_CENTER";
    AlignMode[AlignMode["TAM_MIDDLE_RIGHT"] = 5] = "TAM_MIDDLE_RIGHT";
    AlignMode[AlignMode["TAM_BOTTOM_LEFT"] = 6] = "TAM_BOTTOM_LEFT";
    AlignMode[AlignMode["TAM_BOTTOM_CENTER"] = 7] = "TAM_BOTTOM_CENTER";
    AlignMode[AlignMode["TAM_BOTTOM_RIGHT"] = 8] = "TAM_BOTTOM_RIGHT";
})(AlignMode || (AlignMode = {}));
export var Font;
(function (Font) {
    Font[Font["F_SANS_SERIF"] = 0] = "F_SANS_SERIF";
    Font[Font["F_SERIF"] = 1] = "F_SERIF";
    Font[Font["F_MONOSPACE"] = 2] = "F_MONOSPACE";
})(Font || (Font = {}));
export var Colliders;
(function (Colliders) {
    Colliders[Colliders["CL_NONE"] = 0] = "CL_NONE";
    Colliders[Colliders["CL_POINTER"] = 1] = "CL_POINTER";
    Colliders[Colliders["CL_PHYSICS"] = 2] = "CL_PHYSICS";
    Colliders[Colliders["CL_RESERVED1"] = 4] = "CL_RESERVED1";
    Colliders[Colliders["CL_RESERVED2"] = 8] = "CL_RESERVED2";
    Colliders[Colliders["CL_RESERVED3"] = 16] = "CL_RESERVED3";
    Colliders[Colliders["CL_RESERVED4"] = 32] = "CL_RESERVED4";
    Colliders[Colliders["CL_RESERVED5"] = 64] = "CL_RESERVED5";
    Colliders[Colliders["CL_RESERVED6"] = 128] = "CL_RESERVED6";
    Colliders[Colliders["CL_CUSTOM1"] = 256] = "CL_CUSTOM1";
    Colliders[Colliders["CL_CUSTOM2"] = 512] = "CL_CUSTOM2";
    Colliders[Colliders["CL_CUSTOM3"] = 1024] = "CL_CUSTOM3";
    Colliders[Colliders["CL_CUSTOM4"] = 2048] = "CL_CUSTOM4";
    Colliders[Colliders["CL_CUSTOM5"] = 4096] = "CL_CUSTOM5";
    Colliders[Colliders["CL_CUSTOM6"] = 8192] = "CL_CUSTOM6";
    Colliders[Colliders["CL_CUSTOM7"] = 16384] = "CL_CUSTOM7";
    Colliders[Colliders["CL_CUSTOM8"] = 32768] = "CL_CUSTOM8";
})(Colliders || (Colliders = {}));
export var TextureMovementType;
(function (TextureMovementType) {
    TextureMovementType[TextureMovementType["TMT_OFFSET"] = 0] = "TMT_OFFSET";
    TextureMovementType[TextureMovementType["TMT_TILING"] = 1] = "TMT_TILING";
})(TextureMovementType || (TextureMovementType = {}));
export var ProximityLayer;
(function (ProximityLayer) {
    ProximityLayer["ALL"] = "all";
    ProximityLayer["PLAYER"] = "player";
    ProximityLayer["NON_PLAYER"] = "non_player";
})(ProximityLayer || (ProximityLayer = {}));
export { AdminPermissions, MediaSource } from './constants';
