// @dcl/asset-packs/dist/trigger-enums.js
export var TriggerType;
(function (TriggerType) {
    TriggerType["ON_CLICK"] = "on_click";
    TriggerType["ON_INPUT_ACTION"] = "on_input_action";
    TriggerType["ON_STATE_CHANGE"] = "on_state_change";
    TriggerType["ON_SPAWN"] = "on_spawn";
    TriggerType["ON_TWEEN_END"] = "on_tween_end";
    TriggerType["ON_COUNTER_CHANGE"] = "on_counter_change";
    TriggerType["ON_PLAYER_ENTERS_AREA"] = "on_player_enters_area";
    TriggerType["ON_PLAYER_LEAVES_AREA"] = "on_player_leaves_area";
    TriggerType["ON_DELAY"] = "on_delay";
    TriggerType["ON_LOOP"] = "on_loop";
    TriggerType["ON_CLONE"] = "on_clone";
    TriggerType["ON_CLICK_IMAGE"] = "on_click_image";
    TriggerType["ON_DAMAGE"] = "on_damage";
    TriggerType["ON_GLOBAL_CLICK"] = "on_global_click";
    TriggerType["ON_GLOBAL_PRIMARY"] = "on_global_primary";
    TriggerType["ON_GLOBAL_SECONDARY"] = "on_global_secondary";
    TriggerType["ON_TICK"] = "on_tick";
    TriggerType["ON_HEAL_PLAYER"] = "on_heal_player";
    TriggerType["ON_PLAYER_SPAWN"] = "on_player_spawn";
})(TriggerType || (TriggerType = {}));
export var TriggerConditionType;
(function (TriggerConditionType) {
    TriggerConditionType["WHEN_STATE_IS"] = "when_state_is";
    TriggerConditionType["WHEN_STATE_IS_NOT"] = "when_state_is_not";
    TriggerConditionType["WHEN_COUNTER_EQUALS"] = "when_counter_equals";
    TriggerConditionType["WHEN_COUNTER_IS_GREATER_THAN"] = "when_counter_is_greater_than";
    TriggerConditionType["WHEN_COUNTER_IS_LESS_THAN"] = "when_counter_is_less_than";
    TriggerConditionType["WHEN_DISTANCE_TO_PLAYER_LESS_THAN"] = "when_distance_to_player_less_than";
    TriggerConditionType["WHEN_DISTANCE_TO_PLAYER_GREATER_THAN"] = "when_distance_to_player_greater_than";
    TriggerConditionType["WHEN_PREVIOUS_STATE_IS"] = "when_previous_state_is";
    TriggerConditionType["WHEN_PREVIOUS_STATE_IS_NOT"] = "when_previous_state_is_not";
})(TriggerConditionType || (TriggerConditionType = {}));
export var TriggerConditionOperation;
(function (TriggerConditionOperation) {
    TriggerConditionOperation["AND"] = "and";
    TriggerConditionOperation["OR"] = "or";
})(TriggerConditionOperation || (TriggerConditionOperation = {}));
