// @dcl/asset-packs/dist/types.js
import { BaseComponentNames } from './constants';
export function isAssetPackData(value) {
    return (typeof value === 'object' &&
        value !== null &&
        typeof value.id === 'string' &&
        typeof value.name === 'string');
}
export function isAssetDataWithoutComposite(value) {
    if ('composite' in value) {
        return false;
    }
    return (value &&
        typeof value === 'object' &&
        'id' in value &&
        typeof value.id === 'string' &&
        'name' in value &&
        typeof value.name === 'string' &&
        'category' in value &&
        typeof value.category === 'string' &&
        'tags' in value &&
        Array.isArray(value.tags));
}
export function isAssetData(value) {
    return (value &&
        typeof value === 'object' &&
        'composite' in value &&
        typeof value.composite === 'object' &&
        'version' in value.composite &&
        'components' in value.composite &&
        Array.isArray(value.composite.components));
}
export function isLegacyAssetData(value) {
    return (typeof value === 'object' &&
        value !== null &&
        typeof value.id === 'string' &&
        typeof value.name === 'string' &&
        typeof value.category === 'string' &&
        typeof value.components === 'object' &&
        value.components !== null);
}
export function getTriggerComponent(value) {
    return value.composite.components.find(c => c.name === BaseComponentNames.TRIGGERS)?.data[0]
        ?.json;
}
export function assertValidTriggerComponent(assetName, trigger) {
    trigger.value.forEach(({ type, actions }) => {
        actions.forEach(({ id, name }) => {
            if (!id || !name) {
                throw new Error(`Invalid actions found on "${assetName}" for trigger with type "${type}"`);
            }
        });
    });
}
