// @dcl/asset-packs/dist/ui.js
import { Color4 } from '@dcl/sdk/math';
import { AlignMode, Font } from './enums';
import { CONTENT_URL } from './admin-toolkit-ui/constants';
function getAlignMode(align, isColumn) {
    switch (align) {
        case AlignMode.TAM_TOP_LEFT: {
            return {
                alignItems: 1,
                justifyContent: 0,
            };
        }
        case AlignMode.TAM_TOP_CENTER: {
            return isColumn
                ? {
                    alignItems: 2,
                    justifyContent: 0,
                }
                : {
                    alignItems: 1,
                    justifyContent: 1,
                };
        }
        case AlignMode.TAM_TOP_RIGHT: {
            return isColumn
                ? {
                    alignItems: 3,
                    justifyContent: 0,
                }
                : {
                    alignItems: 1,
                    justifyContent: 2,
                };
        }
        case AlignMode.TAM_MIDDLE_LEFT: {
            return isColumn
                ? {
                    alignItems: 1,
                    justifyContent: 1,
                }
                : {
                    alignItems: 2,
                    justifyContent: 0,
                };
        }
        case AlignMode.TAM_MIDDLE_CENTER: {
            return {
                alignItems: 2,
                justifyContent: 1,
            };
        }
        case AlignMode.TAM_MIDDLE_RIGHT: {
            return isColumn
                ? {
                    alignItems: 3,
                    justifyContent: 1,
                }
                : {
                    alignItems: 2,
                    justifyContent: 2,
                };
        }
        case AlignMode.TAM_BOTTOM_LEFT: {
            return isColumn
                ? {
                    alignItems: 1,
                    justifyContent: 2,
                }
                : {
                    alignItems: 2,
                    justifyContent: 1,
                };
        }
        case AlignMode.TAM_BOTTOM_CENTER: {
            return isColumn
                ? {
                    alignItems: 2,
                    justifyContent: 2,
                }
                : {
                    alignItems: 3,
                    justifyContent: 1,
                };
        }
        case AlignMode.TAM_BOTTOM_RIGHT: {
            return {
                alignItems: 3,
                justifyContent: 2,
            };
        }
    }
}
const validAlignModes = new Set([
    AlignMode.TAM_TOP_LEFT,
    AlignMode.TAM_TOP_CENTER,
    AlignMode.TAM_TOP_RIGHT,
    AlignMode.TAM_MIDDLE_LEFT,
    AlignMode.TAM_MIDDLE_CENTER,
    AlignMode.TAM_MIDDLE_RIGHT,
    AlignMode.TAM_BOTTOM_LEFT,
    AlignMode.TAM_BOTTOM_CENTER,
    AlignMode.TAM_BOTTOM_RIGHT,
]);
export function mapAlignToScreenAlign(align, flexDirection = 0) {
    const isColumn = flexDirection === 1;
    if (validAlignModes.has(align)) {
        return getAlignMode(align, isColumn);
    }
    throw new Error(`Unsupported AlignMode: ${align}`);
}
export function getUITransform(component, entiy, height = 100, width = 100, unit = 2) {
    let uiTransformComponent = component.getMutableOrNull(entiy);
    if (!uiTransformComponent) {
        uiTransformComponent = component.create(entiy);
        uiTransformComponent.heightUnit = unit;
        uiTransformComponent.widthUnit = unit;
        uiTransformComponent.height = height;
        uiTransformComponent.width = width;
        uiTransformComponent.maxHeightUnit = unit;
        uiTransformComponent.maxWidthUnit = unit;
        uiTransformComponent.maxHeight = height;
        uiTransformComponent.maxWidth = width;
    }
    if (entiy === 0) {
        uiTransformComponent.positionType = 1;
    }
    return uiTransformComponent;
}
export function getUIBackground(component, entity, src, textureMode = 0, wrapMode = 1) {
    return component.createOrReplace(entity, {
        textureMode,
        texture: {
            tex: {
                $case: 'texture',
                texture: {
                    src,
                    wrapMode,
                },
            },
        },
        uvs: [],
    });
}
function breakLines(text, linelength) {
    const lineBreak = '\n';
    let counter = 0;
    let line = '';
    let returnText = '';
    let bMatchFound = false;
    const lineLen = linelength ? linelength : 50;
    if (!text)
        return '';
    if (text.length < lineLen + 1) {
        return text;
    }
    while (counter < text.length) {
        line = text.substring(counter, counter + lineLen);
        bMatchFound = false;
        if (line.length == lineLen) {
            for (let i = line.length; i > -1; i--) {
                if (line.substring(i, i + 1) == ' ') {
                    counter += line.substring(0, i).length;
                    line = line.substring(0, i) + lineBreak;
                    returnText += line;
                    bMatchFound = true;
                    break;
                }
            }
            if (!bMatchFound) {
                counter += line.length;
                line = line + lineBreak;
                returnText += line;
            }
        }
        else {
            returnText += line;
            break;
        }
    }
    return returnText;
}
export function getUIText(component, entity, text, fontSize = 10, containerWidth, align = 4, color = Color4.White()) {
    const lineLength = Math.floor(containerWidth / (fontSize / 1.7));
    return component.createOrReplace(entity, {
        value: breakLines(text, lineLength),
        fontSize,
        font: Font.F_MONOSPACE,
        textAlign: align,
        color,
    });
}
export function removeUiTransformEntities(engine, UiTransform, parentEntity) {
    const entitiesWithUiTransform = engine.getEntitiesWith(UiTransform);
    for (const [uiEntity, uiTransform] of entitiesWithUiTransform) {
        if (uiTransform.parent === parentEntity) {
            removeUiTransformEntities(engine, UiTransform, uiEntity);
            engine.removeEntity(uiEntity);
        }
    }
}
const BTN_CLOSE_TEXT_ANNOUNCEMENT = `${CONTENT_URL}/admin_toolkit/assets/icons/text-announcement-close-button.png`;
export function showCaptchaPrompt(engine, UiTransform, UiBackground, UiText, UiInput, UiInputResult, pointerEventsSystem, data, onSubmit) {
    console.log('showCaptchaPrompt');
    const uiStack = engine.addEntity();
    const uiStackTransform = getUITransform(UiTransform, uiStack);
    uiStackTransform.display = 0;
    uiStackTransform.alignItems = 2;
    uiStackTransform.justifyContent = 1;
    const containerEntity = engine.addEntity();
    const containerTransform = getUITransform(UiTransform, containerEntity, 260, 250, 1);
    containerTransform.parent = uiStack;
    containerTransform.display = 0;
    containerTransform.flexDirection = 1;
    containerTransform.alignItems = 2;
    containerTransform.paddingTop = 20;
    containerTransform.paddingTopUnit = 1;
    containerTransform.paddingBottom = 20;
    containerTransform.paddingBottomUnit = 1;
    containerTransform.paddingLeft = 20;
    containerTransform.paddingLeftUnit = 1;
    containerTransform.paddingRight = 20;
    containerTransform.paddingRightUnit = 1;
    containerTransform.pointerFilter = 1;
    UiBackground.createOrReplace(containerEntity, {
        color: Color4.create(0.15, 0.15, 0.15, 0.95),
        textureMode: 0,
        uvs: [],
    });
    const closeButtonEntity = engine.addEntity();
    const closeButtonTransform = getUITransform(UiTransform, closeButtonEntity, 24, 24, 1);
    closeButtonTransform.parent = containerEntity;
    closeButtonTransform.positionType = 1;
    closeButtonTransform.positionRight = 5;
    closeButtonTransform.positionRightUnit = 1;
    closeButtonTransform.positionTop = 5;
    closeButtonTransform.positionTopUnit = 1;
    closeButtonTransform.pointerFilter = 1;
    getUIBackground(UiBackground, closeButtonEntity, BTN_CLOSE_TEXT_ANNOUNCEMENT, 2);
    pointerEventsSystem.onPointerDown({
        entity: closeButtonEntity,
        opts: {
            button: 0,
        },
    }, () => {
        removeUiTransformEntities(engine, UiTransform, uiStack);
        engine.removeEntity(uiStack);
    });
    const titleEntity = engine.addEntity();
    const titleTransform = getUITransform(UiTransform, titleEntity, 20, 200, 1);
    titleTransform.parent = containerEntity;
    titleTransform.marginBottom = 10;
    titleTransform.marginBottomUnit = 1;
    titleTransform.positionType = 1;
    titleTransform.positionTop = 30;
    titleTransform.positionTopUnit = 1;
    getUIText(UiText, titleEntity, 'Enter the captcha', 20, 200, 4, Color4.White());
    const imageEntity = engine.addEntity();
    const imageTransform = getUITransform(UiTransform, imageEntity, 80, 200, 1);
    imageTransform.parent = containerEntity;
    imageTransform.marginBottom = 10;
    imageTransform.marginBottomUnit = 1;
    imageTransform.positionType = 1;
    imageTransform.positionTop = 60;
    imageTransform.positionTopUnit = 1;
    getUIBackground(UiBackground, imageEntity, data.captcha.image, 2);
    const inputEntity = engine.addEntity();
    const inputTransform = getUITransform(UiTransform, inputEntity, 50, 200, 1);
    inputTransform.parent = containerEntity;
    inputTransform.marginBottom = 10;
    inputTransform.marginBottomUnit = 1;
    inputTransform.positionType = 1;
    inputTransform.positionTop = 150;
    inputTransform.positionTopUnit = 1;
    UiInput.createOrReplace(inputEntity, {
        placeholder: 'Enter the captcha',
        color: Color4.White(),
        placeholderColor: Color4.create(0.3, 0.3, 0.3, 1),
        disabled: false,
        textAlign: 4,
        font: Font.F_MONOSPACE,
        fontSize: 10,
    });
    UiInputResult.createOrReplace(inputEntity, {
        value: '',
        isSubmit: false,
    });
    const buttonEntity = engine.addEntity();
    const buttonTransform = getUITransform(UiTransform, buttonEntity, 30, 100, 1);
    buttonTransform.parent = containerEntity;
    buttonTransform.marginBottom = 10;
    buttonTransform.marginBottomUnit = 1;
    buttonTransform.positionType = 1;
    buttonTransform.positionTop = 210;
    buttonTransform.positionTopUnit = 1;
    buttonTransform.pointerFilter = 1;
    getUIText(UiText, buttonEntity, 'Submit', 16, 100, 4, Color4.White());
    UiBackground.createOrReplace(buttonEntity, {
        color: { r: 0.3, g: 0.3, b: 0.3, a: 1 },
        textureMode: 0,
        uvs: [],
    });
    pointerEventsSystem.onPointerDown({
        entity: buttonEntity,
        opts: {
            button: 0,
        },
    }, () => {
        const inputText = UiInputResult.get(inputEntity)?.value || '';
        onSubmit(inputText);
        removeUiTransformEntities(engine, UiTransform, uiStack);
        engine.removeEntity(uiStack);
    });
    return uiStack;
}
