// @dcl/sdk-commands/dist/commands/get-context-files/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.args = void 0;
exports.help = help;
exports.main = main;
const path_1 = __importDefault(require("path"));
const args_1 = require("../../logic/args");
const project_validations_1 = require("../../logic/project-validations");
const error_1 = require("../../logic/error");
const i18next_1 = __importDefault(require("i18next"));
const GITHUB_API_BASE = 'https://api.github.com/repos/decentraland/documentation/contents/ai-sdk-context';
exports.args = (0, args_1.declareArgs)({
    '--help': Boolean,
    '-h': '--help'
});
function help(options) {
    options.components.logger.log(`
  Usage: 'sdk-commands get-context-files [options]'
    Options:
      -h, --help                Gets & updates context files from https://github.com/decentraland/documentation/tree/main/ai-sdk-context only on valid scenes

    Example:
    - Get context files:
      $ sdk-commands get-context-files
  `);
}
async function listFilesFromPath(components, url) {
    const response = await components.fetch.fetch(url);
    if (!response.ok) {
        throw new error_1.CliError('GET_CONTEXT_FILES_LIST_FAILED', i18next_1.default.t('errors.get_context_files.list_failed'));
    }
    return (await response.json());
}
async function downloadFile(components, url) {
    const response = await components.fetch.fetch(url);
    if (!response.ok) {
        throw new error_1.CliError('GET_CONTEXT_FILES_DOWNLOAD_FAILED', i18next_1.default.t('errors.get_context_files.download_failed'));
    }
    return await response.text();
}
async function getAllFiles(components, initialUrl = GITHUB_API_BASE) {
    const files = [];
    const rootFiles = await listFilesFromPath(components, initialUrl);
    for (const file of rootFiles) {
        if (file.type === 'file') {
            if (file.download_url) {
                const filename = file.name;
                const fileInfo = {
                    url: file.download_url,
                    filename: filename,
                    path: file.path
                };
                files.push(fileInfo);
            }
        }
        else if (file.type === 'dir') {
            const subFiles = await getAllFiles(components, file.url);
            files.push(...subFiles);
        }
    }
    return files;
}
async function main(options) {
    const targetDir = process.cwd();
    try {
        await (0, project_validations_1.assertValidProjectFolder)(options.components, targetDir);
        options.components.logger.log('✓ Valid Scene project');
    }
    catch (error) {
        options.components.logger.log('Not a valid Scene...');
        return;
    }
    const contextDir = path_1.default.join(targetDir, 'dclcontext');
    const contextExists = await options.components.fs.directoryExists(contextDir);
    if (contextExists) {
        options.components.logger.log('Context directory exists. Removing old files...');
        await options.components.fs.rm(contextDir, { recursive: true });
    }
    options.components.logger.log('Creating context directory...');
    await options.components.fs.mkdir(contextDir);
    const filesToDownload = await getAllFiles(options.components, GITHUB_API_BASE);
    const successfulDownloads = [];
    const failedDownloads = [];
    const downloadPromises = filesToDownload.map(async ({ url, filename, path: filePath }) => {
        try {
            const content = await downloadFile(options.components, url);
            const localFilePath = path_1.default.join(contextDir, filename);
            await options.components.fs.writeFile(localFilePath, content);
            options.components.logger.log(`✓ Saved ${filePath}`);
            successfulDownloads.push(filePath);
        }
        catch (error) {
            const errorMessage = error instanceof Error ? error.message : String(error);
            options.components.logger.log(`✗ Failed to download ${filePath}: ${errorMessage}`);
            failedDownloads.push({ filePath, error: errorMessage });
        }
    });
    await Promise.all(downloadPromises);
    options.components.logger.log(`\nDownload complete: ${successfulDownloads.length} successful, ${failedDownloads.length} failed`);
    if (successfulDownloads.length > 0) {
        options.components.logger.log(`Successfully downloaded:`);
        successfulDownloads.forEach((filePath) => {
            options.components.logger.log(`  ✓ ${filePath}`);
        });
    }
    if (failedDownloads.length > 0) {
        options.components.logger.log(`Failed downloads:`);
        failedDownloads.forEach(({ filePath, error }) => {
            options.components.logger.log(`  ✗ ${filePath}: ${error}`);
        });
    }
}
