// @dcl/ecs/dist/systems/triggerArea.js
import * as components from '../components';
import { EntityState } from '../engine/entity';
/**
 * @internal
 */
export function createTriggerAreaEventsSystem(engine) {
    const triggerAreaResultComponent = components.TriggerAreaResult(engine);
    const entitiesMap = new Map();
    function hasCallbacksMap(entity) {
        return entitiesMap.has(entity) && entitiesMap.get(entity) !== undefined;
    }
    function addEntityCallback(entity, triggerType, callback) {
        if (hasCallbacksMap(entity)) {
            entitiesMap.get(entity).triggerCallbackMap.set(triggerType, callback);
        }
        else {
            entitiesMap.set(entity, {
                triggerCallbackMap: new Map([[triggerType, callback]]),
                lastConsumedTimestamp: -1
            });
        }
    }
    function removeEntityCallback(entity, triggerType) {
        if (!entitiesMap.has(entity) || !entitiesMap.get(entity).triggerCallbackMap.has(triggerType))
            return;
        const triggerCallbackMap = entitiesMap.get(entity).triggerCallbackMap;
        triggerCallbackMap.delete(triggerType);
        // Remove entity if no more trigger callbacks are registered
        if (triggerCallbackMap.size === 0)
            entitiesMap.delete(entity);
    }
    function onTriggerEnter(entity, cb) {
        addEntityCallback(entity, 0 /* TriggerAreaEventType.TAET_ENTER */, cb);
    }
    function removeOnTriggerEnter(entity) {
        removeEntityCallback(entity, 0 /* TriggerAreaEventType.TAET_ENTER */);
    }
    function onTriggerStay(entity, cb) {
        addEntityCallback(entity, 1 /* TriggerAreaEventType.TAET_STAY */, cb);
    }
    function removeOnTriggerStay(entity) {
        removeEntityCallback(entity, 1 /* TriggerAreaEventType.TAET_STAY */);
    }
    function onTriggerExit(entity, cb) {
        addEntityCallback(entity, 2 /* TriggerAreaEventType.TAET_EXIT */, cb);
    }
    function removeOnTriggerExit(entity) {
        removeEntityCallback(entity, 2 /* TriggerAreaEventType.TAET_EXIT */);
    }
    engine.addSystem(function TriggerAreaResultSystem() {
        const garbageEntries = [];
        for (const [entity, data] of entitiesMap) {
            if (engine.getEntityState(entity) === EntityState.Removed) {
                garbageEntries.push(entity);
                continue;
            }
            const result = triggerAreaResultComponent.get(entity);
            // The Explorer may be taking time before the result component is put
            if (result.size === 0)
                continue;
            const values = Array.from(result.values());
            // determine starting index for new values (more than one could be added between System updates)
            // search backwards to find the anchor at lastConsumedTimestamp
            let startIndex = 0;
            if (data.lastConsumedTimestamp >= 0) {
                const newestTimestamp = values[values.length - 1].timestamp;
                // if nothing new, skip processing
                if (newestTimestamp <= data.lastConsumedTimestamp) {
                    continue;
                }
                // Find index of value with the lastConsumedTimestamp
                let i = values.length - 2;
                while (i >= 0 && values[i].timestamp > data.lastConsumedTimestamp)
                    i--;
                // Mark the following value index as the starting point to trigger all the new value callbacks
                startIndex = i + 1;
            }
            if (startIndex >= values.length)
                continue;
            // Trigger callbacks for all the new values
            for (let i = startIndex; i < values.length; i++) {
                switch (values[i].eventType) {
                    case 0 /* TriggerAreaEventType.TAET_ENTER */:
                        if (!data.triggerCallbackMap.has(0 /* TriggerAreaEventType.TAET_ENTER */))
                            continue;
                        data.triggerCallbackMap.get(0 /* TriggerAreaEventType.TAET_ENTER */)(values[i]);
                        break;
                    case 1 /* TriggerAreaEventType.TAET_STAY */:
                        if (!data.triggerCallbackMap.has(1 /* TriggerAreaEventType.TAET_STAY */))
                            continue;
                        data.triggerCallbackMap.get(1 /* TriggerAreaEventType.TAET_STAY */)(values[i]);
                        break;
                    case 2 /* TriggerAreaEventType.TAET_EXIT */:
                        if (!data.triggerCallbackMap.has(2 /* TriggerAreaEventType.TAET_EXIT */))
                            continue;
                        data.triggerCallbackMap.get(2 /* TriggerAreaEventType.TAET_EXIT */)(values[i]);
                        break;
                }
            }
            data.lastConsumedTimestamp = values[values.length - 1].timestamp;
        }
        // Clean up garbage entries
        garbageEntries.forEach((garbageEntity) => entitiesMap.delete(garbageEntity));
    });
    return {
        onTriggerEnter,
        removeOnTriggerEnter,
        onTriggerStay,
        removeOnTriggerStay,
        onTriggerExit,
        removeOnTriggerExit
    };
}
