// @dcl/schemas/dist/platform/item/emote/adr287/emote-data-adr287.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OutcomeGroup = exports.StartAnimation = exports.EmoteClip = exports.ArmatureId = void 0;
const validation_1 = require("../../../../validation");
var ArmatureId;
(function (ArmatureId) {
    ArmatureId["Armature"] = "Armature";
    ArmatureId["Armature_Prop"] = "Armature_Prop";
    ArmatureId["Armature_Other"] = "Armature_Other";
})(ArmatureId = exports.ArmatureId || (exports.ArmatureId = {}));
(function (ArmatureId) {
    ArmatureId.schema = {
        type: 'string',
        enum: Object.values(ArmatureId)
    };
    ArmatureId.validate = (0, validation_1.generateLazyValidator)(ArmatureId.schema);
})(ArmatureId = exports.ArmatureId || (exports.ArmatureId = {}));
var EmoteClip;
(function (EmoteClip) {
    EmoteClip.schema = {
        type: 'object',
        properties: {
            animation: {
                type: 'string',
                minLength: 1,
                errorMessage: 'animation must be a non-empty string (GLB clip name)'
            }
        },
        required: ['animation'],
        additionalProperties: false,
        errorMessage: {
            required: {
                animation: 'animation is required (GLB clip name)'
            }
        }
    };
    EmoteClip.validate = (0, validation_1.generateLazyValidator)(EmoteClip.schema);
})(EmoteClip = exports.EmoteClip || (exports.EmoteClip = {}));
var StartAnimation;
(function (StartAnimation) {
    StartAnimation.schema = {
        type: 'object',
        properties: {
            loop: {
                type: 'boolean',
                errorMessage: 'startAnimation.loop must be a boolean'
            },
            [ArmatureId.Armature]: Object.assign(Object.assign({}, EmoteClip.schema), { errorMessage: 'startAnimation.Armature is required and must contain valid animation data' }),
            [ArmatureId.Armature_Prop]: Object.assign(Object.assign({}, EmoteClip.schema), { nullable: true, errorMessage: 'startAnimation.Armature_Prop must contain valid animation data when provided' }),
            audio: {
                type: 'string',
                nullable: true,
                errorMessage: 'audio must be a string (audio clip filename) when provided'
            }
        },
        required: ['loop', ArmatureId.Armature],
        additionalProperties: false,
        errorMessage: {
            required: {
                loop: 'startAnimation.loop is required',
                [ArmatureId.Armature]: 'startAnimation.Armature is required'
            }
        }
    };
    StartAnimation.validate = (0, validation_1.generateLazyValidator)(StartAnimation.schema);
})(StartAnimation = exports.StartAnimation || (exports.StartAnimation = {}));
var OutcomeGroup;
(function (OutcomeGroup) {
    OutcomeGroup.schema = {
        type: 'object',
        properties: {
            title: {
                type: 'string',
                minLength: 1,
                errorMessage: 'outcome.title must be a non-empty string'
            },
            loop: {
                type: 'boolean',
                errorMessage: 'outcome.loop must be a boolean'
            },
            clips: {
                type: 'object',
                properties: Object.values(ArmatureId).reduce((properties, armature) => {
                    properties[armature] = Object.assign(Object.assign({}, EmoteClip.schema), { nullable: true, errorMessage: `outcome.clips.${armature} must contain valid animation data when provided` });
                    return properties;
                }, {}),
                additionalProperties: true,
                minProperties: 1,
                errorMessage: 'outcome.clips must contain at least one armature animation'
            },
            audio: {
                type: 'string',
                nullable: true,
                errorMessage: 'audio must be a string (audio clip filename) when provided'
            }
        },
        required: ['title', 'loop', 'clips'],
        additionalProperties: false,
        errorMessage: {
            required: {
                title: 'outcome.title is required',
                loop: 'outcome.loop is required',
                clips: 'outcome.clips is required'
            }
        }
    };
    OutcomeGroup.validate = (0, validation_1.generateLazyValidator)(OutcomeGroup.schema);
})(OutcomeGroup = exports.OutcomeGroup || (exports.OutcomeGroup = {}));
