// @dcl/protocol/out-ts/decentraland/sdk/components/gltf_node_modifiers.gen.ts
/* eslint-disable */
import _m0 from "protobufjs/minimal";
import { PBMaterial } from "./material.gen";

export const protobufPackage = "decentraland.sdk.components";

/**
 * GltfNodeModifiers component is to be used attached to entities that have the GltfContainer component.
 *
 * This allows to override either the Material or the Casting Shadows behaviour of the target GLTF Node.
 *
 * * If the 'path' of the first modifier in the collection is an empty string: the configuration will
 * affect all of the GLTF Nodes (as a global modifier).
 * * Otherwise, for the modifiers whose 'path' is found in the GLTF hierarchy, the modifier will affect only
 * the target Nodes.
 */
export interface PBGltfNodeModifiers {
  modifiers: PBGltfNodeModifiers_GltfNodeModifier[];
}

export interface PBGltfNodeModifiers_GltfNodeModifier {
  /** The GLTF hierarchy path of the target Node to be affected */
  path: string;
  /** The casting shadows enabled override */
  castShadows?:
    | boolean
    | undefined;
  /** The Material that will be overridden on the target Node */
  material?: PBMaterial | undefined;
}

function createBasePBGltfNodeModifiers(): PBGltfNodeModifiers {
  return { modifiers: [] };
}

export namespace PBGltfNodeModifiers {
  export function encode(message: PBGltfNodeModifiers, writer: _m0.Writer = _m0.Writer.create()): _m0.Writer {
    for (const v of message.modifiers) {
      PBGltfNodeModifiers_GltfNodeModifier.encode(v!, writer.uint32(10).fork()).ldelim();
    }
    return writer;
  }

  export function decode(input: _m0.Reader | Uint8Array, length?: number): PBGltfNodeModifiers {
    const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBasePBGltfNodeModifiers();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          if (tag !== 10) {
            break;
          }

          message.modifiers.push(PBGltfNodeModifiers_GltfNodeModifier.decode(reader, reader.uint32()));
          continue;
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skipType(tag & 7);
    }
    return message;
  }

  export function fromJSON(object: any): PBGltfNodeModifiers {
    return {
      modifiers: Array.isArray(object?.modifiers)
        ? object.modifiers.map((e: any) => PBGltfNodeModifiers_GltfNodeModifier.fromJSON(e))
        : [],
    };
  }

  export function toJSON(message: PBGltfNodeModifiers): unknown {
    const obj: any = {};
    if (message.modifiers) {
      obj.modifiers = message.modifiers.map((e) => e ? PBGltfNodeModifiers_GltfNodeModifier.toJSON(e) : undefined);
    } else {
      obj.modifiers = [];
    }
    return obj;
  }

  export function create<I extends Exact<DeepPartial<PBGltfNodeModifiers>, I>>(base?: I): PBGltfNodeModifiers {
    return PBGltfNodeModifiers.fromPartial(base ?? {});
  }

  export function fromPartial<I extends Exact<DeepPartial<PBGltfNodeModifiers>, I>>(object: I): PBGltfNodeModifiers {
    const message = createBasePBGltfNodeModifiers();
    message.modifiers = object.modifiers?.map((e) => PBGltfNodeModifiers_GltfNodeModifier.fromPartial(e)) || [];
    return message;
  }
}

function createBasePBGltfNodeModifiers_GltfNodeModifier(): PBGltfNodeModifiers_GltfNodeModifier {
  return { path: "", castShadows: undefined, material: undefined };
}

export namespace PBGltfNodeModifiers_GltfNodeModifier {
  export function encode(
    message: PBGltfNodeModifiers_GltfNodeModifier,
    writer: _m0.Writer = _m0.Writer.create(),
  ): _m0.Writer {
    if (message.path !== "") {
      writer.uint32(10).string(message.path);
    }
    if (message.castShadows !== undefined) {
      writer.uint32(16).bool(message.castShadows);
    }
    if (message.material !== undefined) {
      PBMaterial.encode(message.material, writer.uint32(26).fork()).ldelim();
    }
    return writer;
  }

  export function decode(input: _m0.Reader | Uint8Array, length?: number): PBGltfNodeModifiers_GltfNodeModifier {
    const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBasePBGltfNodeModifiers_GltfNodeModifier();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          if (tag !== 10) {
            break;
          }

          message.path = reader.string();
          continue;
        case 2:
          if (tag !== 16) {
            break;
          }

          message.castShadows = reader.bool();
          continue;
        case 3:
          if (tag !== 26) {
            break;
          }

          message.material = PBMaterial.decode(reader, reader.uint32());
          continue;
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skipType(tag & 7);
    }
    return message;
  }

  export function fromJSON(object: any): PBGltfNodeModifiers_GltfNodeModifier {
    return {
      path: isSet(object.path) ? String(object.path) : "",
      castShadows: isSet(object.castShadows) ? Boolean(object.castShadows) : undefined,
      material: isSet(object.material) ? PBMaterial.fromJSON(object.material) : undefined,
    };
  }

  export function toJSON(message: PBGltfNodeModifiers_GltfNodeModifier): unknown {
    const obj: any = {};
    message.path !== undefined && (obj.path = message.path);
    message.castShadows !== undefined && (obj.castShadows = message.castShadows);
    message.material !== undefined &&
      (obj.material = message.material ? PBMaterial.toJSON(message.material) : undefined);
    return obj;
  }

  export function create<I extends Exact<DeepPartial<PBGltfNodeModifiers_GltfNodeModifier>, I>>(
    base?: I,
  ): PBGltfNodeModifiers_GltfNodeModifier {
    return PBGltfNodeModifiers_GltfNodeModifier.fromPartial(base ?? {});
  }

  export function fromPartial<I extends Exact<DeepPartial<PBGltfNodeModifiers_GltfNodeModifier>, I>>(
    object: I,
  ): PBGltfNodeModifiers_GltfNodeModifier {
    const message = createBasePBGltfNodeModifiers_GltfNodeModifier();
    message.path = object.path ?? "";
    message.castShadows = object.castShadows ?? undefined;
    message.material = (object.material !== undefined && object.material !== null)
      ? PBMaterial.fromPartial(object.material)
      : undefined;
    return message;
  }
}

type Builtin = Date | Function | Uint8Array | string | number | boolean | undefined;

export type DeepPartial<T> = T extends Builtin ? T
  : T extends Array<infer U> ? Array<DeepPartial<U>> : T extends ReadonlyArray<infer U> ? ReadonlyArray<DeepPartial<U>>
  : T extends { $case: string } ? { [K in keyof Omit<T, "$case">]?: DeepPartial<T[K]> } & { $case: T["$case"] }
  : T extends {} ? { [K in keyof T]?: DeepPartial<T[K]> }
  : Partial<T>;

type KeysOfUnion<T> = T extends T ? keyof T : never;
export type Exact<P, I extends P> = P extends Builtin ? P
  : P & { [K in keyof P]: Exact<P[K], I[K]> } & { [K in Exclude<keyof I, KeysOfUnion<P>>]: never };

function isSet(value: any): boolean {
  return value !== null && value !== undefined;
}
