// @dcl/protocol/out-ts/decentraland/sdk/components/visibility_component.gen.ts
/* eslint-disable */
import _m0 from "protobufjs/minimal";

export const protobufPackage = "decentraland.sdk.components";

export interface PBVisibilityComponent {
  /** default=true */
  visible?:
    | boolean
    | undefined;
  /**
   * Propagation follows certain rules:
   * - Any own visibility component takes priority
   * - If no own visibility component, visibility is determined by visibility of the nearest parent with propagate = true
   * - If no own component and no parent with propagate = true, visibility is "visible"
   * - Visibility is always updated, whenever hierarchy, parent component or own component changes
   */
  propagateToChildren?: boolean | undefined;
}

function createBasePBVisibilityComponent(): PBVisibilityComponent {
  return { visible: undefined, propagateToChildren: undefined };
}

export namespace PBVisibilityComponent {
  export function encode(message: PBVisibilityComponent, writer: _m0.Writer = _m0.Writer.create()): _m0.Writer {
    if (message.visible !== undefined) {
      writer.uint32(8).bool(message.visible);
    }
    if (message.propagateToChildren !== undefined) {
      writer.uint32(16).bool(message.propagateToChildren);
    }
    return writer;
  }

  export function decode(input: _m0.Reader | Uint8Array, length?: number): PBVisibilityComponent {
    const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBasePBVisibilityComponent();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          if (tag !== 8) {
            break;
          }

          message.visible = reader.bool();
          continue;
        case 2:
          if (tag !== 16) {
            break;
          }

          message.propagateToChildren = reader.bool();
          continue;
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skipType(tag & 7);
    }
    return message;
  }

  export function fromJSON(object: any): PBVisibilityComponent {
    return {
      visible: isSet(object.visible) ? Boolean(object.visible) : undefined,
      propagateToChildren: isSet(object.propagateToChildren) ? Boolean(object.propagateToChildren) : undefined,
    };
  }

  export function toJSON(message: PBVisibilityComponent): unknown {
    const obj: any = {};
    message.visible !== undefined && (obj.visible = message.visible);
    message.propagateToChildren !== undefined && (obj.propagateToChildren = message.propagateToChildren);
    return obj;
  }

  export function create<I extends Exact<DeepPartial<PBVisibilityComponent>, I>>(base?: I): PBVisibilityComponent {
    return PBVisibilityComponent.fromPartial(base ?? {});
  }

  export function fromPartial<I extends Exact<DeepPartial<PBVisibilityComponent>, I>>(
    object: I,
  ): PBVisibilityComponent {
    const message = createBasePBVisibilityComponent();
    message.visible = object.visible ?? undefined;
    message.propagateToChildren = object.propagateToChildren ?? undefined;
    return message;
  }
}

type Builtin = Date | Function | Uint8Array | string | number | boolean | undefined;

export type DeepPartial<T> = T extends Builtin ? T
  : T extends Array<infer U> ? Array<DeepPartial<U>> : T extends ReadonlyArray<infer U> ? ReadonlyArray<DeepPartial<U>>
  : T extends { $case: string } ? { [K in keyof Omit<T, "$case">]?: DeepPartial<T[K]> } & { $case: T["$case"] }
  : T extends {} ? { [K in keyof T]?: DeepPartial<T[K]> }
  : Partial<T>;

type KeysOfUnion<T> = T extends T ? keyof T : never;
export type Exact<P, I extends P> = P extends Builtin ? P
  : P & { [K in keyof P]: Exact<P[K], I[K]> } & { [K in Exclude<keyof I, KeysOfUnion<P>>]: never };

function isSet(value: any): boolean {
  return value !== null && value !== undefined;
}
