// @dcl/asset-packs/dist/admin-toolkit-ui/ModerationControl/api.js
import { getDomain, wrapSignedFetch } from '../fetch-utils';
const URLS = () => ({
    SCENE_ADMIN: `https://comms-gatekeeper.decentraland.${getDomain()}/scene-admin`,
    SCENE_BAN: `https://comms-gatekeeper.decentraland.${getDomain()}/scene-bans`,
});
export async function getSceneAdmins() {
    return wrapSignedFetch({ url: URLS().SCENE_ADMIN });
}
export async function postSceneAdmin(adminData) {
    return wrapSignedFetch({
        url: URLS().SCENE_ADMIN,
        init: {
            method: 'POST',
            headers: {},
            body: JSON.stringify(adminData),
        },
    });
}
export async function deleteSceneAdmin(address) {
    return wrapSignedFetch({
        url: URLS().SCENE_ADMIN,
        init: {
            method: 'DELETE',
            headers: {},
            body: JSON.stringify({ admin: address }),
        },
    });
}
export async function postSceneBan(banData) {
    return wrapSignedFetch({
        url: URLS().SCENE_BAN,
        init: {
            method: 'POST',
            headers: {},
            body: JSON.stringify(banData),
        },
    });
}
export async function getSceneBans() {
    return wrapSignedFetch({
        url: URLS().SCENE_BAN,
    });
}
export async function deleteSceneBan(address) {
    return wrapSignedFetch({
        url: URLS().SCENE_BAN,
        init: {
            method: 'DELETE',
            headers: {},
            body: JSON.stringify({ banned_address: address }),
        },
    });
}
