// @dcl/asset-packs/dist/admin-toolkit-ui/ModerationControl/index.js
import ReactEcs, { UiEntity } from '@dcl/react-ecs';
import { Header } from '../Header';
import { AddUserInput, PermissionType } from './AddUserInput';
import { Button } from '../Button';
import { Card } from '../Card';
import { CONTENT_URL } from '../constants';
import { fetchSceneBans } from '..';
import { getModerationControlStyles, getModerationControlColors, } from './styles/ModerationControlStyles';
export const BTN_MODERATION_CONTROL = `${CONTENT_URL}/admin_toolkit/assets/icons/admin-panel-moderation-control-button.png`;
export const MODERATION_CONTROL_ICON = `${CONTENT_URL}/admin_toolkit/assets/icons/moderation-control-icon.png`;
const VERIFIED_USER_ICON = `${CONTENT_URL}/admin_toolkit/assets/icons/admin-panel-verified-user.png`;
const BAN_USER_ICON = `${CONTENT_URL}/admin_toolkit/assets/icons/ban.png`;
export const moderationControlState = {
    showModalAdminList: false,
    showModalBanList: false,
    adminToRemove: undefined,
    unbanMessage: null,
};
export function ModerationControl({ engine, player, sceneAdmins }) {
    const styles = getModerationControlStyles();
    const colors = getModerationControlColors();
    return (ReactEcs.createElement(Card, null,
        ReactEcs.createElement(UiEntity, { uiTransform: styles.container },
            ReactEcs.createElement(Header, { iconSrc: MODERATION_CONTROL_ICON, title: "PERMISSIONS & MODERATION" }),
            ReactEcs.createElement(AddUserInput, { type: PermissionType.ADMIN, sceneAdmins: sceneAdmins }),
            ReactEcs.createElement(Button, { variant: "secondary", id: "moderation_control_admin_list", value: "<b>View Admin List</b>", fontSize: 18, color: colors.white, uiTransform: styles.viewListButton, icon: VERIFIED_USER_ICON, iconTransform: styles.viewListIcon, onMouseDown: () => (moderationControlState.showModalAdminList = true) }),
            ReactEcs.createElement(UiEntity, { uiTransform: styles.divider }),
            ReactEcs.createElement(AddUserInput, { type: PermissionType.BAN, sceneAdmins: sceneAdmins }),
            ReactEcs.createElement(Button, { variant: "secondary", id: "moderation_control_ban_list", value: "<b>View Ban List</b>", fontSize: 18, color: colors.white, uiTransform: styles.viewListButton, icon: BAN_USER_ICON, iconTransform: styles.viewListIcon, onMouseDown: async () => {
                    await fetchSceneBans();
                    moderationControlState.showModalBanList = true;
                } }))));
}
