// @dcl/asset-packs/dist/admin-toolkit-ui/ModerationControl/UsersList.js
import { Color4 } from '@dcl/ecs-math';
import ReactEcs, { UiEntity, Label } from '@dcl/react-ecs';
import { Button } from '../Button';
import { RemoveAdminConfirmation } from './RemoveAdminConfirmation';
import { moderationControlState } from '.';
import { CONTENT_URL } from '../constants';
import { getModalStyles, getModalBackgrounds, getModalColors, getPaginationColor, } from './styles/UsersListStyles';
import { handleUnbanUser } from './utils';
import { clearInterval, setInterval } from '../utils';
export var UserListType;
(function (UserListType) {
    UserListType["ADMIN"] = "admin";
    UserListType["BAN"] = "ban";
})(UserListType || (UserListType = {}));
const USERS_PER_PAGE = 5;
const ICONS = {
    BACK: `${CONTENT_URL}/admin_toolkit/assets/icons/chevron-back.png`,
    NEXT: `${CONTENT_URL}/admin_toolkit/assets/icons/chevron-forward.png`,
    CLOSE: `${CONTENT_URL}/admin_toolkit/assets/icons/close.png`,
};
const getUserKey = (user) => {
    return 'address' in user ? user.address : user.bannedAddress;
};
const getUserAddress = (user) => {
    return 'address' in user ? user.address : user.bannedAddress;
};
const canUserBeRemoved = (user) => {
    return 'canBeRemoved' in user ? user.canBeRemoved !== false : true;
};
const getModalTitle = (type) => {
    return type === UserListType.ADMIN ? '<b>ADMIN LIST</b>' : '<b>SCENE BAN LIST</b>';
};
const getCounterText = (type, count) => {
    const itemType = type === UserListType.ADMIN
        ? count === 1
            ? 'admin'
            : 'admins'
        : count === 1
            ? 'user'
            : 'users';
    return `(${count} ${itemType})`;
};
const getActionButtonText = (type) => {
    return type === UserListType.ADMIN ? '<b>Remove</b>' : '<b>Unban</b>';
};
const closeModal = (type) => {
    if (type === UserListType.ADMIN) {
        moderationControlState.showModalAdminList = false;
    }
    else {
        moderationControlState.showModalBanList = false;
    }
};
export function ModalUserList({ users, engine, type }) {
    const [page, setPage] = ReactEcs.useState(1);
    const styles = getModalStyles();
    const backgrounds = getModalBackgrounds();
    const colors = getModalColors();
    const handleRemoveUser = async (user) => {
        if (type === UserListType.ADMIN) {
            moderationControlState.adminToRemove = user;
        }
        else {
            const bannedUser = user;
            const success = await handleUnbanUser(bannedUser.bannedAddress);
            if (success) {
                const username = bannedUser.name || bannedUser.bannedAddress;
                moderationControlState.unbanMessage = `${username} has been unbanned from your scene`;
            }
            else {
                moderationControlState.unbanMessage = 'We were unable to unban this user';
            }
        }
    };
    ReactEcs.useEffect(() => {
        if (moderationControlState.unbanMessage) {
            let counter = 0;
            const interval = setInterval(engine, () => {
                counter += 100;
                if (counter >= 3000) {
                    moderationControlState.unbanMessage = null;
                }
            }, 100);
            return () => clearInterval(engine, interval);
        }
    }, [moderationControlState.unbanMessage]);
    if (moderationControlState.adminToRemove) {
        return (ReactEcs.createElement(RemoveAdminConfirmation, { admin: moderationControlState.adminToRemove, engine: engine }));
    }
    const startIndex = (page - 1) * USERS_PER_PAGE;
    const endIndex = Math.min(startIndex + USERS_PER_PAGE, users.length);
    const currentPageUsers = users.slice(startIndex, endIndex);
    return (ReactEcs.createElement(UiEntity, { uiTransform: styles.overlay },
        ReactEcs.createElement(UiEntity, { uiTransform: styles.container, uiBackground: backgrounds.container },
            ReactEcs.createElement(UiEntity, { uiTransform: styles.content },
                ReactEcs.createElement(UiEntity, { uiTransform: styles.header },
                    ReactEcs.createElement(UiEntity, { uiTransform: styles.headerIcon, uiBackground: type === UserListType.BAN ? backgrounds.banIcon : backgrounds.headerIcon }),
                    ReactEcs.createElement(Label, { value: getModalTitle(type), fontSize: 24, color: colors.white }),
                    ReactEcs.createElement(Label, { value: getCounterText(type, users.length), fontSize: 16, color: colors.gray, uiTransform: styles.usersCount }),
                    ReactEcs.createElement(Button, { id: "close-modal", onlyIcon: true, icon: ICONS.CLOSE, variant: "secondary", fontSize: 20, uiTransform: styles.closeButton, iconTransform: styles.closeIcon, onMouseDown: () => closeModal(type) })),
                ReactEcs.createElement(UiEntity, { uiTransform: styles.listContainer }, currentPageUsers.map((user, index) => (ReactEcs.createElement(UiEntity, { key: getUserKey(user), uiTransform: styles.userItem },
                    ReactEcs.createElement(UiEntity, { key: `${type}-${user.name || getUserAddress(user)}`, uiTransform: styles.userRow },
                        ReactEcs.createElement(UiEntity, { uiTransform: styles.userInfo },
                            ReactEcs.createElement(UiEntity, { uiTransform: styles.personIconContainer },
                                ReactEcs.createElement(UiEntity, { uiTransform: styles.personIcon, uiBackground: backgrounds.personIcon })),
                            ReactEcs.createElement(UiEntity, { uiTransform: styles.userDetails },
                                user.name && (ReactEcs.createElement(UiEntity, { uiTransform: styles.nameContainer },
                                    ReactEcs.createElement(Label, { value: `<b>${user.name}</b>`, fontSize: 14, color: colors.white }),
                                    !user.name.includes('#') && (ReactEcs.createElement(UiEntity, { uiTransform: styles.verifiedIcon, uiBackground: backgrounds.verifiedIcon })),
                                    'role' in user &&
                                        (user.role === 'owner' || user.role === 'operator') && (ReactEcs.createElement(UiEntity, { uiTransform: styles.roleBadge, uiBackground: backgrounds.roleBadge },
                                        ReactEcs.createElement(Label, { value: `<b>${('role' in user ? (user.role ?? '') : '')
                                                ?.charAt(0)
                                                .toUpperCase() + ('role' in user ? user.role?.slice(1) : '')}</b>`, fontSize: 12, color: colors.black }))))),
                                ReactEcs.createElement(Label, { fontSize: user.name ? 12 : 14, value: user.name ? getUserAddress(user) : getUserAddress(user), color: user.name ? colors.addressGray : colors.white }))),
                        (canUserBeRemoved(user) || type === UserListType.BAN) && (ReactEcs.createElement(Button, { id: `${type}-action-${index}`, value: getActionButtonText(type), variant: "text", fontSize: 14, color: colors.removeRed, labelTransform: styles.removeButton, onMouseDown: () => handleRemoveUser(user) }))),
                    ReactEcs.createElement(UiEntity, { uiTransform: styles.divider, uiBackground: backgrounds.divider })))))),
            users.length > USERS_PER_PAGE && (ReactEcs.createElement(UiEntity, { uiTransform: styles.pagination },
                ReactEcs.createElement(Button, { id: "prev", value: "Prev", variant: "secondary", disabled: page <= 1, fontSize: 18, icon: ICONS.BACK, iconTransform: styles.prevIcon, iconBackground: { color: getPaginationColor(page <= 1) }, color: getPaginationColor(page <= 1), labelTransform: styles.prevLabel, uiTransform: styles.paginationButton, onMouseDown: () => setPage(page - 1) }),
                ReactEcs.createElement(Label, { value: `${page} / ${Math.ceil(users.length / USERS_PER_PAGE)}`, fontSize: 14, color: colors.white }),
                ReactEcs.createElement(Button, { id: "next", value: "<b>Next</b>", variant: "secondary", fontSize: 18, iconRight: ICONS.NEXT, iconRightTransform: styles.nextIcon, labelTransform: styles.nextLabel, iconRightBackground: {
                        color: getPaginationColor(page >= Math.ceil(users.length / USERS_PER_PAGE)),
                    }, color: getPaginationColor(page >= Math.ceil(users.length / USERS_PER_PAGE)), disabled: page >= Math.ceil(users.length / USERS_PER_PAGE), uiTransform: styles.paginationButton, onMouseDown: () => setPage(page + 1) }))),
            type === UserListType.BAN && moderationControlState.unbanMessage && (ReactEcs.createElement(UiEntity, { uiTransform: styles.messageContainer },
                ReactEcs.createElement(Label, { value: moderationControlState.unbanMessage, fontSize: 14, color: Color4.White(), uiTransform: styles.messageLabel }))))));
}
