// @dcl/protocol/out-js/decentraland/sdk/components/audio_stream.gen.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PBAudioStream = exports.protobufPackage = void 0;
/* eslint-disable */
const minimal_1 = __importDefault(require("protobufjs/minimal"));
exports.protobufPackage = "decentraland.sdk.components";
function createBasePBAudioStream() {
    return {
        playing: undefined,
        volume: undefined,
        url: "",
        spatial: undefined,
        spatialMinDistance: undefined,
        spatialMaxDistance: undefined,
    };
}
var PBAudioStream;
(function (PBAudioStream) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.playing !== undefined) {
            writer.uint32(8).bool(message.playing);
        }
        if (message.volume !== undefined) {
            writer.uint32(21).float(message.volume);
        }
        if (message.url !== "") {
            writer.uint32(26).string(message.url);
        }
        if (message.spatial !== undefined) {
            writer.uint32(32).bool(message.spatial);
        }
        if (message.spatialMinDistance !== undefined) {
            writer.uint32(45).float(message.spatialMinDistance);
        }
        if (message.spatialMaxDistance !== undefined) {
            writer.uint32(53).float(message.spatialMaxDistance);
        }
        return writer;
    }
    PBAudioStream.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePBAudioStream();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 8) {
                        break;
                    }
                    message.playing = reader.bool();
                    continue;
                case 2:
                    if (tag !== 21) {
                        break;
                    }
                    message.volume = reader.float();
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.url = reader.string();
                    continue;
                case 4:
                    if (tag !== 32) {
                        break;
                    }
                    message.spatial = reader.bool();
                    continue;
                case 5:
                    if (tag !== 45) {
                        break;
                    }
                    message.spatialMinDistance = reader.float();
                    continue;
                case 6:
                    if (tag !== 53) {
                        break;
                    }
                    message.spatialMaxDistance = reader.float();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    PBAudioStream.decode = decode;
    function fromJSON(object) {
        return {
            playing: isSet(object.playing) ? Boolean(object.playing) : undefined,
            volume: isSet(object.volume) ? Number(object.volume) : undefined,
            url: isSet(object.url) ? String(object.url) : "",
            spatial: isSet(object.spatial) ? Boolean(object.spatial) : undefined,
            spatialMinDistance: isSet(object.spatialMinDistance) ? Number(object.spatialMinDistance) : undefined,
            spatialMaxDistance: isSet(object.spatialMaxDistance) ? Number(object.spatialMaxDistance) : undefined,
        };
    }
    PBAudioStream.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.playing !== undefined && (obj.playing = message.playing);
        message.volume !== undefined && (obj.volume = message.volume);
        message.url !== undefined && (obj.url = message.url);
        message.spatial !== undefined && (obj.spatial = message.spatial);
        message.spatialMinDistance !== undefined && (obj.spatialMinDistance = message.spatialMinDistance);
        message.spatialMaxDistance !== undefined && (obj.spatialMaxDistance = message.spatialMaxDistance);
        return obj;
    }
    PBAudioStream.toJSON = toJSON;
    function create(base) {
        return PBAudioStream.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    PBAudioStream.create = create;
    function fromPartial(object) {
        var _a, _b, _c, _d, _e, _f;
        const message = createBasePBAudioStream();
        message.playing = (_a = object.playing) !== null && _a !== void 0 ? _a : undefined;
        message.volume = (_b = object.volume) !== null && _b !== void 0 ? _b : undefined;
        message.url = (_c = object.url) !== null && _c !== void 0 ? _c : "";
        message.spatial = (_d = object.spatial) !== null && _d !== void 0 ? _d : undefined;
        message.spatialMinDistance = (_e = object.spatialMinDistance) !== null && _e !== void 0 ? _e : undefined;
        message.spatialMaxDistance = (_f = object.spatialMaxDistance) !== null && _f !== void 0 ? _f : undefined;
        return message;
    }
    PBAudioStream.fromPartial = fromPartial;
})(PBAudioStream || (exports.PBAudioStream = PBAudioStream = {}));
function isSet(value) {
    return value !== null && value !== undefined;
}
