// @dcl/protocol/out-js/decentraland/sdk/components/gltf_node_modifiers.gen.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PBGltfNodeModifiers_GltfNodeModifier = exports.PBGltfNodeModifiers = exports.protobufPackage = void 0;
/* eslint-disable */
const minimal_1 = __importDefault(require("protobufjs/minimal"));
const material_gen_1 = require("./material.gen");
exports.protobufPackage = "decentraland.sdk.components";
function createBasePBGltfNodeModifiers() {
    return { modifiers: [] };
}
var PBGltfNodeModifiers;
(function (PBGltfNodeModifiers) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        for (const v of message.modifiers) {
            PBGltfNodeModifiers_GltfNodeModifier.encode(v, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    }
    PBGltfNodeModifiers.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePBGltfNodeModifiers();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.modifiers.push(PBGltfNodeModifiers_GltfNodeModifier.decode(reader, reader.uint32()));
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    PBGltfNodeModifiers.decode = decode;
    function fromJSON(object) {
        return {
            modifiers: Array.isArray(object === null || object === void 0 ? void 0 : object.modifiers)
                ? object.modifiers.map((e) => PBGltfNodeModifiers_GltfNodeModifier.fromJSON(e))
                : [],
        };
    }
    PBGltfNodeModifiers.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        if (message.modifiers) {
            obj.modifiers = message.modifiers.map((e) => e ? PBGltfNodeModifiers_GltfNodeModifier.toJSON(e) : undefined);
        }
        else {
            obj.modifiers = [];
        }
        return obj;
    }
    PBGltfNodeModifiers.toJSON = toJSON;
    function create(base) {
        return PBGltfNodeModifiers.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    PBGltfNodeModifiers.create = create;
    function fromPartial(object) {
        var _a;
        const message = createBasePBGltfNodeModifiers();
        message.modifiers = ((_a = object.modifiers) === null || _a === void 0 ? void 0 : _a.map((e) => PBGltfNodeModifiers_GltfNodeModifier.fromPartial(e))) || [];
        return message;
    }
    PBGltfNodeModifiers.fromPartial = fromPartial;
})(PBGltfNodeModifiers || (exports.PBGltfNodeModifiers = PBGltfNodeModifiers = {}));
function createBasePBGltfNodeModifiers_GltfNodeModifier() {
    return { path: "", castShadows: undefined, material: undefined };
}
var PBGltfNodeModifiers_GltfNodeModifier;
(function (PBGltfNodeModifiers_GltfNodeModifier) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.path !== "") {
            writer.uint32(10).string(message.path);
        }
        if (message.castShadows !== undefined) {
            writer.uint32(16).bool(message.castShadows);
        }
        if (message.material !== undefined) {
            material_gen_1.PBMaterial.encode(message.material, writer.uint32(26).fork()).ldelim();
        }
        return writer;
    }
    PBGltfNodeModifiers_GltfNodeModifier.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePBGltfNodeModifiers_GltfNodeModifier();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.path = reader.string();
                    continue;
                case 2:
                    if (tag !== 16) {
                        break;
                    }
                    message.castShadows = reader.bool();
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.material = material_gen_1.PBMaterial.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    PBGltfNodeModifiers_GltfNodeModifier.decode = decode;
    function fromJSON(object) {
        return {
            path: isSet(object.path) ? String(object.path) : "",
            castShadows: isSet(object.castShadows) ? Boolean(object.castShadows) : undefined,
            material: isSet(object.material) ? material_gen_1.PBMaterial.fromJSON(object.material) : undefined,
        };
    }
    PBGltfNodeModifiers_GltfNodeModifier.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.path !== undefined && (obj.path = message.path);
        message.castShadows !== undefined && (obj.castShadows = message.castShadows);
        message.material !== undefined &&
            (obj.material = message.material ? material_gen_1.PBMaterial.toJSON(message.material) : undefined);
        return obj;
    }
    PBGltfNodeModifiers_GltfNodeModifier.toJSON = toJSON;
    function create(base) {
        return PBGltfNodeModifiers_GltfNodeModifier.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    PBGltfNodeModifiers_GltfNodeModifier.create = create;
    function fromPartial(object) {
        var _a, _b;
        const message = createBasePBGltfNodeModifiers_GltfNodeModifier();
        message.path = (_a = object.path) !== null && _a !== void 0 ? _a : "";
        message.castShadows = (_b = object.castShadows) !== null && _b !== void 0 ? _b : undefined;
        message.material = (object.material !== undefined && object.material !== null)
            ? material_gen_1.PBMaterial.fromPartial(object.material)
            : undefined;
        return message;
    }
    PBGltfNodeModifiers_GltfNodeModifier.fromPartial = fromPartial;
})(PBGltfNodeModifiers_GltfNodeModifier || (exports.PBGltfNodeModifiers_GltfNodeModifier = PBGltfNodeModifiers_GltfNodeModifier = {}));
function isSet(value) {
    return value !== null && value !== undefined;
}
