// @dcl/sdk-commands/dist/commands/code-to-composite/code-commenter.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.commentEntrypoint = commentEntrypoint;
exports.commentSourceFiles = commentSourceFiles;
const path_1 = __importDefault(require("path"));
const glob_1 = require("glob");
const COMMENT_MARKER = 'COMMENTED BY "@dcl/sdk-commands code-to-composite" COMMAND';
/**
 * Checks if a file has already been commented by this command
 */
function isAlreadyCommented(content) {
    return content.includes(COMMENT_MARKER);
}
/**
 * Generates the comment header for migrated files
 */
function generateCommentHeader(content) {
    // use line-by-line commenting to avoid issues with existing /* */ comments
    const commentedLines = content
        .split('\n')
        .map((line) => `// ${line}`)
        .join('\n');
    return `// ============================================================================
// ${COMMENT_MARKER}
//
// This code has been migrated to main.composite and main.crdt files.
// The scene can now be edited visually using the Creator Hub app.
//
// To restore code-based workflow: Uncomment the code below
// For hybrid workflow: Uncomment specific parts you want to keep dynamic
//
// Learn more: https://docs.decentraland.org/creator
// ============================================================================
//
${commentedLines}`;
}
/**
 * Replaces the entrypoint file with a stub main() function
 */
async function commentEntrypoint({ fs }, entrypointPath) {
    const stubContent = `// ${COMMENT_MARKER}
// Scene content is managed by Creator Hub (main.composite)

export function main() {
  // Add dynamic code, systems, or event handlers here
}
`;
    await fs.writeFile(entrypointPath, stubContent);
}
/**
 * Comments out a single source file
 */
async function commentSourceFile({ fs }, filePath) {
    const currentContent = await fs.readFile(filePath, 'utf-8');
    if (isAlreadyCommented(currentContent)) {
        return;
    }
    const commentedContent = generateCommentHeader(currentContent) + '\n';
    await fs.writeFile(filePath, commentedContent);
}
/**
 * Comments out the entrypoint and all source files in the src directory
 *
 * NOTE: Need to comment out all the src files otherwise typechecker will fail preventing deploys, etc.
 */
async function commentSourceFiles(components, sceneCodeEntrypoint, bundleEntrypoint) {
    const normalizedBundleEntrypoint = path_1.default.normalize(bundleEntrypoint);
    const normalizedSceneCodeEntrypoint = path_1.default.normalize(sceneCodeEntrypoint);
    const srcDir = path_1.default.dirname(normalizedSceneCodeEntrypoint);
    const sourceFiles = (0, glob_1.globSync)('**/*.{ts,tsx,js,jsx}', {
        cwd: srcDir,
        absolute: true,
        ignore: ['**/*.d.ts', '**/node_modules/**']
    });
    const filesToComment = sourceFiles.filter((file) => {
        const normalized = path_1.default.normalize(file);
        return normalized !== normalizedBundleEntrypoint && normalized !== normalizedSceneCodeEntrypoint;
    });
    await Promise.all([
        commentEntrypoint(components, normalizedBundleEntrypoint),
        commentSourceFile(components, normalizedSceneCodeEntrypoint),
        ...filesToComment.map((file) => commentSourceFile(components, file))
    ]);
    return filesToComment.length;
}
