// @dcl/sdk-commands/dist/commands/code-to-composite/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.args = void 0;
exports.help = help;
exports.main = main;
const path_1 = __importDefault(require("path"));
const prompts_1 = __importDefault(require("prompts"));
const args_1 = require("../../logic/args");
const workspace_validations_1 = require("../../logic/workspace-validations");
const beautiful_logs_1 = require("../../logic/beautiful-logs");
const scene_executor_1 = require("./scene-executor");
const composite_generator_1 = require("./composite-generator");
const code_commenter_1 = require("./code-commenter");
const asset_migrator_1 = require("./asset-migrator");
const name_generator_1 = require("./name-generator");
exports.args = (0, args_1.declareArgs)({
    '--dir': String,
    '--help': Boolean,
    '-h': '--help',
    '--yes': Boolean,
    '-y': '--yes'
});
function help(options) {
    options.components.logger.log(`
  Usage: 'sdk-commands code-to-composite [options]'

  Description:

    Migrates a code-only scene to Creator Hub-compatible format by generating
    composite/crdt files and commenting out the original code.

  Options:

    -h, --help                Displays complete help
    --dir                     Path to directory to export
    -y, --yes                 Skip the confirmation prompt

  Example:

    - Export a code-only scene to Creator Hub format:
      '$ sdk-commands code-to-composite'

    - Export a specific directory:
      '$ sdk-commands code-to-composite --dir ./my-scene'
  `);
}
async function main(options) {
    const workingDirectory = path_1.default.resolve(process.cwd(), options.args['--dir'] || '.');
    const workspace = await (0, workspace_validations_1.getValidWorkspace)(options.components, workingDirectory);
    const shouldContinue = options.args['--yes'] ? true : await promptForDestructiveAction(options.components);
    if (!shouldContinue)
        return;
    const MAX_STEPS = 5;
    for (const project of workspace.projects) {
        (0, beautiful_logs_1.printCurrentProjectStarting)(options.components.logger, project, workspace);
        if (project.kind === 'scene') {
            await exportSceneToCrdt(options, project, MAX_STEPS);
        }
        else {
            options.components.logger.warn(`Project ${project.workingDirectory} is not a scene, skipping...`);
        }
    }
}
async function promptForDestructiveAction(components) {
    const { logger } = components;
    (0, beautiful_logs_1.printWarning)(logger, 'This command will modify your scene code by commenting it out and creating composite/CRDT files.');
    logger.log('  - Your original code will be backed up with a .backup.ts extension');
    logger.log('  - The scene will be converted to Creator Hub format');
    logger.log('  - Files will be moved around');
    logger.log('  - If you are not using any version control system, undoing this operation will be REALLY cumbersome');
    logger.log('');
    let cancelled = false;
    const onCancel = () => {
        cancelled = true;
    };
    const { confirmed } = await (0, prompts_1.default)({
        type: 'confirm',
        name: 'confirmed',
        message: 'Do you want to continue?',
        initial: false
    }, { onCancel });
    if (cancelled || !confirmed) {
        logger.log('Operation cancelled by user.');
        return false;
    }
    return true;
}
async function exportSceneToCrdt(options, project, maxSteps) {
    const { logger } = options.components;
    let currentStep = 1;
    const compositeFilePath = path_1.default.join(project.workingDirectory, 'assets', 'scene', 'main.composite');
    const crdtFilePath = path_1.default.join(project.workingDirectory, 'main.crdt');
    const entityNamesFilePath = path_1.default.join(project.workingDirectory, 'src', 'entity-names.ts');
    // Step 1: execute scene code to populate engine
    (0, beautiful_logs_1.printProgressStep)(logger, 'Executing scene code to capture state', currentStep++, maxSteps);
    const { engine, sceneCodeEntrypoint } = await (0, scene_executor_1.executeSceneCode)(options.components, project, crdtFilePath);
    (0, beautiful_logs_1.printProgressInfo)(logger, `Engine state captured successfully`);
    // Step 2: migrate assets (before generating composite/crdt)
    (0, beautiful_logs_1.printProgressStep)(logger, 'Organizing assets for Creator Hub', currentStep++, maxSteps);
    const updatedCount = await (0, asset_migrator_1.migrateAssets)(options.components, project, engine);
    (0, beautiful_logs_1.printProgressInfo)(logger, `Migrated ${updatedCount} asset file(s) successfully`);
    // Step 3: generate entity names
    (0, beautiful_logs_1.printProgressStep)(logger, 'Generating names for entities', currentStep++, maxSteps);
    const namedEntitiesCount = await (0, name_generator_1.generateEntityNames)(options.components, engine);
    (0, beautiful_logs_1.printProgressInfo)(logger, `Generated names for ${namedEntitiesCount} entity/entities`);
    // Step 4: generate composite and crdt files (with updated asset paths and names)
    (0, beautiful_logs_1.printProgressStep)(logger, 'Generating composite and crdt files', currentStep++, maxSteps);
    await (0, composite_generator_1.generateCompositeFiles)(options.components, engine, compositeFilePath, crdtFilePath, entityNamesFilePath);
    (0, beautiful_logs_1.printProgressInfo)(logger, `Generated main.composite ✅ (${compositeFilePath})`);
    (0, beautiful_logs_1.printProgressInfo)(logger, `Generated main.crdt ✅ (${crdtFilePath})`);
    (0, beautiful_logs_1.printProgressInfo)(logger, `Generated entity-names.ts ✅ (${entityNamesFilePath})`);
    // Step 5: comment out original code
    (0, beautiful_logs_1.printProgressStep)(logger, 'Commenting out original code', currentStep++, maxSteps);
    const bundleEntrypoint = path_1.default.join(project.workingDirectory, project.scene.main);
    const commentedFiles = await (0, code_commenter_1.commentSourceFiles)(options.components, sceneCodeEntrypoint, bundleEntrypoint);
    const relativeEntrypoint = path_1.default.relative(project.workingDirectory, sceneCodeEntrypoint);
    (0, beautiful_logs_1.printProgressInfo)(logger, `Commented out bundle entrypoint ${project.scene.main} and added stub main() function`);
    (0, beautiful_logs_1.printProgressInfo)(logger, `Commented out scene code entrypoint ${relativeEntrypoint}`);
    if (commentedFiles > 0) {
        (0, beautiful_logs_1.printProgressInfo)(logger, `Commented out ${commentedFiles} additional source file(s)`);
    }
    (0, beautiful_logs_1.printSuccess)(logger, 'Scene successfully migrated to Creator Hub format!', `\nNext steps:
  - Open your scene in the Creator Hub to edit visually
  - Uncomment parts of the code in src/ for hybrid workflows
  - Use "src/entity-names.ts" to reference entities by name in code`);
}
