// @dcl/sdk-commands/dist/commands/code-to-composite/name-generator.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateEntityNames = generateEntityNames;
const path_1 = __importDefault(require("path"));
const dist_cjs_1 = require("@dcl/ecs/dist-cjs");
function capitalize(string) {
    return string.charAt(0).toUpperCase() + string.slice(1);
}
/**
 * Component priority list for name generation
 */
function getNameGenerators() {
    return [
        {
            component: dist_cjs_1.GltfContainer,
            getName: (data) => {
                if (data?.src) {
                    const filename = path_1.default.basename(data.src);
                    return path_1.default.parse(filename).name;
                }
                return null;
            }
        },
        {
            component: dist_cjs_1.MeshRenderer,
            getName: (data) => {
                if (data?.mesh?.['$case']) {
                    const meshType = data.mesh['$case'];
                    return meshType;
                }
                return 'Mesh';
            }
        },
        {
            component: dist_cjs_1.TextShape,
            getName: (data) => {
                if (data?.text) {
                    const text = String(data.text).trim();
                    const ellipsed = text.length > 10 ? text.substring(0, 10) + '...' : text;
                    return `Text: ${ellipsed}`;
                }
                return 'Text';
            }
        },
        {
            component: dist_cjs_1.VideoPlayer,
            getName: (data) => {
                if (data?.src) {
                    const filename = path_1.default.basename(data.src);
                    return path_1.default.parse(filename).name;
                }
                return 'Video';
            }
        },
        {
            component: dist_cjs_1.NftShape,
            getName: () => 'NFT'
        },
        {
            component: dist_cjs_1.LightSource,
            getName: () => 'Light'
        },
        {
            component: dist_cjs_1.AudioSource,
            getName: (data) => {
                if (data?.audioClipUrl) {
                    const filename = path_1.default.basename(data.audioClipUrl);
                    return path_1.default.parse(filename).name;
                }
                return 'Audio';
            }
        },
        {
            component: dist_cjs_1.MeshCollider,
            getName: (data) => {
                const pre = 'Mesh Collider';
                if (data?.mesh?.['$case']) {
                    const meshType = data.mesh['$case'];
                    return `${pre}: ${meshType}`;
                }
                return pre;
            }
        }
    ];
}
/**
 * Generates a name for an entity based on a specific component
 */
function generateNameFromComponent(generator, entity) {
    if (!generator.component.has(entity) || !('getOrNull' in generator.component)) {
        return null;
    }
    const componentData = generator.component.getOrNull(entity);
    const name = generator.getName(componentData);
    return name ? capitalize(name) : null;
}
/**
 * Generates Name components for entities that don't have them
 *
 * Automatically generates Name components for entities based on their other components.
 * This makes entities easier to identify and work with in the Creator Hub.
 */
async function generateEntityNames(components, engine) {
    const { logger } = components;
    let createdCount = 0;
    const processedEntities = new Set();
    // iterate through entities with nameable components in priority order
    for (const generator of getNameGenerators()) {
        for (const [entity] of engine.getEntitiesWith(generator.component)) {
            // skip if already processed (entity may have multiple "nameable" components)
            if (processedEntities.has(entity))
                continue;
            processedEntities.add(entity);
            // skip reserved entities
            if (entity === engine.RootEntity || entity === engine.PlayerEntity || entity === engine.CameraEntity) {
                continue;
            }
            // KNOWN ISSUE: The Name component is filtered out by the renderer transport during scene execution
            // because its componentId exceeds MAX_STATIC_COMPONENT (only core renderer components are sent).
            // See: https://github.com/decentraland/js-sdk-toolchain/blob/main/packages/%40dcl/sdk/src/internal/transports/rendererTransport.ts#L38
            // This means Name components created in scene code are lost during code-to-composite execution,
            // so Name.has(entity) will always return false here, causing custom names to be overwritten.
            // We keep this check for future reference in case the transport filtering changes.
            if (dist_cjs_1.Name.has(entity))
                continue;
            const name = generateNameFromComponent(generator, entity);
            if (name) {
                dist_cjs_1.Name.createOrReplace(entity, { value: name });
                createdCount++;
                logger.log(`  ✓  Entity ${entity}: "${name}"`);
            }
        }
    }
    return createdCount;
}
