// @dcl/sdk-commands/dist/commands/code-to-composite/scene-executor.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.executeSceneCode = executeSceneCode;
const path_1 = __importDefault(require("path"));
const module_1 = __importDefault(require("module"));
const i18next_1 = __importDefault(require("i18next"));
const dist_cjs_1 = require("@dcl/ecs/dist-cjs");
const ByteBuffer_1 = require("@dcl/ecs/dist-cjs/serialization/ByteBuffer");
const inspector_1 = require("@dcl/inspector");
const error_1 = require("../../logic/error");
const bundle_1 = require("../../logic/bundle");
/**
 * Returns the set of component IDs that are used by the @dcl/inspector.
 */
function getInspectorComponentsIds() {
    const { components } = (0, inspector_1.createEngineContext)();
    const ids = new Set();
    for (const [_, component] of Object.entries(components)) {
        ids.add(component.componentId);
    }
    return ids;
}
/**
 * Initializes the ECS engine with a mock transport layer.
 *
 * The transport layer is used to capture CRDT messages sent by the scene code.
 * This allows us to capture the state of all entities and components.
 */
function initEngine() {
    const transport = {
        filter: () => true,
        send: async (_messages) => { }
    };
    dist_cjs_1.engine.addTransport(transport);
    return {
        engine: dist_cjs_1.engine,
        transport
    };
}
/**
 * Creates initial CRDT state with required engine components.
 */
function getInitialCrdtState() {
    function addPlayerEntityTransform() {
        const buffer = new ByteBuffer_1.ReadWriteByteBuffer();
        const transform = dist_cjs_1.Transform.create(dist_cjs_1.engine.PlayerEntity);
        dist_cjs_1.Transform.schema.serialize(transform, buffer);
        const transformData = buffer.toCopiedBinary();
        buffer.resetBuffer();
        dist_cjs_1.PutComponentOperation.write(1, 1, dist_cjs_1.Transform.componentId, transformData, buffer);
        dist_cjs_1.PutComponentOperation.write(2, 1, dist_cjs_1.Transform.componentId, transformData, buffer);
        return buffer.toBinary();
    }
    function addUICanvasOnRootEntity() {
        const buffer = new ByteBuffer_1.ReadWriteByteBuffer();
        const uiCanvasInformation = dist_cjs_1.UiCanvasInformation.create(dist_cjs_1.engine.RootEntity);
        dist_cjs_1.UiCanvasInformation.schema.serialize(uiCanvasInformation, buffer);
        const uiCanvasComponentData = buffer.toCopiedBinary();
        buffer.resetBuffer();
        dist_cjs_1.PutComponentOperation.write(0, 1, dist_cjs_1.UiCanvasInformation.componentId, uiCanvasComponentData, buffer);
        return buffer.toBinary();
    }
    function addCameraMode() {
        const buffer = new ByteBuffer_1.ReadWriteByteBuffer();
        const cameraMode = dist_cjs_1.CameraMode.create(dist_cjs_1.engine.RootEntity);
        dist_cjs_1.CameraMode.schema.serialize(cameraMode, buffer);
        const cameraModeComponentData = buffer.toCopiedBinary();
        buffer.resetBuffer();
        dist_cjs_1.PutComponentOperation.write(2, 1, dist_cjs_1.CameraMode.componentId, cameraModeComponentData, buffer);
        return buffer.toBinary();
    }
    return [addPlayerEntityTransform(), addUICanvasOnRootEntity(), addCameraMode()];
}
/**
 * Creates pre-defined mocks for critical ~system modules.
 *
 * These modules are runtime-specific and don't exist in Node.js.
 * This function provides meaningful mock implementations for the most
 * commonly used system modules to ensure scene code executes correctly.
 */
function createCriticalModuleMocks(engine, transport, crdtState) {
    return {
        '~system/EngineApi': {
            crdtSendToRenderer: async (crdt) => {
                transport.onmessage(crdt.data);
                await engine.update(0);
            },
            crdtGetState: async () => ({
                hasEntities: false,
                data: [...getInitialCrdtState(), crdtState]
            }),
            sendBatch: async () => ({ events: [] }),
            subscribe: async () => ({}),
            unsubscribe: async () => ({}),
            isServer: async () => ({ isServer: false }),
            crdtGetMessageFromRenderer: async () => ({ data: [] })
        },
        '~system/Scene': {
            getSceneInfo: async () => ({
                cid: '',
                metadata: '{}',
                baseUrl: '',
                contents: []
            })
        },
        '~system/UserIdentity': {
            getUserData: async () => ({
                data: {
                    displayName: 'Test User',
                    userId: 'test-user-id',
                    hasConnectedWeb3: false,
                    version: 1,
                    avatar: {
                        bodyShape: 'urn:decentraland:off-chain:base-avatars:BaseMale',
                        skinColor: '#000000',
                        hairColor: '#000000',
                        eyeColor: '#000000',
                        wearables: [],
                        snapshots: {
                            face256: '',
                            body: ''
                        }
                    }
                }
            }),
            getUserPublicKey: async () => ({ address: undefined })
        },
        '~system/EnvironmentApi': {
            isPreviewMode: async () => ({ isPreview: true }),
            getBootstrapData: async () => ({
                id: '',
                baseUrl: '',
                entity: {
                    content: [],
                    metadataJson: '{}'
                },
                useFPSThrottling: false
            }),
            getPlatform: async () => ({ platform: 'desktop' }),
            areUnsafeRequestAllowed: async () => ({ status: false }),
            getCurrentRealm: async () => ({ currentRealm: undefined }),
            getExplorerConfiguration: async () => ({ clientUri: '', configurations: {} }),
            getDecentralandTime: async () => ({ seconds: Date.now() / 1000 })
        },
        '~system/Runtime': {
            getRealm: async () => ({
                realmInfo: {
                    baseUrl: 'http://localhost',
                    realmName: 'localhost',
                    networkId: 1,
                    commsAdapter: 'offline',
                    isPreview: true
                }
            }),
            getWorldTime: async () => ({ seconds: Date.now() / 1000 }),
            readFile: async () => ({ content: new Uint8Array(), hash: '' }),
            getSceneInformation: async () => ({
                urn: '',
                content: [],
                metadataJson: '{}',
                baseUrl: ''
            }),
            getExplorerInformation: async () => ({
                agent: 'code-to-composite',
                platform: 'desktop',
                configurations: {}
            })
        },
        '~system/Players': {
            getPlayerData: async () => ({ data: undefined }),
            getPlayersInScene: async () => ({ players: [] }),
            getConnectedPlayers: async () => ({ players: [] })
        },
        '~system/RestrictedActions': {
            movePlayerTo: async () => ({}),
            teleportTo: async () => ({}),
            triggerEmote: async () => ({}),
            changeRealm: async () => ({ success: false }),
            openExternalUrl: async () => ({ success: false }),
            openNftDialog: async () => ({ success: false }),
            setCommunicationsAdapter: async () => ({ success: false }),
            triggerSceneEmote: async () => ({ success: false })
        },
        '~system/CommsApi': {
            getActiveVideoStreams: async () => ({ streams: [] })
        },
        '~system/CommunicationsController': {
            send: async () => ({}),
            sendBinary: async () => ({ data: [] })
        },
        '~system/EthereumController': {
            requirePayment: async () => ({ jsonAnyResponse: '{}' }),
            signMessage: async () => ({ message: '', hexEncodedMessage: '', signature: '' }),
            convertMessageToObject: async () => ({ dict: {} }),
            sendAsync: async () => ({ jsonAnyResponse: '{}' }),
            getUserAccount: async () => ({ address: undefined })
        },
        '~system/PortableExperiences': {
            spawn: async () => ({ pid: '', parentCid: '', name: '' }),
            kill: async () => ({ status: false }),
            exit: async () => ({ status: false }),
            getPortableExperiencesLoaded: async () => ({ loaded: [] })
        },
        '~system/SignedFetch': {
            signedFetch: async () => ({
                ok: false,
                status: 404,
                statusText: 'Not Found',
                headers: {},
                body: ''
            }),
            getHeaders: async () => ({ headers: {} })
        },
        '~system/Testing': {
            logTestResult: async () => ({}),
            plan: async () => ({}),
            setCameraTransform: async () => ({})
        },
        '~system/UserActionModule': {
            requestTeleport: async () => ({})
        }
    };
}
/**
 * Creates a Proxy that automatically mocks unknown properties and functions.
 *
 * This provides a fallback mechanism for ~system module functions that aren't
 * explicitly mocked.
 */
function createAutoMockProxy(baseMock = {}) {
    return new Proxy(baseMock, {
        get(target, prop) {
            if (prop in target) {
                return target[prop];
            }
            return async (...args) => {
                return {};
            };
        }
    });
}
/**
 * Creates a mock implementation for a ~system/* module.
 *
 * ~system modules are runtime-specific and don't exist in Node.js, so they need
 * to be mocked for scene execution.
 */
function createSystemModuleMock(engine, transport, mainCrdt, moduleId) {
    const criticalModuleMocks = createCriticalModuleMocks(engine, transport, mainCrdt);
    if (moduleId in criticalModuleMocks) {
        return createAutoMockProxy(criticalModuleMocks[moduleId]);
    }
    return createAutoMockProxy();
}
/**
 * Sets up a Node.js require hook to intercept and mock ~system/* module imports.
 *
 * This patches Module.prototype.require to intercept require() calls during
 * scene bundle execution.
 *
 * NOTE: The hook is temporary and should be removed after scene execution using
 * the returned cleanup function.
 */
function setupRequireHook(engine, transport, mainCrdt) {
    const originalRequire = module_1.default.prototype.require;
    module_1.default.prototype.require = function (id) {
        if (id.startsWith('~system/')) {
            return createSystemModuleMock(engine, transport, mainCrdt, id);
        }
        return originalRequire.apply(this, [id]);
    };
    return () => {
        module_1.default.prototype.require = originalRequire;
    };
}
/**
 * Loads and executes the scene bundle to populate the ECS engine.
 */
async function loadAndExecuteBundle(bundlePath) {
    delete require.cache[require.resolve(bundlePath)];
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const sceneModule = require(bundlePath);
    if (typeof sceneModule.onStart === 'function') {
        const onStart = sceneModule.onStart();
        if (typeof onStart.then === 'function') {
            await onStart;
        }
    }
    if (typeof sceneModule.onUpdate === 'function') {
        const onUpdate = sceneModule.onUpdate(0);
        if (typeof onUpdate.then === 'function') {
            await onUpdate;
        }
    }
}
/**
 * Filters out components that are not supported by the @dcl/inspector.
 */
function filterInspectorCompatibleComponents(engine) {
    const validComponentIds = getInspectorComponentsIds();
    const reactEcsComponentIds = new Set([
        dist_cjs_1.UiTransform.componentId,
        dist_cjs_1.UiText.componentId,
        dist_cjs_1.UiBackground.componentId,
        dist_cjs_1.UiInput.componentId,
        dist_cjs_1.UiInputResult.componentId,
        dist_cjs_1.UiDropdown.componentId,
        dist_cjs_1.UiDropdownResult.componentId
    ]);
    // first pass: identify entities with react-ecs components
    const uiEntities = new Set();
    for (const component of engine.componentsIter()) {
        if (reactEcsComponentIds.has(component.componentId)) {
            for (const [entity] of engine.getEntitiesWith(component)) {
                uiEntities.add(entity);
            }
        }
    }
    // second pass: remove components from entities that either:
    // - are root engine entities
    // - have react-ecs components
    for (const component of engine.componentsIter()) {
        if ('deleteFrom' in component) {
            const isValidComponent = validComponentIds.has(component.componentId);
            for (const [entity] of engine.getEntitiesWith(component)) {
                if (uiEntities.has(entity) || !isValidComponent) {
                    component.deleteFrom(entity);
                }
            }
        }
    }
    // BONUS: remove Transform from PlayerEntity to avoid bad placement when running the client
    dist_cjs_1.Transform.deleteFrom(engine.PlayerEntity);
}
/**
 * Loads the main CRDT file if it exists, otherwise returns an empty buffer.
 *
 * The CRDT file contains serialized component state that may already exist
 * in the scene. This is used to initialize the engine with any pre-existing
 * state before executing the scene code.
 */
async function getMainCrdtFile(components, crdtFilePath) {
    let mainCrdt = new Uint8Array();
    if (await components.fs.fileExists(crdtFilePath)) {
        mainCrdt = new Uint8Array(await components.fs.readFile(crdtFilePath));
    }
    return mainCrdt;
}
/**
 * Bundles the scene code and returns the entrypoint.
 */
async function bundle(components, project) {
    try {
        const { inputs } = await (0, bundle_1.bundleProject)(components, {
            workingDirectory: project.workingDirectory,
            watch: false,
            production: false, // keep source maps for better error messages
            emitDeclaration: false,
            ignoreComposite: true, // don't load existing composite files
            customEntryPoint: false
        }, project.scene);
        return inputs[0].entrypoint;
    }
    catch (err) {
        throw new error_1.CliError('CODE_TO_COMPOSITE_BUILD_FAILED', i18next_1.default.t('errors.code_to_composite.build_failed', { error: err.message }));
    }
}
/**
 * Executes a Decentraland scene's code and captures the resulting ECS engine state.
 *
 * This is the main entry point for the code-to-composite command. It runs scene
 * code in a Node.js environment to extract the entity/component structure.
 *
 * The returned engine can then be used to generate composite/CRDT files.
 */
async function executeSceneCode(components, project, crdtFilePath) {
    const { fs, logger } = components;
    logger.log('Building scene...');
    const sceneCodeEntrypoint = await bundle(components, project);
    const bundlePath = path_1.default.join(project.workingDirectory, project.scene.main);
    if (!(await fs.fileExists(bundlePath))) {
        throw new error_1.CliError('CODE_TO_COMPOSITE_BUNDLE_NOT_FOUND', i18next_1.default.t('errors.code_to_composite.bundle_not_found', { bundlePath }));
    }
    logger.log('Executing scene code to capture engine state...');
    const crdtState = await getMainCrdtFile(components, crdtFilePath);
    const { engine, transport } = initEngine();
    const restoreRequire = setupRequireHook(engine, transport, crdtState);
    try {
        await loadAndExecuteBundle(bundlePath);
        filterInspectorCompatibleComponents(engine);
        return { engine: engine, sceneCodeEntrypoint };
    }
    catch (err) {
        throw new error_1.CliError('CODE_TO_COMPOSITE_EXECUTION_FAILED', i18next_1.default.t('errors.code_to_composite.execution_failed', { error: `${err.message}\n${err.stack || ''}` }));
    }
    finally {
        // cleanup: restore original require
        restoreRequire();
    }
}
