// @dcl/asset-packs/dist/admin-toolkit-ui/VideoControl/api.js
import { getDomain, wrapSignedFetch } from '../fetch-utils';
const URLS = () => ({
    STREAM_KEY: `https://comms-gatekeeper.decentraland.${getDomain()}/scene-stream-access`,
    GET_DCL_CAST_INFO: `https://comms-gatekeeper.decentraland.${getDomain()}/cast/generate-stream-link`,
});
export async function getStreamKey() {
    return wrapSignedFetch({ url: URLS().STREAM_KEY }, { toCamelCase: true });
}
export async function generateStreamKey() {
    return wrapSignedFetch({ url: URLS().STREAM_KEY, init: { method: 'POST', headers: {} } }, { toCamelCase: true });
}
export async function revokeStreamKey() {
    return wrapSignedFetch({
        url: URLS().STREAM_KEY,
        init: { method: 'DELETE', headers: {} },
    });
}
export async function resetStreamKey() {
    return wrapSignedFetch({ url: URLS().STREAM_KEY, init: { method: 'PUT', headers: {} } }, { toCamelCase: true });
}
export async function getDclCastInfo() {
    return wrapSignedFetch({ url: URLS().GET_DCL_CAST_INFO }, { toCamelCase: true });
}
