// @well-known-components/interfaces/dist/utils/lifecycle.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Lifecycle = void 0;
const base_component_1 = require("../components/base-component");
const DEBUG = !!process.env.DEBUG;
function log(level, message) {
    if (level == 'DEBUG' && !DEBUG)
        return;
    process.stderr.write(new Date().toISOString() + " [" + level + "] (well-known-components/interfaces): " + message + "\n");
}
async function stopAllComponents(components) {
    const reverseKeys = Object.keys(components).reverse();
    for (let c of reverseKeys) {
        const component = components[c];
        if (!component)
            throw new Error("Component is null: " + c);
        if (typeof component[base_component_1.STOP_COMPONENT] == "function") {
            log('INFO', `Stopping component ${c}`);
            await component[base_component_1.STOP_COMPONENT]();
        }
        // TODO: remove for 3.0.0
        else if (typeof component.stop == "function") {
            log('INFO', `Stopping component ${c}`);
            log('WARN', "IBaseComponent.stop is deprecated (in " + c + ".stop). Use IBaseComponent[STOP_COMPONENT] instead");
            await component.stop();
        }
    }
}
async function allSettled(promises) {
    let mappedPromises = promises.map((p) => {
        let r = p.then((value) => {
            return {
                status: "fulfilled",
                value,
            };
        });
        if ("catch" in p) {
            r = p.catch((reason) => {
                return {
                    status: "rejected",
                    reason,
                };
            });
        }
        return r;
    });
    return Promise.all(mappedPromises);
}
// gracefully finalizes all the components on SIGTERM
async function startComponentsLifecycle(components) {
    log('INFO', "Starting components");
    const pending = [];
    let mutStarted = false;
    let mutLive = false;
    const immutableStartOptions = {
        started() {
            return mutStarted;
        },
        live() {
            return mutLive;
        },
        getComponents() {
            return components;
        },
    };
    for (let c in components) {
        const component = components[c];
        if ((await components[c]) !== components[c]) {
            log('ERROR', "Error initializing components. Component '" +
                c +
                "' is a Promise, it should be an object, did you miss an await in the initComponents?.");
        }
        if (!component)
            throw new Error("Null or empty components are not allowed: " + c);
        var startFn = undefined;
        if (typeof component[base_component_1.START_COMPONENT] == "function") {
            startFn = component[base_component_1.START_COMPONENT].bind(component);
        }
        //TODO remove in 3.0
        else if (component.start && typeof component.start == "function") {
            log("WARN", c + ".start is deprecated (in " + c + ".stop). use IBaseComponent[START_COMPONENT] instead");
            startFn = component.start.bind(component);
        }
        if (startFn) {
            log('DEBUG', `Starting component ${c}`);
            const awaitable = startFn(immutableStartOptions);
            if (awaitable && typeof awaitable == "object" && "then" in awaitable) {
                pending.push(awaitable);
                await awaitable;
                if (awaitable.catch) {
                    // avoid unhanled catch error messages in node.
                    // real catch happens below in `Promise.all(pending)`
                    awaitable.catch((err) => {
                        log('ERROR', `Error initializing component: ${JSON.stringify(c)}. Error will appear in the next line\n${err}\n`);
                    });
                }
            }
        }
    }
    // application started
    mutLive = true;
    if (pending.length == 0)
        return;
    try {
        await Promise.all(pending);
        mutStarted = true;
    }
    catch (e) {
        log("ERROR", "Error initializing components. Stopping components and closing application.");
        await allSettled(pending);
        throw e;
    }
}
/**
 * handles an async function, if it fails the program exits with exit code 1
 */
function asyncTopLevelExceptionHanler(fn) {
    return fn().catch((error) => {
        // print error and stacktrace
        console.error(error);
        // exit program with error
        process.exit(1);
    });
}
/**
 * This namespace handles the basic lifecycle of the components.
 * @public
 */
var Lifecycle;
(function (Lifecycle) {
    /**
     * Program entry point, this should be the one and only top level
     * expression of your program.
     */
    function run(config) {
        return asyncTopLevelExceptionHanler(async () => {
            // pick a componentInitializer
            const componentInitializer = config.initComponents;
            // init ports & components
            log("INFO", "Initializing components");
            const components = Object.freeze(await componentInitializer());
            let componentsStarted;
            // the program can register some handlers to be run before stopping all components
            const beforeStopComponentHandlers = [];
            const beforeStopComponents = async () => {
                while (beforeStopComponentHandlers.length) {
                    try {
                        await (beforeStopComponentHandlers.shift())();
                    }
                    catch (e) {
                        log('ERROR', e);
                        console.error(e);
                    }
                }
            };
            async function stopProgram() {
                process.off("SIGTERM", termHandler);
                process.off("SIGINT", termHandler);
                await beforeStopComponents();
                await stopAllComponents(components);
            }
            function termHandler() {
                log('INFO', "termination signal received");
                stopProgram()
                    .then(() => process.exit())
                    .catch((e) => {
                    log('ERROR', e);
                    console.error(e);
                    process.exit(1);
                });
            }
            const program = {
                get components() {
                    return components;
                },
                beforeStopComponents(callback) {
                    beforeStopComponentHandlers.push(callback);
                },
                async stop() {
                    await stopProgram();
                },
                async startComponents() {
                    if (!componentsStarted) {
                        // start components
                        componentsStarted = startComponentsLifecycle(components);
                    }
                    else {
                        log('WARN', "Warning: startComponents must be called once\n");
                    }
                    return componentsStarted;
                },
            };
            try {
                // wire adapters
                log('INFO', "Wiring app");
                await config.main(program);
                if (!componentsStarted) {
                    log('WARN', "Warning: startComponents was not called inside programEntryPoint.main function\n");
                }
                else {
                    await componentsStarted;
                }
                // gracefully finalizes all the components on SIGINT&SIGTERM
                process.on("SIGTERM", termHandler);
                process.on("SIGINT", termHandler);
            }
            catch (e) {
                try {
                    // gracefully stop all components
                    await program.stop();
                }
                catch (err) {
                    console.error(err);
                }
                finally {
                    // the following throw is handled by asyncTopLevelExceptionHanler
                    // exiting the program
                    throw e;
                }
            }
            return program;
        });
    }
    Lifecycle.run = run;
})(Lifecycle || (exports.Lifecycle = Lifecycle = {}));
