// @dcl/schemas/dist/platform/events/camera.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PhotoTakenEvent = exports.PhotoPrivacyChangedEvent = void 0;
const validation_1 = require("../../validation");
const base_1 = require("./base");
var PhotoPrivacyChangedEvent;
(function (PhotoPrivacyChangedEvent) {
    PhotoPrivacyChangedEvent.schema = {
        type: 'object',
        properties: {
            type: { type: 'string', const: base_1.Events.Type.CAMERA },
            subType: { type: 'string', const: base_1.Events.SubType.Camera.PHOTO_PRIVACY_CHANGED },
            key: { type: 'string' },
            timestamp: { type: 'number', minimum: 0 },
            metadata: {
                type: 'object',
                properties: {
                    userAddress: { type: 'string' },
                    photoId: { type: 'string' },
                    isPublic: { type: 'boolean' }
                },
                required: ['userAddress', 'photoId', 'isPublic']
            }
        },
        required: ['type', 'subType', 'key', 'timestamp', 'metadata'],
        additionalProperties: false
    };
    PhotoPrivacyChangedEvent.validate = (0, validation_1.generateLazyValidator)(PhotoPrivacyChangedEvent.schema);
})(PhotoPrivacyChangedEvent = exports.PhotoPrivacyChangedEvent || (exports.PhotoPrivacyChangedEvent = {}));
var PhotoTakenEvent;
(function (PhotoTakenEvent) {
    PhotoTakenEvent.schema = {
        type: 'object',
        properties: {
            type: { type: 'string', const: base_1.Events.Type.CAMERA },
            subType: { type: 'string', const: base_1.Events.SubType.Camera.PHOTO_TAKEN },
            key: { type: 'string' },
            timestamp: { type: 'number', minimum: 0 },
            metadata: {
                type: 'object',
                properties: {
                    sceneId: { type: 'string' },
                    userAddress: { type: 'string' },
                    realm: { type: 'string' },
                    photoId: { type: 'string' },
                    isPublic: { type: 'boolean' },
                    users: {
                        type: 'array',
                        items: {
                            type: 'object',
                            properties: {
                                address: {
                                    type: 'string'
                                },
                                isEmoting: {
                                    type: 'boolean'
                                }
                            },
                            required: ['address', 'isEmoting'],
                            additionalProperties: false
                        }
                    },
                    placeId: { type: 'string', nullable: true }
                },
                required: ['userAddress', 'realm', 'photoId', 'isPublic', 'users'],
                additionalProperties: false
            }
        },
        required: ['type', 'subType', 'key', 'timestamp', 'metadata'],
        additionalProperties: false
    };
    PhotoTakenEvent.validate = (0, validation_1.generateLazyValidator)(PhotoTakenEvent.schema);
})(PhotoTakenEvent = exports.PhotoTakenEvent || (exports.PhotoTakenEvent = {}));
