// @dcl/schemas/dist/platform/events/comms.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserUnbannedFromSceneEvent = exports.UserBannedFromSceneEvent = exports.UserLeftRoomEvent = exports.UserJoinedRoomEvent = exports.RoomType = void 0;
const validation_1 = require("../../validation");
const base_1 = require("./base");
var RoomType;
(function (RoomType) {
    RoomType["PRIVATE_MESSAGE"] = "private-message";
    RoomType["VOICE_CHAT"] = "voice-chat";
    RoomType["COMMUNITY_VOICE_CHAT"] = "community-voice-chat";
    RoomType["ISLAND"] = "island";
    RoomType["SCENE"] = "scene";
    RoomType["WORLD"] = "world";
    RoomType["UNKNOWN"] = "unknown";
})(RoomType = exports.RoomType || (exports.RoomType = {}));
var UserJoinedRoomEvent;
(function (UserJoinedRoomEvent) {
    UserJoinedRoomEvent.schema = {
        type: 'object',
        properties: {
            type: { type: 'string', const: base_1.Events.Type.COMMS },
            subType: { type: 'string', const: base_1.Events.SubType.Comms.USER_JOINED_ROOM },
            key: { type: 'string' },
            timestamp: { type: 'number', minimum: 0 },
            metadata: {
                type: 'object',
                properties: {
                    parcel: { type: 'string' },
                    sceneId: { type: 'string' },
                    userAddress: { type: 'string' },
                    realmName: { type: 'string' },
                    islandName: { type: 'string', nullable: true },
                    communityId: { type: 'string', nullable: true },
                    voiceChatId: { type: 'string', nullable: true },
                    isWorld: { type: 'boolean' },
                    roomType: { type: 'string', enum: Object.values(RoomType) }
                },
                required: ['parcel', 'sceneId', 'userAddress', 'realmName', 'isWorld', 'roomType']
            }
        },
        required: ['type', 'subType', 'key', 'timestamp', 'metadata']
    };
    UserJoinedRoomEvent.validate = (0, validation_1.generateLazyValidator)(UserJoinedRoomEvent.schema);
})(UserJoinedRoomEvent = exports.UserJoinedRoomEvent || (exports.UserJoinedRoomEvent = {}));
var UserLeftRoomEvent;
(function (UserLeftRoomEvent) {
    UserLeftRoomEvent.schema = {
        type: 'object',
        properties: {
            type: { type: 'string', const: base_1.Events.Type.COMMS },
            subType: { type: 'string', const: base_1.Events.SubType.Comms.USER_LEFT_ROOM },
            key: { type: 'string' },
            timestamp: { type: 'number', minimum: 0 },
            metadata: {
                type: 'object',
                properties: {
                    userAddress: { type: 'string' },
                    realmName: { type: 'string' },
                    isWorld: { type: 'boolean' },
                    roomType: { type: 'string', enum: Object.values(RoomType) },
                    sceneId: { type: 'string', nullable: true },
                    islandName: { type: 'string', nullable: true },
                    communityId: { type: 'string', nullable: true },
                    voiceChatId: { type: 'string', nullable: true }
                },
                required: ['isWorld', 'userAddress', 'realmName', 'roomType']
            }
        },
        required: ['type', 'subType', 'key', 'timestamp', 'metadata'],
        additionalProperties: false
    };
    UserLeftRoomEvent.validate = (0, validation_1.generateLazyValidator)(UserLeftRoomEvent.schema);
})(UserLeftRoomEvent = exports.UserLeftRoomEvent || (exports.UserLeftRoomEvent = {}));
var UserBannedFromSceneEvent;
(function (UserBannedFromSceneEvent) {
    UserBannedFromSceneEvent.schema = {
        type: 'object',
        properties: {
            type: { type: 'string', const: base_1.Events.Type.COMMS },
            subType: { type: 'string', const: base_1.Events.SubType.Comms.USER_BANNED_FROM_SCENE },
            key: { type: 'string' },
            timestamp: { type: 'number', minimum: 0 },
            metadata: {
                type: 'object',
                properties: {
                    placeTitle: { type: 'string' },
                    userAddress: { type: 'string' }
                },
                required: ['placeTitle', 'userAddress']
            }
        },
        required: ['type', 'subType', 'key', 'timestamp', 'metadata'],
        additionalProperties: false
    };
    UserBannedFromSceneEvent.validate = (0, validation_1.generateLazyValidator)(UserBannedFromSceneEvent.schema);
})(UserBannedFromSceneEvent = exports.UserBannedFromSceneEvent || (exports.UserBannedFromSceneEvent = {}));
var UserUnbannedFromSceneEvent;
(function (UserUnbannedFromSceneEvent) {
    UserUnbannedFromSceneEvent.schema = {
        type: 'object',
        properties: {
            type: { type: 'string', const: base_1.Events.Type.COMMS },
            subType: { type: 'string', const: base_1.Events.SubType.Comms.USER_UNBANNED_FROM_SCENE },
            key: { type: 'string' },
            timestamp: { type: 'number', minimum: 0 },
            metadata: {
                type: 'object',
                properties: {
                    placeTitle: { type: 'string' },
                    userAddress: { type: 'string' }
                },
                required: ['placeTitle', 'userAddress']
            }
        },
        required: ['type', 'subType', 'key', 'timestamp', 'metadata'],
        additionalProperties: false
    };
    UserUnbannedFromSceneEvent.validate = (0, validation_1.generateLazyValidator)(UserUnbannedFromSceneEvent.schema);
})(UserUnbannedFromSceneEvent = exports.UserUnbannedFromSceneEvent || (exports.UserUnbannedFromSceneEvent = {}));
