// @dcl/schemas/dist/platform/events/world.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorldsMissingResourcesEvent = exports.WorldsAccessRestrictedEvent = exports.WorldsPermissionGrantedEvent = exports.WorldsPermissionRevokedEvent = exports.WorldsAccessRestoredEvent = exports.WorldDeploymentEvent = exports.WorldScenesUndeploymentEvent = exports.WorldSpawnCoordinateSetEvent = exports.WorldSettingsChangedEvent = void 0;
const auth_chain_1 = require("../../misc/auth-chain");
const validation_1 = require("../../validation");
const base_1 = require("./base");
var WorldSettingsChangedEvent;
(function (WorldSettingsChangedEvent) {
    WorldSettingsChangedEvent.schema = {
        type: 'object',
        properties: {
            type: { type: 'string', const: base_1.Events.Type.WORLD },
            subType: { type: 'string', const: base_1.Events.SubType.Worlds.WORLD_SETTINGS_CHANGED },
            key: { type: 'string' },
            timestamp: { type: 'number', minimum: 0 },
            metadata: {
                type: 'object',
                properties: {
                    title: { type: 'string', nullable: true },
                    description: { type: 'string', nullable: true },
                    contentRating: { type: 'string', nullable: true },
                    skyboxTime: { type: 'number', nullable: true },
                    categories: { type: 'array', items: { type: 'string' }, nullable: true },
                    singlePlayer: { type: 'boolean', nullable: true },
                    showInPlaces: { type: 'boolean', nullable: true },
                    thumbnailUrl: { type: 'string', nullable: true }
                },
                additionalProperties: false
            }
        },
        required: ['type', 'subType', 'key', 'timestamp', 'metadata'],
        additionalProperties: false
    };
})(WorldSettingsChangedEvent = exports.WorldSettingsChangedEvent || (exports.WorldSettingsChangedEvent = {}));
var WorldSpawnCoordinateSetEvent;
(function (WorldSpawnCoordinateSetEvent) {
    WorldSpawnCoordinateSetEvent.schema = {
        type: 'object',
        properties: {
            type: { type: 'string', const: base_1.Events.Type.WORLD },
            subType: { type: 'string', const: base_1.Events.SubType.Worlds.WORLD_SPAWN_COORDINATE_SET },
            key: { type: 'string' },
            timestamp: { type: 'number', minimum: 0 },
            metadata: {
                type: 'object',
                properties: {
                    name: { type: 'string', pattern: '^[a-zA-Z0-9_-]+\\.dcl\\.eth|[a-zA-Z0-9_-]+\\.eth$' },
                    oldCoordinate: {
                        type: 'object',
                        properties: { x: { type: 'number' }, y: { type: 'number' } },
                        required: ['x', 'y'],
                        additionalProperties: false,
                        nullable: true
                    },
                    newCoordinate: {
                        type: 'object',
                        properties: { x: { type: 'number' }, y: { type: 'number' } },
                        required: ['x', 'y'],
                        additionalProperties: false
                    }
                },
                required: ['name', 'oldCoordinate', 'newCoordinate'],
                additionalProperties: false
            }
        },
        required: ['type', 'subType', 'key', 'timestamp', 'metadata'],
        additionalProperties: false
    };
    WorldSpawnCoordinateSetEvent.validate = (0, validation_1.generateLazyValidator)(WorldSpawnCoordinateSetEvent.schema);
})(WorldSpawnCoordinateSetEvent = exports.WorldSpawnCoordinateSetEvent || (exports.WorldSpawnCoordinateSetEvent = {}));
var WorldScenesUndeploymentEvent;
(function (WorldScenesUndeploymentEvent) {
    WorldScenesUndeploymentEvent.schema = {
        type: 'object',
        properties: {
            type: { type: 'string', const: base_1.Events.Type.WORLD },
            subType: { type: 'string', const: base_1.Events.SubType.Worlds.WORLD_SCENES_UNDEPLOYMENT },
            key: { type: 'string' },
            timestamp: { type: 'number', minimum: 0 },
            metadata: {
                type: 'object',
                properties: { entityIds: { type: 'array', items: { type: 'string' }, minItems: 1 } },
                required: ['entityIds'],
                additionalProperties: false
            }
        },
        required: ['type', 'subType', 'key', 'timestamp', 'metadata'],
        additionalProperties: false
    };
})(WorldScenesUndeploymentEvent = exports.WorldScenesUndeploymentEvent || (exports.WorldScenesUndeploymentEvent = {}));
var WorldDeploymentEvent;
(function (WorldDeploymentEvent) {
    WorldDeploymentEvent.schema = {
        type: 'object',
        properties: {
            type: { type: 'string', const: base_1.Events.Type.WORLD },
            subType: { type: 'string', const: base_1.Events.SubType.Worlds.DEPLOYMENT },
            key: { type: 'string' },
            timestamp: { type: 'number', minimum: 0 },
            entity: {
                type: 'object',
                properties: { entityId: { type: 'string' }, authChain: auth_chain_1.AuthChain.schema },
                additionalProperties: true,
                required: ['entityId', 'authChain']
            },
            contentServerUrls: { type: 'array', items: { type: 'string' }, nullable: true },
            force: { type: 'boolean', nullable: true },
            animation: { type: 'string', nullable: true },
            lods: { type: 'array', items: { type: 'string' }, nullable: true }
        },
        required: ['type', 'subType', 'key', 'timestamp', 'entity'],
        additionalProperties: false
    };
    WorldDeploymentEvent.validate = (0, validation_1.generateLazyValidator)(WorldDeploymentEvent.schema);
})(WorldDeploymentEvent = exports.WorldDeploymentEvent || (exports.WorldDeploymentEvent = {}));
var WorldsAccessRestoredEvent;
(function (WorldsAccessRestoredEvent) {
    WorldsAccessRestoredEvent.schema = {
        type: 'object',
        properties: {
            type: { type: 'string', const: base_1.Events.Type.WORLD },
            subType: { type: 'string', const: base_1.Events.SubType.Worlds.WORLDS_ACCESS_RESTORED },
            key: { type: 'string' },
            timestamp: { type: 'number', minimum: 0 },
            metadata: {
                type: 'object',
                properties: {
                    title: { type: 'string' },
                    description: { type: 'string' },
                    url: { type: 'string' },
                    attendee: { type: 'string' }
                },
                required: ['title', 'description', 'url', 'attendee'],
                additionalProperties: false
            }
        },
        required: ['type', 'subType', 'key', 'timestamp', 'metadata'],
        additionalProperties: false
    };
    WorldsAccessRestoredEvent.validate = (0, validation_1.generateLazyValidator)(WorldsAccessRestoredEvent.schema);
})(WorldsAccessRestoredEvent = exports.WorldsAccessRestoredEvent || (exports.WorldsAccessRestoredEvent = {}));
var WorldsPermissionRevokedEvent;
(function (WorldsPermissionRevokedEvent) {
    WorldsPermissionRevokedEvent.schema = {
        type: 'object',
        properties: {
            type: { type: 'string', const: base_1.Events.Type.WORLD },
            subType: { type: 'string', const: base_1.Events.SubType.Worlds.WORLDS_PERMISSION_REVOKED },
            key: { type: 'string' },
            timestamp: { type: 'number', minimum: 0 },
            metadata: {
                type: 'object',
                properties: {
                    title: { type: 'string' },
                    description: { type: 'string' },
                    world: { type: 'string' },
                    permissions: { type: 'array', items: { type: 'string' } },
                    url: { type: 'string' },
                    address: { type: 'string' }
                },
                required: ['title', 'description', 'world', 'permissions', 'url', 'address'],
                additionalProperties: false
            }
        },
        required: ['type', 'subType', 'key', 'timestamp', 'metadata'],
        additionalProperties: false
    };
    WorldsPermissionRevokedEvent.validate = (0, validation_1.generateLazyValidator)(WorldsPermissionRevokedEvent.schema);
})(WorldsPermissionRevokedEvent = exports.WorldsPermissionRevokedEvent || (exports.WorldsPermissionRevokedEvent = {}));
var WorldsPermissionGrantedEvent;
(function (WorldsPermissionGrantedEvent) {
    WorldsPermissionGrantedEvent.schema = {
        type: 'object',
        properties: {
            type: { type: 'string', const: base_1.Events.Type.WORLD },
            subType: { type: 'string', const: base_1.Events.SubType.Worlds.WORLDS_PERMISSION_GRANTED },
            key: { type: 'string' },
            timestamp: { type: 'number', minimum: 0 },
            metadata: {
                type: 'object',
                properties: {
                    title: { type: 'string' },
                    description: { type: 'string' },
                    world: { type: 'string' },
                    permissions: { type: 'array', items: { type: 'string' } },
                    url: { type: 'string' },
                    address: { type: 'string' }
                },
                required: ['title', 'description', 'world', 'permissions', 'url', 'address'],
                additionalProperties: false
            }
        },
        required: ['type', 'subType', 'key', 'timestamp', 'metadata'],
        additionalProperties: false
    };
    WorldsPermissionGrantedEvent.validate = (0, validation_1.generateLazyValidator)(WorldsPermissionGrantedEvent.schema);
})(WorldsPermissionGrantedEvent = exports.WorldsPermissionGrantedEvent || (exports.WorldsPermissionGrantedEvent = {}));
var WorldsAccessRestrictedEvent;
(function (WorldsAccessRestrictedEvent) {
    WorldsAccessRestrictedEvent.schema = {
        type: 'object',
        properties: {
            type: { type: 'string', const: base_1.Events.Type.WORLD },
            subType: { type: 'string', const: base_1.Events.SubType.Worlds.WORLDS_ACCESS_RESTRICTED },
            key: { type: 'string' },
            timestamp: { type: 'number', minimum: 0 },
            metadata: {
                type: 'object',
                properties: {
                    title: { type: 'string' },
                    description: { type: 'string' },
                    when: { type: 'number' },
                    address: { type: 'string' }
                },
                required: ['title', 'description', 'when', 'address'],
                additionalProperties: false
            }
        },
        required: ['type', 'subType', 'key', 'timestamp', 'metadata'],
        additionalProperties: false
    };
    WorldsAccessRestrictedEvent.validate = (0, validation_1.generateLazyValidator)(WorldsAccessRestrictedEvent.schema);
})(WorldsAccessRestrictedEvent = exports.WorldsAccessRestrictedEvent || (exports.WorldsAccessRestrictedEvent = {}));
var WorldsMissingResourcesEvent;
(function (WorldsMissingResourcesEvent) {
    WorldsMissingResourcesEvent.schema = {
        type: 'object',
        properties: {
            type: { type: 'string', const: base_1.Events.Type.WORLD },
            subType: { type: 'string', const: base_1.Events.SubType.Worlds.WORLDS_MISSING_RESOURCES },
            key: { type: 'string' },
            timestamp: { type: 'number', minimum: 0 },
            metadata: {
                type: 'object',
                properties: {
                    title: { type: 'string' },
                    description: { type: 'string' },
                    url: { type: 'string' },
                    when: { type: 'number' },
                    address: { type: 'string' }
                },
                required: ['title', 'description', 'url', 'when', 'address'],
                additionalProperties: false
            }
        },
        required: ['type', 'subType', 'key', 'timestamp', 'metadata'],
        additionalProperties: false
    };
    WorldsMissingResourcesEvent.validate = (0, validation_1.generateLazyValidator)(WorldsMissingResourcesEvent.schema);
})(WorldsMissingResourcesEvent = exports.WorldsMissingResourcesEvent || (exports.WorldsMissingResourcesEvent = {}));
