// @dcl/sdk-commands/dist/logic/bundle.js
"use strict";
// this file is tested extensively to build scenes in the `make test` command
// but since it runs outiside the testing harness, coverage is not collected
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.bundleProject = bundleProject;
exports.bundleSingleProject = bundleSingleProject;
exports.getScriptImportName = getScriptImportName;
exports.generateInitializeScriptsModule = generateInitializeScriptsModule;
const child_process_1 = __importDefault(require("child_process"));
const esbuild_1 = __importDefault(require("esbuild"));
const fp_future_1 = require("fp-future");
const glob_1 = require("glob");
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const url_1 = require("url");
const i18next_1 = __importDefault(require("i18next"));
const log_1 = require("../components/log");
const beautiful_logs_1 = require("./beautiful-logs");
const error_1 = require("./error");
const composite_1 = require("./composite");
const project_validations_1 = require("./project-validations");
const chokidar_1 = require("chokidar");
const debounce_1 = require("./debounce");
const MAX_STEP = 2;
// Keep only the last 10KB of output to prevent memory leaks in watch mode
const MAX_OUTPUT_SIZE = 10 * 1024;
/**
 * Generate the entry-point code for a given original entry-point
 * @param entrypointPath - file to be imported as original entry point
 * @param forceCustomExport
 * @returns the Typescript code
 */
/**
 * Converts a file path to a safe JavaScript module path string.
 * Normalizes to Unix-style forward slashes and escapes special characters.
 */
function toSafeModulePath(filePath) {
    const unixPath = filePath.replace(/\\/g, '/');
    return JSON.stringify(unixPath);
}
function getEntrypointCode(entrypointPath, forceCustomExport, isEditorScene = false) {
    const safeEntrypointPath = toSafeModulePath(entrypointPath);
    if (forceCustomExport)
        return `;"use strict";export * from ${safeEntrypointPath}`;
    return `// BEGIN AUTO GENERATED CODE "~sdk/scene-entrypoint"
"use strict";
import * as entrypoint from ${safeEntrypointPath}
import { engine, NetworkEntity } from '@dcl/sdk/ecs'
import * as sdk from '@dcl/sdk'
import { compositeProvider } from '@dcl/sdk/composite-provider'
import { compositeFromLoader } from '~sdk/all-composites'
import { _initializeScripts } from '~sdk/script-utils'

${isEditorScene &&
        `
import { syncEntity } from '@dcl/sdk/network'
import players from '@dcl/sdk/players'
import { initAssetPacks, setSyncEntity } from '@dcl/asset-packs/dist/scene-entrypoint'
initAssetPacks(engine, { syncEntity }, players)

// TODO: do we need to do this on runtime ?
// I think we have that information at build-time and we avoid to do evaluate this on the worker.
// Read composite.json or main.crdt => If that file has a NetworkEntity import '@dcl/@sdk/network'
`}

if ((entrypoint as any).main !== undefined) {
  function _INTERNAL_startup_system() {
    try {
      // initialize and run all scripts
      _initializeScripts(engine)

      const maybePromise = (entrypoint as any).main()
      if (maybePromise && typeof maybePromise === 'object' && typeof (maybePromise as unknown as Promise<unknown>).then === 'function') {
        maybePromise.catch(console.error)
      }
    } catch (e) {
     console.error(e)
    } finally {
      engine.removeSystem(_INTERNAL_startup_system)
    }
  }
  engine.addSystem(_INTERNAL_startup_system, Infinity)
}

export * from '@dcl/sdk'
export * from ${safeEntrypointPath}
export * from '~sdk/script-utils'
`;
}
async function bundleProject(components, options, sceneJson) {
    const tsconfig = path_1.default.join(options.workingDirectory, 'tsconfig.json');
    /* istanbul ignore if */
    if (!options.single && !sceneJson.main) {
        throw new error_1.CliError('BUNDLE_SCENE_MAIN_REQUIRED', i18next_1.default.t('errors.bundle.scene_main_required'));
    }
    /* istanbul ignore if */
    if (sceneJson.runtimeVersion !== '7') {
        throw new error_1.CliError('BUNDLE_SCENE_RUNTIME_VERSION_REQUIRED', i18next_1.default.t('errors.bundle.scene_runtime_version_required'));
    }
    /* istanbul ignore if */
    if (!(await components.fs.fileExists(tsconfig))) {
        throw new error_1.CliError('BUNDLE_TSCONFIG_REQUIRED', i18next_1.default.t('errors.bundle.tsconfig_required', { tsconfig }));
    }
    const entrypointSource = options.single || 'src/index.ts';
    const entrypoints = (0, glob_1.globSync)(entrypointSource, { cwd: options.workingDirectory, absolute: true });
    /* istanbul ignore if */
    if (!entrypoints.length)
        throw new error_1.CliError('BUNDLE_NO_INPUT_FILES', i18next_1.default.t('errors.bundle.no_input_files', { entrypointSource }));
    // const output = !options.single ? sceneJson.main : options.single.replace(/\.ts$/, '.js')
    // const outfile = path.join(options.workingDirectory, output)
    const inputs = options.single
        ? entrypoints.map((entrypoint) => ({ entrypoint, outputFile: entrypoint.replace(/\.ts$/, '.js') }))
        : [{ entrypoint: entrypoints[0], outputFile: sceneJson.main }];
    for (const input of inputs) {
        await bundleSingleProject(components, {
            ...options,
            tsconfig,
            ...input
        });
    }
    return { sceneJson, inputs };
}
async function bundleSingleProject(components, options) {
    (0, beautiful_logs_1.printProgressStep)(components.logger, `Bundling file ${log_1.colors.bold(options.entrypoint)}`, 1, MAX_STEP);
    const editorScene = await (0, project_validations_1.isEditorScene)(components, options.workingDirectory);
    // Pre-compute composite data so we can inject maxCompositeEntity via esbuild define.
    // This must happen before the esbuild context is created because the define values
    // are baked into the engine at compile time (the entity counter initializer reads it)
    let maxCompositeEntity = 0;
    if (!options.ignoreComposite) {
        const composites = await (0, composite_1.getAllComposites)(components, options.workingDirectory);
        maxCompositeEntity = composites.maxCompositeEntity;
    }
    const sdkPackagePath = (() => {
        try {
            // First try to resolve from project's node_modules
            return path_1.default.dirname(require.resolve('@dcl/sdk/package.json', { paths: [options.workingDirectory] }));
        }
        catch {
            // Fallback to workspace @dcl/sdk
            return path_1.default.dirname(require.resolve('@dcl/sdk/package.json', { paths: [__dirname] }));
        }
    })();
    const context = await esbuild_1.default.context({
        bundle: true,
        platform: 'browser',
        format: 'cjs',
        preserveSymlinks: false,
        outfile: options.outputFile,
        allowOverwrite: false,
        sourcemap: options.production ? 'external' : 'inline',
        minify: options.production,
        minifyIdentifiers: options.production,
        minifySyntax: options.production,
        minifyWhitespace: options.production,
        treeShaking: true,
        metafile: true,
        absWorkingDir: options.workingDirectory,
        target: 'es2020',
        external: ['~system/*', '@dcl/inspector', '@dcl/inspector/*' /* ban importing the inspector from the SDK */],
        alias: {
            // Ensure React is always resolved to the same module to prevent duplication
            react: (() => {
                try {
                    // First try to resolve from project's node_modules
                    return require.resolve('react', { paths: [options.workingDirectory] });
                }
                catch {
                    try {
                        // Fallback to SDK's React dependency
                        return require.resolve('react', { paths: [path_1.default.join(__dirname, '../../../@dcl/sdk')] });
                    }
                    catch {
                        // Final fallback to bundled React
                        return require.resolve('react');
                    }
                }
            })(),
            // Ensure @dcl/sdk is always resolved to workspace version to prevent version conflicts
            '@dcl/sdk': sdkPackagePath,
            // Resolve ecs from sdk dependencies (nested in @dcl/sdk)
            '@dcl/ecs': (() => {
                try {
                    // First try to resolve from project's @dcl/sdk node_modules
                    return path_1.default.dirname(require.resolve('@dcl/ecs/package.json', { paths: [sdkPackagePath] }));
                }
                catch {
                    try {
                        // Fallback: try to resolve from project's node_modules
                        return path_1.default.dirname(require.resolve('@dcl/ecs/package.json', { paths: [options.workingDirectory] }));
                    }
                    catch {
                        // Last resort: try resolving from current directory
                        return path_1.default.dirname(require.resolve('@dcl/ecs/package.json', { paths: [__dirname] }));
                    }
                }
            })(),
            // Resolve asset-packs from the scene's own node_modules (if the user explicitly installed it),
            // otherwise fall back to the version bundled inside @dcl/inspector.
            // NOTE: We use a direct path check (fs.existsSync) instead of require.resolve here because
            // require.resolve walks UP the directory tree from workingDirectory, which would incorrectly
            // pick up @dcl/asset-packs installed next to the scene (e.g. at a monorepo root) rather
            // than the one the user intentionally installed inside the scene.
            '@dcl/asset-packs': (() => {
                const sceneOwnAssetPacks = path_1.default.join(options.workingDirectory, 'node_modules', '@dcl', 'asset-packs', 'package.json');
                if (fs_1.default.existsSync(sceneOwnAssetPacks)) {
                    return path_1.default.dirname(sceneOwnAssetPacks);
                }
                try {
                    // Fallback: resolve from @dcl/inspector's node_modules
                    const inspectorPath = require.resolve('@dcl/inspector/package.json', { paths: [__dirname] });
                    return path_1.default.dirname(require.resolve('@dcl/asset-packs/package.json', { paths: [path_1.default.dirname(inspectorPath)] }));
                }
                catch {
                    // Last resort: try resolving from current directory
                    return path_1.default.dirname(require.resolve('@dcl/asset-packs/package.json', { paths: [__dirname] }));
                }
            })()
        },
        // convert filesystem paths into file:// to enable VSCode debugger
        sourceRoot: options.production ? 'dcl:///' : (0, url_1.pathToFileURL)(path_1.default.dirname(options.outputFile)).toString(),
        define: {
            document: 'undefined',
            window: 'undefined',
            DEBUG: options.production ? 'false' : 'true',
            'globalThis.DEBUG': options.production ? 'false' : 'true',
            'process.env.NODE_ENV': JSON.stringify(options.production ? 'production' : 'development'),
            DCL_MAX_COMPOSITE_ENTITY: String(maxCompositeEntity)
        },
        tsconfig: options.tsconfig,
        supported: {
            'import-assertions': false,
            'import-meta': false,
            'dynamic-import': false,
            hashbang: false
        },
        logOverride: {
            'import-is-undefined': 'silent'
        },
        plugins: [compositeLoader(components, options)],
        stdin: {
            contents: getEntrypointCode(options.entrypoint, options.customEntryPoint, editorScene),
            resolveDir: path_1.default.dirname(options.entrypoint),
            sourcefile: path_1.default.basename(options.entrypoint) + '.entry-point.ts',
            loader: 'ts'
        }
    });
    /* istanbul ignore if */
    if (options.watch) {
        // Instead of using esbuild's watch, we create our own watcher
        const watcher = (0, chokidar_1.watch)(path_1.default.resolve(options.workingDirectory), {
            ignored: ['**/dist/**', '**/*.crdt', '**/*.d.ts', path_1.default.resolve(options.outputFile)],
            ignoreInitial: true
        });
        const debouncedRebuild = (0, debounce_1.debounce)(async () => {
            try {
                await context.rebuild();
                (0, beautiful_logs_1.printProgressInfo)(components.logger, `Bundle saved ${log_1.colors.bold(options.outputFile)}`);
            }
            catch (err) {
                /* istanbul ignore next */
                components.logger.error(err.toString());
            }
        }, 100);
        watcher.on('all', async (event, filePath) => {
            // Rebuild for TypeScript, JavaScript, and composite files
            if (/\.(ts|tsx|js|jsx|composite)$/.test(filePath)) {
                (0, beautiful_logs_1.printProgressInfo)(components.logger, `File ${filePath} changed, rebuilding...`);
                debouncedRebuild();
            }
        });
        // Do initial build
        await context.rebuild();
        (0, beautiful_logs_1.printProgressInfo)(components.logger, `Bundle saved ${log_1.colors.bold(options.outputFile)}`);
        (0, beautiful_logs_1.printProgressInfo)(components.logger, `The compiler is watching for changes`);
    }
    else {
        try {
            await context.rebuild();
            (0, beautiful_logs_1.printProgressInfo)(components.logger, `Bundle saved ${log_1.colors.bold(options.outputFile)}`);
        }
        catch (err) {
            /* istanbul ignore next */
            throw new error_1.CliError('BUNDLE_REBUILD_FAILED', i18next_1.default.t('errors.bundle.rebuild_failed', { error: err.toString() }));
        }
        await context.dispose();
    }
    /* istanbul ignore next */
    if (options.watch)
        (0, beautiful_logs_1.printProgressInfo)(components.logger, `The compiler is watching for changes`);
    await runTypeChecker(components, options);
}
function runTypeChecker(components, options) {
    const tsBin = require.resolve('typescript/lib/tsc');
    const args = [
        '-p',
        'tsconfig.json',
        '--preserveWatchOutput',
        options.emitDeclaration ? '--emitDeclarationOnly' : '--noEmit'
    ];
    /* istanbul ignore if */
    if (options.watch)
        args.push('--watch');
    (0, beautiful_logs_1.printProgressStep)(components.logger, `Running type checker`, 2, MAX_STEP);
    const ts = child_process_1.default.fork(tsBin, args, {
        stdio: 'pipe',
        env: process.env,
        cwd: options.workingDirectory
    });
    const typeCheckerFuture = (0, fp_future_1.future)();
    let stdOutput = '';
    ts.stdout?.on('data', (data) => {
        stdOutput = (stdOutput + data.toString()).slice(-MAX_OUTPUT_SIZE);
    });
    ts.stdout?.pipe(process.stdout);
    ts.stderr?.pipe(process.stderr);
    const cleanup = () => {
        if (!ts.killed)
            ts.kill('SIGTERM');
    };
    process.on('SIGTERM', cleanup);
    process.on('SIGINT', cleanup);
    process.on('exit', cleanup);
    ts.on('close', (code) => {
        // Remove listeners after process exits to prevent memory leaks
        process.off('SIGTERM', cleanup);
        process.off('SIGINT', cleanup);
        process.off('exit', cleanup);
        /* istanbul ignore else */
        if (code === 0) {
            (0, beautiful_logs_1.printProgressInfo)(components.logger, `Type checking completed without errors`);
            typeCheckerFuture.resolve(code);
        }
        else {
            typeCheckerFuture.reject(new error_1.CliError('BUNDLE_TYPE_CHECKER_FAILED', `${stdOutput.replace(/\x1b\[[0-9;]*m/g, '')}\n
          ${i18next_1.default.t('errors.bundle.type_checker_failed', { code })}`));
        }
        stdOutput = '';
    });
    /* istanbul ignore if */
    if (options.watch) {
        typeCheckerFuture.resolve(-1);
    }
    return typeCheckerFuture;
}
function compositeLoader(components, options) {
    let shouldReload = true;
    let compositeData = null;
    return {
        name: 'composite-loader',
        setup(build) {
            build.onStart(() => {
                shouldReload = true;
                compositeData = null;
            });
            // Handle composites virtual module
            build.onResolve({ filter: /~sdk\/all-composites/ }, (_args) => {
                return {
                    namespace: 'sdk-composite',
                    path: 'all-composites'
                };
            });
            build.onLoad({ filter: /.*/, namespace: 'sdk-composite' }, async (_) => {
                // Load compositeData if not already loaded
                if (shouldReload && !compositeData) {
                    if (!options.ignoreComposite) {
                        compositeData = await (0, composite_1.getAllComposites)(components, 
                        // we pass the build.initialOptions.absWorkingDir to build projects with multiple roots at once
                        build.initialOptions.absWorkingDir ?? options.workingDirectory);
                        if (compositeData.withErrors) {
                            (0, beautiful_logs_1.printWarning)(components.logger, 'Some composites are not included because of errors while compiling them. There can be unexpected behavior in the scene, check the errors and try to fix them.');
                        }
                    }
                    shouldReload = false;
                }
                const contents = compositeData
                    ? `export const compositeFromLoader = {${compositeData.compositeLines.join(',')}}`
                    : `export const compositeFromLoader = {}`;
                const watchFiles = compositeData?.watchFiles || [];
                return {
                    loader: 'js',
                    contents,
                    watchFiles
                };
            });
            // Handle scripts virtual module
            build.onResolve({ filter: /~sdk\/script-utils/ }, (_args) => {
                return {
                    namespace: 'sdk-scripts',
                    path: 'script-utils'
                };
            });
            build.onLoad({ filter: /.*/, namespace: 'sdk-scripts' }, async (_) => {
                // ensure compositeData is loaded (in case scripts are imported before composites)
                if (!compositeData && !options.ignoreComposite) {
                    compositeData = await (0, composite_1.getAllComposites)(components, build.initialOptions.absWorkingDir ?? options.workingDirectory);
                }
                const { contents, watchFiles } = await generateInitializeScriptsModule(components, options.workingDirectory, compositeData);
                return {
                    loader: 'js',
                    contents,
                    watchFiles,
                    resolveDir: options.workingDirectory
                };
            });
        }
    };
}
/**
 * Generates a sanitized import name for a script path.
 * Example: "src/scripts/my-script.ts" -> "script_src_scripts_my_script"
 */
function getScriptImportName(scriptPath) {
    return ('script_' +
        scriptPath
            .replace(/\.tsx?$/, '') // remove .ts or .tsx
            .replace(/[^a-zA-Z0-9]/g, '_') // sanitize
    );
}
/**
 * Collects and formats all script data for code generation
 */
function collectScriptData(compositeData, workingDirectory) {
    const runtimeImports = [];
    const typeImports = [];
    const scriptTypes = [];
    const scriptsArray = [];
    const watchFiles = [];
    for (const [scriptPath, scriptInstances] of compositeData.scripts.entries()) {
        const importName = getScriptImportName(scriptPath);
        const absolutePath = path_1.default.join(workingDirectory, scriptPath);
        // For the virtual module runtime
        runtimeImports.push(`import * as ${importName} from ${toSafeModulePath('./' + scriptPath)}`);
        // For .d.ts file using import() type syntax (works in ambient modules)
        typeImports.push(`  type ${importName} = typeof import(${toSafeModulePath(absolutePath)})`);
        // Add to ScriptRegistry interface
        scriptTypes.push(`  ${toSafeModulePath(scriptPath)}: ExtractScriptType<${importName}>`);
        // Add each script instance to the array
        for (const script of scriptInstances) {
            scriptsArray.push(`  { ...${JSON.stringify(script)}, module: ${importName} }`);
        }
        watchFiles.push(absolutePath);
    }
    return {
        runtimeImports: runtimeImports.join('\n'),
        typeImports: typeImports.join('\n'),
        scriptTypes: `interface ScriptRegistry {\n${scriptTypes.join(',\n')}\n}`,
        scriptsArray: `[\n${scriptsArray.join(',\n')}\n]`,
        watchFiles
    };
}
/**
 * Reads and prepares the runtime script code for inlining
 */
async function prepareRuntimeCode(fs) {
    const runtimeCodePath = require.resolve('./runtime-script');
    const runtimeCode = await fs.readFile(runtimeCodePath, 'utf-8');
    // Strip CommonJS/module system code
    return (runtimeCode
        .replace(/"use strict";?\s*/g, '')
        .replace(/Object\.defineProperty\(exports,.*?\);?\s*/g, '')
        .replace(/exports\.\w+\s*=\s*void 0;?\s*/g, '')
        .replace(/exports\.\w+\s*=\s*/g, '')
        .replace(/^export\s+/gm, '')
        .replace(/^import\s+.*$/gm, '')
        // fix nested asset-packs path (importing from @dcl/inspector is banned in runtime, but @dcl/asset-packs not)
        .replace(/@dcl\/inspector\/node_modules\/@dcl\/asset-packs/g, '@dcl/asset-packs')
        .replace(/\n{3,}/g, '\n\n')
        .trim());
}
/**
 * Generates the TypeScript declaration for the ~sdk/script-utils module
 */
async function generateScriptModuleDeclaration(components, typeImports, scriptTypes) {
    const templatePath = path_1.default.join(__dirname, 'script-module.d.ts.template');
    const template = await components.fs.readFile(templatePath, 'utf-8');
    return template.replace('__TYPE_IMPORTS__', typeImports).replace('__SCRIPT_TYPES__', scriptTypes);
}
/**
 * Updates the sdk.d.ts file with script type declarations
 */
async function updateSdkTypeDeclarations(components, workingDirectory, typeImports, scriptTypes, scriptCount) {
    try {
        const jsRuntimePath = require.resolve('@dcl/js-runtime/sdk.d.ts', { paths: [workingDirectory] });
        let existingSdkDts = await components.fs.readFile(jsRuntimePath, 'utf-8');
        // remove any existing ~sdk/script-utils module declaration
        existingSdkDts = existingSdkDts
            .replace(/\/\/ @internal[\s]*\ndeclare module '~sdk\/script-utils'[\s\S]*?(?=\n\/\/|$)/, '')
            .trim();
        // generate and append the new module declaration
        const scriptModuleDeclaration = await generateScriptModuleDeclaration(components, typeImports, scriptTypes);
        const updatedSdkDts = existingSdkDts + '\n' + scriptModuleDeclaration;
        await components.fs.writeFile(jsRuntimePath, updatedSdkDts);
        components.logger.info(`✓ Updated sdk.d.ts with ${scriptCount} script type(s) at: ${jsRuntimePath}`);
    }
    catch (err) {
        components.logger.error('⚠ Could not update sdk.d.ts with script types:', err);
    }
}
/**
 * Generates the virtual module content with script initialization and helper functions
 */
function generateVirtualModuleContent(runtimeImports, runtimeCode, scriptsArray) {
    return `
${runtimeImports}

${runtimeCode}

export function _initializeScripts(engine) {
  const scriptsArray = ${scriptsArray}
  return runScripts(engine, scriptsArray)
}

// export helper functions that are defined in the inlined runtime code
export { getScriptInstance, getScriptInstancesByPath, getAllScriptInstances, callScriptMethod }
`;
}
async function generateInitializeScriptsModule(components, workingDirectory, compositeData) {
    // prepare runtime code (always needed for helper functions)
    const runtimeCode = await prepareRuntimeCode(components.fs);
    // default empty implementation if no scripts
    if (!compositeData || compositeData.scripts.size === 0) {
        return {
            contents: generateVirtualModuleContent('', runtimeCode, '[]'),
            watchFiles: []
        };
    }
    // Step 1: Collect all script data
    const scriptData = collectScriptData(compositeData, workingDirectory);
    // Step 2: Update TypeScript declarations
    await updateSdkTypeDeclarations(components, workingDirectory, scriptData.typeImports, scriptData.scriptTypes, compositeData.scripts.size);
    // Step 3: Generate virtual module content
    const contents = generateVirtualModuleContent(scriptData.runtimeImports, runtimeCode, scriptData.scriptsArray);
    return { contents, watchFiles: scriptData.watchFiles };
}
