// @dcl/sdk-commands/dist/logic/lang.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Language = void 0;
exports.getLanguage = getLanguage;
exports.initLanguage = initLanguage;
const path_1 = require("path");
const arg_1 = __importDefault(require("arg"));
const i18next_1 = __importDefault(require("i18next"));
const i18next_fs_backend_1 = __importDefault(require("i18next-fs-backend"));
var Language;
(function (Language) {
    Language["EN"] = "en";
    Language["ES"] = "es";
    Language["ZH"] = "zh";
})(Language || (exports.Language = Language = {}));
function getLanguage(args) {
    try {
        const language = (0, arg_1.default)({ '--language': String }, { permissive: true, argv: args })['--language'];
        return Language[language?.toUpperCase()] || Language.EN;
    }
    catch (e) {
        return Language.EN;
    }
}
function initLanguage(language) {
    return i18next_1.default.use(i18next_fs_backend_1.default).init({
        lng: language,
        fallbackLng: 'en',
        showSupportNotice: false,
        interpolation: {
            escapeValue: false
        },
        backend: {
            loadPath: (0, path_1.join)(__dirname, '../locales/{{lng}}.json')
        },
    });
}
