// @dcl/sdk-commands/dist/logic/project-files.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.b64HashingFunction = exports.machineId = void 0;
exports.getPublishableFiles = getPublishableFiles;
exports.normalizeDecentralandFilename = normalizeDecentralandFilename;
exports.projectFilesToContentMappings = projectFilesToContentMappings;
exports.getProjectPublishableFilesWithHashes = getProjectPublishableFilesWithHashes;
exports.getPackageJson = getPackageJson;
const dcl_ignore_1 = require("./dcl-ignore");
const glob_1 = require("glob");
const ignore_1 = __importDefault(require("ignore"));
const i18next_1 = __importDefault(require("i18next"));
const os_1 = __importDefault(require("os"));
const path_1 = __importStar(require("path"));
const error_1 = require("./error");
/**
 * Returns an array of the publishable files for a given folder.
 *
 */
async function getPublishableFiles(components, projectRoot) {
    const ignorePatterns = await (0, dcl_ignore_1.getDCLIgnorePatterns)(components, projectRoot);
    const ig = (0, ignore_1.default)().add(ignorePatterns);
    const allFiles = (0, glob_1.sync)('**/*', {
        cwd: projectRoot,
        absolute: false,
        dot: false,
        ignore: ignorePatterns,
        nodir: true
    });
    return ig.filter(allFiles);
}
/**
 * This function converts paths to decentraland-compatible paths.
 * - From windows separators to unix separators.
 * - All to lowercase
 */
function normalizeDecentralandFilename(projectRoot, filename) {
    const newAbsolute = path_1.default.resolve(projectRoot, filename);
    const relativePath = path_1.default.relative(projectRoot, newAbsolute);
    // 1. win->unix style
    // 2. remove heading /
    return relativePath.replace(/(\\)/g, '/').replace(/^\/+/, '').toLowerCase();
}
/**
 * This function normalizes the content mappings of a project to be used by the
 * Decentraland file system
 */
function projectFilesToContentMappings(projectRoot, files) {
    return files.map((file) => {
        return {
            file: normalizeDecentralandFilename(projectRoot, file.absolutePath),
            hash: file.hash
        };
    });
}
/**
 * Returns the content mappings for a specific project folder.
 * NOTE: the result of this function IS NOT NORMALIZED. Paths sould be normalized
 * with normalizeDecentralandFilename before usage
 *
 * TODO: Unit test this function
 */
async function getProjectPublishableFilesWithHashes(components, projectRoot, hashingFunction) {
    const projectFiles = await getPublishableFiles(components, projectRoot);
    const ret = [];
    const usedFilenames = new Set();
    for (const file of projectFiles) {
        const absolutePath = path_1.default.resolve(projectRoot, file);
        /* istanbul ignore if */
        if (!(await components.fs.fileExists(absolutePath)))
            continue;
        const normalizedFile = normalizeDecentralandFilename(projectRoot, file);
        /* istanbul ignore if */
        if (usedFilenames.has(normalizedFile)) {
            throw new error_1.CliError('PROJECT_FILES_DUPLICATE_FILE', i18next_1.default.t('errors.project_files.duplicate_file', { file }));
        }
        usedFilenames.add(normalizedFile);
        ret.push({
            absolutePath,
            hash: await hashingFunction(absolutePath)
        });
    }
    return ret;
}
exports.machineId = os_1.default.hostname() || os_1.default.userInfo().username;
const b64HashingFunction = (str) => {
    const unique = `${str}-${exports.machineId}`;
    return 'b64-' + Buffer.from(unique).toString('base64');
};
exports.b64HashingFunction = b64HashingFunction;
/* istanbul ignore next */
async function getPackageJson(components, projectRoot) {
    try {
        const packageJsonRaw = await components.fs.readFile((0, path_1.resolve)(projectRoot, 'package.json'), 'utf8');
        const packageJson = JSON.parse(packageJsonRaw);
        return packageJson;
    }
    catch (err) {
        throw new error_1.CliError('PROJECT_FILES_INVALID_PACKAGE_JSON', i18next_1.default.t('errors.project_files.invalid_package_json', { error: err.message }));
    }
}
