// @dcl/sdk-commands/dist/logic/runtime-script.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getScriptInstance = getScriptInstance;
exports.getScriptInstancesByPath = getScriptInstancesByPath;
exports.getAllScriptInstances = getAllScriptInstances;
exports.callScriptMethod = callScriptMethod;
exports.runScripts = runScripts;
/* eslint-disable no-console */
const dist_cjs_1 = require("@dcl/ecs/dist-cjs");
const asset_packs_1 = require("@dcl/inspector/node_modules/@dcl/asset-packs");
if (!globalThis.__DCL_SCRIPT_INSTANCES__) {
    globalThis.__DCL_SCRIPT_INSTANCES__ = new Map();
}
function entityIsRemoved(engine, entity) {
    return engine.getEntityState(entity) === dist_cjs_1.EntityState.Removed;
}
/**
 * Get a specific script instance by entity and script path.
 * @param entity - The entity ID
 * @param scriptPath - The script file path
 * @returns The script instance or null if not found
 */
function getScriptInstance(entity, scriptPath) {
    const key = `${entity}:${scriptPath}`;
    const entry = globalThis.__DCL_SCRIPT_INSTANCES__.get(key);
    return entry?.instance || null;
}
/**
 * Get all instances of a specific script (across all entities).
 * @param scriptPath - The script file path
 * @returns Array of { entity, instance } objects
 */
function getScriptInstancesByPath(scriptPath) {
    const results = [];
    for (const [_, value] of globalThis.__DCL_SCRIPT_INSTANCES__) {
        if (value.path === scriptPath) {
            results.push({ entity: value.entity, instance: value.instance });
        }
    }
    return results;
}
/**
 * Get all script instances attached to a specific entity
 * @param entity - The entity ID
 * @returns Array of { path, instance } objects
 */
function getAllScriptInstances(entity) {
    const results = [];
    for (const [_, value] of globalThis.__DCL_SCRIPT_INSTANCES__) {
        if (value.entity === entity) {
            results.push({ path: value.path, instance: value.instance });
        }
    }
    return results;
}
/**
 * Call a method on a script instance (with safety checks).
 * @param entity - The entity ID
 * @param scriptPath - The script file path
 * @param methodName - The method name to call
 * @param args - Arguments to pass to the method
 * @returns The method's return value or undefined if method not found
 */
function callScriptMethod(entity, scriptPath, methodName, ...args) {
    const instance = getScriptInstance(entity, scriptPath);
    if (instance && typeof instance[methodName] === 'function') {
        return instance[methodName](...args);
    }
    console.error(`Method ${methodName} not found on script ${scriptPath} for entity ${entity}`);
    return undefined;
}
/**
 * Creates an ActionCallback function from an ActionRef.
 * The returned function, when called, will trigger the specified action on the entity.
 *
 * @param actionRef - The action reference containing entity and action name
 * @returns A function that triggers the action when called
 */
function createActionCallback(actionRef) {
    return () => {
        if (!actionRef.entity || !actionRef.action) {
            console.error('[Script] ActionCallback called with invalid action reference:', actionRef);
            return;
        }
        const actionEvents = (0, asset_packs_1.getActionEvents)(actionRef.entity);
        actionEvents.emit(actionRef.action, {});
    };
}
/**
 * Resolves script parameters, converting ActionRef values to ActionCallback functions.
 *
 * @param params - The raw parameters from the script layout
 * @returns Array of resolved parameter values
 */
function resolveScriptParams(params) {
    return Object.values(params).map((param) => {
        if (param.type === 'action' && param.value && typeof param.value === 'object') {
            const actionRef = param.value;
            return createActionCallback(actionRef);
        }
        return param.value;
    });
}
/**
 * Initializes and runs all scripts organized by priority.
 * Supports both functional-style scripts (with start/update functions) and class-based scripts.
 * Scripts are extracted at build time from composites.
 *
 * @internal This function is called automatically by the SDK entry point.
 * Users should not call this function directly.
 *
 * @param engine - The ECS engine instance
 * @param scripts - Scripts with their modules, extracted at build time
 */
function runScripts(engine, scripts) {
    const scriptsByPriority = groupScriptsByPriority(scripts);
    const classInstances = new Map();
    const functionScripts = new Map();
    for (const [priority, instances] of Object.entries(scriptsByPriority)) {
        for (const script of instances) {
            if (entityIsRemoved(engine, script.entity))
                continue;
            const module = script.module;
            if (!module) {
                console.error('[Script] Unknown module:', script.path);
                continue;
            }
            const src = script.path.split('/').slice(0, -1).join('/');
            const layout = script.layout ? JSON.parse(script.layout) : {};
            const params = resolveScriptParams(layout.params || {});
            const registryKey = `${script.entity}:${script.path}`;
            if (typeof module.start === 'function') {
                try {
                    module.start(src, script.entity, ...params);
                }
                catch (e) {
                    console.error('[Script Error] ' + script.path + ' start() failed:', e);
                    throw e;
                }
                functionScripts.set(script.key, {
                    src,
                    module: module,
                    entity: script.entity,
                    params
                });
                globalThis.__DCL_SCRIPT_INSTANCES__.set(registryKey, {
                    instance: module,
                    entity: script.entity,
                    path: script.path
                });
            }
            else {
                const ScriptClass = Object.values(module).find((exp) => typeof exp === 'function');
                if (!ScriptClass) {
                    console.error('[Script] No class found in module:', script.path);
                    continue;
                }
                try {
                    const instance = new ScriptClass(src, script.entity, ...params);
                    if (typeof instance.start === 'function') {
                        instance.start();
                    }
                    classInstances.set(script.key, { instance, entity: script.entity });
                    globalThis.__DCL_SCRIPT_INSTANCES__.set(registryKey, {
                        instance: instance,
                        entity: script.entity,
                        path: script.path
                    });
                }
                catch (e) {
                    console.error('[Script Error] ' + script.path + ' class initialization failed:', e);
                    throw e;
                }
            }
        }
        engine.addSystem((dt) => {
            for (const scriptData of instances) {
                try {
                    const classInstance = classInstances.get(scriptData.key);
                    if (classInstance && !entityIsRemoved(engine, classInstance.entity)) {
                        if (typeof classInstance.instance.update === 'function') {
                            classInstance.instance.update(dt);
                        }
                    }
                    else {
                        const functionScript = functionScripts.get(scriptData.key);
                        if (functionScript && !entityIsRemoved(engine, functionScript.entity)) {
                            if (typeof functionScript.module.update === 'function') {
                                functionScript.module.update(functionScript.src, functionScript.entity, dt, ...functionScript.params);
                            }
                        }
                    }
                }
                catch (e) {
                    console.error('[Script Error] update() failed:', e);
                }
            }
        }, Number(priority));
    }
}
function groupScriptsByPriority(scripts) {
    const scriptsByPriority = {};
    for (let i = 0; i < scripts.length; i++) {
        const script = scripts[i];
        const key = `${script.entity}:${script.path}:${i}`;
        if (!scriptsByPriority[script.priority]) {
            scriptsByPriority[script.priority] = [];
        }
        scriptsByPriority[script.priority].push({ ...script, key });
    }
    return scriptsByPriority;
}
