// @dcl/sdk-commands/dist/logic/scene-validations.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MAX_FILE_SIZE_BYTES = exports.SCENE_FILE = void 0;
exports.getSceneFilePath = getSceneFilePath;
exports.assertValidScene = assertValidScene;
exports.getValidSceneJson = getValidSceneJson;
exports.getBaseCoords = getBaseCoords;
exports.getFiles = getFiles;
exports.validateFilesSizes = validateFilesSizes;
const path_1 = require("path");
const schemas_1 = require("@dcl/schemas");
const dist_cjs_1 = require("@dcl/ecs/dist-cjs");
const i18next_1 = __importDefault(require("i18next"));
const project_1 = require("../commands/init/project");
const error_1 = require("./error");
const coordinates_1 = require("./coordinates");
const project_files_1 = require("./project-files");
const beautiful_logs_1 = require("./beautiful-logs");
exports.SCENE_FILE = 'scene.json';
exports.MAX_FILE_SIZE_BYTES = 50 * 1e6; // 50mb
/**
 * Composes the path to the `scene.json` file based on the provided path.
 * @param projectRoot The path to the directory containing the scene file.
 */
function getSceneFilePath(projectRoot) {
    return (0, path_1.resolve)(projectRoot, exports.SCENE_FILE);
}
function getWorldRangesConstraintsMessage() {
    const ranges = (0, schemas_1.getWorld)().validWorldRanges;
    let str = '';
    for (const range of ranges) {
        str += `"x" from ${range.xMin} to ${range.xMax} and "y" from ${range.yMin} to ${range.yMax}\n`;
    }
    return str;
}
function checkMissingOrDefault(obj, defaults) {
    const missingKeys = Object.entries(defaults).reduce((acc, [key, value]) => {
        return obj[key] && obj[key] !== value ? acc : acc.concat(key);
    }, []);
    return missingKeys;
}
function assertValidScene(components, scene, opts = { log: false }) {
    if (!schemas_1.Scene.validate(scene)) {
        const errors = [];
        if (schemas_1.Scene.validate.errors) {
            for (const error of schemas_1.Scene.validate.errors) {
                const errorPath = error.instancePath.slice(1).replace(/\//g, ' => ');
                errors.push(`Error validating scene.json: ${errorPath} ${error.message}`);
            }
        }
        throw new error_1.CliError('SCENE_VALIDATIONS_INVALID_SCENE_JSON', i18next_1.default.t('errors.scene_validations.invalid_scene_json', { errors: errors.join('\n') }));
    }
    const parcelSet = new Set(scene.scene?.parcels);
    if (parcelSet.size < scene.scene?.parcels?.length) {
        throw new error_1.CliError('SCENE_VALIDATIONS_DUPLICATE_PARCELS', i18next_1.default.t('errors.scene_validations.duplicate_parcels'));
    }
    if (!parcelSet.has(scene.scene?.base)) {
        throw new error_1.CliError('SCENE_VALIDATIONS_BASE_PARCEL_REQUIRED', i18next_1.default.t('errors.scene_validations.base_parcel_required', { base: scene.scene?.base }), scene.scene?.base);
    }
    const objParcels = scene.scene?.parcels?.map(coordinates_1.getObject);
    objParcels.forEach(({ x, y }) => {
        if ((0, schemas_1.isInsideWorldLimits)(x, y)) {
            return;
        }
        const constraints = getWorldRangesConstraintsMessage();
        throw new error_1.CliError('SCENE_VALIDATIONS_COORDINATES_OUTSIDE_LIMITS', i18next_1.default.t('errors.scene_validations.coordinates_outside_limits', { x, y, constraints }));
    });
    if (!(0, dist_cjs_1.areConnected)(objParcels)) {
        throw new error_1.CliError('SCENE_VALIDATIONS_PARCELS_NOT_CONNECTED', i18next_1.default.t('errors.scene_validations.parcels_not_connected'));
    }
    if (!scene.main?.endsWith('.js')) {
        throw new error_1.CliError('SCENE_VALIDATIONS_INVALID_MAIN_FORMAT', i18next_1.default.t('errors.scene_validations.invalid_main_format', { main: scene.main }), scene.main);
    }
    const minimalScene = (0, project_1.getMinimalSceneJson)();
    const defaults = { ...minimalScene.display, navmapThumbnail: 'images/scene-thumbnail.png' };
    const missingKeys = checkMissingOrDefault(scene.display ?? {}, defaults);
    if (missingKeys.length && opts.log) {
        const missingKeysMsg = missingKeys.join(', ');
        (0, beautiful_logs_1.printWarning)(components.logger, `Don't forget to update your scene.json metadata: [${missingKeysMsg}]
https://docs.decentraland.org/creator/development-guide/scene-metadata/#scene-title-description-and-image`);
    }
}
/**
 * Get valid Scene JSON
 */
async function getValidSceneJson(components, projectRoot, opts) {
    try {
        const sceneJsonRaw = await components.fs.readFile(getSceneFilePath(projectRoot), 'utf8');
        const sceneJson = JSON.parse(sceneJsonRaw);
        assertValidScene(components, sceneJson, opts);
        return sceneJson;
    }
    catch (err) {
        throw new error_1.CliError('SCENE_VALIDATIONS_INVALID_SCENE_JSON', i18next_1.default.t('errors.scene_validations.invalid_scene_json', { error: err.message }));
    }
}
function getBaseCoords(scene) {
    const [x, y] = scene.scene.base
        .replace(/\ /g, '')
        .split(',')
        .map((a) => parseInt(a));
    return { x, y };
}
/**
 * Returns a promise of an array of objects containing the path and the content for all the files in the project.
 * All the paths added to the `.dclignore` file will be excluded from the results.
 * Windows directory separators are replaced for POSIX separators.
 * @param ignoreFile The contents of the .dclignore file
 */
async function getFiles(components, dir) {
    const files = await (0, project_files_1.getPublishableFiles)(components, dir);
    const data = [];
    for (let i = 0; i < files.length; i++) {
        const file = files[i];
        const filePath = (0, path_1.resolve)(dir, file);
        const stat = await components.fs.stat(filePath);
        const content = await components.fs.readFile(filePath);
        data.push({
            path: file.replace(/\\/g, '/'),
            content: Buffer.from(content),
            size: stat.size
        });
    }
    return data;
}
function validateFilesSizes(files) {
    for (const { path, size } of files) {
        if (size > exports.MAX_FILE_SIZE_BYTES) {
            const maxSizeInMb = exports.MAX_FILE_SIZE_BYTES / 1e6;
            throw new error_1.CliError('SCENE_VALIDATIONS_MAX_FILE_SIZE_EXCEEDED', i18next_1.default.t('errors.scene_validations.max_file_size_exceeded', { path, maxSizeInMb }));
        }
    }
}
