// @dcl/protocol/out-ts/decentraland/social_service/errors.gen.ts
/* eslint-disable */
import _m0 from "protobufjs/minimal";

export const protobufPackage = "decentraland.social_service";

export interface InvalidFriendshipAction {
  message?: string | undefined;
}

export interface InternalServerError {
  message?: string | undefined;
}

export interface InvalidRequest {
  message?: string | undefined;
}

export interface ProfileNotFound {
  message?: string | undefined;
}

export interface ConflictingError {
  message?: string | undefined;
}

export interface ForbiddenError {
  message?: string | undefined;
}

export interface NotFoundError {
  message?: string | undefined;
}

function createBaseInvalidFriendshipAction(): InvalidFriendshipAction {
  return { message: undefined };
}

export namespace InvalidFriendshipAction {
  export function encode(message: InvalidFriendshipAction, writer: _m0.Writer = _m0.Writer.create()): _m0.Writer {
    if (message.message !== undefined) {
      writer.uint32(10).string(message.message);
    }
    return writer;
  }

  export function decode(input: _m0.Reader | Uint8Array, length?: number): InvalidFriendshipAction {
    const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseInvalidFriendshipAction();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          if (tag !== 10) {
            break;
          }

          message.message = reader.string();
          continue;
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skipType(tag & 7);
    }
    return message;
  }

  export function fromJSON(object: any): InvalidFriendshipAction {
    return { message: isSet(object.message) ? String(object.message) : undefined };
  }

  export function toJSON(message: InvalidFriendshipAction): unknown {
    const obj: any = {};
    message.message !== undefined && (obj.message = message.message);
    return obj;
  }

  export function create<I extends Exact<DeepPartial<InvalidFriendshipAction>, I>>(base?: I): InvalidFriendshipAction {
    return InvalidFriendshipAction.fromPartial(base ?? {});
  }

  export function fromPartial<I extends Exact<DeepPartial<InvalidFriendshipAction>, I>>(
    object: I,
  ): InvalidFriendshipAction {
    const message = createBaseInvalidFriendshipAction();
    message.message = object.message ?? undefined;
    return message;
  }
}

function createBaseInternalServerError(): InternalServerError {
  return { message: undefined };
}

export namespace InternalServerError {
  export function encode(message: InternalServerError, writer: _m0.Writer = _m0.Writer.create()): _m0.Writer {
    if (message.message !== undefined) {
      writer.uint32(10).string(message.message);
    }
    return writer;
  }

  export function decode(input: _m0.Reader | Uint8Array, length?: number): InternalServerError {
    const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseInternalServerError();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          if (tag !== 10) {
            break;
          }

          message.message = reader.string();
          continue;
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skipType(tag & 7);
    }
    return message;
  }

  export function fromJSON(object: any): InternalServerError {
    return { message: isSet(object.message) ? String(object.message) : undefined };
  }

  export function toJSON(message: InternalServerError): unknown {
    const obj: any = {};
    message.message !== undefined && (obj.message = message.message);
    return obj;
  }

  export function create<I extends Exact<DeepPartial<InternalServerError>, I>>(base?: I): InternalServerError {
    return InternalServerError.fromPartial(base ?? {});
  }

  export function fromPartial<I extends Exact<DeepPartial<InternalServerError>, I>>(object: I): InternalServerError {
    const message = createBaseInternalServerError();
    message.message = object.message ?? undefined;
    return message;
  }
}

function createBaseInvalidRequest(): InvalidRequest {
  return { message: undefined };
}

export namespace InvalidRequest {
  export function encode(message: InvalidRequest, writer: _m0.Writer = _m0.Writer.create()): _m0.Writer {
    if (message.message !== undefined) {
      writer.uint32(10).string(message.message);
    }
    return writer;
  }

  export function decode(input: _m0.Reader | Uint8Array, length?: number): InvalidRequest {
    const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseInvalidRequest();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          if (tag !== 10) {
            break;
          }

          message.message = reader.string();
          continue;
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skipType(tag & 7);
    }
    return message;
  }

  export function fromJSON(object: any): InvalidRequest {
    return { message: isSet(object.message) ? String(object.message) : undefined };
  }

  export function toJSON(message: InvalidRequest): unknown {
    const obj: any = {};
    message.message !== undefined && (obj.message = message.message);
    return obj;
  }

  export function create<I extends Exact<DeepPartial<InvalidRequest>, I>>(base?: I): InvalidRequest {
    return InvalidRequest.fromPartial(base ?? {});
  }

  export function fromPartial<I extends Exact<DeepPartial<InvalidRequest>, I>>(object: I): InvalidRequest {
    const message = createBaseInvalidRequest();
    message.message = object.message ?? undefined;
    return message;
  }
}

function createBaseProfileNotFound(): ProfileNotFound {
  return { message: undefined };
}

export namespace ProfileNotFound {
  export function encode(message: ProfileNotFound, writer: _m0.Writer = _m0.Writer.create()): _m0.Writer {
    if (message.message !== undefined) {
      writer.uint32(10).string(message.message);
    }
    return writer;
  }

  export function decode(input: _m0.Reader | Uint8Array, length?: number): ProfileNotFound {
    const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseProfileNotFound();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          if (tag !== 10) {
            break;
          }

          message.message = reader.string();
          continue;
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skipType(tag & 7);
    }
    return message;
  }

  export function fromJSON(object: any): ProfileNotFound {
    return { message: isSet(object.message) ? String(object.message) : undefined };
  }

  export function toJSON(message: ProfileNotFound): unknown {
    const obj: any = {};
    message.message !== undefined && (obj.message = message.message);
    return obj;
  }

  export function create<I extends Exact<DeepPartial<ProfileNotFound>, I>>(base?: I): ProfileNotFound {
    return ProfileNotFound.fromPartial(base ?? {});
  }

  export function fromPartial<I extends Exact<DeepPartial<ProfileNotFound>, I>>(object: I): ProfileNotFound {
    const message = createBaseProfileNotFound();
    message.message = object.message ?? undefined;
    return message;
  }
}

function createBaseConflictingError(): ConflictingError {
  return { message: undefined };
}

export namespace ConflictingError {
  export function encode(message: ConflictingError, writer: _m0.Writer = _m0.Writer.create()): _m0.Writer {
    if (message.message !== undefined) {
      writer.uint32(10).string(message.message);
    }
    return writer;
  }

  export function decode(input: _m0.Reader | Uint8Array, length?: number): ConflictingError {
    const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseConflictingError();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          if (tag !== 10) {
            break;
          }

          message.message = reader.string();
          continue;
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skipType(tag & 7);
    }
    return message;
  }

  export function fromJSON(object: any): ConflictingError {
    return { message: isSet(object.message) ? String(object.message) : undefined };
  }

  export function toJSON(message: ConflictingError): unknown {
    const obj: any = {};
    message.message !== undefined && (obj.message = message.message);
    return obj;
  }

  export function create<I extends Exact<DeepPartial<ConflictingError>, I>>(base?: I): ConflictingError {
    return ConflictingError.fromPartial(base ?? {});
  }

  export function fromPartial<I extends Exact<DeepPartial<ConflictingError>, I>>(object: I): ConflictingError {
    const message = createBaseConflictingError();
    message.message = object.message ?? undefined;
    return message;
  }
}

function createBaseForbiddenError(): ForbiddenError {
  return { message: undefined };
}

export namespace ForbiddenError {
  export function encode(message: ForbiddenError, writer: _m0.Writer = _m0.Writer.create()): _m0.Writer {
    if (message.message !== undefined) {
      writer.uint32(10).string(message.message);
    }
    return writer;
  }

  export function decode(input: _m0.Reader | Uint8Array, length?: number): ForbiddenError {
    const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseForbiddenError();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          if (tag !== 10) {
            break;
          }

          message.message = reader.string();
          continue;
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skipType(tag & 7);
    }
    return message;
  }

  export function fromJSON(object: any): ForbiddenError {
    return { message: isSet(object.message) ? String(object.message) : undefined };
  }

  export function toJSON(message: ForbiddenError): unknown {
    const obj: any = {};
    message.message !== undefined && (obj.message = message.message);
    return obj;
  }

  export function create<I extends Exact<DeepPartial<ForbiddenError>, I>>(base?: I): ForbiddenError {
    return ForbiddenError.fromPartial(base ?? {});
  }

  export function fromPartial<I extends Exact<DeepPartial<ForbiddenError>, I>>(object: I): ForbiddenError {
    const message = createBaseForbiddenError();
    message.message = object.message ?? undefined;
    return message;
  }
}

function createBaseNotFoundError(): NotFoundError {
  return { message: undefined };
}

export namespace NotFoundError {
  export function encode(message: NotFoundError, writer: _m0.Writer = _m0.Writer.create()): _m0.Writer {
    if (message.message !== undefined) {
      writer.uint32(10).string(message.message);
    }
    return writer;
  }

  export function decode(input: _m0.Reader | Uint8Array, length?: number): NotFoundError {
    const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseNotFoundError();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          if (tag !== 10) {
            break;
          }

          message.message = reader.string();
          continue;
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skipType(tag & 7);
    }
    return message;
  }

  export function fromJSON(object: any): NotFoundError {
    return { message: isSet(object.message) ? String(object.message) : undefined };
  }

  export function toJSON(message: NotFoundError): unknown {
    const obj: any = {};
    message.message !== undefined && (obj.message = message.message);
    return obj;
  }

  export function create<I extends Exact<DeepPartial<NotFoundError>, I>>(base?: I): NotFoundError {
    return NotFoundError.fromPartial(base ?? {});
  }

  export function fromPartial<I extends Exact<DeepPartial<NotFoundError>, I>>(object: I): NotFoundError {
    const message = createBaseNotFoundError();
    message.message = object.message ?? undefined;
    return message;
  }
}

type Builtin = Date | Function | Uint8Array | string | number | boolean | undefined;

export type DeepPartial<T> = T extends Builtin ? T
  : T extends Array<infer U> ? Array<DeepPartial<U>> : T extends ReadonlyArray<infer U> ? ReadonlyArray<DeepPartial<U>>
  : T extends { $case: string } ? { [K in keyof Omit<T, "$case">]?: DeepPartial<T[K]> } & { $case: T["$case"] }
  : T extends {} ? { [K in keyof T]?: DeepPartial<T[K]> }
  : Partial<T>;

type KeysOfUnion<T> = T extends T ? keyof T : never;
export type Exact<P, I extends P> = P extends Builtin ? P
  : P & { [K in keyof P]: Exact<P[K], I[K]> } & { [K in Exclude<keyof I, KeysOfUnion<P>>]: never };

function isSet(value: any): boolean {
  return value !== null && value !== undefined;
}
