// @dcl/react-ecs/dist/system.js
import { EntityState } from '@dcl/ecs';
import * as ecsComponents from '@dcl/ecs/dist/components';
import React from 'react';
import { createReconciler } from './reconciler';
import { getUiScaleFactor, resetUiScaleFactor, setUiScaleFactor } from './components/utils';
/**
 * @public
 */
export function createReactBasedUiSystem(engine, pointerSystem) {
    const renderer = createReconciler(engine, pointerSystem);
    let uiComponent = undefined;
    let virtualSize = undefined;
    const additionalRenderers = new Map();
    const UiCanvasInformation = ecsComponents.UiCanvasInformation(engine);
    // Unique owner to prevent other UI systems resetting this scale factor.
    const uiScaleFactorOwner = Symbol('react-ecs-ui-scale');
    function getActiveVirtualSize() {
        // Main renderer options win; otherwise use the first additional renderer option.
        if (virtualSize)
            return virtualSize;
        for (const entry of additionalRenderers.values()) {
            if (entry.options)
                return entry.options;
        }
        return undefined;
    }
    function ReactBasedUiSystem() {
        const components = [];
        // Add main UI component
        if (uiComponent) {
            components.push(React.createElement(uiComponent, { key: '__main__' }));
        }
        const entitiesToRemove = [];
        for (const [entity, entry] of additionalRenderers) {
            // Check for entity-based cleanup
            if (engine.getEntityState(entity) === EntityState.Removed) {
                entitiesToRemove.push(entity);
            }
            else {
                components.push(React.createElement(entry.ui, { key: `__entity_${entity}__` }));
            }
        }
        // Entity-based cleanup
        for (const entity of entitiesToRemove) {
            additionalRenderers.delete(entity);
        }
        // Always update the renderer - pass null when empty to clear the UI
        if (components.length > 0) {
            renderer.update(React.createElement(React.Fragment, null, ...components));
        }
        else {
            renderer.update(null);
        }
    }
    function UiScaleSystem() {
        const activeVirtualSize = getActiveVirtualSize();
        if (!activeVirtualSize) {
            // Reset only if this system owns the scale factor.
            resetUiScaleFactor(uiScaleFactorOwner);
            return;
        }
        const canvasInfo = UiCanvasInformation.getOrNull(engine.RootEntity);
        if (!canvasInfo)
            return;
        const { width, height } = canvasInfo;
        const { virtualWidth, virtualHeight } = activeVirtualSize;
        if (!virtualWidth || !virtualHeight)
            return;
        const nextScale = Math.min(width / virtualWidth, height / virtualHeight);
        if (Number.isFinite(nextScale) && nextScale !== getUiScaleFactor()) {
            // Track ownership when updating to avoid cross-system conflicts.
            setUiScaleFactor(nextScale, uiScaleFactorOwner);
        }
    }
    engine.addSystem(UiScaleSystem, 100e3 + 1, '@dcl/react-ecs-ui-scale');
    engine.addSystem(ReactBasedUiSystem, 100e3, '@dcl/react-ecs');
    return {
        destroy() {
            engine.removeSystem(UiScaleSystem);
            engine.removeSystem(ReactBasedUiSystem);
            resetUiScaleFactor(uiScaleFactorOwner);
            for (const entity of renderer.getEntities()) {
                engine.removeEntity(entity);
            }
        },
        setUiRenderer(ui, options) {
            uiComponent = ui;
            virtualSize = options;
        },
        addUiRenderer(entity, ui, options) {
            additionalRenderers.set(entity, { ui, options });
        },
        removeUiRenderer(entity) {
            additionalRenderers.delete(entity);
        }
    };
}
