// @dcl/asset-packs/dist/admin-toolkit-ui/index.js
import { Color4 } from '@dcl/sdk/math';
import ReactEcs, { Label, Button as DCLButton, UiEntity } from '@dcl/react-ecs';
import { getComponents } from '../definitions';
import { VideoControl } from './VideoControl';
import { TextAnnouncementsControl } from './TextAnnouncementsControl';
import { SmartItemsControl } from './SmartItemsControl';
import { Button } from './Button';
import { TextAnnouncements } from './TextAnnouncements';
import { CONTENT_URL } from './constants';
import { TabType } from './types';
import { BTN_MODERATION_CONTROL, ModerationControl, moderationControlState, } from './ModerationControl';
import { getSceneAdmins, getSceneBans } from './ModerationControl/api';
import { ModalUserList, UserListType } from './ModerationControl/UsersList';
import { isPreview } from './fetch-utils';
export const nextTickFunctions = [];
const ADMIN_TOOLKIT_VIRTUAL_UI_SIZE = { virtualWidth: 1920, virtualHeight: 1080 };
export let state = {
    adminToolkitUiEntity: 0,
    panelOpen: false,
    activeTab: TabType.NONE,
    videoControl: {
        selectedVideoPlayer: undefined,
        selectedStream: undefined,
        dclCast: undefined,
    },
    smartItemsControl: {
        selectedSmartItem: undefined,
        smartItems: new Map(),
    },
    textAnnouncementControl: {
        entity: undefined,
        text: undefined,
        messageRateTracker: new Map(),
        announcements: [],
        maxAnnouncements: 4,
    },
    rewardsControl: {
        selectedRewardItem: undefined,
    },
};
let sceneAdminsCache = [];
let sceneBansCache = [];
const BTN_VIDEO_CONTROL = `${CONTENT_URL}/admin_toolkit/assets/icons/admin-panel-video-control-button.png`;
const BTN_SMART_ITEM_CONTROL = `${CONTENT_URL}/admin_toolkit/assets/icons/admin-panel-smart-item-control-button.png`;
const BTN_TEXT_ANNOUNCEMENT_CONTROL = `${CONTENT_URL}/admin_toolkit/assets/icons/admin-panel-text-announcement-control-button.png`;
const BTN_ADMIN_TOOLKIT_CONTROL = `${CONTENT_URL}/admin_toolkit/assets/icons/admin-panel-control-button.png`;
const BTN_ADMIN_TOOLKIT_BACKGROUND = `${CONTENT_URL}/admin_toolkit/assets/backgrounds/admin-tool-background.png`;
export const containerBackgroundColor = Color4.create(0, 0, 0, 0.75);
const ADMIN_TOOLS_ENTITY = 8000;
function getAdminToolkitEntity(engine) {
    const { AdminTools } = getComponents(engine);
    return Array.from(engine.getEntitiesWith(AdminTools))[0][0];
}
function getAdminToolkitComponent(engine) {
    const { AdminTools } = getComponents(engine);
    return Array.from(engine.getEntitiesWith(AdminTools))[0][1];
}
export async function fetchSceneAdmins() {
    const [error, response] = await getSceneAdmins();
    if (error) {
        console.log(JSON.stringify({ error }));
        sceneAdminsCache = [];
        return;
    }
    sceneAdminsCache = (response ?? [])
        .map($ => ({
        name: $.name,
        address: $.admin,
        role: 'admin',
        verified: !$.name.includes('#'),
        canBeRemoved: !!$.canBeRemoved,
    }))
        .sort(a => (a.canBeRemoved ? 1 : -1));
}
export async function fetchSceneBans() {
    const [error, response] = await getSceneBans();
    if (error) {
        sceneBansCache = [];
        return;
    }
    sceneBansCache = response?.results ?? [];
}
export function clearSceneBansCache() {
    sceneBansCache = [];
}
export function getSmartItems(engine) {
    const adminToolkitComponent = getAdminToolkitComponent(engine);
    return Array.from(adminToolkitComponent.smartItemsControl.smartItems ?? []);
}
function getRewards(engine) {
    const adminToolkitComponent = getAdminToolkitComponent(engine);
    return Array.from(adminToolkitComponent?.rewardsControl?.rewardItems ?? []);
}
function initTextAnnouncementSync(engine) {
    const { TextAnnouncements } = getComponents(engine);
    TextAnnouncements.createOrReplace(state.adminToolkitUiEntity, {
        text: '',
        author: '',
        id: '',
    });
}
let adminDataInitialized = false;
export async function initializeAdminData(engine, sdkHelpers) {
    if (!adminDataInitialized) {
        const { TextAnnouncements, VideoControlState } = getComponents(engine);
        state.adminToolkitUiEntity = getAdminToolkitEntity(engine) ?? engine.addEntity();
        initTextAnnouncementSync(engine);
        if (!VideoControlState.getOrNull(state.adminToolkitUiEntity)) {
            VideoControlState.create(state.adminToolkitUiEntity);
        }
        sdkHelpers?.syncEntity?.(state.adminToolkitUiEntity, [VideoControlState.componentId, TextAnnouncements.componentId], ADMIN_TOOLS_ENTITY);
        engine.addSystem(() => {
            if (nextTickFunctions.length > 0) {
                const nextTick = nextTickFunctions.shift();
                if (nextTick) {
                    nextTick();
                }
            }
        }, Number.POSITIVE_INFINITY);
        await Promise.all([fetchSceneAdmins(), fetchSceneBans()]);
        adminDataInitialized = true;
        console.log('initializeAdminData - initialized');
    }
}
export function createAdminToolkitUI(engine, pointerEventsSystem, reactBasedUiSystem, sdkHelpers, playersHelper) {
    initializeAdminData(engine, sdkHelpers).then(() => {
        console.log('createAdminToolkitUI - initialized');
        reactBasedUiSystem.setUiRenderer(() => uiComponent(engine, pointerEventsSystem, sdkHelpers, playersHelper), ADMIN_TOOLKIT_VIRTUAL_UI_SIZE);
    });
}
function isAllowedAdmin(_engine, adminToolkitEntitie, player) {
    if (!player)
        return false;
    const playerAddress = player.userId.toLowerCase();
    const isAdmin = sceneAdminsCache.find($ => $.address === playerAddress);
    return isAdmin || isPreview();
}
const uiComponent = (engine, pointerEventsSystem, sdkHelpers, playersHelper) => {
    const adminToolkitEntity = getAdminToolkitComponent(engine);
    const player = playersHelper?.getPlayer();
    const isPlayerAdmin = isAllowedAdmin(engine, adminToolkitEntity, player);
    return [
        ReactEcs.createElement(UiEntity, { uiTransform: {
                positionType: 'absolute',
                height: '100%',
                width: '100%',
            } },
            isPlayerAdmin ? (ReactEcs.createElement(UiEntity, { uiTransform: {
                    positionType: 'absolute',
                    flexDirection: 'row',
                    position: { top: 120, right: 10 },
                } },
                ReactEcs.createElement(UiEntity, { uiTransform: {
                        display: state.panelOpen ? 'flex' : 'none',
                        width: 500,
                        pointerFilter: 'block',
                        flexDirection: 'column',
                        margin: { right: 8 },
                    } },
                    ReactEcs.createElement(UiEntity, { uiTransform: {
                            width: '100%',
                            height: 50,
                            flexDirection: 'row',
                            alignItems: 'center',
                            borderRadius: 12,
                            padding: {
                                left: 12,
                                right: 12,
                            },
                        }, uiBackground: { color: containerBackgroundColor } },
                        ReactEcs.createElement(Label, { value: "ADMIN TOOLS", fontSize: 20, color: Color4.create(160, 155, 168, 1), uiTransform: { flexGrow: 1 } }),
                        ReactEcs.createElement(Button, { id: "admin_toolkit_moderation_control", variant: state.activeTab === TabType.MODERATION_CONTROL ? 'primary' : 'text', icon: BTN_MODERATION_CONTROL, onlyIcon: true, uiTransform: {
                                display: adminToolkitEntity.moderationControl.isEnabled && !isPreview()
                                    ? 'flex'
                                    : 'none',
                                width: 49,
                                height: 42,
                                margin: { right: 8 },
                                alignItems: 'center',
                                justifyContent: 'center',
                            }, iconBackground: {
                                color: state.activeTab === TabType.MODERATION_CONTROL
                                    ? Color4.Black()
                                    : Color4.White(),
                            }, iconTransform: { height: '100%', width: '100%' }, onMouseDown: () => {
                                if (state.activeTab !== TabType.MODERATION_CONTROL) {
                                    state.activeTab = TabType.NONE;
                                    nextTickFunctions.push(() => {
                                        state.activeTab = TabType.MODERATION_CONTROL;
                                    });
                                }
                                else {
                                    state.activeTab = TabType.NONE;
                                }
                            } }),
                        ReactEcs.createElement(Button, { id: "admin_toolkit_panel_video_control", variant: state.activeTab === TabType.VIDEO_CONTROL ? 'primary' : 'text', icon: BTN_VIDEO_CONTROL, iconBackground: {
                                color: state.activeTab === TabType.VIDEO_CONTROL ? Color4.Black() : Color4.White(),
                            }, onlyIcon: true, uiTransform: {
                                display: adminToolkitEntity.videoControl.isEnabled ? 'flex' : 'none',
                                width: 49,
                                height: 42,
                                margin: { right: 8 },
                                alignItems: 'center',
                                justifyContent: 'center',
                            }, iconTransform: {
                                height: '100%',
                                width: '100%',
                            }, onMouseDown: () => {
                                if (state.activeTab !== TabType.VIDEO_CONTROL) {
                                    state.activeTab = TabType.NONE;
                                    nextTickFunctions.push(() => {
                                        state.activeTab = TabType.VIDEO_CONTROL;
                                    });
                                }
                                else {
                                    state.activeTab = TabType.NONE;
                                }
                            } }),
                        ReactEcs.createElement(Button, { id: "admin_toolkit_panel_smart_items_control", variant: state.activeTab === TabType.SMART_ITEMS_CONTROL ? 'primary' : 'text', icon: BTN_SMART_ITEM_CONTROL, iconBackground: {
                                color: state.activeTab === TabType.SMART_ITEMS_CONTROL
                                    ? Color4.Black()
                                    : Color4.White(),
                            }, onlyIcon: true, uiTransform: {
                                display: adminToolkitEntity.smartItemsControl.isEnabled ? 'flex' : 'none',
                                width: 49,
                                height: 42,
                                margin: { right: 8 },
                                alignItems: 'center',
                                justifyContent: 'center',
                            }, iconTransform: {
                                height: '100%',
                                width: '100%',
                            }, onMouseDown: () => {
                                if (state.activeTab !== TabType.SMART_ITEMS_CONTROL) {
                                    state.activeTab = TabType.NONE;
                                    nextTickFunctions.push(() => {
                                        state.activeTab = TabType.SMART_ITEMS_CONTROL;
                                    });
                                }
                                else {
                                    state.activeTab = TabType.NONE;
                                }
                            } }),
                        ReactEcs.createElement(Button, { id: "admin_toolkit_panel_text_announcement_control", variant: state.activeTab === TabType.TEXT_ANNOUNCEMENT_CONTROL ? 'primary' : 'text', icon: BTN_TEXT_ANNOUNCEMENT_CONTROL, iconBackground: {
                                color: state.activeTab === TabType.TEXT_ANNOUNCEMENT_CONTROL
                                    ? Color4.Black()
                                    : Color4.White(),
                            }, onlyIcon: true, uiTransform: {
                                display: adminToolkitEntity.textAnnouncementControl.isEnabled ? 'flex' : 'none',
                                width: 49,
                                height: 42,
                                margin: { right: 8 },
                                alignItems: 'center',
                                justifyContent: 'center',
                            }, iconTransform: {
                                height: '100%',
                                width: '100%',
                            }, onMouseDown: () => {
                                if (state.activeTab !== TabType.TEXT_ANNOUNCEMENT_CONTROL) {
                                    state.activeTab = TabType.NONE;
                                    nextTickFunctions.push(() => {
                                        state.activeTab = TabType.TEXT_ANNOUNCEMENT_CONTROL;
                                    });
                                }
                                else {
                                    state.activeTab = TabType.NONE;
                                }
                            } })),
                    state.activeTab === TabType.TEXT_ANNOUNCEMENT_CONTROL ? (ReactEcs.createElement(TextAnnouncementsControl, { engine: engine, state: state, player: player })) : null,
                    state.activeTab === TabType.VIDEO_CONTROL ? (ReactEcs.createElement(VideoControl, { engine: engine, state: state })) : null,
                    state.activeTab === TabType.SMART_ITEMS_CONTROL ? (ReactEcs.createElement(SmartItemsControl, { engine: engine, state: state })) : null,
                    state.activeTab === TabType.MODERATION_CONTROL && (ReactEcs.createElement(ModerationControl, { engine: engine, player: player, sceneAdmins: sceneAdminsCache }))),
                ReactEcs.createElement(UiEntity, { uiTransform: {
                        display: 'flex',
                        height: 42,
                        width: 42,
                        alignItems: 'center',
                        alignContent: 'center',
                        justifyContent: 'center',
                        pointerFilter: 'block',
                    }, uiBackground: {
                        texture: {
                            src: BTN_ADMIN_TOOLKIT_BACKGROUND,
                        },
                        textureMode: 'stretch',
                        color: Color4.create(1, 1, 1, 1),
                    } },
                    ReactEcs.createElement(DCLButton, { value: "", uiTransform: {
                            height: 40,
                            width: 40,
                            alignItems: 'center',
                            alignContent: 'center',
                            justifyContent: 'center',
                        }, uiBackground: {
                            texture: {
                                src: BTN_ADMIN_TOOLKIT_CONTROL,
                            },
                            textureMode: 'stretch',
                            color: Color4.create(1, 1, 1, 1),
                        }, onMouseDown: () => {
                            state.panelOpen = !state.panelOpen;
                        } })))) : null,
            ReactEcs.createElement(TextAnnouncements, { engine: engine, state: state })),
        moderationControlState.showModalAdminList && (ReactEcs.createElement(ModalUserList, { users: sceneAdminsCache ?? [], engine: engine, type: UserListType.ADMIN })),
        moderationControlState.showModalBanList && (ReactEcs.createElement(ModalUserList, { users: sceneBansCache ?? [], engine: engine, type: UserListType.BAN })),
    ];
};
