// @dcl/asset-packs/dist/admin-toolkit-ui/Loading.js
import { Color4 } from '@dcl/ecs-math';
import ReactEcs, { UiEntity } from '@dcl/react-ecs';
import { clearInterval, setInterval } from './utils';
export function LoadingDots({ uiTransform, engine, width = 10, height = 10 }) {
    let __frame = 0;
    const [frame, setFrame] = ReactEcs.useState(0);
    ReactEcs.useEffect(() => {
        const interval = setInterval(engine, () => {
            __frame = (__frame + 1) % 4;
            setFrame(__frame);
        }, 340);
        return () => clearInterval(engine, interval);
    }, []);
    return (ReactEcs.createElement(UiEntity, { uiTransform: {
            display: 'flex',
            flexDirection: 'row',
            justifyContent: 'center',
            alignItems: 'center',
            ...uiTransform,
        } }, [1, 2, 3].map(i => (ReactEcs.createElement(UiEntity, { key: `dot-${i}`, uiTransform: {
            width,
            height,
            borderRadius: width / 2,
            margin: { right: 8 },
        }, uiBackground: {
            color: frame >= i ? Color4.fromHexString('#FF2D55') : Color4.fromHexString('#43404A'),
        } })))));
}
