// @dcl/asset-packs/dist/admin-toolkit-ui/RewardsControl.js
import ReactEcs, { UiEntity, Label, Dropdown } from '@dcl/react-ecs';
import { Color4 } from '@dcl/sdk/math';
import { getActionEvents, getComponents, getPayload } from '../definitions';
import { Button } from './Button';
import { CONTENT_URL } from './constants';
const ICONS = {
    REWARDS_CONTROL: `${CONTENT_URL}/admin_toolkit/assets/icons/rewards-control.png`,
    SEND: `${CONTENT_URL}/admin_toolkit/assets/icons/rewards-send.png`,
};
function getAdminToolkitRewardsControl(engine) {
    const { AdminTools } = getComponents(engine);
    const adminToolkitEntities = Array.from(engine.getEntitiesWith(AdminTools));
    return adminToolkitEntities.length > 0 ? adminToolkitEntities[0][1].rewardsControl : null;
}
function getRewardItems(engine) {
    const adminToolkitRewardsControl = getAdminToolkitRewardsControl(engine);
    if (!adminToolkitRewardsControl ||
        !adminToolkitRewardsControl.rewardItems ||
        adminToolkitRewardsControl.rewardItems.length === 0)
        return [];
    return Array.from(adminToolkitRewardsControl.rewardItems);
}
export function RewardsControl({ engine, state }) {
    const rewardItems = getRewardItems(engine);
    return (ReactEcs.createElement(UiEntity, { uiTransform: {
            width: '100%',
            height: '100%',
            flexDirection: 'column',
        } },
        ReactEcs.createElement(UiEntity, { uiTransform: {
                flexDirection: 'row',
                margin: { bottom: 32 },
                height: 30,
            } },
            ReactEcs.createElement(UiEntity, { uiTransform: { width: 30, height: 30 }, uiBackground: {
                    color: Color4.White(),
                    textureMode: 'stretch',
                    texture: { src: ICONS.REWARDS_CONTROL },
                } }),
            ReactEcs.createElement(Label, { value: "<b>AIRDROPS</b>", fontSize: 24, color: Color4.White() })),
        ReactEcs.createElement(UiEntity, { uiTransform: {
                flexDirection: 'column',
                margin: { bottom: 32 },
            } },
            ReactEcs.createElement(Label, { value: "<b>Selected Airdrop</b>", fontSize: 16, color: Color4.White(), uiTransform: { margin: { bottom: 16 } } }),
            ReactEcs.createElement(Dropdown, { key: "RewardsItemSelector", acceptEmpty: true, emptyLabel: "Select your airdrop", options: [
                    ...rewardItems.map((item) => item.customName),
                ], selectedIndex: state.rewardsControl.selectedRewardItem ?? -1, onChange: idx => (state.rewardsControl.selectedRewardItem = idx), textAlign: "middle-left", fontSize: 14, uiTransform: {
                    width: '100%',
                    height: 40,
                }, uiBackground: { color: Color4.White() }, color: Color4.Black() })),
        ReactEcs.createElement(UiEntity, { uiTransform: { flexDirection: 'column' } },
            ReactEcs.createElement(Label, { value: "<b>Actions</b>", fontSize: 16, color: Color4.White(), uiTransform: { margin: { bottom: 16 } } }),
            ReactEcs.createElement(UiEntity, { uiTransform: { flexDirection: 'row' } },
                ReactEcs.createElement(Button, { id: "rewards_control_release", value: "<b>Release</b>", fontSize: 16, uiTransform: {
                        margin: { right: 16 },
                        alignItems: 'center',
                        justifyContent: 'center',
                    }, icon: ICONS.SEND, iconTransform: {
                        height: 25,
                        width: 25,
                    }, onMouseDown: () => handleRelease(engine, state), disabled: state.rewardsControl.selectedRewardItem === undefined }),
                ReactEcs.createElement(Button, { id: "rewards_control_clear", value: "<b>Clear</b>", fontSize: 16, onMouseDown: () => handleClear(engine, state), disabled: state.rewardsControl.selectedRewardItem === undefined })))));
}
function handleRelease(engine, state) {
    const { Actions, Rewards } = getComponents(engine);
    const rewardItems = getRewardItems(engine);
    const selectedRewardItem = rewardItems[state.rewardsControl.selectedRewardItem];
    const rewardItem = Rewards.getOrNull(selectedRewardItem.entity);
    if (!rewardItem)
        return;
    const action = Actions.getOrNull(selectedRewardItem.entity)?.value.find(($) => $.name === 'Airdrop');
    if (action) {
        const actionEvents = getActionEvents(selectedRewardItem.entity);
        actionEvents.emit(action.name, getPayload(action));
    }
}
function handleClear(engine, state) {
    const { Actions, Rewards } = getComponents(engine);
    const rewardItems = getRewardItems(engine);
    const selectedRewardItem = rewardItems[state.rewardsControl.selectedRewardItem];
    const rewardItem = Rewards.getOrNull(selectedRewardItem.entity);
    if (!rewardItem)
        return;
    const action = Actions.getOrNull(selectedRewardItem.entity)?.value.find(($) => $.name === 'Invisible');
    if (action) {
        const actionEvents = getActionEvents(selectedRewardItem.entity);
        actionEvents.emit(action.name, getPayload(action));
    }
}
