// @dcl/asset-packs/dist/admin-toolkit-ui/SmartItemsControl.js
import ReactEcs, { Dropdown, Label, UiEntity } from '@dcl/react-ecs';
import { Color4 } from '@dcl/sdk/math';
import { getActionEvents, getComponents, getPayload } from '../definitions';
import { getExplorerComponents } from '../components';
import { Button } from './Button';
import { CONTENT_URL } from './constants';
import { Header } from './Header';
import { Card } from './Card';
import { getSmartItems } from '.';
const ICONS = {
    SMART_ITEM_CONTROL: `${CONTENT_URL}/admin_toolkit/assets/icons/smart-item-control.png`,
};
function getSmartItemActions(engine, smartItem) {
    const { Actions } = getComponents(engine);
    if (!smartItem || !Actions.has(smartItem.entity))
        return [];
    const actions = Array.from(Actions.get(smartItem.entity).value);
    return actions;
}
function handleExecuteAction(smartItem, action) {
    const actionEvents = getActionEvents(smartItem.entity);
    actionEvents.emit(action.name, getPayload(action));
}
function handleSelectSmartItem(state, smartItems, idx) {
    state.smartItemsControl.selectedSmartItem = idx;
    const smartItem = smartItems[idx];
    if (!state.smartItemsControl.smartItems.has(smartItem.entity)) {
        const stateSmartItems = new Map(state.smartItemsControl.smartItems);
        stateSmartItems.set(smartItem.entity, {
            visible: true,
            selectedAction: smartItem.defaultAction,
        });
        state.smartItemsControl = {
            ...state.smartItemsControl,
            smartItems: new Map(stateSmartItems),
        };
    }
}
function handleSelectAction(state, smartItem, action) {
    const stateSmartItems = new Map(state.smartItemsControl.smartItems);
    stateSmartItems.set(smartItem.entity, {
        ...stateSmartItems.get(smartItem.entity),
        selectedAction: action.name,
    });
    state.smartItemsControl = {
        ...state.smartItemsControl,
        smartItems: new Map(stateSmartItems),
    };
}
function handleHideShowEntity(engine, state, smartItems) {
    const { VisibilityComponent } = getExplorerComponents(engine);
    const smartItemEntity = smartItems[state.smartItemsControl.selectedSmartItem].entity;
    const smartItem = state.smartItemsControl.smartItems.get(smartItemEntity);
    const toggleVisibility = !smartItem.visible;
    state.smartItemsControl.smartItems.get(smartItemEntity).visible = toggleVisibility;
    const visibility = VisibilityComponent.getOrCreateMutable(smartItemEntity);
    visibility.visible = toggleVisibility;
}
function SmartItemSelector({ engine, smartItems, selectedIndex, onSelect, }) {
    return (ReactEcs.createElement(UiEntity, { key: "SmartItemsControlDropdownWrapper", uiTransform: {
            flexDirection: 'column',
            margin: { bottom: 32 },
        } },
        ReactEcs.createElement(Label, { value: "<b>Selected Smart Item</b>", fontSize: 16, color: Color4.White(), uiTransform: {
                margin: { bottom: 16 },
            } }),
        ReactEcs.createElement(Dropdown, { key: "SmartItemsControlDropdownSelector", acceptEmpty: true, emptyLabel: "Select Smart Item", options: smartItems.map((item) => item.customName), selectedIndex: selectedIndex ?? -1, onChange: onSelect, textAlign: "middle-left", fontSize: 14, uiTransform: {
                height: 40,
                width: '100%',
            }, uiBackground: { color: Color4.White() }, color: Color4.Black() })));
}
function ActionSelector({ engine, actions, selectedIndex, disabled, onChange, }) {
    return (ReactEcs.createElement(UiEntity, { uiTransform: {
            flexDirection: 'column',
            margin: { bottom: 32 },
        } },
        ReactEcs.createElement(Label, { value: "<b>Actions</b>", fontSize: 16, color: Color4.White(), uiTransform: {
                margin: { bottom: 16 },
            } }),
        actions.length ? (ReactEcs.createElement(Dropdown, { acceptEmpty: true, emptyLabel: "Select Action", options: actions.map((action) => action.name), selectedIndex: selectedIndex, onChange: onChange, disabled: disabled, textAlign: "middle-left", fontSize: 14, uiTransform: {
                height: 40,
                width: '100%',
            }, uiBackground: {
                color: disabled ? Color4.Gray() : Color4.White(),
            }, color: Color4.Black() })) : (ReactEcs.createElement(UiEntity, null))));
}
function ActionButtons({ engine, state, smartItems, actions, selectedAction, }) {
    const selectedSmartItem = state.smartItemsControl.selectedSmartItem !== undefined
        ? smartItems[state.smartItemsControl.selectedSmartItem]
        : undefined;
    const isVisible = selectedSmartItem &&
        state.smartItemsControl.smartItems.get(selectedSmartItem.entity)?.visible;
    return (ReactEcs.createElement(UiEntity, { uiTransform: {
            flexDirection: 'row',
        } },
        ReactEcs.createElement(Button, { id: "smart_items_control_restart", value: "<b>Play Action</b>", variant: "text", fontSize: 16, color: Color4.White(), uiTransform: {
                margin: { right: 8 },
                height: 40,
            }, labelTransform: {
                margin: { left: 10, right: 10 },
            }, disabled: !selectedSmartItem || !selectedAction, onMouseDown: () => {
                if (selectedSmartItem && selectedAction) {
                    handleExecuteAction(selectedSmartItem, selectedAction);
                }
            } }),
        ReactEcs.createElement(Button, { id: "smart_items_control_hide_show", value: `<b>${isVisible ? 'Hide' : 'Show'} Entity</b>`, variant: "text", fontSize: 16, color: Color4.White(), onMouseDown: () => handleHideShowEntity(engine, state, smartItems), disabled: !selectedSmartItem, uiTransform: { margin: { right: 8 } }, labelTransform: {
                margin: { left: 10, right: 10 },
            } })));
}
export function SmartItemsControl({ engine, state }) {
    const smartItems = getSmartItems(engine);
    const actions = state.smartItemsControl.selectedSmartItem !== undefined
        ? getSmartItemActions(engine, smartItems[state.smartItemsControl.selectedSmartItem])
        : [];
    const selectedActionIndex = state.smartItemsControl.selectedSmartItem !== undefined
        ? actions.findIndex((action) => {
            const selectedSmartItem = smartItems[state.smartItemsControl.selectedSmartItem];
            const stateSelectedAction = state.smartItemsControl.smartItems.get(selectedSmartItem.entity)?.selectedAction;
            return action.name === (stateSelectedAction ?? selectedSmartItem.defaultAction);
        })
        : undefined;
    return (ReactEcs.createElement(Card, null,
        ReactEcs.createElement(UiEntity, { key: "SmartItemsControl", uiTransform: {
                height: '100%',
                width: '100%',
                flexDirection: 'column',
            } },
            ReactEcs.createElement(Header, { iconSrc: ICONS.SMART_ITEM_CONTROL, title: "SMART ITEM ACTIONS" }),
            ReactEcs.createElement(SmartItemSelector, { engine: engine, smartItems: smartItems, selectedIndex: state.smartItemsControl.selectedSmartItem, onSelect: idx => {
                    handleSelectSmartItem(state, smartItems, idx);
                } }),
            ReactEcs.createElement(ActionSelector, { engine: engine, actions: actions, selectedIndex: selectedActionIndex, disabled: state.smartItemsControl.selectedSmartItem === undefined, onChange: idx => {
                    if (state.smartItemsControl.selectedSmartItem !== undefined) {
                        handleSelectAction(state, smartItems[state.smartItemsControl.selectedSmartItem], actions[idx]);
                    }
                } }),
            ReactEcs.createElement(ActionButtons, { engine: engine, state: state, smartItems: smartItems, actions: actions, selectedAction: selectedActionIndex !== undefined ? actions[selectedActionIndex] : undefined }))));
}
