// @dcl/asset-packs/dist/admin-toolkit-ui/TextAnnouncementsControl.js
import ReactEcs, { Label, UiEntity, Input } from '@dcl/react-ecs';
import { Color4 } from '@dcl/sdk/math';
import { getComponents } from '../definitions';
import { Button } from './Button';
import { CONTENT_URL } from './constants';
import { Header } from './Header';
import { Card } from './Card';
const ICONS = {
    TEXT_ANNOUNCEMENT_CONTROL: `${CONTENT_URL}/admin_toolkit/assets/icons/text-announcement-control.png`,
    CHECK: `${CONTENT_URL}/admin_toolkit/assets/icons/text-announcement-check.png`,
    BTN_CLOSE_TEXT_ANNOUNCEMENT: `${CONTENT_URL}/admin_toolkit/assets/icons/text-announcement-close-button.png`,
    CHAT_MESSAGE_ICON: `${CONTENT_URL}/admin_toolkit/assets/icons/text-announcement-chat-message.png`,
};
let ANNOUNCEMENT_STATE;
export function TextAnnouncementsControl({ engine, state, player, }) {
    return (ReactEcs.createElement(Card, null,
        ReactEcs.createElement(UiEntity, { uiTransform: {
                width: '100%',
                height: '100%',
                flexDirection: 'column',
            } },
            ReactEcs.createElement(Header, { iconSrc: ICONS.TEXT_ANNOUNCEMENT_CONTROL, title: "TEXT ANNOUNCEMENTS" }),
            ReactEcs.createElement(UiEntity, { uiTransform: { flexDirection: 'column' } },
                ReactEcs.createElement(Label, { value: "<b>Message window</b>", fontSize: 16, color: Color4.White(), uiTransform: { margin: { bottom: 16 } } }),
                ReactEcs.createElement(Input, { onSubmit: value => {
                        handleSendTextAnnouncement(engine, state, value, player);
                    }, onChange: value => {
                        state.textAnnouncementControl.text = value;
                    }, fontSize: 16, placeholder: 'Write your announcement here', placeholderColor: Color4.create(160 / 255, 155 / 255, 168 / 255, 1), color: Color4.Black(), uiBackground: { color: Color4.White() }, uiTransform: {
                        width: '100%',
                        height: 80,
                        margin: { bottom: 16 },
                    } }),
                ReactEcs.createElement(UiEntity, { uiTransform: {
                        width: '100%',
                        height: 40,
                        flexDirection: 'row',
                        margin: {
                            bottom: 10,
                            top: 0,
                            right: 0,
                            left: 0,
                        },
                    } },
                    ReactEcs.createElement(Label, { value: `${state.textAnnouncementControl.text?.length ?? 0} / 90`, fontSize: 14, color: Color4.create(187 / 255, 187 / 255, 187 / 255, 1), uiTransform: { flexGrow: 1 }, textAlign: "top-left" }),
                    ReactEcs.createElement(Button, { id: "text_announcement_control_clear", value: "<b>Clear Announcements</b>", variant: "text", fontSize: 16, color: Color4.White(), uiTransform: {
                            height: 40,
                            margin: { right: 8 },
                        }, onMouseDown: () => {
                            handleClearTextAnnouncement(engine, state);
                        } }),
                    ReactEcs.createElement(Button, { id: "text_announcement_control_share", value: "<b>Share</b>", variant: "primary", fontSize: 16, labelTransform: {
                            margin: { left: 20, right: 20 },
                        }, uiTransform: { height: 40 }, onMouseDown: () => {
                            handleSendTextAnnouncement(engine, state, state.textAnnouncementControl.text, player);
                        } }))),
            ReactEcs.createElement(UiEntity, { uiTransform: { minHeight: 30 } },
                ReactEcs.createElement(UiEntity, { uiTransform: {
                        display: ANNOUNCEMENT_STATE !== undefined ? 'flex' : 'none',
                        width: 30,
                        height: 30,
                    }, uiBackground: {
                        texture: { src: ICONS.CHECK },
                        textureMode: 'stretch',
                    } }),
                ReactEcs.createElement(Label, { uiTransform: {
                        display: ANNOUNCEMENT_STATE !== undefined ? 'flex' : 'none',
                    }, value: `Message ${ANNOUNCEMENT_STATE === 'sent' ? 'sent' : 'cleared'}!`, fontSize: 14, color: Color4.create(187 / 255, 187 / 255, 187 / 255, 1) })))));
}
function handleClearTextAnnouncement(engine, state) {
    const { TextAnnouncements } = getComponents(engine);
    const textAnnouncement = TextAnnouncements.getMutableOrNull(state.adminToolkitUiEntity);
    if (textAnnouncement) {
        textAnnouncement.text = '';
        textAnnouncement.id = '';
        textAnnouncement.author = '';
    }
    state.textAnnouncementControl.announcements = [];
    ANNOUNCEMENT_STATE = 'cleared';
}
function handleSendTextAnnouncement(engine, state, text, player) {
    if (!text) {
        return;
    }
    const { TextAnnouncements } = getComponents(engine);
    const textAnnouncement = TextAnnouncements.getMutableOrNull(state.adminToolkitUiEntity);
    if (textAnnouncement) {
        const author = player?.name;
        let timestamp = Date.now();
        textAnnouncement.author = author;
        textAnnouncement.id = `${timestamp}-${author}`;
        textAnnouncement.text = text.slice(0, 90);
    }
    state.textAnnouncementControl.text = '';
    ANNOUNCEMENT_STATE = 'sent';
    return;
}
