// @dcl/asset-packs/dist/admin-toolkit-ui/TextAnnouncements.js
import ReactEcs, { Label, UiEntity } from '@dcl/react-ecs';
import { getComponents } from '../definitions';
import { CONTENT_URL } from './constants';
const ICONS = {
    BTN_CLOSE_TEXT_ANNOUNCEMENT: `${CONTENT_URL}/admin_toolkit/assets/icons/text-announcement-close-button.png`,
    CHAT_MESSAGE_ICON: `${CONTENT_URL}/admin_toolkit/assets/icons/text-announcement-chat-message.png`,
};
let textAnnouncementsHidden = new Set();
export function TextAnnouncements({ engine, state }) {
    const { TextAnnouncements } = getComponents(engine);
    const textAnnouncements = TextAnnouncements.getOrNull(state.adminToolkitUiEntity);
    if (!textAnnouncements?.text || textAnnouncementsHidden.has(textAnnouncements.id)) {
        return null;
    }
    return (ReactEcs.createElement(UiEntity, { uiTransform: {
            display: 'flex',
            flexDirection: 'column',
            alignItems: 'center',
            justifyContent: 'flex-end',
            height: '100%',
            width: '100%',
        } },
        ReactEcs.createElement(UiEntity, { key: textAnnouncements.id, uiTransform: {
                display: 'flex',
                flexDirection: 'column',
                height: 150,
                width: 400,
                margin: {
                    bottom: 10,
                },
                padding: {
                    top: 10,
                    bottom: 10,
                    left: 10,
                    right: 10,
                },
            }, uiBackground: { color: { r: 0.15, g: 0.15, b: 0.15, a: 0.95 } } },
            ReactEcs.createElement(UiEntity, { uiTransform: {
                    alignSelf: 'center',
                    justifyContent: 'center',
                    height: 50,
                    width: 50,
                }, uiBackground: {
                    texture: {
                        src: ICONS.CHAT_MESSAGE_ICON,
                    },
                    textureMode: 'stretch',
                    color: { r: 1, g: 1, b: 1, a: 1 },
                } }),
            ReactEcs.createElement(Label, { uiTransform: {
                    alignItems: 'center',
                    justifyContent: 'center',
                }, fontSize: 18, value: textAnnouncements.text }),
            textAnnouncements.author ? (ReactEcs.createElement(Label, { uiTransform: {
                    alignItems: 'flex-end',
                    justifyContent: 'flex-end',
                }, fontSize: 14, color: { r: 0.7, g: 0.7, b: 0.7, a: 1 }, value: `- ${textAnnouncements.author}` })) : null,
            ReactEcs.createElement(UiEntity, { uiTransform: {
                    height: 24,
                    width: 24,
                    positionType: 'absolute',
                    position: {
                        top: 5,
                        right: 5,
                    },
                }, uiBackground: {
                    texture: {
                        src: ICONS.BTN_CLOSE_TEXT_ANNOUNCEMENT,
                    },
                    textureMode: 'stretch',
                }, onMouseDown: () => {
                    textAnnouncementsHidden.add(textAnnouncements.id);
                } }))));
}
