// @dcl/schemas/dist/platform/item/emote/adr74/emote-data-adr74.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmoteDataADR74 = void 0;
const emote_category_1 = require("../emote-category");
const validation_1 = require("../../../../validation");
const representation_adr74_1 = require("./representation-adr74");
const emote_data_adr287_1 = require("../adr287/emote-data-adr287");
var EmoteDataADR74;
(function (EmoteDataADR74) {
    EmoteDataADR74.schema = {
        type: 'object',
        properties: {
            tags: {
                type: 'array',
                items: {
                    type: 'string',
                    minLength: 1
                }
            },
            representations: {
                type: 'array',
                items: representation_adr74_1.EmoteRepresentationADR74.schema,
                minItems: 1
            },
            category: emote_category_1.EmoteCategory.schema,
            loop: {
                type: 'boolean'
            },
            startAnimation: Object.assign(Object.assign({}, emote_data_adr287_1.StartAnimation.schema), { nullable: true }),
            randomizeOutcomes: {
                type: 'boolean',
                nullable: true
            },
            outcomes: {
                type: 'array',
                items: emote_data_adr287_1.OutcomeGroup.schema,
                minItems: 1,
                maxItems: 3,
                nullable: true
            }
        },
        required: ['category', 'tags', 'representations', 'loop'],
        dependencies: {
            startAnimation: ['randomizeOutcomes', 'outcomes'],
            randomizeOutcomes: ['startAnimation', 'outcomes'],
            outcomes: ['startAnimation', 'randomizeOutcomes']
        },
        additionalProperties: true
    };
    EmoteDataADR74.validate = (0, validation_1.generateLazyValidator)(EmoteDataADR74.schema);
})(EmoteDataADR74 = exports.EmoteDataADR74 || (exports.EmoteDataADR74 = {}));
