// @dcl/asset-packs/dist/admin-toolkit-ui/VideoControl/LiveStream/DeleteStreamKey.js
import { Color4 } from '@dcl/ecs-math';
import ReactEcs, { UiEntity, Label } from '@dcl/react-ecs';
import { Button } from '../../Button';
import { LoadingDots } from '../../Loading';
import { Error } from '../../Error';
import { resetStreamKey } from '../api';
import { getComponents } from '../../../definitions';
import { state } from '../..';
export function DeleteStreamKeyConfirmation({ engine, onCancel, onReset, }) {
    const [isLoading, setIsLoading] = ReactEcs.useState(false);
    const [error, setError] = ReactEcs.useState('');
    const { VideoControlState } = getComponents(engine);
    return (ReactEcs.createElement(UiEntity, { uiTransform: {
            minHeight: 479,
            display: 'flex',
            flexDirection: 'column',
            alignItems: 'center',
            justifyContent: 'center',
            borderRadius: 12,
        } },
        ReactEcs.createElement(Label, { value: "<b>Are you sure you want to reset your Stream Key?</b>", fontSize: 16, color: Color4.fromHexString('#FCFCFC') }),
        ReactEcs.createElement(Label, { value: "Active streams using this stream key will be disconnected.", fontSize: 14, color: Color4.fromHexString('#FFFFFFB2'), uiTransform: {
                margin: { top: 6, bottom: 24 },
            } }),
        ReactEcs.createElement(UiEntity, { uiTransform: {
                display: 'flex',
                flexDirection: 'row',
                justifyContent: 'center',
                alignItems: 'center',
                width: '100%',
            } },
            !isLoading && (ReactEcs.createElement(Button, { id: "stream_key_cancel_remove", value: "<b>Cancel</b>", variant: "primary", fontSize: 18, color: Color4.Black(), uiTransform: {
                    width: 90,
                    height: 40,
                    display: 'flex',
                    justifyContent: 'center',
                    alignItems: 'center',
                    margin: { right: 30, left: 30 },
                }, onMouseDown: () => onCancel() })),
            !isLoading && (ReactEcs.createElement(Button, { id: "stream_key_confirm_remove", value: '<b>Reset</b>', variant: "primary", fontSize: 18, color: Color4.White(), uiTransform: {
                    padding: 8,
                    height: 40,
                    justifyContent: 'center',
                    alignItems: 'center',
                }, uiBackground: { color: Color4.fromHexString('#FB3B3B') }, onMouseDown: async () => {
                    setIsLoading(true);
                    const [error, data] = await resetStreamKey();
                    if (error) {
                        setError(error);
                        setIsLoading(false);
                    }
                    else {
                        const videoControl = VideoControlState.getMutable(state.adminToolkitUiEntity);
                        videoControl.endsAt = data?.endsAt;
                        onReset();
                        setIsLoading(false);
                    }
                } }))),
        isLoading && ReactEcs.createElement(LoadingDots, { engine: engine }),
        error && ReactEcs.createElement(Error, { text: error })));
}
