// @dcl/asset-packs/dist/admin-toolkit-ui/VideoControl/LiveStream/GenerateStreamKey.js
import { Color4 } from '@dcl/sdk/math';
import ReactEcs, { UiEntity, Label } from '@dcl/react-ecs';
import { Button } from '../../Button';
import { Error } from '../../Error';
import { generateStreamKey } from '../api';
import { LoadingDots } from '../../Loading';
import { getComponents } from '../../../definitions';
import { state } from '../..';
export function GenerateStreamKey({ engine, onGenerate, }) {
    const [loading, setLoading] = ReactEcs.useState(false);
    const [error, setError] = ReactEcs.useState('');
    const { VideoControlState } = getComponents(engine);
    return (ReactEcs.createElement(UiEntity, { uiTransform: {
            width: '100%',
            height: 460,
            justifyContent: 'center',
            alignItems: 'center',
        } }, loading ? (ReactEcs.createElement(UiEntity, { uiTransform: { flexDirection: 'column' } },
        ReactEcs.createElement(LoadingDots, { engine: engine }),
        ReactEcs.createElement(Label, { value: "Generating your Stream Key", color: Color4.fromHexString('#A09BA8'), fontSize: 16, uiTransform: { margin: { top: 16 } } }))) : (ReactEcs.createElement(UiEntity, { uiTransform: { flexDirection: 'column' } },
        ReactEcs.createElement(UiEntity, { uiTransform: { justifyContent: 'center', width: '100%' } },
            ReactEcs.createElement(Button, { id: "video_control_live_generate_key", value: "<b>Get Stream Key</b>", fontSize: 18, uiTransform: {
                    height: 52,
                    padding: 16,
                    margin: { bottom: 16 },
                }, onMouseDown: async () => {
                    setLoading(true);
                    const [error, data] = await generateStreamKey();
                    setLoading(false);
                    if (error) {
                        setError(error);
                    }
                    else if (data) {
                        const videoControl = VideoControlState.getMutable(state.adminToolkitUiEntity);
                        videoControl.endsAt = data.endsAt;
                        onGenerate();
                    }
                } })),
        error && (ReactEcs.createElement(Error, { text: error, uiTransform: { margin: { top: 16 } } }))))));
}
