// @dcl/gltf-validator-ts/dist/cjs/validators/extension-validator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseExtensionValidator = void 0;
const types_1 = require("../types");
class BaseExtensionValidator {
    validateValueInRange(value, min, max, pointer, addMessage, exclusive = false) {
        if (typeof value !== "number")
            return true;
        const inRange = exclusive
            ? value > min && value < max
            : value >= min && value <= max;
        if (!inRange) {
            addMessage({
                code: "VALUE_NOT_IN_RANGE",
                message: `Value ${value} is out of range.`,
                severity: types_1.Severity.ERROR,
                pointer,
            });
            return false;
        }
        return true;
    }
    validateValueInList(value, validValues, pointer, addMessage) {
        if (!validValues.includes(value)) {
            const validValuesStr = validValues.map((v) => `'${v}'`).join(", ");
            addMessage({
                code: "VALUE_NOT_IN_LIST",
                message: `Invalid value '${value}'. Valid values are (${validValuesStr}).`,
                severity: types_1.Severity.WARNING,
                pointer,
            });
            return false;
        }
        return true;
    }
    addIncompleteExtensionWarning(pointer, addMessage) {
        addMessage({
            code: "INCOMPLETE_EXTENSION_SUPPORT",
            message: "Validation support for this extension is incomplete; the asset may have undetected issues.",
            severity: types_1.Severity.INFO,
            pointer,
        });
    }
    escapeJsonPointer(str) {
        return str.replace(/~/g, "~0").replace(/\//g, "~1");
    }
}
exports.BaseExtensionValidator = BaseExtensionValidator;
