// @dcl/gltf-validator-ts/dist/cjs/validators/texture-validator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TextureValidator = void 0;
const types_1 = require("../types");
class TextureValidator {
    validate(texture, index, gltf) {
        const messages = [];
        // Check source
        if (texture.source !== undefined) {
            if (typeof texture.source !== "number" || texture.source < 0) {
                messages.push({
                    code: "INVALID_VALUE",
                    message: "Texture source must be a non-negative integer.",
                    severity: types_1.Severity.ERROR,
                    pointer: `/textures/${index}/source`,
                });
            }
            else if (!gltf.images || texture.source >= gltf.images.length) {
                messages.push({
                    code: "UNRESOLVED_REFERENCE",
                    message: "Unresolved reference: " + texture.source + ".",
                    severity: types_1.Severity.ERROR,
                    pointer: `/textures/${index}/source`,
                });
            }
        }
        // Check sampler
        if (texture.sampler !== undefined) {
            if (typeof texture.sampler !== "number" || texture.sampler < 0) {
                messages.push({
                    code: "INVALID_VALUE",
                    message: "Texture sampler must be a non-negative integer.",
                    severity: types_1.Severity.ERROR,
                    pointer: `/textures/${index}/sampler`,
                });
            }
            else if (!gltf.samplers || texture.sampler >= gltf.samplers.length) {
                messages.push({
                    code: "UNRESOLVED_REFERENCE",
                    message: "Unresolved reference: " + texture.sampler + ".",
                    severity: types_1.Severity.ERROR,
                    pointer: `/textures/${index}/sampler`,
                });
            }
        }
        // Check for unexpected properties
        const expectedProperties = [
            "source",
            "sampler",
            "name",
            "extensions",
            "extras",
        ];
        for (const key in texture) {
            if (!expectedProperties.includes(key)) {
                messages.push({
                    code: "UNEXPECTED_PROPERTY",
                    message: "Unexpected property.",
                    severity: types_1.Severity.WARNING,
                    pointer: `/textures/${index}/${key}`,
                });
            }
        }
        // Validate extensions
        if (texture["extensions"]) {
            const extensions = texture["extensions"];
            for (const extensionName in extensions) {
                const extension = extensions[extensionName];
                this.validateExtension(extensionName, extension, index, gltf, messages);
            }
        }
        return messages;
    }
    validateExtension(extensionName, extension, textureIndex, gltf, messages) {
        const basePointer = `/textures/${textureIndex}/extensions/${extensionName}`;
        // Validate EXT_texture_webp
        if (extensionName === "EXT_texture_webp" &&
            extension &&
            typeof extension === "object" &&
            !Array.isArray(extension)) {
            const extObj = extension;
            const expectedProperties = ["source"];
            for (const key in extObj) {
                if (!expectedProperties.includes(key)) {
                    messages.push({
                        code: "UNEXPECTED_PROPERTY",
                        message: "Unexpected property.",
                        severity: types_1.Severity.WARNING,
                        pointer: `${basePointer}/${key}`,
                    });
                }
            }
            // Validate source property if present
            if (extObj.source !== undefined) {
                if (typeof extObj.source !== "number" || extObj.source < 0) {
                    messages.push({
                        code: "INVALID_VALUE",
                        message: "Extension source must be a non-negative integer.",
                        severity: types_1.Severity.ERROR,
                        pointer: `${basePointer}/source`,
                    });
                }
                else if (!gltf.images || extObj.source >= gltf.images.length) {
                    messages.push({
                        code: "UNRESOLVED_REFERENCE",
                        message: `Unresolved reference: ${extObj.source}.`,
                        severity: types_1.Severity.ERROR,
                        pointer: `${basePointer}/source`,
                    });
                }
            }
        }
        // Add validation for other texture extensions here as needed
    }
}
exports.TextureValidator = TextureValidator;
