// @dcl/sdk-commands/dist/commands/start/data-layer/fs.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.pathToPosix = pathToPosix;
exports.createFileSystemInterfaceFromFsComponent = createFileSystemInterfaceFromFsComponent;
const path_1 = __importDefault(require("path"));
/**
 * Convert paths to posix stlye
 * .i.e: scene\\assets\\main.composite -> scene/assets/main.composite
 */
function pathToPosix(value) {
    return value.replace(/\\/g, '/');
}
function createFileSystemInterfaceFromFsComponent({ fs }, projectWorkingDirectory = process.cwd()) {
    return {
        dirname(value) {
            return pathToPosix(path_1.default.dirname(value));
        },
        basename(value) {
            return pathToPosix(path_1.default.basename(value));
        },
        join(...paths) {
            return path_1.default.join(...paths);
        },
        async existFile(filePath) {
            const resolvedPath = path_1.default.isAbsolute(filePath) ? filePath : path_1.default.resolve(projectWorkingDirectory, filePath);
            return fs.fileExists(resolvedPath);
        },
        async readFile(filePath) {
            const resolvedPath = path_1.default.isAbsolute(filePath) ? filePath : path_1.default.resolve(projectWorkingDirectory, filePath);
            return fs.readFile(resolvedPath);
        },
        async writeFile(filePath, content) {
            const resolvedPath = path_1.default.isAbsolute(filePath) ? filePath : path_1.default.resolve(projectWorkingDirectory, filePath);
            const folder = path_1.default.dirname(resolvedPath);
            if (!(await fs.directoryExists(folder))) {
                await fs.mkdir(folder, { recursive: true });
            }
            await fs.writeFile(resolvedPath, content);
        },
        async rm(filePath) {
            const resolvedPath = path_1.default.isAbsolute(filePath) ? filePath : path_1.default.resolve(projectWorkingDirectory, filePath);
            await fs.rm(resolvedPath);
        },
        async rmdir(dirPath) {
            const resolvedPath = path_1.default.isAbsolute(dirPath) ? dirPath : path_1.default.resolve(projectWorkingDirectory, dirPath);
            await fs.rm(resolvedPath, { recursive: true });
        },
        async readdir(dirPath) {
            if (dirPath.indexOf('/../') !== -1) {
                throw new Error('The usage of /../ is not allowed');
            }
            const root = dirPath === '.' || dirPath === './' || dirPath === '';
            const resolvedPath = root ? projectWorkingDirectory : dirPath;
            const result = await fs.readdir(resolvedPath);
            return Promise.all(result.map(async (name) => ({
                name: pathToPosix(name),
                isDirectory: await fs.directoryExists(path_1.default.resolve(dirPath, name))
            })));
        },
        cwd() {
            return pathToPosix(projectWorkingDirectory);
        },
        async stat(filePath) {
            const resolvedPath = path_1.default.isAbsolute(filePath) ? filePath : path_1.default.resolve(projectWorkingDirectory, filePath);
            const stats = await fs.stat(resolvedPath);
            return { size: Number(stats.size) };
        }
    };
}
