// @dcl/sdk-commands/dist/commands/deploy/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.args = void 0;
exports.help = help;
exports.main = main;
const path_1 = require("path");
const schemas_1 = require("@dcl/schemas");
const dcl_catalyst_client_1 = require("dcl-catalyst-client");
const fp_future_1 = __importDefault(require("fp-future"));
const i18next_1 = __importDefault(require("i18next"));
const scene_validations_1 = require("../../logic/scene-validations");
const args_1 = require("../../logic/args");
const error_1 = require("../../logic/error");
const beautiful_logs_1 = require("../../logic/beautiful-logs");
const project_files_1 = require("../../logic/project-files");
const utils_1 = require("./utils");
const build_1 = require("../build");
const workspace_validations_1 = require("../../logic/workspace-validations");
const analytics_features_1 = require("./analytics-features");
const config_1 = require("../../logic/config");
exports.args = (0, args_1.declareArgs)({
    '--dir': String,
    '--help': Boolean,
    '-h': '--help',
    '--target': String,
    '-t': '--target',
    '--target-content': String,
    '-tc': '--target-content',
    '--skip-validations': Boolean,
    '--skip-version-checks': Boolean,
    '--skip-build': Boolean,
    '--https': Boolean,
    '--force-upload': Boolean,
    '--yes': Boolean,
    '--no-browser': Boolean,
    '-b': '--no-browser',
    '--port': Number,
    '-p': '--port',
    '--programmatic': Boolean,
    '--multi-scene': Boolean
});
function help(options) {
    options.components.logger.log(`
  Usage: 'sdk-commands build [options]'
    Options:
      -h, --help                Displays complete help
      -p, --port        [port]  Select a custom port for the development server
      -t, --target              Specifies the address and port for the target catalyst server. Defaults to peer.decentraland.org
      -t, --target-content      Specifies the address and port for the target content server. Example: 'peer.decentraland.org/content'. Can't be set together with --target
      -b, --no-browser          Do not open a new browser window
      --skip-version-checks     Skip the ECS and CLI version checks, avoid the warning message and launch anyway
      --skip-build              Skip build before deploy
      --skip-validations        Skip permissions verifications on the client side when deploying content
      --programmatic            Enable programmatic mode - returns a promise that resolves when deployment completes
      --multi-scene             Deploy alongside existing scenes in a world (additive deploy, no deletion)

    Example:
    - Deploy your scene:
      $ sdk-commands deploy
    - Deploy your scene to a specific content server:
      $ sdk-commands deploy --target my-favorite-catalyst-server.org:2323
    - Deploy programmatically:
      $ sdk-commands deploy --programmatic
`);
}
async function main(options) {
    const projectRoot = (0, path_1.resolve)(process.cwd(), options.args['--dir'] || '.');
    const workspace = await (0, workspace_validations_1.getValidWorkspace)(options.components, projectRoot);
    const project = workspace.projects[0];
    const openBrowser = !options.args['--no-browser'];
    const skipBuild = options.args['--skip-build'];
    const linkerPort = options.args['--port'];
    const isProgrammatic = options.args['--programmatic'];
    const multiScene = !!options.args['--multi-scene'];
    const autoYes = !!options.args['--yes'];
    const targetContent = options.args['--target-content'];
    if (workspace.projects.length !== 1) {
        throw new error_1.CliError('DEPLOY_WORKSPACE_NOT_SUPPORTED', i18next_1.default.t('errors.deploy.workspace_not_supported'));
    }
    if (project.kind !== 'scene') {
        throw new error_1.CliError('DEPLOY_INVALID_PROJECT_TYPE', i18next_1.default.t('errors.deploy.invalid_project_type'));
    }
    if (options.args['--target'] && options.args['--target-content']) {
        throw new error_1.CliError('DEPLOY_INVALID_ARGUMENTS', i18next_1.default.t('errors.deploy.invalid_arguments'));
    }
    const sdkVersion = await (0, config_1.getInstalledPackageVersion)(options.components, '@dcl/sdk', projectRoot);
    const sceneJson = { sdkVersion, ...(await (0, scene_validations_1.getValidSceneJson)(options.components, projectRoot, { log: true })) };
    const coords = (0, scene_validations_1.getBaseCoords)(sceneJson);
    const isWorld = (0, utils_1.sceneHasWorldCfg)(sceneJson);
    const worldName = sceneJson.worldConfiguration?.name;
    if (isWorld && !targetContent && !options.args['--target']) {
        throw new error_1.CliError('DEPLOY_MISSING_TARGET_CONTENT', `Your scene.json has worldConfiguration set (world: "${worldName}"). ` +
            `You must specify the worlds content server using --target-content.\n` +
            `Example: sdk-commands deploy --target-content https://worlds-content-server.decentraland.org`);
    }
    const trackProps = {
        projectHash: (0, project_files_1.b64HashingFunction)(projectRoot),
        coords,
        isWorld
    };
    const packageJson = await (0, project_files_1.getPackageJson)(options.components, projectRoot);
    // NOTE: keeping 👇 for query backward-compatibility
    const dependencies = Array.from(new Set([...Object.keys(packageJson.dependencies || {}), ...Object.keys(packageJson.devDependencies || {})]));
    const dependencyVersions = {};
    for (const dep of dependencies) {
        dependencyVersions[dep] = await (0, config_1.getInstalledPackageVersion)(options.components, dep, projectRoot);
    }
    options.components.analytics.track('Scene deploy started', trackProps);
    let needsDelete = false;
    if (isWorld && !multiScene && worldName) {
        try {
            const existingScenes = await (0, utils_1.fetchWorldScenes)(options.components.logger, worldName, targetContent);
            if (existingScenes.length > 0) {
                const scenesOnOtherParcels = (0, utils_1.getScenesOnOtherParcels)(existingScenes, sceneJson.scene.parcels);
                if (scenesOnOtherParcels.length > 0) {
                    options.components.logger.warn(`World "${worldName}" has ${scenesOnOtherParcels.length} other scene(s) that will be removed:`);
                    for (const scene of scenesOnOtherParcels) {
                        const title = scene.entity?.metadata?.display?.title || 'Untitled';
                        const parcels = scene.parcels?.join(', ') || 'unknown parcels';
                        options.components.logger.log(`  - "${title}" at parcels ${parcels}`);
                    }
                    options.components.logger.log('');
                    options.components.logger.warn('Deploying without --multi-scene will DELETE all existing scenes in world first.');
                    if (!autoYes) {
                        const confirmed = await (0, utils_1.promptUser)('Continue? (y/N) ');
                        if (!confirmed) {
                            throw new error_1.CliError('DEPLOY_CANCELLED', 'Deployment cancelled by user.');
                        }
                    }
                    needsDelete = true;
                }
            }
        }
        catch (e) {
            if (e instanceof error_1.CliError)
                throw e;
            options.components.logger.warn(`Could not check existing world scenes: ${e.message}`);
        }
    }
    if (!skipBuild) {
        await (0, build_1.buildScene)({ ...options, args: { '--dir': projectRoot, '--watch': false, '--production': true, _: [] } }, project);
    }
    // Obtain list of files to deploy
    const files = await (0, scene_validations_1.getFiles)(options.components, projectRoot);
    (0, scene_validations_1.validateFilesSizes)(files);
    const contentFiles = new Map(files.map((file) => [file.path, new Uint8Array(file.content)]));
    const trackFeatures = await (0, analytics_features_1.analyticsFeatures)(options.components, (0, path_1.resolve)(projectRoot, sceneJson.main));
    const { entityId, files: entityFiles } = await dcl_catalyst_client_1.DeploymentBuilder.buildEntity({
        type: schemas_1.EntityType.SCENE,
        pointers: sceneJson.scene.parcels,
        files: contentFiles,
        metadata: sceneJson
    });
    const messageToSign = entityId;
    const deleteScenesFromWorldPayload = needsDelete && worldName ? (0, utils_1.buildDeleteScenesFromWorldPayload)(worldName) : undefined;
    const awaitResponse = (0, fp_future_1.default)();
    const { program } = await (0, utils_1.getAddressAndSignature)(options.components, awaitResponse, messageToSign, sceneJson, files, !!options.args['--skip-validations'] || !!options.args['--target'] || !!options.args['--target-content'], {
        openBrowser,
        linkerPort,
        isHttps: !!options.args['--https']
    }, deployEntity, deleteScenesFromWorldPayload, targetContent, multiScene);
    // Programmatic mode early return
    if (isProgrammatic) {
        return {
            finish: async () => {
                const result = await awaitResponse;
                void program?.stop();
                return result;
            },
            stop: async () => {
                void program?.stop();
            }
        };
    }
    try {
        // Keep the CLI live till the user signs the payload
        await awaitResponse;
    }
    finally {
        void program?.stop();
    }
    async function deployEntity(linkerResponse) {
        const { authChain, chainId } = linkerResponse;
        if (isWorld && worldName && targetContent && linkerResponse.address) {
            options.components.logger.log(`[DEPLOY] Checking permissions for wallet ${linkerResponse.address}...`);
            try {
                const permissionCheck = await (0, utils_1.checkWorldDeploymentPermission)(options.components.logger, worldName, linkerResponse.address, targetContent, sceneJson.scene.parcels);
                if (!permissionCheck.allowed) {
                    const deniedMsg = permissionCheck.deniedParcels?.length
                        ? `\nYou don't have permission for parcels: ${permissionCheck.deniedParcels.join(', ')}.`
                        : '';
                    throw new error_1.CliError('DEPLOY_NO_PERMISSION', `Wallet ${linkerResponse.address} does not have permission to deploy to world "${worldName}".${deniedMsg}\n` +
                        `Contact the world owner to grant deployment permissions.`);
                }
                options.components.logger.log(`[DEPLOY] Permissions verified.`);
            }
            catch (e) {
                if (e instanceof error_1.CliError)
                    throw e;
                options.components.logger.warn(`Could not verify deployment permissions: ${e.message}`);
            }
        }
        // Uploading data
        const { client, url } = await (0, utils_1.getCatalyst)(chainId, options.args['--target'], options.args['--target-content']);
        if (needsDelete && worldName && !linkerResponse.deleteSignature) {
            throw new error_1.CliError('DEPLOY_DELETE_FAILED', `Cannot delete existing scenes from "${worldName}": missing signature for deleting the scenes.`);
        }
        if (needsDelete && worldName && linkerResponse.deleteSignature) {
            options.components.logger.info(`[DEPLOY] deleting scenes for "${worldName}"`);
            try {
                const deleteResponse = await (0, utils_1.deleteWorldScenes)(options.components, worldName, linkerResponse.deleteSignature, targetContent);
                if (deleteResponse.ok) {
                    options.components.logger.info(`[DEPLOY] existing scenes for "${worldName} deleted successfully"`);
                }
                else {
                    const errorText = await deleteResponse.text();
                    options.components.logger.info(`[DEPLOY] DELETE FAILED: status=${deleteResponse.status} body=${errorText}`);
                    throw new error_1.CliError('DEPLOY_DELETE_FAILED', `Failed to delete existing scenes from "${worldName}" (status ${deleteResponse.status}): ${errorText}\n`);
                }
            }
            catch (e) {
                if (e instanceof error_1.CliError)
                    throw e;
                throw new error_1.CliError('DEPLOY_DELETE_FAILED', `Error deleting existing scenes from "${worldName}": ${e.message}\n` +
                    `Hint: Use --multi-scene to deploy alongside existing scenes without deleting them.`);
            }
        }
        (0, beautiful_logs_1.printProgressInfo)(options.components.logger, `Uploading data to: ${url}...`);
        const deployData = { entityId, files: entityFiles, authChain };
        const position = sceneJson.scene.base;
        const network = chainId === schemas_1.ChainId.ETHEREUM_SEPOLIA ? 'sepolia' : 'mainnet';
        const worldRealm = isWorld ? `&realm=${sceneJson.worldConfiguration?.name}` : '';
        const domain = chainId === schemas_1.ChainId.ETHEREUM_MAINNET ? 'https://play.decentraland.org' : 'https://play.decentraland.zone';
        const sceneUrl = `${domain}/?NETWORK=${network}&position=${position}${worldRealm}`;
        try {
            const response = (await client.deploy(deployData, {
                timeout: 600000
            }));
            let responseData;
            if (response.status !== 200) {
                responseData = await response.text();
                throw new Error(responseData);
            }
            responseData = (await response.json());
            if (responseData.message) {
                (0, beautiful_logs_1.printProgressInfo)(options.components.logger, responseData.message);
            }
            (0, beautiful_logs_1.printSuccess)(options.components.logger, 'Content uploaded successfully', sceneUrl);
            options.components.analytics.track('Scene deploy success', {
                ...trackProps,
                sceneId: entityId,
                targetContentServer: url,
                worldName: sceneJson.worldConfiguration?.name,
                isPortableExperience: !!sceneJson.isPortableExperience,
                dependencies,
                dependencyVersions,
                serverlessMultiplayer: trackFeatures.serverlessMultiplayer
            });
            if (!isWorld) {
                options.components.logger.info('You can close the terminal now');
            }
        }
        catch (e) {
            options.components.logger.error('Could not upload content:');
            options.components.logger.error(e.message);
            options.components.analytics.track('Scene deploy failure', { ...trackProps, error: e.message ?? '' });
            throw new error_1.CliError('DEPLOY_UPLOAD_FAILED', i18next_1.default.t('errors.deploy.failed_to_upload', { error: e.message }), e.stack);
        }
    }
}
