// @dcl/sdk-commands/dist/commands/deploy/utils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCatalyst = getCatalyst;
exports.buildDeleteScenesFromWorldPayload = buildDeleteScenesFromWorldPayload;
exports.deleteWorldScenes = deleteWorldScenes;
exports.getAddressAndSignature = getAddressAndSignature;
exports.sceneHasWorldCfg = sceneHasWorldCfg;
exports.getSceneInfo = getSceneInfo;
exports.getScenesOnOtherParcels = getScenesOnOtherParcels;
exports.fetchWorldScenes = fetchWorldScenes;
exports.promptUser = promptUser;
exports.fetchWorldPermissions = fetchWorldPermissions;
exports.fetchParcelPermissions = fetchParcelPermissions;
exports.checkWorldDeploymentPermission = checkWorldDeploymentPermission;
const readline = __importStar(require("readline"));
const schemas_1 = require("@dcl/schemas");
const dcl_catalyst_client_1 = require("dcl-catalyst-client");
const contracts_snapshots_1 = require("dcl-catalyst-client/dist/contracts-snapshots");
const fetch_component_1 = require("@well-known-components/fetch-component");
const eth_connect_1 = require("eth-connect");
const crypto_1 = require("@dcl/crypto/dist/crypto");
const crypto_2 = require("@dcl/crypto");
const routes_1 = require("../../linker-dapp/routes");
const account_1 = require("../../logic/account");
const config_1 = require("../../logic/config");
const coordinates_1 = require("../../logic/coordinates");
const catalyst_requests_1 = require("../../logic/catalyst-requests");
const run_dapp_1 = require("../../run-dapp");
async function getCatalyst(chainId = schemas_1.ChainId.ETHEREUM_MAINNET, target, targetContent) {
    if (target) {
        const catalyst = (0, dcl_catalyst_client_1.createCatalystClient)({
            url: target.endsWith('/') ? target.slice(0, -1) : target,
            fetcher: (0, fetch_component_1.createFetchComponent)()
        });
        const content = await catalyst.getContentClient();
        const { lambdas: { publicUrl } } = await catalyst.fetchAbout();
        return { client: content, url: publicUrl };
    }
    if (targetContent) {
        return { client: (0, dcl_catalyst_client_1.createContentClient)({ url: targetContent, fetcher: (0, fetch_component_1.createFetchComponent)() }), url: targetContent };
    }
    let catalystClient;
    if (chainId === schemas_1.ChainId.ETHEREUM_SEPOLIA) {
        catalystClient = (0, dcl_catalyst_client_1.createCatalystClient)({ url: 'peer.decentraland.zone', fetcher: (0, fetch_component_1.createFetchComponent)() });
    }
    else {
        const catalysts = (0, contracts_snapshots_1.getCatalystServersFromCache)('mainnet');
        for (const catalyst of catalysts) {
            try {
                const client = (0, dcl_catalyst_client_1.createCatalystClient)({ url: catalyst.address, fetcher: (0, fetch_component_1.createFetchComponent)() });
                const about = await client.fetchAbout();
                if (about.healthy) {
                    catalystClient = client;
                    break;
                }
            }
            catch (_) {
                // Unreachable or invalid response — skip and try next catalyst
                continue;
            }
        }
    }
    return {
        client: await catalystClient.getContentClient(),
        url: (await catalystClient.fetchAbout()).lambdas.publicUrl
    };
}
function buildDeleteScenesFromWorldPayload(worldName) {
    const encodedName = encodeURIComponent(worldName);
    const path = `/entities/${encodedName}`;
    const timestamp = String(Date.now());
    const metadata = '{}';
    return ['delete', path, timestamp, metadata].join(':').toLowerCase();
}
async function deleteWorldScenes(components, worldName, deleteSignature, targetContent) {
    const { logger } = components;
    const encodedName = encodeURIComponent(worldName);
    const deleteUrl = `${targetContent}/entities/${encodedName}`;
    const authChain = JSON.parse(deleteSignature);
    const lastLink = authChain[authChain.length - 1];
    const payloadParts = lastLink.payload.split(':');
    const timestamp = payloadParts[2] || String(Date.now());
    const metadata = payloadParts[3] || '{}';
    const headers = {
        'x-identity-timestamp': timestamp,
        'x-identity-metadata': metadata
    };
    authChain.forEach((link, i) => {
        headers[`x-identity-auth-chain-${i}`] = JSON.stringify(link);
    });
    const response = await fetch(deleteUrl, { method: 'DELETE', headers });
    logger.info(`[DELETE] deleting world scenes status=${response.status}`);
    return response;
}
async function getAddressAndSignature(components, awaitResponse, messageToSign, scene, files, skipValidations, linkOptions, deployCallback, deleteScenesFromWorldPayload, targetContent, multiScene) {
    if (process.env.DCL_PRIVATE_KEY) {
        const wallet = (0, account_1.createWallet)(process.env.DCL_PRIVATE_KEY);
        const authChain = crypto_2.Authenticator.createSimpleAuthChain(messageToSign, wallet.address, (0, crypto_1.ethSign)((0, eth_connect_1.hexToBytes)(wallet.privateKey), messageToSign));
        let deleteSignature;
        if (deleteScenesFromWorldPayload) {
            const deleteAuthChain = crypto_2.Authenticator.createSimpleAuthChain(deleteScenesFromWorldPayload, wallet.address, (0, crypto_1.ethSign)((0, eth_connect_1.hexToBytes)(wallet.privateKey), deleteScenesFromWorldPayload));
            deleteSignature = JSON.stringify(deleteAuthChain);
        }
        const linkerResponse = { authChain, address: wallet.address, deleteSignature };
        await deployCallback(linkerResponse);
        awaitResponse.resolve();
        return {};
    }
    const sceneInfo = await getSceneInfo(components, scene, messageToSign, skipValidations, deleteScenesFromWorldPayload, targetContent, multiScene);
    const { router: commonRouter } = (0, routes_1.setRoutes)(components, sceneInfo);
    const router = setDeployRoutes(commonRouter, components, awaitResponse, sceneInfo, files, deployCallback);
    components.logger.info('You need to sign the content before the deployment:');
    const { program } = await (0, run_dapp_1.runDapp)(components, router, { ...linkOptions, uri: `/` });
    return { program };
}
function setDeployRoutes(router, components, awaitResponse, sceneInfo, files, deployCallback) {
    const { logger } = components;
    const resolveLinkerPromise = () => setTimeout(() => awaitResponse.resolve(), 100);
    const rejectLinkerPromise = (e) => setTimeout(() => awaitResponse.reject(e), 100);
    let linkerResponse;
    router.get('/api/files', async () => ({
        body: files.map((file) => ({
            name: file.path,
            size: file.size
        }))
    }));
    router.get('/api/catalyst-pointers', async () => {
        const { x, y } = (0, coordinates_1.getObject)(sceneInfo.baseParcel);
        const pointer = `${x},${y}`;
        const chainId = linkerResponse?.chainId || 1;
        const network = chainId === schemas_1.ChainId.ETHEREUM_MAINNET ? 'mainnet' : 'sepolia';
        const value = await (0, catalyst_requests_1.getPointers)(components, pointer, network);
        const deployedToAll = new Set(value.map((c) => c.entityId)).size === 1;
        if (deployedToAll)
            resolveLinkerPromise();
        return {
            body: { catalysts: value }
        };
    });
    router.get('/api/world-parcel-permissions/:address', async (ctx) => {
        if (!sceneInfo.isWorld || !sceneInfo.world || !sceneInfo.targetContent) {
            return { status: 400, body: { message: 'Not a world deployment or missing targetContent' } };
        }
        try {
            const address = ctx.params.address;
            const permissions = await fetchWorldPermissions(sceneInfo.world, sceneInfo.targetContent);
            const lowerAddress = address.toLowerCase();
            const allGranted = sceneInfo.parcels.map((p) => {
                const [x, y] = p.split(',');
                return { x: parseInt(x, 10), y: parseInt(y, 10), isUpdateAuthorized: true };
            });
            if (permissions.owner?.toLowerCase() === lowerAddress) {
                return { body: { authorizations: allGranted, worldWidePermission: true } };
            }
            const walletSummary = permissions.summary?.[lowerAddress];
            const deploymentSummary = walletSummary?.find((s) => s.permission === 'deployment');
            const hasWorldWide = deploymentSummary?.world_wide ?? false;
            if (hasWorldWide) {
                return { body: { authorizations: allGranted, worldWidePermission: true } };
            }
            const allowedParcels = await fetchParcelPermissions(sceneInfo.world, address, sceneInfo.targetContent);
            const allowedSet = new Set(allowedParcels);
            return {
                body: {
                    authorizations: sceneInfo.parcels.map((p) => {
                        const [x, y] = p.split(',');
                        return { x: parseInt(x, 10), y: parseInt(y, 10), isUpdateAuthorized: allowedSet.has(p) };
                    }),
                    worldWidePermission: false
                }
            };
        }
        catch (e) {
            logger.error(`Error fetching world parcel permissions: ${e.message}`);
            return { status: 500, body: { message: e.message } };
        }
    });
    router.post('/api/deploy', async (ctx) => {
        const value = (await ctx.request.json());
        if (!value.address || !value.authChain || !value.chainId) {
            const errorMessage = `Invalid payload: ${Object.keys(value).join(' - ')}`;
            logger.error(errorMessage);
            resolveLinkerPromise();
            return { status: 400, body: { message: errorMessage } };
        }
        linkerResponse = value;
        try {
            await deployCallback(value);
            if (sceneInfo.isWorld) {
                resolveLinkerPromise();
            }
            return {};
        }
        catch (e) {
            rejectLinkerPromise(e);
            return { status: 400, body: { message: e.message } };
        }
    });
    return router;
}
function sceneHasWorldCfg(scene) {
    return !!Object.keys(scene.worldConfiguration || {}).length;
}
async function getSceneInfo(components, scene, rootCID, skipValidations, deleteScenesFromWorldPayload, targetContent, multiScene) {
    const { scene: { parcels, base }, display, isPortableExperience } = scene;
    return {
        baseParcel: base,
        parcels,
        rootCID,
        landRegistry: await (0, config_1.getLandRegistry)(components),
        estateRegistry: await (0, config_1.getEstateRegistry)(components),
        debug: !!process.env.DEBUG,
        title: display?.title,
        description: display?.description,
        skipValidations,
        isPortableExperience: !!isPortableExperience,
        isWorld: sceneHasWorldCfg(scene),
        world: scene.worldConfiguration?.name,
        deleteScenesFromWorldPayload,
        targetContent,
        multiScene
    };
}
function getScenesOnOtherParcels(existingScenes, deployingParcels) {
    const parcelsSet = new Set(deployingParcels);
    return existingScenes.filter((scene) => {
        const sceneParcels = scene.parcels || [];
        return sceneParcels.every((p) => !parcelsSet.has(p));
    });
}
async function fetchWorldScenes(logger, worldName, targetContent) {
    const encodedName = encodeURIComponent(worldName);
    const url = `${targetContent}/world/${encodedName}/scenes`;
    const response = await fetch(url);
    if (!response.ok) {
        const text = await response.text();
        logger.info(`[DEPLOY] fetching scenes from world -  error: ${text}`);
        throw new Error(`Failed to fetch world scenes: ${response.status} ${response.statusText}`);
    }
    const data = (await response.json());
    logger.info(`[DEPLOY] fetching scenes from world success: total=${data.total}, scenes=${data.scenes?.length}`);
    return data.scenes || [];
}
function promptUser(question) {
    const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
    return new Promise((resolve) => {
        rl.question(question, (answer) => {
            rl.close();
            resolve(answer.toLowerCase() === 'y' || answer.toLowerCase() === 'yes');
        });
    });
}
async function fetchWorldPermissions(worldName, targetContent) {
    const encodedName = encodeURIComponent(worldName);
    const url = `${targetContent}/world/${encodedName}/permissions`;
    const response = await fetch(url);
    if (!response.ok) {
        throw new Error(`Failed to fetch world permissions: ${response.status} ${response.statusText}`);
    }
    return (await response.json());
}
async function fetchParcelPermissions(worldName, address, targetContent) {
    const encodedName = encodeURIComponent(worldName);
    const url = `${targetContent}/world/${encodedName}/permissions/deployment/address/${address.toLowerCase()}/parcels`;
    const response = await fetch(url);
    if (!response.ok) {
        throw new Error(`Failed to fetch parcel permissions: ${response.status} ${response.statusText}`);
    }
    const data = (await response.json());
    return data.parcels || [];
}
async function checkWorldDeploymentPermission(logger, worldName, address, targetContent, deployingParcels) {
    const permissions = await fetchWorldPermissions(worldName, targetContent);
    const lowerAddress = address.toLowerCase();
    // Check if wallet is the world owner
    if (permissions.owner?.toLowerCase() === lowerAddress) {
        return { allowed: true };
    }
    // Check world-wide deployment permission
    const deploymentPermission = permissions.permissions?.deployment;
    if (deploymentPermission) {
        if (deploymentPermission.type === 'unrestricted') {
            return { allowed: true };
        }
        const allowedWallets = (deploymentPermission.wallets || []).map((w) => w.toLowerCase());
        if (allowedWallets.includes(lowerAddress)) {
            // Check if wallet has world-wide permission or only per-parcel
            const walletSummary = permissions.summary?.[lowerAddress];
            const deploymentSummary = walletSummary?.find((s) => s.permission === 'deployment');
            if (!deploymentSummary || deploymentSummary.world_wide) {
                return { allowed: true };
            }
        }
    }
    // Check per-parcel permissions
    const allowedParcels = await fetchParcelPermissions(worldName, address, targetContent);
    const allowedSet = new Set(allowedParcels);
    const deniedParcels = deployingParcels.filter((p) => !allowedSet.has(p));
    if (deniedParcels.length === 0) {
        return { allowed: true };
    }
    return { allowed: false, deniedParcels };
}
