// @dcl/asset-packs/dist/admin-toolkit-ui/VideoControl/DclCast/DclCastInfo.js
import ReactEcs, { Label, UiEntity } from '@dcl/react-ecs';
import { copyToClipboard } from '~system/RestrictedActions';
import { Color4 } from '@dcl/sdk/math';
import { Button } from '../../Button';
import { FeedbackButton } from '../../FeedbackButton';
import { VideoControlVolume } from '../VolumeControl';
import { createVideoPlayerControls, isDclCast } from '../utils';
import { LIVEKIT_STREAM_SRC } from '../LiveStream';
import { getDclCastStyles, getDclCastColors, getDclCastBackgrounds } from './styles';
import { CONTENT_URL } from '../../constants';
const ICONS = {
    COPY_TO_CLIPBOARD_ICON: `${CONTENT_URL}/admin_toolkit/assets/icons/copy-to-clipboard.png`,
};
const DclCastInfo = ({ state, engine, onResetRoomId, entity, video, }) => {
    const controls = createVideoPlayerControls(entity, engine);
    const styles = getDclCastStyles();
    const colors = getDclCastColors();
    const backgrounds = getDclCastBackgrounds();
    return (ReactEcs.createElement(UiEntity, { uiTransform: styles.fullContainer },
        ReactEcs.createElement(UiEntity, { uiTransform: styles.mainBorderedContainer },
            ReactEcs.createElement(UiEntity, { uiTransform: styles.headerRow },
                ReactEcs.createElement(UiEntity, { uiTransform: styles.columnFlexStart },
                    ReactEcs.createElement(Label, { value: '<b>Room ID</b>', fontSize: 24, color: Color4.White() }),
                    ReactEcs.createElement(Label, { value: `Expires in ${state.videoControl.dclCast?.expiresInDays} days`, fontSize: 14, color: colors.gray, uiTransform: styles.marginTopSmall })),
                video?.src && isDclCast(video.src) ? (ReactEcs.createElement(Button, { id: "dcl_cast_deactivate", value: "<b>Deactivate</b>", variant: "text", fontSize: 16, color: colors.white, uiTransform: styles.activateButton, onMouseDown: () => {
                        controls.setSource('');
                        state.videoControl.selectedStream = undefined;
                    } })) : (ReactEcs.createElement(Button, { id: "dcl_cast_activate", value: "<b>Activate</b>", labelTransform: styles.activateButtonLabel, uiTransform: styles.activateButton, fontSize: 16, uiBackground: backgrounds.success, color: colors.black, onMouseDown: () => {
                        controls.setSource(LIVEKIT_STREAM_SRC);
                        state.videoControl.selectedStream = 'dcl-cast';
                    } }))),
            ReactEcs.createElement(UiEntity, { uiTransform: styles.columnContainer },
                ReactEcs.createElement(UiEntity, { uiTransform: styles.rowCenterSpaceBetween },
                    ReactEcs.createElement(UiEntity, { uiTransform: {
                            display: 'flex',
                            flexDirection: 'column',
                        } },
                        ReactEcs.createElement(Label, { value: '<b>Cast speakers</b>', fontSize: 18, color: colors.white }),
                        ReactEcs.createElement(UiEntity, { uiText: {
                                value: 'This link grants streaming access.',
                                fontSize: 14,
                                color: colors.gray,
                                textAlign: 'top-left',
                            } })),
                    ReactEcs.createElement(FeedbackButton, { id: "dcl_cast_copy_stream_link", value: "<b>Copy Link</b>", variant: "text", fontSize: 18, color: colors.white, iconRight: ICONS.COPY_TO_CLIPBOARD_ICON, iconRightTransform: styles.iconSmall, labelTransform: styles.marginRightSmall, uiTransform: styles.copyLinkButton, onMouseDown: () => {
                            state.videoControl.dclCast?.streamLink &&
                                copyToClipboard({
                                    text: state.videoControl.dclCast.streamLink,
                                });
                        } })),
                ReactEcs.createElement(UiEntity, { uiTransform: styles.separatorLine }),
                ReactEcs.createElement(UiEntity, { uiTransform: styles.rowCenterSpaceBetween },
                    ReactEcs.createElement(UiEntity, { uiTransform: styles.textInfoContainer },
                        ReactEcs.createElement(Label, { value: '<b>Viewers</b>', fontSize: 18, color: colors.white }),
                        ReactEcs.createElement(UiEntity, { uiText: {
                                value: 'This link grants viewing access.',
                                fontSize: 14,
                                color: colors.gray,
                                textAlign: 'top-left',
                            } })),
                    ReactEcs.createElement(FeedbackButton, { id: "dcl_cast_copy_watcher_link", value: "<b>Copy Link</b>", variant: "text", fontSize: 18, color: colors.white, iconRight: ICONS.COPY_TO_CLIPBOARD_ICON, iconRightTransform: styles.iconSmall, labelTransform: styles.marginRightSmall, uiTransform: styles.copyLinkButton, onMouseDown: () => {
                            state.videoControl.dclCast?.watcherLink &&
                                copyToClipboard({
                                    text: state.videoControl.dclCast.watcherLink,
                                });
                        } })))),
        ReactEcs.createElement(UiEntity, { uiTransform: styles.columnWithMarginTop },
            ReactEcs.createElement(VideoControlVolume, { engine: engine, entity: entity, video: video, label: "<b>Cast volume</b>" }),
            ReactEcs.createElement(UiEntity, null,
                ReactEcs.createElement(Button, { id: "dcl_cast_reset_room_id", value: "<b>Reset Room</b>", variant: "text", fontSize: 16, color: colors.danger, uiTransform: styles.resetButton, onMouseDown: onResetRoomId })))));
};
export default DclCastInfo;
