// @dcl/asset-packs/dist/admin-toolkit-ui/VideoControl/DclCast/index.js
import ReactEcs, { UiEntity } from '@dcl/react-ecs';
import { Color4 } from '@dcl/sdk/math';
import { getDclCastInfo, resetStreamKey } from '../api';
import { CONTENT_URL } from '../../constants';
import { Header } from '../../Header';
import DclCastInfo from './DclCastInfo';
import { LoadingDots } from '../../Loading';
import { Button } from '../../Button';
import { getDclCastStyles, getDclCastColors } from './styles';
import { getComponents } from '../../../definitions';
const ICONS = {
    DCL_CAST_ICON: `${CONTENT_URL}/admin_toolkit/assets/icons/video-control-dcl-cast.png`,
};
export async function handleGetDclCastInfo(state) {
    const [error, data] = await getDclCastInfo();
    if (error) {
        console.error(error);
        return null;
    }
    else {
        if (data) {
            state.videoControl.dclCast = data;
            return data;
        }
    }
}
const DclCast = ({ engine, state, entity, video, }) => {
    const { VideoControlState } = getComponents(engine);
    const styles = getDclCastStyles();
    const colors = getDclCastColors();
    const [isLoading, setIsLoading] = ReactEcs.useState(false);
    const [error, setError] = ReactEcs.useState(false);
    const fetchDclCastInfo = async () => {
        console.log('fetching DCL cast info...');
        setIsLoading(true);
        setError(false);
        const result = await handleGetDclCastInfo(state);
        if (!result) {
            setError(true);
        }
        setIsLoading(false);
    };
    const handleResetRoomId = async () => {
        setIsLoading(true);
        const [error, data] = await resetStreamKey();
        if (error) {
            setIsLoading(false);
            setError(true);
        }
        else {
            const videoControl = VideoControlState.getMutable(state.adminToolkitUiEntity);
            videoControl.endsAt = data?.endsAt;
            fetchDclCastInfo();
        }
    };
    ReactEcs.useEffect(() => {
        fetchDclCastInfo();
    }, []);
    return (ReactEcs.createElement(UiEntity, { uiTransform: styles.fullContainer },
        ReactEcs.createElement(Header, { iconSrc: ICONS.DCL_CAST_ICON, title: "DCL Cast" }),
        ReactEcs.createElement(UiEntity, { uiTransform: styles.fullWidthWithBottomMargin },
            ReactEcs.createElement(UiEntity, { uiText: {
                    value: 'Use a browser-based DCL Cast room to easily stream camera and screen feed to a screen in your scene.',
                    fontSize: 16,
                    color: Color4.fromHexString('#A09BA8'),
                    textAlign: 'top-left',
                    textWrap: 'wrap',
                }, uiTransform: styles.marginBottomSmall })),
        isLoading && (ReactEcs.createElement(LoadingDots, { uiTransform: { minHeight: 400 }, engine: engine })),
        error && (ReactEcs.createElement(UiEntity, { uiTransform: styles.columnCentered },
            ReactEcs.createElement(UiEntity, { uiText: {
                    value: '<b>Failed to fetch DCL Cast info</b>',
                    fontSize: 16,
                    color: Color4.White(),
                }, uiTransform: { margin: { bottom: 8 } } }),
            ReactEcs.createElement(UiEntity, { uiText: {
                    value: 'Please retry.',
                    fontSize: 16,
                    color: Color4.Gray(),
                } }),
            ReactEcs.createElement(Button, { id: "dcl_cast_retry", value: "<b>Retry</b>", variant: "secondary", fontSize: 16, color: colors.white, onMouseDown: () => {
                    handleGetDclCastInfo(state);
                }, uiTransform: styles.retryButton }))),
        !isLoading && !error && (ReactEcs.createElement(DclCastInfo, { state: state, entity: entity, engine: engine, video: video, onResetRoomId: handleResetRoomId }))));
};
export default DclCast;
