// @dcl/gltf-validator-ts/dist/esm/parser.js
export class GLTFParser {
    parse(data) {
        // Check for BOM at the beginning of JSON file (UTF-8 BOM is EF BB BF)
        if (data.length >= 3 &&
            data[0] === 0xEF &&
            data[1] === 0xBB &&
            data[2] === 0xBF) {
            throw new Error(`Invalid JSON data. Parser output: BOM found at the beginning of UTF-8 stream.`);
        }
        const text = new TextDecoder().decode(data);
        try {
            const json = JSON.parse(text);
            // Check if parsed JSON is an object (not array or primitive)
            if (typeof json !== 'object' || json === null || Array.isArray(json)) {
                const typeDescription = Array.isArray(json) ? '[]' : typeof json;
                throw new Error(`TYPE_MISMATCH:Type mismatch. Property value ${typeDescription} is not a 'object'.:`);
            }
            return this.validateAndNormalize(json);
        }
        catch (error) {
            throw new Error(`Failed to parse GLTF JSON: ${error}`);
        }
    }
    validateAndNormalize(json) {
        // Basic structure validation - don't throw, let validator handle it
        // if (!json.asset || !json.asset.version) {
        //   throw new Error('GLTF must have an asset object with version');
        // }
        // Normalize arrays to ensure they exist
        const jsonObj = json;
        const gltf = {
            asset: jsonObj['asset'],
            ...(jsonObj['scene'] !== undefined && { scene: jsonObj['scene'] }),
            scenes: Array.isArray(jsonObj['scenes']) ? jsonObj['scenes'] : [],
            nodes: Array.isArray(jsonObj['nodes']) ? jsonObj['nodes'] : [],
            materials: Array.isArray(jsonObj['materials']) ? jsonObj['materials'] : [],
            accessors: Array.isArray(jsonObj['accessors']) ? jsonObj['accessors'] : [],
            animations: Array.isArray(jsonObj['animations']) ? jsonObj['animations'] : [],
            buffers: Array.isArray(jsonObj['buffers']) ? jsonObj['buffers'] : [],
            bufferViews: Array.isArray(jsonObj['bufferViews']) ? jsonObj['bufferViews'] : [],
            cameras: Array.isArray(jsonObj['cameras']) ? jsonObj['cameras'] : [],
            images: Array.isArray(jsonObj['images']) ? jsonObj['images'] : [],
            meshes: Array.isArray(jsonObj['meshes']) ? jsonObj['meshes'] : [],
            samplers: (jsonObj['samplers'] || []),
            skins: Array.isArray(jsonObj['skins']) ? jsonObj['skins'] : [],
            textures: Array.isArray(jsonObj['textures']) ? jsonObj['textures'] : [],
            extensions: jsonObj['extensions'],
            extras: jsonObj['extras']
        };
        // Track which properties were explicitly defined in the original JSON
        gltf._explicitlyDefined = {
            scenes: 'scenes' in jsonObj,
            nodes: 'nodes' in jsonObj,
            materials: 'materials' in jsonObj,
            accessors: 'accessors' in jsonObj,
            animations: 'animations' in jsonObj,
            buffers: 'buffers' in jsonObj,
            bufferViews: 'bufferViews' in jsonObj,
            cameras: 'cameras' in jsonObj,
            images: 'images' in jsonObj,
            meshes: 'meshes' in jsonObj,
            samplers: 'samplers' in jsonObj,
            skins: 'skins' in jsonObj,
            textures: 'textures' in jsonObj
        };
        // Copy any additional properties
        for (const key in jsonObj) {
            if (!(key in gltf)) {
                gltf[key] = jsonObj[key];
            }
        }
        return gltf;
    }
}
