// @dcl/gltf-validator-ts/dist/esm/type-guards.js
/**
 * Advanced TypeScript type guards for runtime type safety
 * These provide compile-time and runtime type checking for GLTF validation
 */
import { GLTF_MIME_TYPES, IMAGE_MIME_TYPES, ComponentType, AccessorType } from './types';
// Basic type guards
export function isObject(value) {
    return typeof value === 'object' && value !== null && !Array.isArray(value);
}
export function isNonEmptyArray(value) {
    return Array.isArray(value) && value.length > 0;
}
export function isString(value) {
    return typeof value === 'string';
}
export function isNumber(value) {
    return typeof value === 'number' && Number.isFinite(value);
}
export function isPositiveNumber(value) {
    return isNumber(value) && value > 0;
}
export function isNonNegativeNumber(value) {
    return isNumber(value) && value >= 0;
}
export function isInteger(value) {
    return isNumber(value) && Number.isInteger(value);
}
export function isPositiveInteger(value) {
    return isInteger(value) && value > 0;
}
export function isNonNegativeInteger(value) {
    return isInteger(value) && value >= 0;
}
// Branded type constructors
export function createJSONPointer(pointer) {
    if (!pointer.startsWith('/')) {
        throw new Error(`Invalid JSON pointer: ${pointer}. Must start with '/'`);
    }
    return pointer;
}
export function createMimeType(mimeType) {
    if (!mimeType.includes('/')) {
        throw new Error(`Invalid MIME type: ${mimeType}. Must contain '/'`);
    }
    return mimeType;
}
export function createBase64String(base64) {
    try {
        atob(base64);
        return base64;
    }
    catch {
        throw new Error(`Invalid base64 string: ${base64}`);
    }
}
// GLTF component type guards
export function isGLTFMaterial(value) {
    if (!isObject(value))
        return false;
    // Check for material-specific properties
    const material = value;
    return (
    // At least one material property should exist
    material['alphaMode'] !== undefined ||
        material['alphaCutoff'] !== undefined ||
        material['doubleSided'] !== undefined ||
        material['pbrMetallicRoughness'] !== undefined ||
        material['normalTexture'] !== undefined ||
        material['occlusionTexture'] !== undefined ||
        material['emissiveTexture'] !== undefined ||
        material['emissiveFactor'] !== undefined ||
        material['name'] !== undefined ||
        material['extensions'] !== undefined ||
        material['extras'] !== undefined);
}
export function isMaterialWithExtensions(value) {
    return isGLTFMaterial(value);
}
export function isGLTFMesh(value) {
    if (!isObject(value))
        return false;
    const mesh = value;
    return Array.isArray(mesh['primitives']);
}
export function isMeshWithExtensions(value) {
    return isGLTFMesh(value);
}
export function isGLTFAnimation(value) {
    if (!isObject(value))
        return false;
    const animation = value;
    return Array.isArray(animation['channels']) && Array.isArray(animation['samplers']);
}
export function isAnimationWithExtensions(value) {
    return isGLTFAnimation(value);
}
export function isGLTFAccessor(value) {
    if (!isObject(value))
        return false;
    const accessor = value;
    return (isNumber(accessor['componentType']) &&
        isNumber(accessor['count']) &&
        isString(accessor['type']));
}
export function isGLTFBufferView(value) {
    if (!isObject(value))
        return false;
    const bufferView = value;
    return (isNonNegativeInteger(bufferView['buffer']) &&
        isPositiveInteger(bufferView['byteLength']));
}
export function isGLTFNode(value) {
    if (!isObject(value))
        return false;
    // Node can be empty or have various properties
    const node = value;
    // Check that if properties exist, they have correct types
    if (node['camera'] !== undefined && !isNonNegativeInteger(node['camera']))
        return false;
    if (node['children'] !== undefined && !Array.isArray(node['children']))
        return false;
    if (node['skin'] !== undefined && !isNonNegativeInteger(node['skin']))
        return false;
    if (node['matrix'] !== undefined && (!Array.isArray(node['matrix']) || node['matrix'].length !== 16))
        return false;
    if (node['mesh'] !== undefined && !isNonNegativeInteger(node['mesh']))
        return false;
    if (node['rotation'] !== undefined && (!Array.isArray(node['rotation']) || node['rotation'].length !== 4))
        return false;
    if (node['scale'] !== undefined && (!Array.isArray(node['scale']) || node['scale'].length !== 3))
        return false;
    if (node['translation'] !== undefined && (!Array.isArray(node['translation']) || node['translation'].length !== 3))
        return false;
    if (node['weights'] !== undefined && !Array.isArray(node['weights']))
        return false;
    return true;
}
// Validation context type guards
export function hasExtensions(value) {
    return isObject(value) && isObject(value['extensions']);
}
// Extension type guards
export function isExtensionName(value) {
    const validExtensions = [
        'KHR_materials_anisotropy',
        'KHR_materials_clearcoat',
        'KHR_materials_dispersion',
        'KHR_materials_emissive_strength',
        'KHR_materials_ior',
        'KHR_materials_iridescence',
        'KHR_materials_pbrSpecularGlossiness',
        'KHR_materials_sheen',
        'KHR_materials_specular',
        'KHR_materials_transmission',
        'KHR_materials_volume',
        'KHR_texture_transform',
        'KHR_lights_punctual',
        'KHR_materials_variants'
    ];
    return validExtensions.includes(value);
}
export function isExtensionOfType(extension, extensionName) {
    if (!isObject(extension))
        return false;
    // Type-specific validation based on extension name
    switch (extensionName) {
        case 'KHR_materials_anisotropy':
            return isKHRMaterialsAnisotropyExtension(extension);
        case 'KHR_materials_clearcoat':
            return isKHRMaterialsClearcoatExtension(extension);
        // Add more cases as needed
        default:
            return true; // Fallback for unknown extensions
    }
}
// Specific extension type guards
function isKHRMaterialsAnisotropyExtension(value) {
    if (!isObject(value))
        return false;
    const ext = value;
    return ((ext['anisotropyStrength'] === undefined || isNonNegativeNumber(ext['anisotropyStrength'])) &&
        (ext['anisotropyRotation'] === undefined || isNumber(ext['anisotropyRotation'])) &&
        (ext['anisotropyTexture'] === undefined || isObject(ext['anisotropyTexture'])));
}
function isKHRMaterialsClearcoatExtension(value) {
    if (!isObject(value))
        return false;
    const ext = value;
    return ((ext['clearcoatFactor'] === undefined || isNonNegativeNumber(ext['clearcoatFactor'])) &&
        (ext['clearcoatRoughnessFactor'] === undefined || isNonNegativeNumber(ext['clearcoatRoughnessFactor'])) &&
        (ext['clearcoatTexture'] === undefined || isObject(ext['clearcoatTexture'])) &&
        (ext['clearcoatRoughnessTexture'] === undefined || isObject(ext['clearcoatRoughnessTexture'])) &&
        (ext['clearcoatNormalTexture'] === undefined || isObject(ext['clearcoatNormalTexture'])));
}
// MIME type guards
export function isImageMimeType(value) {
    return IMAGE_MIME_TYPES.includes(value);
}
export function isGLTFMimeType(value) {
    return GLTF_MIME_TYPES.includes(value);
}
// Component type guards
export function isValidComponentType(value) {
    const validTypes = [
        ComponentType.BYTE,
        ComponentType.UNSIGNED_BYTE,
        ComponentType.SHORT,
        ComponentType.UNSIGNED_SHORT,
        ComponentType.UNSIGNED_INT,
        ComponentType.FLOAT,
        ComponentType.DOUBLE
    ];
    return validTypes.includes(value);
}
export function isValidAccessorType(value) {
    const validTypes = [
        AccessorType.SCALAR,
        AccessorType.VEC2,
        AccessorType.VEC3,
        AccessorType.VEC4,
        AccessorType.MAT2,
        AccessorType.MAT3,
        AccessorType.MAT4
    ];
    return validTypes.includes(value);
}
// Array validation helpers
export function isArrayOfType(value, itemGuard) {
    return Array.isArray(value) && value.every(itemGuard);
}
export function isNonEmptyArrayOfType(value, itemGuard) {
    return isArrayOfType(value, itemGuard) && value.length > 0;
}
// Utility assertion functions
export function assertIsObject(value, context) {
    if (!isObject(value)) {
        throw new Error(`Expected object in ${context}, got ${typeof value}`);
    }
}
export function assertIsNumber(value, context) {
    if (!isNumber(value)) {
        throw new Error(`Expected number in ${context}, got ${typeof value}`);
    }
}
export function assertIsString(value, context) {
    if (!isString(value)) {
        throw new Error(`Expected string in ${context}, got ${typeof value}`);
    }
}
