// @dcl/gltf-validator-ts/dist/esm/types.js
// Severity levels
export var Severity;
(function (Severity) {
    Severity[Severity["ERROR"] = 0] = "ERROR";
    Severity[Severity["WARNING"] = 1] = "WARNING";
    Severity[Severity["INFO"] = 2] = "INFO";
    Severity[Severity["HINT"] = 3] = "HINT";
})(Severity || (Severity = {}));
// Component type constants
export const ComponentType = {
    BYTE: 5120,
    UNSIGNED_BYTE: 5121,
    SHORT: 5122,
    UNSIGNED_SHORT: 5123,
    UNSIGNED_INT: 5125,
    FLOAT: 5126,
    DOUBLE: 5130
};
// Accessor type constants
export const AccessorType = {
    SCALAR: 'SCALAR',
    VEC2: 'VEC2',
    VEC3: 'VEC3',
    VEC4: 'VEC4',
    MAT2: 'MAT2',
    MAT3: 'MAT3',
    MAT4: 'MAT4'
};
// Animation path constants
export const AnimationPath = {
    TRANSLATION: 'translation',
    ROTATION: 'rotation',
    SCALE: 'scale',
    WEIGHTS: 'weights'
};
// Animation interpolation constants
export const AnimationInterpolation = {
    LINEAR: 'LINEAR',
    STEP: 'STEP',
    CUBICSPLINE: 'CUBICSPLINE'
};
// Material alpha mode constants
export const AlphaMode = {
    OPAQUE: 'OPAQUE',
    MASK: 'MASK',
    BLEND: 'BLEND'
};
// Primitive mode constants
export const PrimitiveMode = {
    POINTS: 0,
    LINES: 1,
    LINE_LOOP: 2,
    LINE_STRIP: 3,
    TRIANGLES: 4,
    TRIANGLE_STRIP: 5,
    TRIANGLE_FAN: 6
};
// Filter constants
export const Filter = {
    NEAREST: 9728,
    LINEAR: 9729,
    NEAREST_MIPMAP_NEAREST: 9984,
    LINEAR_MIPMAP_NEAREST: 9985,
    NEAREST_MIPMAP_LINEAR: 9986,
    LINEAR_MIPMAP_LINEAR: 9987
};
// Wrap mode constants
export const WrapMode = {
    CLAMP_TO_EDGE: 33071,
    MIRRORED_REPEAT: 33648,
    REPEAT: 10497
};
// Camera type constants
export const CameraType = {
    PERSPECTIVE: 'perspective',
    ORTHOGRAPHIC: 'orthographic'
};
// Const assertion helpers
export const GLTF_MIME_TYPES = ['model/gltf+json', 'model/gltf-binary'];
export const IMAGE_MIME_TYPES = ['image/jpeg', 'image/png', 'image/webp', 'image/ktx2'];
export const SEVERITY_LEVELS = ['ERROR', 'WARNING', 'INFO', 'HINT'];
// Validation code constants with const assertions
export const VALIDATION_CODES = {
    // Type validation codes
    TYPE_MISMATCH: 'TYPE_MISMATCH',
    UNDEFINED_PROPERTY: 'UNDEFINED_PROPERTY',
    UNEXPECTED_PROPERTY: 'UNEXPECTED_PROPERTY',
    INVALID_VALUE: 'INVALID_VALUE',
    VALUE_NOT_IN_RANGE: 'VALUE_NOT_IN_RANGE',
    VALUE_NOT_IN_LIST: 'VALUE_NOT_IN_LIST',
    UNRESOLVED_REFERENCE: 'UNRESOLVED_REFERENCE',
    // GLTF specific codes
    EMPTY_ENTITY: 'EMPTY_ENTITY',
    UNUSED_OBJECT: 'UNUSED_OBJECT',
    ACCESSOR_NORMALIZED_INVALID: 'ACCESSOR_NORMALIZED_INVALID',
    BUFFER_DATA_URI_MIME_TYPE_INVALID: 'BUFFER_DATA_URI_MIME_TYPE_INVALID',
    IMAGE_NON_ENABLED_MIME_TYPE: 'IMAGE_NON_ENABLED_MIME_TYPE',
    MATERIAL_ALPHA_CUTOFF_INVALID_MODE: 'MATERIAL_ALPHA_CUTOFF_INVALID_MODE',
    NODE_MATRIX_TRS: 'NODE_MATRIX_TRS',
    CAMERA_YFOV_GEQUAL_PI: 'CAMERA_YFOV_GEQUAL_PI',
    // Extension codes
    UNDECLARED_EXTENSION: 'UNDECLARED_EXTENSION',
    UNEXPECTED_EXTENSION_OBJECT: 'UNEXPECTED_EXTENSION_OBJECT',
    INCOMPLETE_EXTENSION_SUPPORT: 'INCOMPLETE_EXTENSION_SUPPORT'
};
// Component type constants with const assertions
export const COMPONENT_TYPE_SIZES = {
    [ComponentType.BYTE]: 1,
    [ComponentType.UNSIGNED_BYTE]: 1,
    [ComponentType.SHORT]: 2,
    [ComponentType.UNSIGNED_SHORT]: 2,
    [ComponentType.UNSIGNED_INT]: 4,
    [ComponentType.FLOAT]: 4,
    [ComponentType.DOUBLE]: 8
};
export const ACCESSOR_TYPE_COMPONENT_COUNTS = {
    [AccessorType.SCALAR]: 1,
    [AccessorType.VEC2]: 2,
    [AccessorType.VEC3]: 3,
    [AccessorType.VEC4]: 4,
    [AccessorType.MAT2]: 4,
    [AccessorType.MAT3]: 9,
    [AccessorType.MAT4]: 16
};
// Type-safe object keys
export function getTypedKeys(obj) {
    return Object.keys(obj);
}
// Type-safe object entries
export function getTypedEntries(obj) {
    return Object.entries(obj);
}
