// @dcl/gltf-validator-ts/dist/cjs/type-guards.js
"use strict";
/**
 * Advanced TypeScript type guards for runtime type safety
 * These provide compile-time and runtime type checking for GLTF validation
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.isObject = isObject;
exports.isNonEmptyArray = isNonEmptyArray;
exports.isString = isString;
exports.isNumber = isNumber;
exports.isPositiveNumber = isPositiveNumber;
exports.isNonNegativeNumber = isNonNegativeNumber;
exports.isInteger = isInteger;
exports.isPositiveInteger = isPositiveInteger;
exports.isNonNegativeInteger = isNonNegativeInteger;
exports.createJSONPointer = createJSONPointer;
exports.createMimeType = createMimeType;
exports.createBase64String = createBase64String;
exports.isGLTFMaterial = isGLTFMaterial;
exports.isMaterialWithExtensions = isMaterialWithExtensions;
exports.isGLTFMesh = isGLTFMesh;
exports.isMeshWithExtensions = isMeshWithExtensions;
exports.isGLTFAnimation = isGLTFAnimation;
exports.isAnimationWithExtensions = isAnimationWithExtensions;
exports.isGLTFAccessor = isGLTFAccessor;
exports.isGLTFBufferView = isGLTFBufferView;
exports.isGLTFNode = isGLTFNode;
exports.hasExtensions = hasExtensions;
exports.isExtensionName = isExtensionName;
exports.isExtensionOfType = isExtensionOfType;
exports.isImageMimeType = isImageMimeType;
exports.isGLTFMimeType = isGLTFMimeType;
exports.isValidComponentType = isValidComponentType;
exports.isValidAccessorType = isValidAccessorType;
exports.isArrayOfType = isArrayOfType;
exports.isNonEmptyArrayOfType = isNonEmptyArrayOfType;
exports.assertIsObject = assertIsObject;
exports.assertIsNumber = assertIsNumber;
exports.assertIsString = assertIsString;
const types_1 = require("./types");
// Basic type guards
function isObject(value) {
    return typeof value === 'object' && value !== null && !Array.isArray(value);
}
function isNonEmptyArray(value) {
    return Array.isArray(value) && value.length > 0;
}
function isString(value) {
    return typeof value === 'string';
}
function isNumber(value) {
    return typeof value === 'number' && Number.isFinite(value);
}
function isPositiveNumber(value) {
    return isNumber(value) && value > 0;
}
function isNonNegativeNumber(value) {
    return isNumber(value) && value >= 0;
}
function isInteger(value) {
    return isNumber(value) && Number.isInteger(value);
}
function isPositiveInteger(value) {
    return isInteger(value) && value > 0;
}
function isNonNegativeInteger(value) {
    return isInteger(value) && value >= 0;
}
// Branded type constructors
function createJSONPointer(pointer) {
    if (!pointer.startsWith('/')) {
        throw new Error(`Invalid JSON pointer: ${pointer}. Must start with '/'`);
    }
    return pointer;
}
function createMimeType(mimeType) {
    if (!mimeType.includes('/')) {
        throw new Error(`Invalid MIME type: ${mimeType}. Must contain '/'`);
    }
    return mimeType;
}
function createBase64String(base64) {
    try {
        atob(base64);
        return base64;
    }
    catch {
        throw new Error(`Invalid base64 string: ${base64}`);
    }
}
// GLTF component type guards
function isGLTFMaterial(value) {
    if (!isObject(value))
        return false;
    // Check for material-specific properties
    const material = value;
    return (
    // At least one material property should exist
    material['alphaMode'] !== undefined ||
        material['alphaCutoff'] !== undefined ||
        material['doubleSided'] !== undefined ||
        material['pbrMetallicRoughness'] !== undefined ||
        material['normalTexture'] !== undefined ||
        material['occlusionTexture'] !== undefined ||
        material['emissiveTexture'] !== undefined ||
        material['emissiveFactor'] !== undefined ||
        material['name'] !== undefined ||
        material['extensions'] !== undefined ||
        material['extras'] !== undefined);
}
function isMaterialWithExtensions(value) {
    return isGLTFMaterial(value);
}
function isGLTFMesh(value) {
    if (!isObject(value))
        return false;
    const mesh = value;
    return Array.isArray(mesh['primitives']);
}
function isMeshWithExtensions(value) {
    return isGLTFMesh(value);
}
function isGLTFAnimation(value) {
    if (!isObject(value))
        return false;
    const animation = value;
    return Array.isArray(animation['channels']) && Array.isArray(animation['samplers']);
}
function isAnimationWithExtensions(value) {
    return isGLTFAnimation(value);
}
function isGLTFAccessor(value) {
    if (!isObject(value))
        return false;
    const accessor = value;
    return (isNumber(accessor['componentType']) &&
        isNumber(accessor['count']) &&
        isString(accessor['type']));
}
function isGLTFBufferView(value) {
    if (!isObject(value))
        return false;
    const bufferView = value;
    return (isNonNegativeInteger(bufferView['buffer']) &&
        isPositiveInteger(bufferView['byteLength']));
}
function isGLTFNode(value) {
    if (!isObject(value))
        return false;
    // Node can be empty or have various properties
    const node = value;
    // Check that if properties exist, they have correct types
    if (node['camera'] !== undefined && !isNonNegativeInteger(node['camera']))
        return false;
    if (node['children'] !== undefined && !Array.isArray(node['children']))
        return false;
    if (node['skin'] !== undefined && !isNonNegativeInteger(node['skin']))
        return false;
    if (node['matrix'] !== undefined && (!Array.isArray(node['matrix']) || node['matrix'].length !== 16))
        return false;
    if (node['mesh'] !== undefined && !isNonNegativeInteger(node['mesh']))
        return false;
    if (node['rotation'] !== undefined && (!Array.isArray(node['rotation']) || node['rotation'].length !== 4))
        return false;
    if (node['scale'] !== undefined && (!Array.isArray(node['scale']) || node['scale'].length !== 3))
        return false;
    if (node['translation'] !== undefined && (!Array.isArray(node['translation']) || node['translation'].length !== 3))
        return false;
    if (node['weights'] !== undefined && !Array.isArray(node['weights']))
        return false;
    return true;
}
// Validation context type guards
function hasExtensions(value) {
    return isObject(value) && isObject(value['extensions']);
}
// Extension type guards
function isExtensionName(value) {
    const validExtensions = [
        'KHR_materials_anisotropy',
        'KHR_materials_clearcoat',
        'KHR_materials_dispersion',
        'KHR_materials_emissive_strength',
        'KHR_materials_ior',
        'KHR_materials_iridescence',
        'KHR_materials_pbrSpecularGlossiness',
        'KHR_materials_sheen',
        'KHR_materials_specular',
        'KHR_materials_transmission',
        'KHR_materials_volume',
        'KHR_texture_transform',
        'KHR_lights_punctual',
        'KHR_materials_variants'
    ];
    return validExtensions.includes(value);
}
function isExtensionOfType(extension, extensionName) {
    if (!isObject(extension))
        return false;
    // Type-specific validation based on extension name
    switch (extensionName) {
        case 'KHR_materials_anisotropy':
            return isKHRMaterialsAnisotropyExtension(extension);
        case 'KHR_materials_clearcoat':
            return isKHRMaterialsClearcoatExtension(extension);
        // Add more cases as needed
        default:
            return true; // Fallback for unknown extensions
    }
}
// Specific extension type guards
function isKHRMaterialsAnisotropyExtension(value) {
    if (!isObject(value))
        return false;
    const ext = value;
    return ((ext['anisotropyStrength'] === undefined || isNonNegativeNumber(ext['anisotropyStrength'])) &&
        (ext['anisotropyRotation'] === undefined || isNumber(ext['anisotropyRotation'])) &&
        (ext['anisotropyTexture'] === undefined || isObject(ext['anisotropyTexture'])));
}
function isKHRMaterialsClearcoatExtension(value) {
    if (!isObject(value))
        return false;
    const ext = value;
    return ((ext['clearcoatFactor'] === undefined || isNonNegativeNumber(ext['clearcoatFactor'])) &&
        (ext['clearcoatRoughnessFactor'] === undefined || isNonNegativeNumber(ext['clearcoatRoughnessFactor'])) &&
        (ext['clearcoatTexture'] === undefined || isObject(ext['clearcoatTexture'])) &&
        (ext['clearcoatRoughnessTexture'] === undefined || isObject(ext['clearcoatRoughnessTexture'])) &&
        (ext['clearcoatNormalTexture'] === undefined || isObject(ext['clearcoatNormalTexture'])));
}
// MIME type guards
function isImageMimeType(value) {
    return types_1.IMAGE_MIME_TYPES.includes(value);
}
function isGLTFMimeType(value) {
    return types_1.GLTF_MIME_TYPES.includes(value);
}
// Component type guards
function isValidComponentType(value) {
    const validTypes = [
        types_1.ComponentType.BYTE,
        types_1.ComponentType.UNSIGNED_BYTE,
        types_1.ComponentType.SHORT,
        types_1.ComponentType.UNSIGNED_SHORT,
        types_1.ComponentType.UNSIGNED_INT,
        types_1.ComponentType.FLOAT,
        types_1.ComponentType.DOUBLE
    ];
    return validTypes.includes(value);
}
function isValidAccessorType(value) {
    const validTypes = [
        types_1.AccessorType.SCALAR,
        types_1.AccessorType.VEC2,
        types_1.AccessorType.VEC3,
        types_1.AccessorType.VEC4,
        types_1.AccessorType.MAT2,
        types_1.AccessorType.MAT3,
        types_1.AccessorType.MAT4
    ];
    return validTypes.includes(value);
}
// Array validation helpers
function isArrayOfType(value, itemGuard) {
    return Array.isArray(value) && value.every(itemGuard);
}
function isNonEmptyArrayOfType(value, itemGuard) {
    return isArrayOfType(value, itemGuard) && value.length > 0;
}
// Utility assertion functions
function assertIsObject(value, context) {
    if (!isObject(value)) {
        throw new Error(`Expected object in ${context}, got ${typeof value}`);
    }
}
function assertIsNumber(value, context) {
    if (!isNumber(value)) {
        throw new Error(`Expected number in ${context}, got ${typeof value}`);
    }
}
function assertIsString(value, context) {
    if (!isString(value)) {
        throw new Error(`Expected string in ${context}, got ${typeof value}`);
    }
}
