// @dcl/schemas/dist/platform/events/base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Events = void 0;
var Events;
(function (Events) {
    let Type;
    (function (Type) {
        Type["BLOCKCHAIN"] = "blockchain";
        Type["CATALYST_DEPLOYMENT"] = "catalyst-deployment";
        Type["WORLD"] = "world";
        Type["CLIENT"] = "client";
        Type["WEB"] = "web";
        Type["MARKETPLACE"] = "marketplace";
        Type["REWARDS"] = "rewards";
        Type["BADGE"] = "badge";
        Type["ASSET_BUNDLE"] = "asset-bundle";
        Type["SOCIAL_SERVICE"] = "social-service";
        Type["CREDITS_SERVICE"] = "credits-service";
        Type["STREAMING"] = "streaming";
        Type["COMMS"] = "comms";
        Type["REFERRAL"] = "referral";
        Type["COMMUNITY"] = "community";
        Type["CAMERA"] = "camera";
        Type["EVENT"] = "event";
        Type["GOVERNANCE"] = "governance";
        Type["MODERATION"] = "moderation";
    })(Type = Events.Type || (Events.Type = {}));
    let SubType;
    (function (SubType) {
        let Blockchain;
        (function (Blockchain) {
            Blockchain["BID_ACCEPTED"] = "bid-accepted";
            Blockchain["ITEM_SOLD"] = "item-sold";
            Blockchain["RENTAL_ENDED"] = "land-rental-ended";
            Blockchain["RENTAL_STARTED"] = "land-rental-started";
            Blockchain["ROYALTIES_EARNED"] = "royalties-earned";
            Blockchain["COLLECTION_CREATED"] = "collection-created";
            Blockchain["ITEM_PUBLISHED"] = "item-published";
            Blockchain["TRANSFER_RECEIVED"] = "transfer-received";
            Blockchain["TIP_RECEIVED"] = "tip-received";
        })(Blockchain = SubType.Blockchain || (SubType.Blockchain = {}));
        let Marketplace;
        (function (Marketplace) {
            Marketplace["BID_RECEIVED"] = "bid-received";
        })(Marketplace = SubType.Marketplace || (SubType.Marketplace = {}));
        let CatalystDeployment;
        (function (CatalystDeployment) {
            CatalystDeployment["SCENE"] = "scene";
            CatalystDeployment["PROFILE"] = "profile";
            CatalystDeployment["WEARABLE"] = "wearable";
            CatalystDeployment["STORE"] = "store";
            CatalystDeployment["EMOTE"] = "emote";
            CatalystDeployment["OUTFITS"] = "outfits";
        })(CatalystDeployment = SubType.CatalystDeployment || (SubType.CatalystDeployment = {}));
        let Worlds;
        (function (Worlds) {
            Worlds["DEPLOYMENT"] = "deployment";
            Worlds["WORLD_SETTINGS_CHANGED"] = "world_settings_changed";
            Worlds["WORLD_SCENES_UNDEPLOYMENT"] = "world_scenes_undeployment";
            Worlds["WORLD_UNDEPLOYMENT"] = "world_undeployment";
            Worlds["WORLD_SPAWN_COORDINATE_SET"] = "world_spawn_coordinate_set";
            Worlds["WORLDS_PERMISSION_GRANTED"] = "worlds_permission_granted";
            Worlds["WORLDS_PERMISSION_REVOKED"] = "worlds_permission_revoked";
            Worlds["WORLDS_ACCESS_RESTORED"] = "worlds_access_restored";
            Worlds["WORLDS_ACCESS_RESTRICTED"] = "worlds_access_restricted";
            Worlds["WORLDS_MISSING_RESOURCES"] = "worlds_missing_resources";
        })(Worlds = SubType.Worlds || (SubType.Worlds = {}));
        let Client;
        (function (Client) {
            Client["LOGGED_IN"] = "logged-in";
            Client["LOGGED_IN_CACHED"] = "logged-in-cached";
            Client["MOVE_TO_PARCEL"] = "move-to-parcel";
            Client["USED_EMOTE"] = "used-emote";
            Client["PASSPORT_OPENED"] = "passport-opened";
            Client["WALKED_DISTANCE"] = "walked-distance";
            Client["VERTICAL_HEIGHT_REACHED"] = "vertical-height-reached";
        })(Client = SubType.Client || (SubType.Client = {}));
        let Web;
        (function (Web) {
            Web["AUTH_IDENTIFY"] = "auth-identify";
        })(Web = SubType.Web || (SubType.Web = {}));
        let Rewards;
        (function (Rewards) {
            Rewards["REWARD_IN_PROGRESS"] = "reward-in-progress";
            Rewards["REWARD_ASSIGNED"] = "reward-assigned";
            Rewards["CAMPAIGN_OUT_OF_FUNDS"] = "campaign-out-of-funds";
            Rewards["CAMPAIGN_GAS_PRICE_HIGHER_THAN_EXPECTED"] = "campaign-gas-price-higher-than-expected";
            Rewards["CAMPAIGN_OUT_OF_STOCK"] = "campaign-out-of-stock";
            Rewards["REWARD_DELAYED"] = "reward-delayed";
        })(Rewards = SubType.Rewards || (SubType.Rewards = {}));
        let Badge;
        (function (Badge) {
            Badge["GRANTED"] = "badge-granted";
        })(Badge = SubType.Badge || (SubType.Badge = {}));
        let AssetBundle;
        (function (AssetBundle) {
            AssetBundle["CONVERTED"] = "converted";
            AssetBundle["MANUALLY_QUEUED"] = "manually-queued";
        })(AssetBundle = SubType.AssetBundle || (SubType.AssetBundle = {}));
        let SocialService;
        (function (SocialService) {
            SocialService["FRIENDSHIP_REQUEST"] = "friendship-request";
            SocialService["FRIENDSHIP_ACCEPTED"] = "friendship-accepted";
        })(SocialService = SubType.SocialService || (SubType.SocialService = {}));
        let CreditsService;
        (function (CreditsService) {
            CreditsService["CREDITS_GOAL_COMPLETED"] = "credits-goal-completed";
            CreditsService["COMPLETE_GOALS_REMINDER"] = "complete-goals-reminder";
            CreditsService["CLAIM_CREDITS_REMINDER"] = "claim-credits-reminder";
            CreditsService["DO_NOT_MISS_OUT_REMINDER"] = "do-not-miss-out-reminder";
            CreditsService["USAGE_REMINDER"] = "usage-reminder";
            CreditsService["USAGE_24_HOURS_REMINDER"] = "usage-24-hours-reminder";
            CreditsService["ON_DEMAND_CREDITS_GRANTED"] = "on-demand-credits-granted";
            CreditsService["NEW_SEASON_REMINDER"] = "new-season-reminder";
        })(CreditsService = SubType.CreditsService || (SubType.CreditsService = {}));
        let Streaming;
        (function (Streaming) {
            Streaming["STREAMING_KEY_RESET"] = "streaming-key-reset";
            Streaming["STREAMING_KEY_REVOKE"] = "streaming-key-revoke";
            Streaming["STREAMING_KEY_EXPIRED"] = "streaming-key-expired";
            Streaming["STREAMING_TIME_EXCEEDED"] = "streaming-time-exceeded";
            Streaming["STREAMING_PLACE_UPDATED"] = "streaming-place-updated";
            Streaming["COMMUNITY_STREAMING_ENDED"] = "community-streaming-ended";
        })(Streaming = SubType.Streaming || (SubType.Streaming = {}));
        let Comms;
        (function (Comms) {
            Comms["USER_JOINED_ROOM"] = "user-joined-room";
            Comms["USER_LEFT_ROOM"] = "user-left-room";
            Comms["USER_BANNED_FROM_SCENE"] = "user-banned-from-scene";
            Comms["USER_UNBANNED_FROM_SCENE"] = "user-unbanned-from-scene";
        })(Comms = SubType.Comms || (SubType.Comms = {}));
        let Referral;
        (function (Referral) {
            Referral["REFERRAL_INVITED_USERS_ACCEPTED"] = "referral-invited-users-accepted";
            Referral["REFERRAL_NEW_TIER_REACHED"] = "referral-new-tier-reached";
        })(Referral = SubType.Referral || (SubType.Referral = {}));
        let Community;
        (function (Community) {
            Community["DELETED"] = "community-deleted";
            Community["DELETED_CONTENT_VIOLATION"] = "community-deleted-content-violation";
            Community["RENAMED"] = "community-renamed";
            Community["MEMBER_BANNED"] = "community-member-banned";
            Community["MEMBER_LEFT"] = "community-member-left";
            Community["MEMBER_REMOVED"] = "community-member-removed";
            Community["REQUEST_TO_JOIN_RECEIVED"] = "community-request-to-join-received";
            Community["REQUEST_TO_JOIN_ACCEPTED"] = "community-request-to-join-accepted";
            Community["INVITE_RECEIVED"] = "community-invite-received";
            Community["OWNERSHIP_TRANSFERRED"] = "community-ownership-transferred";
            Community["POST_ADDED"] = "community-post-added";
            Community["VOICE_CHAT_STARTED"] = "community-voice-chat-started";
        })(Community = SubType.Community || (SubType.Community = {}));
        let Camera;
        (function (Camera) {
            Camera["PHOTO_TAKEN"] = "photo-taken";
            Camera["PHOTO_PRIVACY_CHANGED"] = "photo-privacy-changed";
        })(Camera = SubType.Camera || (SubType.Camera = {}));
        let Event;
        (function (Event) {
            Event["EVENT_CREATED"] = "event-created";
            Event["EVENT_STARTS_SOON"] = "event-starts-soon";
            Event["EVENT_STARTED"] = "event-started";
            Event["EVENT_ENDED"] = "event-ended";
        })(Event = SubType.Event || (SubType.Event = {}));
        let Governance;
        (function (Governance) {
            Governance["PROPOSAL_ENACTED"] = "governance_proposal_enacted";
            Governance["COAUTHOR_REQUESTED"] = "governance_coauthor_requested";
            Governance["PITCH_PASSED"] = "governance_pitch_passed";
            Governance["TENDER_PASSED"] = "governance_tender_passed";
            Governance["AUTHORED_PROPOSAL_FINISHED"] = "governance_authored_proposal_finished";
            Governance["VOTING_ENDED_VOTER"] = "governance_voting_ended_voter";
            Governance["NEW_COMMENT_ON_PROPOSAL"] = "governance_new_comment_on_proposal";
            Governance["NEW_COMMENT_ON_PROJECT_UPDATED"] = "governance_new_comment_on_project_update";
            Governance["WHALE_VOTE"] = "governance_whale_vote";
            Governance["VOTED_ON_BEHALF"] = "governance_voted_on_behalf";
            Governance["CLIFF_ENDED"] = "governance_cliff_ended";
        })(Governance = SubType.Governance || (SubType.Governance = {}));
        let Moderation;
        (function (Moderation) {
            Moderation["USER_BAN_CREATED"] = "user-ban-created";
            Moderation["USER_BAN_LIFTED"] = "user-ban-lifted";
            Moderation["USER_WARNING_CREATED"] = "user-warning-created";
        })(Moderation = SubType.Moderation || (SubType.Moderation = {}));
    })(SubType = Events.SubType || (Events.SubType = {}));
})(Events = exports.Events || (exports.Events = {}));
