// @dcl/schemas/dist/platform/events/communities.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CommunityVoiceChatStartedEvent = exports.CommunityOwnershipTransferredEvent = exports.CommunityPostAddedEvent = exports.CommunityInviteReceivedEvent = exports.CommunityRequestToJoinAcceptedEvent = exports.CommunityRequestToJoinReceivedEvent = exports.CommunityMemberRemovedEvent = exports.CommunityMemberLeftEvent = exports.CommunityMemberBannedEvent = exports.CommunityRenamedEvent = exports.CommunityDeletedContentViolationEvent = exports.CommunityDeletedEvent = void 0;
const validation_1 = require("../../validation");
const base_1 = require("./base");
var CommunityDeletedEvent;
(function (CommunityDeletedEvent) {
    CommunityDeletedEvent.schema = {
        type: 'object',
        properties: {
            type: { type: 'string', const: base_1.Events.Type.COMMUNITY },
            subType: { type: 'string', const: base_1.Events.SubType.Community.DELETED },
            key: { type: 'string' },
            timestamp: { type: 'number', minimum: 1 },
            metadata: {
                type: 'object',
                properties: {
                    id: { type: 'string' },
                    name: { type: 'string' },
                    memberAddresses: { type: 'array', items: { type: 'string' } },
                    thumbnailUrl: { type: 'string' }
                },
                required: ['id', 'name', 'memberAddresses', 'thumbnailUrl'],
                additionalProperties: false
            }
        },
        required: ['type', 'subType', 'key', 'timestamp', 'metadata'],
        additionalProperties: false
    };
    CommunityDeletedEvent.validate = (0, validation_1.generateLazyValidator)(CommunityDeletedEvent.schema);
})(CommunityDeletedEvent = exports.CommunityDeletedEvent || (exports.CommunityDeletedEvent = {}));
var CommunityDeletedContentViolationEvent;
(function (CommunityDeletedContentViolationEvent) {
    CommunityDeletedContentViolationEvent.schema = {
        type: 'object',
        properties: {
            type: { type: 'string', const: base_1.Events.Type.COMMUNITY },
            subType: { type: 'string', const: base_1.Events.SubType.Community.DELETED_CONTENT_VIOLATION },
            key: { type: 'string' },
            timestamp: { type: 'number', minimum: 1 },
            metadata: {
                type: 'object',
                properties: {
                    id: { type: 'string' },
                    name: { type: 'string' },
                    ownerAddress: { type: 'string' },
                    thumbnailUrl: { type: 'string' }
                },
                required: ['id', 'name', 'ownerAddress', 'thumbnailUrl'],
                additionalProperties: false
            }
        },
        required: ['type', 'subType', 'key', 'timestamp', 'metadata'],
        additionalProperties: false
    };
    CommunityDeletedContentViolationEvent.validate = (0, validation_1.generateLazyValidator)(CommunityDeletedContentViolationEvent.schema);
})(CommunityDeletedContentViolationEvent = exports.CommunityDeletedContentViolationEvent || (exports.CommunityDeletedContentViolationEvent = {}));
var CommunityRenamedEvent;
(function (CommunityRenamedEvent) {
    CommunityRenamedEvent.schema = {
        type: 'object',
        properties: {
            type: { type: 'string', const: base_1.Events.Type.COMMUNITY },
            subType: { type: 'string', const: base_1.Events.SubType.Community.RENAMED },
            key: { type: 'string' },
            timestamp: { type: 'number', minimum: 1 },
            metadata: {
                type: 'object',
                properties: {
                    id: { type: 'string' },
                    oldName: { type: 'string' },
                    newName: { type: 'string' },
                    memberAddresses: { type: 'array', items: { type: 'string' } },
                    thumbnailUrl: { type: 'string' }
                },
                required: ['id', 'oldName', 'newName', 'memberAddresses', 'thumbnailUrl'],
                additionalProperties: false
            }
        },
        required: ['type', 'subType', 'key', 'timestamp', 'metadata'],
        additionalProperties: false
    };
    CommunityRenamedEvent.validate = (0, validation_1.generateLazyValidator)(CommunityRenamedEvent.schema);
})(CommunityRenamedEvent = exports.CommunityRenamedEvent || (exports.CommunityRenamedEvent = {}));
var CommunityMemberBannedEvent;
(function (CommunityMemberBannedEvent) {
    CommunityMemberBannedEvent.schema = {
        type: 'object',
        properties: {
            type: { type: 'string', const: base_1.Events.Type.COMMUNITY },
            subType: { type: 'string', const: base_1.Events.SubType.Community.MEMBER_BANNED },
            key: { type: 'string' },
            timestamp: { type: 'number', minimum: 1 },
            metadata: {
                type: 'object',
                properties: {
                    id: { type: 'string' },
                    name: { type: 'string' },
                    memberAddress: { type: 'string' },
                    thumbnailUrl: { type: 'string' }
                },
                required: ['id', 'name', 'memberAddress', 'thumbnailUrl'],
                additionalProperties: false
            }
        },
        required: ['type', 'subType', 'key', 'timestamp', 'metadata'],
        additionalProperties: false
    };
    CommunityMemberBannedEvent.validate = (0, validation_1.generateLazyValidator)(CommunityMemberBannedEvent.schema);
})(CommunityMemberBannedEvent = exports.CommunityMemberBannedEvent || (exports.CommunityMemberBannedEvent = {}));
var CommunityMemberLeftEvent;
(function (CommunityMemberLeftEvent) {
    CommunityMemberLeftEvent.schema = {
        type: 'object',
        properties: {
            type: { type: 'string', const: base_1.Events.Type.COMMUNITY },
            subType: { type: 'string', const: base_1.Events.SubType.Community.MEMBER_LEFT },
            key: { type: 'string' },
            timestamp: { type: 'number', minimum: 1 },
            metadata: {
                type: 'object',
                properties: {
                    id: { type: 'string' },
                    memberAddress: { type: 'string' }
                },
                required: ['id', 'memberAddress'],
                additionalProperties: false
            }
        },
        required: ['type', 'subType', 'key', 'timestamp', 'metadata'],
        additionalProperties: false
    };
    CommunityMemberLeftEvent.validate = (0, validation_1.generateLazyValidator)(CommunityMemberLeftEvent.schema);
})(CommunityMemberLeftEvent = exports.CommunityMemberLeftEvent || (exports.CommunityMemberLeftEvent = {}));
var CommunityMemberRemovedEvent;
(function (CommunityMemberRemovedEvent) {
    CommunityMemberRemovedEvent.schema = {
        type: 'object',
        properties: {
            type: { type: 'string', const: base_1.Events.Type.COMMUNITY },
            subType: { type: 'string', const: base_1.Events.SubType.Community.MEMBER_REMOVED },
            key: { type: 'string' },
            timestamp: { type: 'number', minimum: 1 },
            metadata: {
                type: 'object',
                properties: {
                    id: { type: 'string' },
                    name: { type: 'string' },
                    memberAddress: { type: 'string' },
                    thumbnailUrl: { type: 'string' }
                },
                required: ['id', 'name', 'memberAddress', 'thumbnailUrl'],
                additionalProperties: false
            }
        },
        required: ['type', 'subType', 'key', 'timestamp', 'metadata'],
        additionalProperties: false
    };
    CommunityMemberRemovedEvent.validate = (0, validation_1.generateLazyValidator)(CommunityMemberRemovedEvent.schema);
})(CommunityMemberRemovedEvent = exports.CommunityMemberRemovedEvent || (exports.CommunityMemberRemovedEvent = {}));
var CommunityRequestToJoinReceivedEvent;
(function (CommunityRequestToJoinReceivedEvent) {
    CommunityRequestToJoinReceivedEvent.schema = {
        type: 'object',
        properties: {
            type: { type: 'string', const: base_1.Events.Type.COMMUNITY },
            subType: { type: 'string', const: base_1.Events.SubType.Community.REQUEST_TO_JOIN_RECEIVED },
            key: { type: 'string' },
            timestamp: { type: 'number', minimum: 1 },
            metadata: {
                type: 'object',
                properties: {
                    communityId: { type: 'string' },
                    communityName: { type: 'string' },
                    memberAddress: { type: 'string' },
                    memberName: { type: 'string' },
                    thumbnailUrl: { type: 'string' },
                    addressesToNotify: { type: 'array', items: { type: 'string' } }
                },
                required: ['communityId', 'communityName', 'memberAddress', 'memberName', 'thumbnailUrl', 'addressesToNotify'],
                additionalProperties: false
            }
        },
        required: ['type', 'subType', 'key', 'timestamp', 'metadata'],
        additionalProperties: false
    };
    CommunityRequestToJoinReceivedEvent.validate = (0, validation_1.generateLazyValidator)(CommunityRequestToJoinReceivedEvent.schema);
})(CommunityRequestToJoinReceivedEvent = exports.CommunityRequestToJoinReceivedEvent || (exports.CommunityRequestToJoinReceivedEvent = {}));
var CommunityRequestToJoinAcceptedEvent;
(function (CommunityRequestToJoinAcceptedEvent) {
    CommunityRequestToJoinAcceptedEvent.schema = {
        type: 'object',
        properties: {
            type: { type: 'string', const: base_1.Events.Type.COMMUNITY },
            subType: { type: 'string', const: base_1.Events.SubType.Community.REQUEST_TO_JOIN_ACCEPTED },
            key: { type: 'string' },
            timestamp: { type: 'number', minimum: 1 },
            metadata: {
                type: 'object',
                properties: {
                    communityId: { type: 'string' },
                    communityName: { type: 'string' },
                    memberAddress: { type: 'string' },
                    thumbnailUrl: { type: 'string' }
                },
                required: ['communityId', 'communityName', 'memberAddress', 'thumbnailUrl'],
                additionalProperties: false
            }
        },
        required: ['type', 'subType', 'key', 'timestamp', 'metadata'],
        additionalProperties: false
    };
    CommunityRequestToJoinAcceptedEvent.validate = (0, validation_1.generateLazyValidator)(CommunityRequestToJoinAcceptedEvent.schema);
})(CommunityRequestToJoinAcceptedEvent = exports.CommunityRequestToJoinAcceptedEvent || (exports.CommunityRequestToJoinAcceptedEvent = {}));
var CommunityInviteReceivedEvent;
(function (CommunityInviteReceivedEvent) {
    CommunityInviteReceivedEvent.schema = {
        type: 'object',
        properties: {
            type: { type: 'string', const: base_1.Events.Type.COMMUNITY },
            subType: { type: 'string', const: base_1.Events.SubType.Community.INVITE_RECEIVED },
            key: { type: 'string' },
            timestamp: { type: 'number', minimum: 1 },
            metadata: {
                type: 'object',
                properties: {
                    communityId: { type: 'string' },
                    communityName: { type: 'string' },
                    memberAddress: { type: 'string' },
                    thumbnailUrl: { type: 'string' }
                },
                required: ['communityId', 'communityName', 'memberAddress', 'thumbnailUrl'],
                additionalProperties: false
            }
        },
        required: ['type', 'subType', 'key', 'timestamp', 'metadata'],
        additionalProperties: false
    };
    CommunityInviteReceivedEvent.validate = (0, validation_1.generateLazyValidator)(CommunityInviteReceivedEvent.schema);
})(CommunityInviteReceivedEvent = exports.CommunityInviteReceivedEvent || (exports.CommunityInviteReceivedEvent = {}));
var CommunityPostAddedEvent;
(function (CommunityPostAddedEvent) {
    CommunityPostAddedEvent.schema = {
        type: 'object',
        properties: {
            type: { type: 'string', const: base_1.Events.Type.COMMUNITY },
            subType: { type: 'string', const: base_1.Events.SubType.Community.POST_ADDED },
            key: { type: 'string' },
            timestamp: { type: 'number', minimum: 1 },
            metadata: {
                type: 'object',
                properties: {
                    communityId: { type: 'string' },
                    communityName: { type: 'string' },
                    thumbnailUrl: { type: 'string' },
                    postId: { type: 'string' },
                    authorAddress: { type: 'string' },
                    addressesToNotify: { type: 'array', items: { type: 'string' } }
                },
                required: ['addressesToNotify', 'communityId', 'communityName', 'thumbnailUrl', 'postId', 'authorAddress'],
                additionalProperties: false
            }
        },
        required: ['type', 'subType', 'key', 'timestamp', 'metadata'],
        additionalProperties: false
    };
    CommunityPostAddedEvent.validate = (0, validation_1.generateLazyValidator)(CommunityPostAddedEvent.schema);
})(CommunityPostAddedEvent = exports.CommunityPostAddedEvent || (exports.CommunityPostAddedEvent = {}));
var CommunityOwnershipTransferredEvent;
(function (CommunityOwnershipTransferredEvent) {
    CommunityOwnershipTransferredEvent.schema = {
        type: 'object',
        properties: {
            type: { type: 'string', const: base_1.Events.Type.COMMUNITY },
            subType: { type: 'string', const: base_1.Events.SubType.Community.OWNERSHIP_TRANSFERRED },
            key: { type: 'string' },
            timestamp: { type: 'number', minimum: 1 },
            metadata: {
                type: 'object',
                properties: {
                    communityId: { type: 'string' },
                    communityName: { type: 'string' },
                    oldOwnerAddress: { type: 'string' },
                    newOwnerAddress: { type: 'string' },
                    thumbnailUrl: { type: 'string' }
                },
                required: ['communityId', 'communityName', 'oldOwnerAddress', 'newOwnerAddress', 'thumbnailUrl'],
                additionalProperties: false
            }
        },
        required: ['type', 'subType', 'key', 'timestamp', 'metadata'],
        additionalProperties: false
    };
    CommunityOwnershipTransferredEvent.validate = (0, validation_1.generateLazyValidator)(CommunityOwnershipTransferredEvent.schema);
})(CommunityOwnershipTransferredEvent = exports.CommunityOwnershipTransferredEvent || (exports.CommunityOwnershipTransferredEvent = {}));
var CommunityVoiceChatStartedEvent;
(function (CommunityVoiceChatStartedEvent) {
    CommunityVoiceChatStartedEvent.schema = {
        type: 'object',
        properties: {
            type: { type: 'string', const: base_1.Events.Type.COMMUNITY },
            subType: { type: 'string', const: base_1.Events.SubType.Community.VOICE_CHAT_STARTED },
            key: { type: 'string' },
            timestamp: { type: 'number', minimum: 1 },
            metadata: {
                type: 'object',
                properties: {
                    communityId: { type: 'string' },
                    communityName: { type: 'string' },
                    thumbnailUrl: { type: 'string' },
                    addressesToNotify: { type: 'array', items: { type: 'string' } }
                },
                required: ['communityId', 'communityName', 'thumbnailUrl', 'addressesToNotify'],
                additionalProperties: false
            }
        },
        required: ['type', 'subType', 'key', 'timestamp', 'metadata'],
        additionalProperties: false
    };
    CommunityVoiceChatStartedEvent.validate = (0, validation_1.generateLazyValidator)(CommunityVoiceChatStartedEvent.schema);
})(CommunityVoiceChatStartedEvent = exports.CommunityVoiceChatStartedEvent || (exports.CommunityVoiceChatStartedEvent = {}));
