// @dcl/schemas/dist/platform/events/event.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventEndedEvent = exports.EventCreatedEvent = exports.EventStartsSoonEvent = exports.EventStartedEvent = void 0;
const validation_1 = require("../../validation");
const base_1 = require("./base");
var EventStartedEvent;
(function (EventStartedEvent) {
    EventStartedEvent.schema = {
        type: 'object',
        properties: {
            type: { type: 'string', const: base_1.Events.Type.EVENT },
            subType: { type: 'string', const: base_1.Events.SubType.Event.EVENT_STARTED },
            key: { type: 'string' },
            timestamp: { type: 'number', minimum: 0 },
            metadata: {
                type: 'object',
                properties: {
                    name: { type: 'string' },
                    image: { type: 'string' },
                    link: { type: 'string' },
                    communityId: { type: 'string' },
                    communityThumbnail: { type: 'string', nullable: true },
                    title: { type: 'string' },
                    description: { type: 'string' },
                    attendee: { type: 'string' }
                },
                required: ['name', 'image', 'link'],
                additionalProperties: false
            }
        },
        required: ['type', 'subType', 'key', 'timestamp', 'metadata'],
        additionalProperties: false
    };
    EventStartedEvent.validate = (0, validation_1.generateLazyValidator)(EventStartedEvent.schema);
})(EventStartedEvent = exports.EventStartedEvent || (exports.EventStartedEvent = {}));
var EventStartsSoonEvent;
(function (EventStartsSoonEvent) {
    EventStartsSoonEvent.schema = {
        type: 'object',
        properties: {
            type: { type: 'string', const: base_1.Events.Type.EVENT },
            subType: { type: 'string', const: base_1.Events.SubType.Event.EVENT_STARTS_SOON },
            key: { type: 'string' },
            timestamp: { type: 'number', minimum: 0 },
            metadata: {
                type: 'object',
                properties: {
                    name: { type: 'string' },
                    image: { type: 'string' },
                    link: { type: 'string' },
                    startsAt: { type: 'string' },
                    endsAt: { type: 'string' },
                    title: { type: 'string' },
                    description: { type: 'string' },
                    attendee: { type: 'string' }
                },
                required: ['name', 'image', 'link', 'startsAt', 'endsAt'],
                additionalProperties: false
            }
        },
        required: ['type', 'subType', 'key', 'timestamp', 'metadata'],
        additionalProperties: false
    };
    EventStartsSoonEvent.validate = (0, validation_1.generateLazyValidator)(EventStartsSoonEvent.schema);
})(EventStartsSoonEvent = exports.EventStartsSoonEvent || (exports.EventStartsSoonEvent = {}));
var EventCreatedEvent;
(function (EventCreatedEvent) {
    EventCreatedEvent.schema = {
        type: 'object',
        properties: {
            type: { type: 'string', const: base_1.Events.Type.EVENT },
            subType: { type: 'string', const: base_1.Events.SubType.Event.EVENT_CREATED },
            key: { type: 'string' },
            timestamp: { type: 'number', minimum: 0 },
            metadata: {
                type: 'object',
                properties: {
                    name: { type: 'string' },
                    image: { type: 'string' },
                    communityId: { type: 'string' },
                    communityName: { type: 'string' },
                    communityThumbnail: { type: 'string', nullable: true },
                    title: { type: 'string' },
                    description: { type: 'string' },
                    attendee: { type: 'string' }
                },
                required: ['name', 'image'],
                additionalProperties: false
            }
        },
        required: ['type', 'subType', 'key', 'timestamp', 'metadata'],
        additionalProperties: false
    };
    EventCreatedEvent.validate = (0, validation_1.generateLazyValidator)(EventCreatedEvent.schema);
})(EventCreatedEvent = exports.EventCreatedEvent || (exports.EventCreatedEvent = {}));
var EventEndedEvent;
(function (EventEndedEvent) {
    EventEndedEvent.schema = {
        type: 'object',
        properties: {
            type: { type: 'string', const: base_1.Events.Type.EVENT },
            subType: { type: 'string', const: base_1.Events.SubType.Event.EVENT_ENDED },
            key: { type: 'string' },
            timestamp: { type: 'number', minimum: 0 },
            metadata: {
                type: 'object',
                properties: {
                    communityId: { type: 'string', nullable: true },
                    totalAttendees: { type: 'number' }
                },
                required: ['totalAttendees'],
                additionalProperties: false
            }
        },
        required: ['type', 'subType', 'key', 'timestamp', 'metadata'],
        additionalProperties: false
    };
    EventEndedEvent.validate = (0, validation_1.generateLazyValidator)(EventEndedEvent.schema);
})(EventEndedEvent = exports.EventEndedEvent || (exports.EventEndedEvent = {}));
