// @dcl/schemas/dist/platform/events/governance.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GovernanceCliffEndedEvent = exports.GovernanceVotedOnBehalfEvent = exports.GovernanceWhaleVoteEvent = exports.GovernanceNewCommentOnProjectUpdatedEvent = exports.GovernanceNewCommentOnProposalEvent = exports.GovernanceVotingEndedVoterEvent = exports.GovernanceAuthoredProposalFinishedEvent = exports.GovernanceTenderPassedEvent = exports.GovernancePitchPassedEvent = exports.GovernanceCoauthorRequestedEvent = exports.GovernanceProposalEnactedEvent = void 0;
const validation_1 = require("../../validation");
const base_1 = require("./base");
const utils_1 = require("./utils");
const governanceMetadataSchema = {
    type: 'object',
    properties: {
        proposalId: { type: 'string' },
        proposalTitle: { type: 'string' },
        title: { type: 'string' },
        description: { type: 'string' },
        link: { type: 'string' },
        address: { type: 'string' }
    },
    required: ['proposalId', 'proposalTitle', 'title', 'description', 'link', 'address'],
    additionalProperties: false
};
var GovernanceProposalEnactedEvent;
(function (GovernanceProposalEnactedEvent) {
    GovernanceProposalEnactedEvent.schema = (0, utils_1.createEventSchema)(base_1.Events.Type.GOVERNANCE, base_1.Events.SubType.Governance.PROPOSAL_ENACTED, governanceMetadataSchema);
    GovernanceProposalEnactedEvent.validate = (0, validation_1.generateLazyValidator)(GovernanceProposalEnactedEvent.schema);
})(GovernanceProposalEnactedEvent = exports.GovernanceProposalEnactedEvent || (exports.GovernanceProposalEnactedEvent = {}));
var GovernanceCoauthorRequestedEvent;
(function (GovernanceCoauthorRequestedEvent) {
    GovernanceCoauthorRequestedEvent.schema = (0, utils_1.createEventSchema)(base_1.Events.Type.GOVERNANCE, base_1.Events.SubType.Governance.COAUTHOR_REQUESTED, governanceMetadataSchema);
    GovernanceCoauthorRequestedEvent.validate = (0, validation_1.generateLazyValidator)(GovernanceCoauthorRequestedEvent.schema);
})(GovernanceCoauthorRequestedEvent = exports.GovernanceCoauthorRequestedEvent || (exports.GovernanceCoauthorRequestedEvent = {}));
var GovernancePitchPassedEvent;
(function (GovernancePitchPassedEvent) {
    GovernancePitchPassedEvent.schema = (0, utils_1.createEventSchema)(base_1.Events.Type.GOVERNANCE, base_1.Events.SubType.Governance.PITCH_PASSED, governanceMetadataSchema);
    GovernancePitchPassedEvent.validate = (0, validation_1.generateLazyValidator)(GovernancePitchPassedEvent.schema);
})(GovernancePitchPassedEvent = exports.GovernancePitchPassedEvent || (exports.GovernancePitchPassedEvent = {}));
var GovernanceTenderPassedEvent;
(function (GovernanceTenderPassedEvent) {
    GovernanceTenderPassedEvent.schema = (0, utils_1.createEventSchema)(base_1.Events.Type.GOVERNANCE, base_1.Events.SubType.Governance.TENDER_PASSED, governanceMetadataSchema);
    GovernanceTenderPassedEvent.validate = (0, validation_1.generateLazyValidator)(GovernanceTenderPassedEvent.schema);
})(GovernanceTenderPassedEvent = exports.GovernanceTenderPassedEvent || (exports.GovernanceTenderPassedEvent = {}));
var GovernanceAuthoredProposalFinishedEvent;
(function (GovernanceAuthoredProposalFinishedEvent) {
    GovernanceAuthoredProposalFinishedEvent.schema = (0, utils_1.createEventSchema)(base_1.Events.Type.GOVERNANCE, base_1.Events.SubType.Governance.AUTHORED_PROPOSAL_FINISHED, governanceMetadataSchema);
    GovernanceAuthoredProposalFinishedEvent.validate = (0, validation_1.generateLazyValidator)(GovernanceAuthoredProposalFinishedEvent.schema);
})(GovernanceAuthoredProposalFinishedEvent = exports.GovernanceAuthoredProposalFinishedEvent || (exports.GovernanceAuthoredProposalFinishedEvent = {}));
var GovernanceVotingEndedVoterEvent;
(function (GovernanceVotingEndedVoterEvent) {
    GovernanceVotingEndedVoterEvent.schema = (0, utils_1.createEventSchema)(base_1.Events.Type.GOVERNANCE, base_1.Events.SubType.Governance.VOTING_ENDED_VOTER, governanceMetadataSchema);
    GovernanceVotingEndedVoterEvent.validate = (0, validation_1.generateLazyValidator)(GovernanceVotingEndedVoterEvent.schema);
})(GovernanceVotingEndedVoterEvent = exports.GovernanceVotingEndedVoterEvent || (exports.GovernanceVotingEndedVoterEvent = {}));
var GovernanceNewCommentOnProposalEvent;
(function (GovernanceNewCommentOnProposalEvent) {
    GovernanceNewCommentOnProposalEvent.schema = (0, utils_1.createEventSchema)(base_1.Events.Type.GOVERNANCE, base_1.Events.SubType.Governance.NEW_COMMENT_ON_PROPOSAL, governanceMetadataSchema);
    GovernanceNewCommentOnProposalEvent.validate = (0, validation_1.generateLazyValidator)(GovernanceNewCommentOnProposalEvent.schema);
})(GovernanceNewCommentOnProposalEvent = exports.GovernanceNewCommentOnProposalEvent || (exports.GovernanceNewCommentOnProposalEvent = {}));
var GovernanceNewCommentOnProjectUpdatedEvent;
(function (GovernanceNewCommentOnProjectUpdatedEvent) {
    GovernanceNewCommentOnProjectUpdatedEvent.schema = (0, utils_1.createEventSchema)(base_1.Events.Type.GOVERNANCE, base_1.Events.SubType.Governance.NEW_COMMENT_ON_PROJECT_UPDATED, governanceMetadataSchema);
    GovernanceNewCommentOnProjectUpdatedEvent.validate = (0, validation_1.generateLazyValidator)(GovernanceNewCommentOnProjectUpdatedEvent.schema);
})(GovernanceNewCommentOnProjectUpdatedEvent = exports.GovernanceNewCommentOnProjectUpdatedEvent || (exports.GovernanceNewCommentOnProjectUpdatedEvent = {}));
var GovernanceWhaleVoteEvent;
(function (GovernanceWhaleVoteEvent) {
    GovernanceWhaleVoteEvent.schema = (0, utils_1.createEventSchema)(base_1.Events.Type.GOVERNANCE, base_1.Events.SubType.Governance.WHALE_VOTE, governanceMetadataSchema);
    GovernanceWhaleVoteEvent.validate = (0, validation_1.generateLazyValidator)(GovernanceWhaleVoteEvent.schema);
})(GovernanceWhaleVoteEvent = exports.GovernanceWhaleVoteEvent || (exports.GovernanceWhaleVoteEvent = {}));
var GovernanceVotedOnBehalfEvent;
(function (GovernanceVotedOnBehalfEvent) {
    GovernanceVotedOnBehalfEvent.schema = (0, utils_1.createEventSchema)(base_1.Events.Type.GOVERNANCE, base_1.Events.SubType.Governance.VOTED_ON_BEHALF, governanceMetadataSchema);
    GovernanceVotedOnBehalfEvent.validate = (0, validation_1.generateLazyValidator)(GovernanceVotedOnBehalfEvent.schema);
})(GovernanceVotedOnBehalfEvent = exports.GovernanceVotedOnBehalfEvent || (exports.GovernanceVotedOnBehalfEvent = {}));
var GovernanceCliffEndedEvent;
(function (GovernanceCliffEndedEvent) {
    GovernanceCliffEndedEvent.schema = (0, utils_1.createEventSchema)(base_1.Events.Type.GOVERNANCE, base_1.Events.SubType.Governance.CLIFF_ENDED, governanceMetadataSchema);
    GovernanceCliffEndedEvent.validate = (0, validation_1.generateLazyValidator)(GovernanceCliffEndedEvent.schema);
})(GovernanceCliffEndedEvent = exports.GovernanceCliffEndedEvent || (exports.GovernanceCliffEndedEvent = {}));
