// @dcl/schemas/dist/platform/events/moderation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserWarningCreatedEvent = exports.UserBanLiftedEvent = exports.UserBanCreatedEvent = void 0;
const validation_1 = require("../../validation");
const base_1 = require("./base");
var UserBanCreatedEvent;
(function (UserBanCreatedEvent) {
    UserBanCreatedEvent.schema = {
        type: 'object',
        properties: {
            type: { type: 'string', const: base_1.Events.Type.MODERATION },
            subType: { type: 'string', const: base_1.Events.SubType.Moderation.USER_BAN_CREATED },
            key: { type: 'string' },
            timestamp: { type: 'number', minimum: 0 },
            metadata: {
                type: 'object',
                properties: {
                    id: { type: 'string' },
                    bannedAddress: { type: 'string' },
                    bannedBy: { type: 'string' },
                    reason: { type: 'string' },
                    bannedAt: { type: 'number', minimum: 0 },
                    expiresAt: { type: 'number', nullable: true, minimum: 0 },
                    customMessage: { type: 'string', nullable: true }
                },
                required: ['id', 'bannedAddress', 'bannedBy', 'reason', 'bannedAt', 'expiresAt']
            }
        },
        required: ['type', 'subType', 'key', 'timestamp', 'metadata'],
        additionalProperties: false
    };
    UserBanCreatedEvent.validate = (0, validation_1.generateLazyValidator)(UserBanCreatedEvent.schema);
})(UserBanCreatedEvent = exports.UserBanCreatedEvent || (exports.UserBanCreatedEvent = {}));
var UserBanLiftedEvent;
(function (UserBanLiftedEvent) {
    UserBanLiftedEvent.schema = {
        type: 'object',
        properties: {
            type: { type: 'string', const: base_1.Events.Type.MODERATION },
            subType: { type: 'string', const: base_1.Events.SubType.Moderation.USER_BAN_LIFTED },
            key: { type: 'string' },
            timestamp: { type: 'number', minimum: 0 },
            metadata: {
                type: 'object',
                properties: {
                    id: { type: 'string' },
                    bannedAddress: { type: 'string' },
                    liftedBy: { type: 'string' },
                    liftedAt: { type: 'number', minimum: 0 }
                },
                required: ['id', 'bannedAddress', 'liftedBy', 'liftedAt']
            }
        },
        required: ['type', 'subType', 'key', 'timestamp', 'metadata'],
        additionalProperties: false
    };
    UserBanLiftedEvent.validate = (0, validation_1.generateLazyValidator)(UserBanLiftedEvent.schema);
})(UserBanLiftedEvent = exports.UserBanLiftedEvent || (exports.UserBanLiftedEvent = {}));
var UserWarningCreatedEvent;
(function (UserWarningCreatedEvent) {
    UserWarningCreatedEvent.schema = {
        type: 'object',
        properties: {
            type: { type: 'string', const: base_1.Events.Type.MODERATION },
            subType: { type: 'string', const: base_1.Events.SubType.Moderation.USER_WARNING_CREATED },
            key: { type: 'string' },
            timestamp: { type: 'number', minimum: 0 },
            metadata: {
                type: 'object',
                properties: {
                    id: { type: 'string' },
                    warnedAddress: { type: 'string' },
                    warnedBy: { type: 'string' },
                    reason: { type: 'string' },
                    warnedAt: { type: 'number', minimum: 0 }
                },
                required: ['id', 'warnedAddress', 'warnedBy', 'reason', 'warnedAt']
            }
        },
        required: ['type', 'subType', 'key', 'timestamp', 'metadata'],
        additionalProperties: false
    };
    UserWarningCreatedEvent.validate = (0, validation_1.generateLazyValidator)(UserWarningCreatedEvent.schema);
})(UserWarningCreatedEvent = exports.UserWarningCreatedEvent || (exports.UserWarningCreatedEvent = {}));
